---
title: "PlatypusDB vignette"
author: "Alexander Yermanos, Victor Kreiner, Andreas Agrafiotis"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{PlatypusDB vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
fig_width: 16
fig_height: 16
---

## 1. Introduction

PlatypusDB is a publicly available database for combined gene expression and immune receptor sequencing datasets. The package Platypus provides and easy-to-use R interface to source available data, integrate it with locally saved datasets and comprehensively analyse VDJ and GEX data. For more information on the analysis function of the Platypus package please refer to the corresponding vignette. Here only the functions concerning interaction with the PlatypusDB will be covered.

Source code can be found at github.com/alexyermanos/Platypus 
More information and direct download links can be found on the joint Platypus and PlatypusDB website: <SOME URL>

## 2. Installation (Platypus v3 which contains all functions needed to access PlatypusDB)

Due to the recent changes of the default clonotyping strategy in Cellranger (version 5 and version 6) we are currently rebuilding v3 of Platypus to revolve around the VDJ_GEX_matrix function (vgm for short). This function integrates both repertoire and transcriptome information and will serve as the input to all secondary functions in future iterations of the package. The advantage of this is having all repertoire and transcriptome information at a per-cell level.

The change in clonotyping can be found here - https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/algorithms/clonotyping

The VDJ_GEX_matrix function will soon be found in the newest version of the R package (v3.0) with special thanks to Victor Kreiner. The current functions and documentation can be found already at https://github.com/alexyermanos/Platypus/blob/Functions/R/VDJ_GEX_matrix.R
All other functions are already or in the process of being updated. A function which is available for V3 as a new parameter "platypus.version" which can be set to either "v2", for backcompatibility or "v3". Few new functions are only compatible with "v3"

Can stay tuned for updates https://twitter.com/AlexYermanos

```{r setup, eval = TRUE, include=FALSE}
gc()
knitr::opts_chunk$set(fig.width=7, fig.height=7) 
```

```{r, eval = TRUE, include=FALSE}
#this is for sourcing all functions under development and is not included in the knitting
source_code_dir <- "C:/Dokumente usw/Master/Reddy Lab/1_thesis/3_code/4_DB project/PlatypusDB_admin/R" 
file_path_vec <- list.files(source_code_dir, full.names = T)
for(f_path in file_path_vec){
  print(f_path)
  tryCatch({source(f_path)}, error = function(e){e})
}
gc()
knitr::opts_chunk$set(fig.width=7, fig.height=7) 

library(tidyverse)
library(Seurat)
```

```{r, eval = TRUE, fig.show='hold', message=FALSE}

### Removing any previous versions of the package
#First can ensure that there is no previous version installed locally
#detach("package:Platypus", unload=TRUE)
#remove.packages("Platypus")

### Dependencies 
#install.packages("stringr")


### Downloading and installing Platypus

# First we need to download the most recent version from the master branch at https://github.com/alexyermanos/Platypus we can install the package using the following command. 
# WARNING: This needs to be replaced with your own directory where the downloaded package is found

# For MacOS users it may look like this
#install.packages("~/Downloads/Platypus_3.1.tar.gz", repos = NULL, type="source")

# For windows it will likely look something like this. 
# WARNING: You will need to replace 'YourPCName' with your user name for the windows account in the directory. 
#install.packages("C:/Users/YourPCName/Downloads/Platypus_3.1.tar.gz", repos = NULL, type="source")

# Now we can load the installed package into the R environment. In case of problems with installing other R packages that are used in Platypus, please see the README file at the https://github.com/alexyermanos/Platypus, where we outline how to install the other R packages for both Windows and MacOS.
library(Platypus)



# The individual R functions can additionally be found on the github in the Functions branch. Within this branch, there is a folder "R" which contains the individual functions. This can similarly be downloaded and loaded into the R environment in case not all functions are desired. These functions are actively updated and may include more features than the in original tar.gz file. 


```

## 3. List available datasets

To get familiar with the database, we download a list of the hosted projects and their metadata. 

```{r, eval = TRUE, fig.show='hold'}

projects_metadata <- PlatypusDB_list_projects()

```

Get the project IDs, which will used for downloading projects below:

Project IDs are composed of the first authors surname, the year of publication and optionally a letter for unique identification

```{r, eval = TRUE, fig.show='hold'}

names(projects_metadata)

```

For this example, we are interested in a dataset of central nervous system-infiltrating lymphocytes in experimental autoimmune enchepalomyelitis (EAE), a mouse model of multiple sclerosis. 

Two datasets are available for this project: kreiner2021a and kreiner2021b

We view the metadata to get information specific to these projects and their samples

```{r, eval = TRUE, fig.show='hold'}
#use View(projects_metadata[["kreiner2021a"]]) for a full overview

projects_metadata[["kreiner2021a"]][1:2,]

projects_metadata[["kreiner2021b"]][1:2,]

```

kreiner2021a contains data of T cells, while kreiner2021b concernes B cells. 

Sample specific information in the table are below

```{r, eval = TRUE, fig.show='hold'}

projects_metadata[["kreiner2021a"]][5:9,]

```

This dataset therefore contains three samples s1 - s3 with different immunization procedures. 

## 4. Download data

To download data we used the PlatypusDB_fetch function.
In the following several use cases are exemplified

The basis for all are PlatypusDB.links, with which any combination of files can be easily specified
They are composes of three section divided by "/"
1. project_id (Author2019letter, ... )
2. sample_id (s1, s2, s3, ...)
3. filetype (VDJmatrix, GEXmatrix, metadata, VDJ, GEX)

Any of these components can be substituted with "ALL". Further sample_id can be substituted with "". (See examples below) 

### 4.1. Download full VDJ_GEX_matrix

Given that the VDJ_GEX_matrix is a project-level file, we do not have to specify a sample_id. Further, because we wish to get a full VDJ_GEX_matrix, we have to download both the GEXmatrix as well as the VDJmatrix file and combine them. This is done automatically.

```{r, eval = TRUE, fig.show='hold'}

PlatypusDB_fetch(
    PlatypusDB.links = c("kreiner2021a//ALL"), #specifing no sample id and downloading all project level data available
    save.to.disk = F, #Whether to save it to a specified path. Necessary if download size exceeds available RAM
    load.to.enviroment = T, #Whether to load into the enviroment directly
    load.to.list = F, #Whether to return a list of loaded objects
    combine.objects = T, #Whether to combine objects appropriately. Needed here to get the full VDJ_GEX_matrix
    path.to.save = "~/Downloads/PlatypusDB_downloads/") #optional argument

```

Within the R Enviroment a new object has appeared: 
kreiner2021a__VDJGEXmatrix

This has the same structure as the output of the VDJ_GEX_matrix function and can be directly used as input to platypus functions

```{r, eval = TRUE, fig.show='hold'}

plots <- VDJ_clonal_expansion(VDJ = kreiner2021a__VDJGEXmatrix[[1]], clones = 30, color.by = "seurat_clusters", group.by = "sample_id")

plots[[1]][[2]]

```

### 4.2. Loading the VDJ matrix only

Here we also exemplify loading objects as a list

```{r, eval = TRUE, fig.show='hold'}

loaded_list <- PlatypusDB_fetch(
    PlatypusDB.links = c("kreiner2021a//VDJmatrix"),
    save.to.disk = F, 
    load.to.enviroment = F, 
    load.to.list = T) #Returns a list containing the downloaded objects

```

Again, we can directly call platypus functions on this output

```{r, eval = TRUE, fig.show='hold'}

plots <- VDJ_clonal_donut(VDJ = loaded_list[[1]][[1]])

plots[1]

```

### 4.3. Loading VDJ and GEX Rdata to run the VDJ_GEX_matrix

If we wish to recalculate the VDJ_GEX_matrix for a single sample, we can download the appropriate files and input those to the function

```{r, eval = TRUE, fig.show='hold'}

data_for_VGM <- PlatypusDB_fetch(
                PlatypusDB.links = c("kreiner2021a/s1/ALL"),
                save.to.disk = F, 
                load.to.enviroment = F, 
                load.to.list = T, 
                combine.objects = T)


s2_VGM <- VDJ_GEX_matrix(Data.in = data_for_VGM)

```

This object can again be analyzed further in Platypus

```{r, eval = TRUE, fig.show='hold'}

FeaturePlot(s2_VGM[[2]], features = c("PTPRC","CD4"))

```

### 4.3. Integrating database data and local data via the VDJ_GEX_matrix function

Here we download the VDJ and GEX files of all samples and additionally load locally saved samples 

```{r, eval = TRUE, fig.show='hold'}

downloaded_data <- PlatypusDB_fetch(
                   PlatypusDB.links = c("kreiner2021a/s1/ALL","kreiner2021a/s3/ALL"),
                   save.to.disk = F, 
                   load.to.enviroment = F, 
                   load.to.list = T, 
                   combine.objects = T)

```

As a local dataset we use T cells from the CNS of 4 naive 18 months old mice. This dataset is also available on PlatypusDB as yermanos2021a (https://doi.org/10.1098/rspb.2020.2793)

```{r, eval = TRUE, fig.show='hold'}

#load in local data
local_data <- PlatypusDB_load_from_disk(
              VDJ.out.directory.list = list("~/Downloads/Local_CNS_data/VDJ_S3"),
              GEX.out.directory.list = list("~/Downloads/Local_CNS_data/GEX_S3"))
  #To process data with Feature barcode technology refer to the Platypus Feature Barcode vignette

```

With both datasets loaded we can provide these as a list input to the VDJ_GEX_matrix function. 

Within this function several methods for GEX data integration are available. We suggest to test these out individually to ensure that batch effects are not the main source of variation between datasets. 
Available options for integration.method: 
'scale.data' corresponds to the ScaleData function internal to harmony package. All raw matrices are merged and then scaled. 
'anchors' scales data individually and then finds and align cells in similar states as described here: https://satijalab.org/seurat/articles/integration_introduction.html. 
'sct' specifies SCTransform from the Seurat package. 
'harmony' should be specificied to perform harmony integration. This method requires the harmony package from bioconductor.

```{r, eval = TRUE, fig.show='hold'}

comb_VGM <- VDJ_GEX_matrix(
            Data.in = list(downloaded_data, local_data),
            group.id = c("EAE rMOG","EAE MOG35-55", "18m aged CNS"),
            integration.method = "harmony")

DimPlot(comb_VGM[[2]], group.by = "group_id")

```

### 4.4. (VDJ only) Integrating database data and local data via the VDJ_GEX_matrix function

Here we download the VDJ files of all samples and additionally load locally saved samples

```{r, eval = TRUE, fig.show='hold'}

downloaded_data <- PlatypusDB_fetch(
                   PlatypusDB.links = c("kreiner2021a/s1/VDJ","kreiner2021a/s3/VDJ"),
                   save.to.disk = F, 
                   load.to.enviroment = F, 
                   load.to.list = T, 
                   combine.objects = T)


#load in local data
local_data <- PlatypusDB_load_from_disk(
              VDJ.out.directory.list = list("~/Downloads/Local_CNS_data/VDJ_S3"))

comb_VGM <- VDJ_GEX_matrix(
        Data.in = list(downloaded_data, local_data),
        group.id = c("EAE rMOG","EAE MOG35-55", "18m aged CNS"))

plots <- VDJ_Vgene_usage_stacked_barplot(VDJ = comb_VGM[[1]], platypus.version = "v3", HC.gene.number = 10)

plots

```

## 5 Searching for public CDR3s

In case that shared CDR3s sequences and clones are of interest, PlatypusDB allows searching for such without downloading all VDJ files within the database. 
If both VDJ and VJ CDR3s are provided, the function will also look for cells containing both sequences

```{r, eval = FALSE, fig.show='hold'}

public_clones <- PlatypusDB_find_CDR3s(VDJ.cdr3s.aa = c("CMRYGNYWYFDVW") , VJ.cdr3s.aa = c("CLQHGESPFTF"), projects.to.search = "ALL")

head(public_clones[[1]]) #subset of VDJ dataframes with query VDJ CDR3s
head(public_clones[[2]]) #subset of VDJ dataframes with query VJ CDR3s
head(public_clones[[3]]) #subset of VDJ dataframes with query VDJ AND VJ CDR3s
nrow(public_clones[[3]]) #Nr of cells in database containing query VDJ and VJ CDR3s

```

## 6 Version information

```{r, eval = TRUE, fig.show='hold'}
sessionInfo()
```


