---
title: "PlatypusDB vignette"
author: "Alexander Yermanos, Victor Kreiner, Andreas Agrafiotis"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{PlatypusDB vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
fig_width: 16
fig_height: 16
---

## 1. Introduction

PlatypusDB is a publicly available database for combined gene expression and immune receptor sequencing datasets. The package Platypus provides and easy-to-use R interface to source available data, integrate it with locally saved datasets and comprehensively analyse VDJ and GEX data. For more information on the analysis function of the Platypus package please refer to the corresponding vignettes. Here the functions concerning interaction with the PlatypusDB will be covered.

When using datasets from PlatypusDB please make sure to cite the source of the dataset as well as PlatypusDB at:

```{r eval = FALSE}

Yermanos A et al. Platypus: an open-access software for integrating lymphocyte single-cell immune repertoires with transcriptomes. NAR Genomics Bioinforma (2021) 3: doi:10.1093/NARGAB/LQAB023

*** DB BioRxiv Citation *** 
  
```

```{r setup, include=FALSE, eval = TRUE}
#to avoid of the dplyr summarise warnings
library(dplyr, warn.conflicts = FALSE)
options(dplyr.summarise.inform = FALSE)
library(stats, warn.conflicts = F)
knitr::opts_chunk$set(fig.width=7, fig.height=7) 
library(tidyverse)
library(Seurat)
library(Platypus)

# REMOVE FOR FINAL VIGNETTE ONCE BUGFIX IS ON CRAN 
#source("~/GitHub/Platypus/R/PlatypusDB_fetch.R")
#source("~/GitHub/Platypus/R/PlatypusDB_find_CDR3s.R")
```


```{r, fig.show='hold', message=FALSE, eval = TRUE}

library(Platypus)

```

## 2. List available datasets

For an overview and to read specific information about available dataset, a list of metadata table can be downloaded.

```{r, eval = TRUE, fig.show='hold'}

projects_metadata <- PlatypusDB_list_projects()

```
This is a list of named dataframes. 
Names correspond to project_ids, which are later used for downloading files.

```{r, eval = TRUE, fig.show='hold'}
names(projects_metadata)
```
Use list indexing to view the metadata to get information specific to these projects and their samples

```{r, eval = TRUE, fig.show='hold'}

projects_metadata[["yermanos2021a"]]$dataset_info

```

Sample specific information in the table are below

```{r, eval = TRUE, fig.show='hold'}

projects_metadata[["yermanos2021a"]]$sample_info
```

The dataset yermanos2021a contains data from T cells in the CNS of aged mice. Analysis of this dataset is showcased in the Platypus Quickstart vignette. 

! Note: In some cases, datasets contain GEX and VDJ for both B and T cells. Because VDJ libraries were prepared and sequenced separately, these are uploaded as two different projects, with the same GEX dataset.
e.g. yermanos2021a contains T cell VDJ and GEX data across cells, while yermanos2021b contains B cell VDJ and again GEX data across cells. To filter out certain celltypes in GEX before clustering, please refer to the VGM vignette and the parameter "exclude.on.cell.state.marker" 

## 3 Downloading data

The PlatypusDB_fetch function allows easy and flexible downloads using PlatypusDB.links. 

The basis for all are PlatypusDB.links, with which any combination of files can be easily specified
They are composes of three section divided by "/"
1. project_id (Author2019letter, ... )
2. sample_id (s1, s2, s3, ...) -> SEE METADATA
3. filetype (VDJmatrix, GEXmatrix, metadata, VDJ, GEX)

Any of these components can be substituted with "ALL". Further sample_id can be substituted with "". (See examples below)

## 3.1 Download raw data and integrate via VGM

Here we download the raw VDJ and GEX Cellranger output data for sample 1 and 2 of the dataset yermanos2021a 
(To download data from all samples, substitute "s1" with "ALL")

```{r, eval = TRUE, fig.show='hold', message=FALSE}

data_for_VGM <- PlatypusDB_fetch(
                PlatypusDB.links = c("yermanos2021a/Aged.CNS.pool.3m.Tcell.S1/ALL",
                                     "shlesinger2022b/TFH.LCMV.acuteLCMV.10dpi.S1/ALL"),
                save.to.disk = F, 
                load.to.enviroment = F, 
                load.to.list = T)

#Now we can process this via the VGM function.
integrated_VGM <- VDJ_GEX_matrix(Data.in = data_for_VGM, verbose = T, 
                          group.id = c("yermanos2021a","shlesinger2022b"))

#And plot a umap to check the output 
Seurat::DimPlot(integrated_VGM[[2]], group.by = "group_id")
Seurat::DimPlot(integrated_VGM[[2]], group.by = "VDJ_available")

```
## 3.2 Download processed VDJ_GEX_matrix

For each dataset, we generate a VDJ_GEX_matrix under default settings. This is to give quick access to a dataset which is ready for analysis, but comes at the cost of processing flexibility. 

If the sample level Platypus.link element is empty (as below), the data element can be either "VDJmatrix", "GEXmatrix" or "ALL".
In case of ALL both VDJ and GEX matrices are loaded and returned as one object. 

Indipendently of what matrix is loaded, the classical list output format of the VDJ_GEX_matrix is conserved. 

```{r, eval = TRUE, fig.show='hold', message=FALSE}

PlatypusDB_fetch(
                PlatypusDB.links = c("yermanos2021b//ALL"),
                save.to.disk = F, 
                load.to.enviroment = T, 
                load.to.list = F, 
                combine.objects = T) #Setting combine objects to true to return a single VGM

#Pass VGM onto Platypus analysis functions
VDJ_clonal_expansion(yermanos2021b__VDJGEXmatrix[[1]], color.by = "seurat_clusters")[[1]][[1]]

```
## 4. Integrating local and PlatypusDB data

Comparing local data to established datasets is one of the core functionalities of PlatypusDB. 

The intregration is based on a standard by-sample data format and the VDJ_GEX_matrix function.

Here we first load in a local dataset via the PlatypusDB_load_from_disk() function
(For showcasing purposes we load in 2 samples from the yermanos2021a dataset)

```{r, eval = TRUE, fig.show='hold'}

#load in local data
local_data <- PlatypusDB_load_from_disk(
              VDJ.out.directory.list = list("C:/Users/PlatypusDB/yermanos2021b__VDJ_RAW/Aged.CNS.pool.3m.Bcell.S1",                     "C:/Users/PlatypusDB/yermanos2021b__VDJ_RAW/Aged.CNS.pool.12m.Bcell.S2"),
              GEX.out.directory.list = list("C:/Users/PlatypusDB/yermanos2021b__GEX_RAW/Aged.CNS.pool.3m.Bcell.S1",                     "C:/Users/PlatypusDB/yermanos2021b__GEX_RAW/Aged.CNS.pool.12m.Bcell.S2"))
  #To process data with Feature barcode technology refer to the Platypus Feature Barcode vignette

```

We next load samples 3 and 4 of the same dataset from PlatypusDB

```{r, eval = TRUE, fig.show='hold'}

PlatypusDB_data <- PlatypusDB_fetch(
                PlatypusDB.links = c("yermanos2021b/Aged.CNS.pool.18m.Bcell.S3/ALL",
                                     "yermanos2021b/Aged.CNS.single.18m.Bcell.S4/ALL"),
                save.to.disk = F, 
                load.to.enviroment = F, 
                load.to.list = T)

#Again this format contains per sample data in a nested list format. 
names(PlatypusDB_data) #List level 1: Samples
names(PlatypusDB_data[[1]]) #List level 2: VDJ GEX FB data

```
Given the same format of loaded local data and downloaded data, the VDJ_GEX_matrix function can now be used to integrate both

```{r, eval = TRUE, fig.show='hold', message=FALSE}

vgm <- VDJ_GEX_matrix(Data.in = list(local_data , PlatypusDB_data))

#Pass VGM onto Platypus analysis functions
VDJ_clonal_expansion(vgm[[1]], color.by = "seurat_clusters")[[1]][[1]]

```


## 5. Searching for public CDR3s

In case that shared CDR3s sequences and clones are of interest, PlatypusDB allows searching for such without downloading all VDJ files within the database. 
If both VDJ and VJ CDR3s are provided, the function will also look for cells containing both sequences

```{r, eval = FALSE, fig.show='hold'}

public_clones <- PlatypusDB_find_CDR3s(VDJ.cdr3s.aa = c("CMRYGNYWYFDVW") , VJ.cdr3s.aa = c("CLQHGESPFTF"), projects.to.search = "ALL")

head(public_clones[[1]]) #subset of VDJ dataframes with query VDJ CDR3s
head(public_clones[[2]]) #subset of VDJ dataframes with query VJ CDR3s
head(public_clones[[3]]) #subset of VDJ dataframes with query VDJ AND VJ CDR3s
nrow(public_clones[[3]]) #Nr of cells in database containing query VDJ and VJ CDR3s

```

## 6. Version information

```{r, eval = TRUE, fig.show='hold'}
sessionInfo()
```

