<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single-cell immune repertoire analysis • Platypus</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Single-cell immune repertoire analysis">
<meta property="og:description" content="Allows the analysis of single cell immune repertoire data. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/vignette.html">FOR TESTING ONLY / Platypus vignette / FOR TESTING ONLY</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="https://repository-images.githubusercontent.com/297313954/10e0a180-713e-11eb-9a23-ef93a9d86e8b" alt="alt text"> # Platypus</p>
<p>Platypus is an R toolkit designed to facilitate the data analysis of single-cell immune repertoire sequencing experiments. The manuscript corresponding to Platypus v2 can be found here at Yermanos et al NARGAB 2021 <a href="https://doi.org/10.1093/nargab/lqab023" class="external-link uri">https://doi.org/10.1093/nargab/lqab023</a></p>
<div id="ongoing-updates-in-the-platypus-pipeline" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#ongoing-updates-in-the-platypus-pipeline" class="anchor" aria-hidden="true"></a>Ongoing updates in the Platypus pipeline</h1></div>
<p>Due to the recent changes of the default clonotyping strategy in Cellranger (version 5 and version 6) we are currently rebuilding v3 of Platypus to revolve around the VDJ_GEX_matrix function. This function integrates both repertoire and transcriptome information and will serve as the input to all secondary functions in future iterations of the package. The advantage of this is having all repertoire and transcriptome information at a per-cell level.</p>
<p>The change in clonotyping can be found here - <a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/algorithms/clonotyping" class="external-link uri">https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/algorithms/clonotyping</a></p>
<p>The VDJ_GEX_matrix function will soon be found in the newest version of the R package (v3.0) with special thanks to Victor Kreiner. The current functions and documentation can be found already at <a href="https://github.com/alexyermanos/Platypus/tree/Functions/R" class="external-link uri">https://github.com/alexyermanos/Platypus/tree/Functions/R</a></p>
<p>Stay tuned for updates <a href="https://twitter.com/AlexYermanos" class="external-link uri">https://twitter.com/AlexYermanos</a></p>
</div>
<div id="system-requirements" class="section level1">
<h1 class="hasAnchor">
<a href="#system-requirements" class="anchor" aria-hidden="true"></a>System requirements</h1>
<p>Platypus has been successfully installed on MacOS X (v10.14.6) and Windows 10 Pro (v1909), and used on R versions 4.0.0 and 3.6.1</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor" aria-hidden="true"></a>Installation</h1>
<p>The package can be installed directly from the tar.gz file on this GitHub. Please see the vignette for examples of how the package can be used.</p>
<p>Please scroll down for instructions on how to install the necessary dependencies.</p>
<pre class="{r}"><code>
### Removing any previous versions of the package
# First we will ensure that there is no previous version installed locally
#detach("package:Platypus", unload=TRUE)
#remove.packages("Platypus")

### Downloading and installing Platypus


# First we need to download the most recent version from the master branch at https://github.com/alexyermanos/Platypus we can install the package using the following command. 
# WARNING: This needs to be replaced with your own directory where the downloaded package is found

# For MacOS ( users it may look like this
#install.packages("~/Downloads/Platypus_2.0.4.tar.gz", repos = NULL, type="source")

# For windows it will likely look something like this. 
# WARNING: You will need to replace YourPCName with your User name for the windows account in the directory. 
# install.packages("C:\Users\YourPCName\Downloads\Platypus_2.0.4.tar.gz", repos = NULL, type="source")

# Now we can load the installed package into the R environment. 
#library(Platypus)

# The individual R functions can additionally be found on the github in the Functions branch. Within this branch, there is a folder "R" which contains the individual functions. This can similarly be downloaded and loaded into the R environment incase not all functions are desired. Similarly, these functions are actively updated and may include more features than the in original tar.gz file. 


### Downloading the test data
# The COVID-19 data (~136 MB size of the zip file) can be found at the following link https://polybox.ethz.ch/index.php/s/fxQJ3NrRSwiPSSo This dataset contains VDJ (separate libraries for B and T cells) and GEX libraries from two convalescent COVID-19 patients.

# After downloading the zip file named "PlatypusTestData.zip", please unzip the file and find the path to the newly formed folder. Typically this will be in the Downloads folder, so the code below should work on MacOS. For windows please uncomment the code and change the user name to match your PC.
</code></pre>
<p>Platypus uses a number of different R packages, some of which need prior installation. These can be installed either from CRAN, Bioconductor using “BiocManager” or GitHub using “devtools”:</p>
<div id="cran" class="section level2">
<h2 class="hasAnchor">
<a href="#cran" class="anchor" aria-hidden="true"></a>CRAN</h2>
<div id="ape" class="section level3">
<h3 class="hasAnchor">
<a href="#ape" class="anchor" aria-hidden="true"></a>ape</h3>
</div>
<div id="circlize" class="section level3">
<h3 class="hasAnchor">
<a href="#circlize" class="anchor" aria-hidden="true"></a>circlize</h3>
</div>
<div id="do" class="section level3">
<h3 class="hasAnchor">
<a href="#do" class="anchor" aria-hidden="true"></a>do</h3>
</div>
<div id="dplyr" class="section level3">
<h3 class="hasAnchor">
<a href="#dplyr" class="anchor" aria-hidden="true"></a>dplyr</h3>
</div>
<div id="ggplot2" class="section level3">
<h3 class="hasAnchor">
<a href="#ggplot2" class="anchor" aria-hidden="true"></a>ggplot2</h3>
</div>
<div id="ggseqlogo" class="section level3">
<h3 class="hasAnchor">
<a href="#ggseqlogo" class="anchor" aria-hidden="true"></a>ggseqlogo</h3>
</div>
<div id="igraph" class="section level3">
<h3 class="hasAnchor">
<a href="#igraph" class="anchor" aria-hidden="true"></a>igraph</h3>
</div>
<div id="jsonlite" class="section level3">
<h3 class="hasAnchor">
<a href="#jsonlite" class="anchor" aria-hidden="true"></a>jsonlite</h3>
</div>
<div id="phytools" class="section level3">
<h3 class="hasAnchor">
<a href="#phytools" class="anchor" aria-hidden="true"></a>phytools</h3>
</div>
<div id="reshape2" class="section level3">
<h3 class="hasAnchor">
<a href="#reshape2" class="anchor" aria-hidden="true"></a>reshape2</h3>
</div>
<div id="seqinr" class="section level3">
<h3 class="hasAnchor">
<a href="#seqinr" class="anchor" aria-hidden="true"></a>seqinr</h3>
</div>
<div id="seurat" class="section level3">
<h3 class="hasAnchor">
<a href="#seurat" class="anchor" aria-hidden="true"></a>Seurat</h3>
</div>
<div id="scales" class="section level3">
<h3 class="hasAnchor">
<a href="#scales" class="anchor" aria-hidden="true"></a>scales</h3>
</div>
<div id="stringdist" class="section level3">
<h3 class="hasAnchor">
<a href="#stringdist" class="anchor" aria-hidden="true"></a>stringdist</h3>
</div>
<div id="stringr" class="section level3">
<h3 class="hasAnchor">
<a href="#stringr" class="anchor" aria-hidden="true"></a>stringr</h3>
</div>
<div id="tibble" class="section level3">
<h3 class="hasAnchor">
<a href="#tibble" class="anchor" aria-hidden="true"></a>tibble</h3>
</div>
<div id="tidyverse" class="section level3">
<h3 class="hasAnchor">
<a href="#tidyverse" class="anchor" aria-hidden="true"></a>tidyverse</h3>
</div>
<div id="useful" class="section level3">
<h3 class="hasAnchor">
<a href="#useful" class="anchor" aria-hidden="true"></a>useful</h3>
</div>
<div id="utils" class="section level3">
<h3 class="hasAnchor">
<a href="#utils" class="anchor" aria-hidden="true"></a>utils</h3>
<p>Code to install the packages from CRAN:</p>
<pre class="{r}"><code>install.packages("ape")
install.packages("circlize")
install.packages("do")
install.packages("dplyr")
install.packages("ggplot2")
install.packages("ggseqlogo")
install.packages("igraph")
install.packages("jsonlite")
install.packages("phytools")
install.packages("reshape2")
install.packages("seqinr")
install.packages("Seurat")
install.packages("scales")
install.packages("stringdist")
install.packages("stringr")
install.packages("tibble")
install.packages("tidyverse")
install.packages("useful")
#install.packages("utils")</code></pre>
</div>
</div>
<div id="github" class="section level2">
<h2 class="hasAnchor">
<a href="#github" class="anchor" aria-hidden="true"></a>GitHub</h2>
<div id="harmony-but-not-required-if-not-using-the-harmony-integration-method" class="section level3">
<h3 class="hasAnchor">
<a href="#harmony-but-not-required-if-not-using-the-harmony-integration-method" class="anchor" aria-hidden="true"></a>Harmony (but not required if not using the Harmony integration method)</h3>
<p>Code for installing harmony:</p>
<pre class="{r}"><code>install.packages("devtools")
library(devtools)
install_github("immunogenomics/harmony")</code></pre>
<p>##Bioconductor</p>
</div>
<div id="biostrings" class="section level3">
<h3 class="hasAnchor">
<a href="#biostrings" class="anchor" aria-hidden="true"></a>Biostrings</h3>
</div>
<div id="orgmmegdb" class="section level3">
<h3 class="hasAnchor">
<a href="#orgmmegdb" class="anchor" aria-hidden="true"></a>org.Mm.eg.db</h3>
</div>
<div id="edger" class="section level3">
<h3 class="hasAnchor">
<a href="#edger" class="anchor" aria-hidden="true"></a>edgeR</h3>
</div>
<div id="fgsea" class="section level3">
<h3 class="hasAnchor">
<a href="#fgsea" class="anchor" aria-hidden="true"></a>fgsea</h3>
</div>
<div id="msa" class="section level3">
<h3 class="hasAnchor">
<a href="#msa" class="anchor" aria-hidden="true"></a>msa</h3>
<p>Code for installing the packages from Bioconductor:</p>
<pre class="{r}"><code>if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
    
BiocManager::install("Biostrings")
BiocManager::install("org.Mm.eg.db")
BiocManager::install("edgeR")
BiocManager::install("fgsea")
BiocManager::install("msa")</code></pre>
<p>Please post on the github page with any questions or if you would like to contribute.</p>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2">GPL-2</a></li>
</ul>
</div>


<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Alexander Yermanos <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Alexander Yermanos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
