<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CellPhoneDB • Platypus</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CellPhoneDB">
<meta property="og:description" content="Platypus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/quickstart.html">Get started</a>
</li>
<li>
  <a href="../vignette_index.html">Vignettes</a>
</li>
<li>
  <a href="../PlatypusDB_index.html">PlatypusDB</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://immunesim.readthedocs.io/en/latest/" class="external-link">ImmuneSIM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../cute_monotremes.html">Cute Monotremes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/alexyermanos" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>s
<li>
  <a href="https://github.com/alexyermanos/Platypus" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CellPhoneDB</h1>
                        <h4 data-toc-skip class="author">Vittoria
Martinolli D’Arcy, Andreas Agrafiotis, Victor Kreiner, Alexander
Yermanos</h4>
            
            <h4 data-toc-skip class="date">25/03/2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alexyermanos/Platypus/tree/mastervignettes/CellPhoneDB.Rmd" class="external-link"><code>vignettes/CellPhoneDB.Rmd</code></a></small>
      <div class="hidden name"><code>CellPhoneDB.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1 Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The Platypus family of packages are meant to provide potential
pipelines and examples relevant to the broad field of computational
immunology. The core set of functions can be found at <a href="https://github.com/alexyermanos/Platypus" class="external-link uri">https://github.com/alexyermanos/Platypus</a> and examples of
use can be found in the publications <a href="https://doi.org/10.1093/nargab/lqab023" class="external-link uri">https://doi.org/10.1093/nargab/lqab023</a> and
<strong>insert biorxiv database manuscript here</strong></p>
<p>Stay tuned for updates <a href="https://twitter.com/AlexYermanos" class="external-link uri">https://twitter.com/AlexYermanos</a></p>
<p>CellPhoneDB is a repository of curated receptors, ligands and their
interactions and includes tools which identify and analyse interactions
between a given pair of cells from single cell RNA sequencing data. We
have developed a pipeline for the implementation and adaptation of
CellPhoneDB functons in the Platypus package that can be used directly
on the GEX output of VDJ_GEX_matrix function.<br>
More specifically, CellPhoneDB_analyse is a function that allows the
user to organize the data so that it can be read and processed by
CellPhoneDB. The results of the analysis are then saved in a directory
“out” and are also added in a new vgm list item (CellPhoneDB). The
function requires prior installation of Python on the computer and the
running time can take several minutes.</p>
</div>
<div class="section level2">
<h2 id="loading-the-vgm">2. Loading the VGM<a class="anchor" aria-label="anchor" href="#loading-the-vgm"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Platypus</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#download raw data</span></span>
<span><span class="va">yermanos2021_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlatypusDB_fetch.html">PlatypusDB_fetch</a></span><span class="op">(</span>PlatypusDB.links <span class="op">=</span> </span>
<span>                                     <span class="st">"yermanos2021b/ALL/ALL"</span>, load.to.list <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#convert it into vgm object</span></span>
<span><span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>Data.in <span class="op">=</span> <span class="va">yermanos2021_raw</span>,</span>
<span>                      GEX.integrate <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                      VDJ.combine <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                      integrate.GEX.to.VDJ <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                      integrate.VDJ.to.GEX <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                      verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The output of the VDJ_GEX_matrix is the main object for all
downstream functions in Platypus, as well as CellPhoneDB. We can create
this directly into the R session by using the public data available on
PlatypusDB. <strong>ADD DATASET DESCRIPTION</strong>.</p>
</div>
<div class="section level2">
<h2 id="statistical-analysis-of-cells-interaction">3. Statistical analysis of cells’ interaction<a class="anchor" aria-label="anchor" href="#statistical-analysis-of-cells-interaction"></a>
</h2>
<div class="section level3">
<h3 id="running-the-cellphonedb_analyse-function">3.1. Running the CellPhoneDB_analyse function<a class="anchor" aria-label="anchor" href="#running-the-cellphonedb_analyse-function"></a>
</h3>
<p>Requirement for the following is the installation of Python 3 on the
local computer. The CellPhoneDB_analyse function takes the
VDJ_GEX_matrix function output as first input parameter (vgm.input). The
second input parameter (column) selects the column of the vgm object
that is used for filtering the groups/pairs to be analysed, with the
name of these being specified by the third input parameter (groups). The
fourth input parameter (organism) indicates the organism the data is
from. If this is set to “mouse”, the function converts the gene names
into humanised ones, because CellPhoneDB can only be applied to human
genes. The other input parameters are CellPhoneDB specific arguments and
allow the user to modify some analysis parameters, such as the gene
identity, the results precision, the possibility of subsampling the
cells, the p-value threshold and other.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vgm_cellphonedb</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/CellPhoneDB_analyse.html">CellPhoneDB_analyse</a></span><span class="op">(</span>vgm.input <span class="op">=</span> <span class="va">vgm</span>, <span class="co">#the function takes vgm as an input</span></span>
<span>                            column <span class="op">=</span> <span class="st">"seurat_clusters"</span>, <span class="co">#column name of vgm[[2]] where the groups to be tested for interactions are located</span></span>
<span>                            groups<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">9</span><span class="op">)</span>, <span class="co">#Vector with groups of interest in the indicated column</span></span>
<span>                            organism <span class="op">=</span> <span class="st">"mouse"</span>, <span class="co">#if == "mouse" the function needs to convert the gene's name, By default "human" </span></span>
<span>                            </span>
<span>                            <span class="co">##the next argumemts are the CellPhoneDB specific method arguments </span></span>
<span>                            gene.id <span class="op">=</span> <span class="st">"ensembl"</span>, <span class="co">#possible arguments: "ensembl" , "hgnc_symbol", "gene_name". By default "ensembl"</span></span>
<span>                            analysis.method <span class="op">=</span><span class="st">"statistical_analysis"</span>,  <span class="co">#possible arguments: "statistical_analysis", in the future also "degs_analysis" . By default "statistical_analysis"</span></span>
<span>                            project.name <span class="op">=</span> <span class="st">"subsampled"</span>, <span class="co">#Name of the project. If specified, a subfolder with this name is created in the output folder</span></span>
<span>                            iterations <span class="op">=</span><span class="fl">1000</span>, <span class="co">#Number of iterations for the statistical analysis. By default 1000</span></span>
<span>                            threshold <span class="op">=</span> <span class="fl">0.2</span>, <span class="co">#% of cells expressing the specific ligand/receptor. By default not specified. Range: 0&lt;= threshold &lt;=1.</span></span>
<span>                            result.precision <span class="op">=</span> <span class="fl">2</span>, <span class="co">#Number of decimal digits in results. By default 3</span></span>
<span></span>
<span>                            subsampling <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#If set to TRUE it enables subsampling. By default == FALSE </span></span>
<span>                            subsampling.num.pc <span class="op">=</span><span class="fl">100</span>, <span class="co">#Subsampling NumPC argument (number of PCs to use). By default 100, if subsampling == TRUE</span></span>
<span>                            subsampling.num.cells <span class="op">=</span> <span class="fl">150</span>, <span class="co">#Number of cells to subsample to. By default 1/3 of cells, if subsampling == TRUE.</span></span>
<span>                            subsampling.log <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co">#If set to TRUE it enables subsampling log1p for non log-transformed data inputs. Mandatory with no default, if subsampling == TRUE.  </span></span>
<span>                            </span>
<span>                            <span class="co">## statistical method arguments </span></span>
<span>                            pvalue <span class="op">=</span> <span class="fl">0.05</span>, <span class="co">#P-value threshold. By default 0.05.</span></span>
<span>                            debug.seed <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, <span class="co">#Debug random seed -1. To disable it please use a value &gt;=0. By default -1</span></span>
<span>                            threads <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, <span class="co">#Number of threads to use. &gt;=1. By default -1</span></span>
<span>                            install.cellphonedb <span class="op">=</span><span class="cn">FALSE</span>, <span class="co">#Boolean argument that installs the CellPhoneDB Python package if set ==TRUE. By default FALSE. </span></span>
<span>                            platypus.version <span class="op">=</span> <span class="st">"v3"</span>,</span>
<span>                            pre.code <span class="op">=</span> <span class="st">"conda init cmd.exe &amp;&amp; activate cpdb"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="looking-into-the-outpout-of-cellphonedb_analyse">3.2. Looking into the outpout of CellPhoneDB_analyse<a class="anchor" aria-label="anchor" href="#looking-into-the-outpout-of-cellphonedb_analyse"></a>
</h3>
<p>CellPhoneDB_analyse adds the list item $CellPhoneDB to the VGM as the
10th element of the VGM object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vgm_cellphonedb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "VDJ"            "GEX"            "VDJ.GEX.stats"  "Running params"</span></span>
<span><span class="co">##  [5] "sessionInfo"    ""               ""               ""              </span></span>
<span><span class="co">##  [9] ""               "CellPhoneDB"</span></span></code></pre>
<p>CellPhoneDB is a list where the results and plots of the CellPhoneDB
analysis are stored, with the following names:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vgm_cellphonedb</span><span class="op">$</span><span class="va">CellPhoneDB</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "deconvoluted"      "means"             "significant.means"</span></span>
<span><span class="co">## [4] "pvalues"           "interaction.count" "count.network"    </span></span>
<span><span class="co">## [7] "heatmap_log_count" "heatmap_count"     "dot_plot"</span></span></code></pre>
<p>The analysis results are summarised in .txt files in the current
directory of the user and saved in the form of data.frame in
CellPhoneDB.</p>
<p>More specifically, the “means” data.frame contains mean values for
each ligand-receptor interaction.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm_cellphonedb</span><span class="op">$</span><span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">means</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   id_cp_interaction                    interacting_pair</span></span>
<span><span class="co">## 1   CPI-CS0C6F86DF5  2arachidonoylglycerol_byDAGLA_CNR2</span></span>
<span><span class="co">## 2   CPI-CS030959ED1  2arachidonoylglycerol_byDAGLB_CNR2</span></span>
<span><span class="co">## 3   CPI-CS045C37E64 2arachidonoylglycerol_byDAGLA_GPR55</span></span>
<span><span class="co">## 4   CPI-CS01F7D06F1 2arachidonoylglycerol_byDAGLB_GPR55</span></span>
<span><span class="co">## 5   CPI-SC06D7BD0F6           FN1_integrin_aVb3_complex</span></span>
<span><span class="co">## 6   CPI-SC07763D025         LAMC1_integrin_aVb3_complex</span></span>
<span><span class="co">##                               partner_a                     partner_b</span></span>
<span><span class="co">## 1 complex:2arachidonoylglycerol_byDAGLA                 simple:P34972</span></span>
<span><span class="co">## 2 complex:2arachidonoylglycerol_byDAGLB                 simple:P34972</span></span>
<span><span class="co">## 3 complex:2arachidonoylglycerol_byDAGLA                 simple:Q9Y2T6</span></span>
<span><span class="co">## 4 complex:2arachidonoylglycerol_byDAGLB                 simple:Q9Y2T6</span></span>
<span><span class="co">## 5                         simple:P02751 complex:integrin_aVb3_complex</span></span>
<span><span class="co">## 6                         simple:P11047 complex:integrin_aVb3_complex</span></span>
<span><span class="co">##            gene_a          gene_b secreted receptor_a receptor_b</span></span>
<span><span class="co">## 1                 ENSG00000188822     True      False       True</span></span>
<span><span class="co">## 2                 ENSG00000188822     True      False       True</span></span>
<span><span class="co">## 3                 ENSG00000135898     True      False       True</span></span>
<span><span class="co">## 4                 ENSG00000135898     True      False       True</span></span>
<span><span class="co">## 5 ENSG00000115414                     True      False      False</span></span>
<span><span class="co">## 6 ENSG00000135862                     True      False      False</span></span>
<span><span class="co">##   annotation_strategy is_integrin group_1_3.group_1_3 group_1_3.group_2_6</span></span>
<span><span class="co">## 1             curated       False                0.02                0.00</span></span>
<span><span class="co">## 2             curated       False                0.04                0.00</span></span>
<span><span class="co">## 3             curated       False                0.19                0.07</span></span>
<span><span class="co">## 4             curated       False                0.21                0.10</span></span>
<span><span class="co">## 5             curated        True                0.00                0.00</span></span>
<span><span class="co">## 6             curated        True                0.26                0.00</span></span>
<span><span class="co">##   group_1_3.group_3_9 group_2_6.group_1_3 group_2_6.group_2_6</span></span>
<span><span class="co">## 1                0.00                0.02                0.00</span></span>
<span><span class="co">## 2                0.00                0.03                0.00</span></span>
<span><span class="co">## 3                0.07                0.19                0.08</span></span>
<span><span class="co">## 4                0.10                0.19                0.08</span></span>
<span><span class="co">## 5                0.00                0.00                0.00</span></span>
<span><span class="co">## 6                0.28                0.05                0.00</span></span>
<span><span class="co">##   group_2_6.group_3_9 group_3_9.group_1_3 group_3_9.group_2_6</span></span>
<span><span class="co">## 1                0.00                0.00                0.00</span></span>
<span><span class="co">## 2                0.00                0.02                0.00</span></span>
<span><span class="co">## 3                0.08                0.00                0.00</span></span>
<span><span class="co">## 4                0.08                0.19                0.08</span></span>
<span><span class="co">## 5                0.00                0.10                0.00</span></span>
<span><span class="co">## 6                0.07                0.13                0.00</span></span>
<span><span class="co">##   group_3_9.group_3_9</span></span>
<span><span class="co">## 1                0.00</span></span>
<span><span class="co">## 2                0.00</span></span>
<span><span class="co">## 3                0.00</span></span>
<span><span class="co">## 4                0.08</span></span>
<span><span class="co">## 5                0.12</span></span>
<span><span class="co">## 6                0.14</span></span></code></pre>
<p>The “deconvoluted” data.frame gives additional information for each
of the interacting partners. This is important as some of the
interacting partners are heteromers.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm_cellphonedb</span><span class="op">$</span><span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">deconvoluted</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   gene_name uniprot is_complex protein_name                  complex_name</span></span>
<span><span class="co">## 1     DAGLA  Q9Y4D2       True   DGLA_HUMAN 2arachidonoylglycerol_byDAGLA</span></span>
<span><span class="co">## 2     DAGLA  Q9Y4D2       True   DGLA_HUMAN 2arachidonoylglycerol_byDAGLA</span></span>
<span><span class="co">## 3     DAGLB  Q8NCG7       True   DGLB_HUMAN 2arachidonoylglycerol_byDAGLB</span></span>
<span><span class="co">## 4     DAGLB  Q8NCG7       True   DGLB_HUMAN 2arachidonoylglycerol_byDAGLB</span></span>
<span><span class="co">## 5   UBASH3B  Q8TF42       True  UBS3B_HUMAN  Dehydroepiandrosterone_bySTS</span></span>
<span><span class="co">## 6     DHCR7  Q9UBM7       True  DHCR7_HUMAN           Desmosterol_byDHCR7</span></span>
<span><span class="co">##   id_cp_interaction group_1_3 group_2_6 group_3_9</span></span>
<span><span class="co">## 1   CPI-CS0C6F86DF5      0.02      0.03      0.00</span></span>
<span><span class="co">## 2   CPI-CS045C37E64      0.02      0.03      0.00</span></span>
<span><span class="co">## 3   CPI-CS030959ED1      0.07      0.03      0.03</span></span>
<span><span class="co">## 4   CPI-CS01F7D06F1      0.07      0.03      0.03</span></span>
<span><span class="co">## 5   CPI-CS09B8977D7      0.07      0.00      0.15</span></span>
<span><span class="co">## 6   CPI-CS086D3E39F      0.08      0.07      0.16</span></span></code></pre>
<p>The “pvalues” and “significant.means” data.frames are the output of
the statistical inference analysis and they report the values of the
significant interactions.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm_cellphonedb</span><span class="op">$</span><span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">significant.means</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   id_cp_interaction interacting_pair     partner_a     partner_b</span></span>
<span><span class="co">## 1   CPI-SS0E9A581F8      ICAM1_ITGAL simple:P05362 simple:P20701</span></span>
<span><span class="co">## 2   CPI-SS09C4FCB90        ICAM1_SPN simple:P05362 simple:P16150</span></span>
<span><span class="co">## 3   CPI-SS09808DEAE         CD74_APP simple:P04233 simple:P05067</span></span>
<span><span class="co">## 4   CPI-SS061A7E099      CD55_ADGRE5 simple:P08174 simple:P48960</span></span>
<span><span class="co">## 5   CPI-SS0339AE950       SIRPA_CD47 simple:P78324 simple:Q08722</span></span>
<span><span class="co">## 6   CPI-SS00D9A0644       FGFR2_CD83 simple:P21802 simple:Q01151</span></span>
<span><span class="co">##            gene_a          gene_b secreted receptor_a receptor_b</span></span>
<span><span class="co">## 1 ENSG00000090339 ENSG00000005844    False       True       True</span></span>
<span><span class="co">## 2 ENSG00000090339 ENSG00000197471    False       True      False</span></span>
<span><span class="co">## 3 ENSG00000019582 ENSG00000142192    False       True      False</span></span>
<span><span class="co">## 4 ENSG00000196352 ENSG00000123146     True       True       True</span></span>
<span><span class="co">## 5 ENSG00000198053 ENSG00000196776    False       True       True</span></span>
<span><span class="co">## 6 ENSG00000066468 ENSG00000112149     True       True       True</span></span>
<span><span class="co">##   annotation_strategy is_integrin  rank group_1_3.group_1_3 group_1_3.group_2_6</span></span>
<span><span class="co">## 1             curated       False 0.111                0.43                  NA</span></span>
<span><span class="co">## 2             curated       False 0.111                  NA                  NA</span></span>
<span><span class="co">## 3             curated       False 0.111                  NA                  NA</span></span>
<span><span class="co">## 4             curated       False 0.111                  NA                  NA</span></span>
<span><span class="co">## 5             curated       False 0.222                0.63                  NA</span></span>
<span><span class="co">## 6        InnateDB-All       False 0.222                0.55                  NA</span></span>
<span><span class="co">##   group_1_3.group_3_9 group_2_6.group_1_3 group_2_6.group_2_6</span></span>
<span><span class="co">## 1                  NA                  NA                  NA</span></span>
<span><span class="co">## 2                0.31                  NA                  NA</span></span>
<span><span class="co">## 3                2.64                  NA                  NA</span></span>
<span><span class="co">## 4                  NA                  NA                  NA</span></span>
<span><span class="co">## 5                0.77                  NA                  NA</span></span>
<span><span class="co">## 6                  NA                  NA                  NA</span></span>
<span><span class="co">##   group_2_6.group_3_9 group_3_9.group_1_3 group_3_9.group_2_6</span></span>
<span><span class="co">## 1                  NA                  NA                  NA</span></span>
<span><span class="co">## 2                  NA                  NA                  NA</span></span>
<span><span class="co">## 3                  NA                  NA                  NA</span></span>
<span><span class="co">## 4                  NA                0.51                  NA</span></span>
<span><span class="co">## 5                  NA                  NA                  NA</span></span>
<span><span class="co">## 6                  NA                0.52                  NA</span></span>
<span><span class="co">##   group_3_9.group_3_9</span></span>
<span><span class="co">## 1                  NA</span></span>
<span><span class="co">## 2                  NA</span></span>
<span><span class="co">## 3                  NA</span></span>
<span><span class="co">## 4                  NA</span></span>
<span><span class="co">## 5                  NA</span></span>
<span><span class="co">## 6                  NA</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm_cellphonedb</span><span class="op">$</span><span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">pvalues</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   id_cp_interaction                    interacting_pair</span></span>
<span><span class="co">## 1   CPI-CS0C6F86DF5  2arachidonoylglycerol_byDAGLA_CNR2</span></span>
<span><span class="co">## 2   CPI-CS030959ED1  2arachidonoylglycerol_byDAGLB_CNR2</span></span>
<span><span class="co">## 3   CPI-CS045C37E64 2arachidonoylglycerol_byDAGLA_GPR55</span></span>
<span><span class="co">## 4   CPI-CS01F7D06F1 2arachidonoylglycerol_byDAGLB_GPR55</span></span>
<span><span class="co">## 5   CPI-SC06D7BD0F6           FN1_integrin_aVb3_complex</span></span>
<span><span class="co">## 6   CPI-SC07763D025         LAMC1_integrin_aVb3_complex</span></span>
<span><span class="co">##                               partner_a                     partner_b</span></span>
<span><span class="co">## 1 complex:2arachidonoylglycerol_byDAGLA                 simple:P34972</span></span>
<span><span class="co">## 2 complex:2arachidonoylglycerol_byDAGLB                 simple:P34972</span></span>
<span><span class="co">## 3 complex:2arachidonoylglycerol_byDAGLA                 simple:Q9Y2T6</span></span>
<span><span class="co">## 4 complex:2arachidonoylglycerol_byDAGLB                 simple:Q9Y2T6</span></span>
<span><span class="co">## 5                         simple:P02751 complex:integrin_aVb3_complex</span></span>
<span><span class="co">## 6                         simple:P11047 complex:integrin_aVb3_complex</span></span>
<span><span class="co">##            gene_a          gene_b secreted receptor_a receptor_b</span></span>
<span><span class="co">## 1                 ENSG00000188822     True      False       True</span></span>
<span><span class="co">## 2                 ENSG00000188822     True      False       True</span></span>
<span><span class="co">## 3                 ENSG00000135898     True      False       True</span></span>
<span><span class="co">## 4                 ENSG00000135898     True      False       True</span></span>
<span><span class="co">## 5 ENSG00000115414                     True      False      False</span></span>
<span><span class="co">## 6 ENSG00000135862                     True      False      False</span></span>
<span><span class="co">##   annotation_strategy is_integrin group_1_3.group_1_3 group_1_3.group_2_6</span></span>
<span><span class="co">## 1             curated       False                   1                   1</span></span>
<span><span class="co">## 2             curated       False                   1                   1</span></span>
<span><span class="co">## 3             curated       False                   1                   1</span></span>
<span><span class="co">## 4             curated       False                   1                   1</span></span>
<span><span class="co">## 5             curated        True                   1                   1</span></span>
<span><span class="co">## 6             curated        True                   1                   1</span></span>
<span><span class="co">##   group_1_3.group_3_9 group_2_6.group_1_3 group_2_6.group_2_6</span></span>
<span><span class="co">## 1                   1                   1                   1</span></span>
<span><span class="co">## 2                   1                   1                   1</span></span>
<span><span class="co">## 3                   1                   1                   1</span></span>
<span><span class="co">## 4                   1                   1                   1</span></span>
<span><span class="co">## 5                   1                   1                   1</span></span>
<span><span class="co">## 6                   1                   1                   1</span></span>
<span><span class="co">##   group_2_6.group_3_9 group_3_9.group_1_3 group_3_9.group_2_6</span></span>
<span><span class="co">## 1                   1                   1                   1</span></span>
<span><span class="co">## 2                   1                   1                   1</span></span>
<span><span class="co">## 3                   1                   1                   1</span></span>
<span><span class="co">## 4                   1                   1                   1</span></span>
<span><span class="co">## 5                   1                   1                   1</span></span>
<span><span class="co">## 6                   1                   1                   1</span></span>
<span><span class="co">##   group_3_9.group_3_9</span></span>
<span><span class="co">## 1                   1</span></span>
<span><span class="co">## 2                   1</span></span>
<span><span class="co">## 3                   1</span></span>
<span><span class="co">## 4                   1</span></span>
<span><span class="co">## 5                   1</span></span>
<span><span class="co">## 6                   1</span></span></code></pre>
<p>The “interaction.count” and “count.network” data.frames are generated
when plotting the heatmap and they report the output network and the
interactions count.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm_cellphonedb</span><span class="op">$</span><span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">interaction.count</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           X all_sum</span></span>
<span><span class="co">## 1 group_1_3      18</span></span>
<span><span class="co">## 2 group_2_6       1</span></span>
<span><span class="co">## 3 group_3_9      12</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm_cellphonedb</span><span class="op">$</span><span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">count.network</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      SOURCE    TARGET count</span></span>
<span><span class="co">## 1 group_1_3 group_1_3     8</span></span>
<span><span class="co">## 2 group_1_3 group_2_6     0</span></span>
<span><span class="co">## 3 group_1_3 group_3_9    10</span></span>
<span><span class="co">## 4 group_2_6 group_1_3     0</span></span>
<span><span class="co">## 5 group_2_6 group_2_6     0</span></span>
<span><span class="co">## 6 group_2_6 group_3_9     1</span></span></code></pre>
<p>The plots obtained with the CellPhoneDB analysis results
(heatmap_log_count, heatmap_count and dot_plot) are saved as .png files
in the current directory and included in the list item CellPhoneDB as
well and can be accessed as following. The dot_plot is not displayed
because it is too big for the vignette.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vgm_cellphonedb</span><span class="op">[[</span><span class="st">"CellPhoneDB"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"heatmap_log_count"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">vgm_cellphonedb</span><span class="op">[[</span><span class="st">"CellPhoneDB"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"heatmap_count"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#vgm_cellphonedb[["CellPhoneDB"]][["dot_plot"]]</span></span></code></pre></div>
<p>When more groups need to be compared, they can simply be included in
the vector argument “groups”. Most function arguments have their default
and do not need to be specified unless the user wants to modify the
default parameters. The dot_plot is not displayed because it is too big
for the vignette.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vgm_cellphonedb</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/CellPhoneDB_analyse.html">CellPhoneDB_analyse</a></span><span class="op">(</span>vgm.input <span class="op">=</span> <span class="va">vgm</span>, </span>
<span>                            groups<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">8</span>, <span class="fl">9</span><span class="op">)</span>, </span>
<span>                            organism <span class="op">=</span> <span class="st">"mouse"</span></span>
<span>                            <span class="op">)</span></span>
<span><span class="va">vgm_cellphonedb</span><span class="op">[[</span><span class="st">"CellPhoneDB"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"heatmap_count"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">vgm_cellphonedb</span><span class="op">[[</span><span class="st">"CellPhoneDB"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"heatmap_log_count"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#vgm_cellphonedb[["CellPhoneDB"]][["dot_plot"]]</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="customise-dot-plot">4 Customise Dot Plot<a class="anchor" aria-label="anchor" href="#customise-dot-plot"></a>
</h2>
<p>Since the Dot Plot obtained by CellPhoneDB analysis is particularly
large and contains data that is non-relevant for visualisation purposes,
the function dot_plot allows the user to customize the plot and create a
reduced one, with only user-specified information. The user can select
the rows and/or columns of interest to be plotted, as well as set a
p-value/log2means threshold, according to which specific genes to plot
are selected. Only genes that, for at least one interaction pair,
fulfill the threshold criteria, are plotted.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interacting_genes</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PVR_TNFSF9"</span>,  <span class="st">"PVR_CD96"</span> ,  <span class="st">"PVR_CD226"</span> ,  <span class="st">"PVR_TIGIT"</span> ,  <span class="st">"CADM1_CADM1"</span> , </span>
<span> <span class="st">"TIGIT_NECTIN2"</span> ,  <span class="st">"CD226_NECTIN2"</span> ,  <span class="st">"COL6A6_a10b1 complex"</span> ,  <span class="st">"FN1_a10b1 complex"</span> , <span class="st">" COL6A1_a10b1 complex"</span>,  <span class="st">"COL11A2_a10b1 complex"</span> ,  <span class="st">"COL8A2_a10b1 complex"</span> ,  <span class="st">"COL15A1_a10b1 complex"</span> ,  <span class="st">"COL14A1_a10b1 complex"</span> ,  <span class="st">"COL27A1_a10b1 complex"</span> ,  <span class="st">"COL12A1_a10b1 complex"</span> ,  <span class="st">"COL11A2_a1b1 complex"</span> ,  <span class="st">"COL8A2_a1b1 complex"</span> ,  <span class="st">"COL15A1_a1b1 complex"</span> ,  <span class="st">"COL14A1_a1b1 complex"</span> ,  <span class="st">"COL27A1_a1b1 complex"</span> ,  <span class="st">"COL12A1_a1b1 complex"</span> , <span class="st">" COL6A6_a2b1 complex"</span> ,  <span class="st">"FN1_a2b1 complex"</span> ,  <span class="st">"COL6A1_a2b1 complex"</span> ,  <span class="st">"COL6A2_a2b1 complex"</span> ,  <span class="st">"COL11A2_a2b1 complex"</span> ,   <span class="st">"COL27A1_a2b1 complex"</span> ,  <span class="st">"COL12A1_a2b1 complex"</span> ,  <span class="st">"LAMC1_a2b1 complex"</span> ,  <span class="st">"CDH1_a2b1 complex"</span> ,  <span class="st">"FN1_aVb3 complex"</span> ,  <span class="st">"LAMC1_aVb3 complex"</span> ,  <span class="st">"FGF2_aVb3 complex"</span> ,  <span class="st">"SPP1_aVb3 complex"</span> ,  <span class="st">"THY1_aVb3 complex"</span> ,  <span class="st">"MMP2_aVb3 complex"</span> ,  <span class="st">"L1CAM_aVb3 complex"</span> ,  <span class="st">"PLAUR_aVb3 complex"</span> ,  <span class="st">"FN1_a4b1 complex"</span> ,  <span class="st">"SPP1_a4b1 complex"</span> , <span class="st">" PLAUR_a4b1 complex"</span>,  <span class="st">"VCAM1_a4b1 complex"</span>,<span class="st">"JAM2_a4b1 complex"</span>,<span class="st">"FN1_a4b7 complex"</span>, <span class="st">"VCAM1_a4b7 complex"</span> , <span class="st">"LIFR_OSM"</span>, <span class="st">"NKG2D II receptor_MICA"</span>,<span class="st">"NKG2D II receptor_RAET1E"</span>,<span class="st">"CCL24_CCR2"</span>, <span class="st">"CCL24_CCR3"</span>, <span class="st">"CCL3L1_CCR3"</span>, <span class="st">"CCL3L1_CCR1"</span>,<span class="st">"CCL22_CCR4"</span>,<span class="st">"CCL3L1_DPP4"</span>,<span class="st">"CCL22_DPP4"</span>,<span class="st">"CXCL10_DPP4"</span> ,<span class="st">"CXCL10_CXCR3"</span> , <span class="st">"CXCL13_CXCR5"</span>,<span class="st">"PDGFB_LRP1"</span>,  <span class="st">"MDK_LRP1"</span>, <span class="st">"SIRPA_CD47"</span>, <span class="st">"MDK_SORL1"</span>, <span class="st">"PDGFB_ADGRV1"</span> , <span class="st">"XCL1_ADGRV1"</span>,<span class="st">"EFNA5_EPHA7"</span>,<span class="st">"GDF11_ANTXR1"</span>, <span class="st">"FGFR2_CD83"</span> , <span class="st">"FGFR2_XPR1"</span> , <span class="st">"PLXNB2_PTN"</span>,                   <span class="st">"PLXNB2_SEMA4D"</span>, <span class="st">"CD72_SEMA4D"</span> ,<span class="st">"PLXNB2_SEMA4G"</span> , <span class="st">"LTA_LTBR"</span>, <span class="st">"LTA_TNFRSF14"</span>, <span class="st">"TNFSF13_TNFRSF14"</span>,             <span class="st">"CD160_TNFRSF14"</span>, <span class="st">"MIF_TNFRSF14"</span>, <span class="st">"BTLA_TNFRSF14"</span>, <span class="st">"TNFSF13_TNFRSF17"</span>, <span class="st">"TNFSF10_TNFRSF10D"</span>,            <span class="st">"MIF_TNFRSF10D"</span>,<span class="st">"GDF11_TGFR_AVR2A"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vgm_cellphonedb</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/dot_plot.html">dot_plot</a></span><span class="op">(</span>vgm.input <span class="op">=</span> <span class="va">vgm_cellphonedb</span>, <span class="co">#vgm where to save the dotplot</span></span>
<span>                          selected.rows<span class="op">=</span> <span class="va">interacting_genes</span>, <span class="co">#vector of rows to plot (interacting genes pairs). By default NULL.</span></span>
<span>                          selected.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"group_1_3|group_2_6"</span>, <span class="st">"group_1_3|group_3_9"</span>, <span class="st">"group_2_6|group_1_3"</span>, <span class="st">"group_2_6|group_3_9"</span>, <span class="st">"group_3_9|group_1_3"</span>, <span class="st">"group_3_9|group_2_6"</span><span class="op">)</span>, <span class="co">#vector of columns (interacting group pairs). By default NULL.</span></span>
<span>                          threshold.type<span class="op">=</span><span class="st">"log2means_topn"</span>, <span class="co">#thresholding criterion. By default NULL.</span></span>
<span>                          threshold.value<span class="op">=</span><span class="fl">50</span>, <span class="co">#Number of top values to select or value below (for pvalue)/above (for log2mean) which to select the genes. By default  NULL.</span></span>
<span>                          project.name <span class="op">=</span> <span class="st">"subsampled"</span>, <span class="co">#By default to NULL.</span></span>
<span>                          filename <span class="op">=</span> <span class="st">"selection_plot.png"</span>, <span class="co">#name of the output file saved</span></span>
<span>                          height <span class="op">=</span> <span class="fl">20</span>, <span class="co">#height of the plot</span></span>
<span>                          width<span class="op">=</span><span class="fl">8</span>, <span class="co">#width of the plot</span></span>
<span>                          text.size <span class="op">=</span> <span class="fl">9</span>, <span class="co">#size of the text in the plot</span></span>
<span>                          return.vector<span class="op">=</span><span class="cn">TRUE</span> <span class="co">#Defaults to FALSE. If set to TRUE, it includes the vector of genes_pairs present in the dot_plot in the vgm CellPhoneDB list item. </span></span>
<span>                          <span class="op">)</span></span></code></pre></div>
<p>The resulting plot is saved in the current user directory in the
subfolder of the output of the CellPhoneDB analysis, as well as in the
vgm object as an additional list item of CellPhoneDB. If the user is
interested in a specific interaction group pair only, this can be
selected alone with the selected.columns argument.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vgm_cellphonedb</span><span class="op">$</span><span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">selection_dot_plot</span></span></code></pre></div>
<p><img src="CellPhoneDB_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>If selected rows/columns are not specified, all columns and rows are
taken into consideration. In this case, results might appear more
chaotic, especially with threshold.type=“pvalue” and
threshold.type=“log2means”.</p>
<p>In order to get the vector of interacting genes pairs displayed in
the dot_plot, one needs to set the argument “return.vector” to TRUE. The
vector will be included in the vgm CellPhoneDB list item as
$interacting_pairs and can be accessed as following:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vgm_cellphonedb</span><span class="op">[[</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">interacting_pairs</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "SIRPA_CD47"    "MIF_TNFRSF14"  "FGFR2_CD83"    "CXCL10_CXCR3" </span></span>
<span><span class="co">##  [5] "FGFR2_XPR1"    "CXCL10_DPP4"   "PLXNB2_SEMA4D" "PVR_TIGIT"    </span></span>
<span><span class="co">##  [9] "BTLA_TNFRSF14" "PLXNB2_SEMA4G"</span></span></code></pre>
<p>Moreover, this argument can also be useful to have the vector of all
the possible interacting gene pairs in the right format, choose the ones
of interest and include them in the selected.rows parameter. This can be
achieved by running the function without rows or threshold restriction
and setting the argument return.vector to TRUE.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vgm_cellphonedb</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/dot_plot.html">dot_plot</a></span><span class="op">(</span>vgm.input <span class="op">=</span> <span class="va">vgm_cellphonedb</span>, return.vector<span class="op">=</span><span class="cn">TRUE</span>, project.name <span class="op">=</span> <span class="st">"subsampled"</span><span class="op">)</span></span>
<span><span class="va">vgm_cellphonedb</span><span class="op">[[</span><span class="st">"CellPhoneDB"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">interacting_pairs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "2arachidonoylglycerol_byDAGLA_CNR2"  "2arachidonoylglycerol_byDAGLB_CNR2" </span></span>
<span><span class="co">##  [3] "2arachidonoylglycerol_byDAGLA_GPR55" "2arachidonoylglycerol_byDAGLB_GPR55"</span></span>
<span><span class="co">##  [5] "FN1_integrin_aVb3_complex"           "LAMC1_integrin_aVb3_complex"        </span></span>
<span><span class="co">##  [7] "THBS1_integrin_aVb3_complex"         "FCER2_integrin_aVb3_complex"        </span></span>
<span><span class="co">##  [9] "FGF1_integrin_aVb3_complex"          "PLAUR_integrin_aVb3_complex"</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alexander Yermanos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
