<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PlatypusV3 Clonotyping and multiple chains • Platypus</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="PlatypusV3 Clonotyping and multiple chains">
<meta property="og:description" content="Platypus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/quickstart.html">Get started</a>
</li>
<li>
  <a href="../vignette_index.html">Vignettes</a>
</li>
<li>
  <a href="../PlatypusDB_index.html">PlatypusDB</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://immunesim.readthedocs.io/en/latest/" class="external-link">ImmuneSIM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../cute_monotremes.html">Cute Monotremes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/alexyermanos" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/alexyermanos/Platypus" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>PlatypusV3 Clonotyping and multiple chains</h1>
                        <h4 data-toc-skip class="author">Victor Kreiner,
Tudor-Stefan Cotet, Alexander Yermanos</h4>
            
            <h4 data-toc-skip class="date">2022-08-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alexyermanos/Platypus/tree/mastervignettes/clonotyping.Rmd" class="external-link"><code>vignettes/clonotyping.Rmd</code></a></small>
      <div class="hidden name"><code>clonotyping.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Clonotyping describes the process of assigning cells to clonal
groups. These groups encompas all cells carrying T or B cell receptors
that were generated in the same recombination event in a single
anchestoral cell.</p>
<p>Classical clonotyping is based on identical sequences of VJ and VDJ
CDR3 regions. But this strategy may not be the best strategy given
somatic hypermutation may occur in the CDR3 region. Therefore, there
could be highly similar clones that likely bind the same antigen that
are officially part of different clonal families. To address this, we
have added a function that allows for various heuristic clonotyping
strategies. This involves clonotyping by identical amino acid CDRH3 +
CDRL3 sequence, identical germline usage, or seqeunce homology
requirements. This function works both for V3 and V2 of platypus. The
version needs to be specified. For V3 (vgm) platypus.version a new
clonotyping function was added to improve the existing one. For the full
documentation see VDJ_clonotype().</p>
<p>Three new parameters have been added therein:</p>
<ol style="list-style-type: decimal">
<li><p>global.clonotype If set to TRUE, clonotypes will be generated
across all samples of the vgm. This may be useful if more than one
sample has been taken from the same animal (e.g. spleen and bone marrow)
and processed in separate Cellranger runs. Moreover this can provide an
alternative approach to investigate clone sharing across donors or mice.
After global clonotyping, the resulting clonotype_id can be used as a
grouping variable and summary statistics can be generated per clone
(including its presence in multiple repertoires.)</p></li>
<li><p>Hierarchical clonotyping for cells with aberrant chain numbers A
key issue in Clonotyping is how to deal with cells with aberrant chain
numbers (e.g. 0VDJ 1VJ, 1VDJ 0VJ, 1VDJ 2VJ, …) Cells with only a single
chain are common due to stochasitcal sampling of mRNA. Cells with more
than 2 chains are uncommon. These originate from either cell doublets,
from contamination with free cellular mRNA during capture or lysis.
Further it is not to disregard, that cells carrying 2VJ chains are
described in literature.</p></li>
</ol>
<p>Hierarchical clonotyping in Platypus was build to adress these issues
while giving the user ample flexibility. The underlying concept is to
first define “high confidence” clones based on 1VDJ 1VJ cells and later
slot in aberrant cells. A key update as part of the hierarchical
clonotyping strategy in v3 is the same-sample-criterion. This is
explained below.</p>
<p>For cells with aberrant numbers of VDJ and VJ chains the functions
implements a hierarchical clonotyping. In brief, clonotypes are
determined based on all cells with exactly 1VDJ and 1VJ chain. Other
cells are then “joined in” in post. Three different modes are
available:</p>
<ol style="list-style-type: decimal">
<li><p>hierarchical = “none” Create clonotypes from 1 VDJ 1 VJ cells.
Aberrant cells are be grouped separately (two cells with the same VJ
chain but both without a VDJ chain are still grouped as one
clonotype.)</p></li>
<li><p>hierarchical = “single.chains” Cells with a single chain
(e.g. 0VDJ and 1VJ or 1VDJ and 0VJ) are merged into existing clonotypes
containing the same or a homologous VJ chain, depending on the
clonotyping strategy. Here the same-sample-criterion is applied when
using global clonotyping. If multiple matching clonotypes have been
found for that cell then ones within the same sample_id are prioritized.
If further multiple matches exist in that sample_id, the cell is merged
into the clone with the highest frequency.</p></li>
<li><p>hierarchical = “double.and.single.chains” A combination of 1VDJ
and 2VJ chain will first be checked for frequency in the dataset (by
exact nucleotide CDR3 matching). Only if this count exceeds the
triple.chain.count.threshold, the clone is used as a “hub clone”. This
protects from merging clonotypes on the basis of rare doublets. Existing
clonotypes are then merged into the 1 VDJ 2 VJ clonotypes as they match
with the assumption that e.g. a cell with 1 VDJ 1 VJ is part of that
same clonotype, but missing a VJ chain due to stochastical
sampling.</p></li>
</ol>
<p>!Sidenote on aberrant cell filtering: The VDJ_GEX_matrix function
offers the feature to filter cells with aberrant chains or to select
chains in cells with 2 VDJ or VJ chains based on the contig UMI count.
For usage examples see the VDJ_GEX_matrix documentation or the
VDJ_GEX_matrix (VGM) vignette.</p>
<p>In the example that follows, we use VDJ_clonotype to group cells into
clones based on identical CDRH3 + CDRL3 amino acid sequence. We will
compare this to the case in which we group B cells by using the same
germline genes (both heavy chain and light chain).</p>
</div>
<div class="section level2">
<h2 id="examing-effects-of-different-clonotyping-strategies">2. Examing effects of different clonotyping strategies<a class="anchor" aria-label="anchor" href="#examing-effects-of-different-clonotyping-strategies"></a>
</h2>
<p>For this example we download a processed VGM from the Aged CNS
dataset also used in the quickstart vignettes.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Platypus</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/PlatypusDB_fetch.html">PlatypusDB_fetch</a></span><span class="op">(</span>PlatypusDB.links <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"yermanos2021a//VDJmatrix"</span><span class="op">)</span>, </span>
<span>                                     load.to.enviroment <span class="op">=</span> <span class="cn">T</span>, combine.objects <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "yermanos2021a__VDJmatrix"</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Checking clonal expansion from 10x default clonotyping</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_donut.html">VDJ_clonal_donut</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                          counts.to.use <span class="op">=</span> <span class="st">"clonotype_id_10x"</span> <span class="co">#Column name of clonotype ids to plot</span></span>
<span>                          <span class="op">)</span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="clonotyping_files/figure-html/unnamed-chunk-2-1.png" width="672"></p>
<p>We see only 6 expanded clones and a total of 16 cells within these
clones based on 10x default clonotyping.</p>
</div>
<div class="section level2">
<h2 id="clonotyping-by-homology">2.1 Clonotyping by homology<a class="anchor" aria-label="anchor" href="#clonotyping-by-homology"></a>
</h2>
<p>Options for clonotyping strategies are: 10x.default, cdr3.nt,
cdr3.aa, VDJJ.VJJ, VDJJ.VJJ.cdr3length,
VDJJ.VJJ.cdr3length.cdr3.homology, VDJJ.VJJ.cdr3length.VDJcdr3.homology,
cdr3.homology, VDJcdr3.homology</p>
<p>We therefore assess the effects of a more inclusive clonotyping
strategy, which is based on homology of VDJ CDR3 regions</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/VDJ_clonotype.html">VDJ_clonotype</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  VDJ.VJ.1chain <span class="op">=</span> <span class="cn">F</span>, <span class="co">#Keeping cells with aberrant chain numbers</span></span>
<span>  clone.strategy <span class="op">=</span> <span class="st">"VDJcdr3.homology"</span>,</span>
<span>  homology.threshold <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="co"># corresponds to 80% sequence similarity</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_donut.html">VDJ_clonal_donut</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                          counts.to.use <span class="op">=</span> <span class="st">"clonotype_id_VDJcdr3.homology"</span> <span class="co">#Column name of clonotype ids to plot</span></span>
<span>                          <span class="op">)</span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="clonotyping_files/figure-html/unnamed-chunk-3-1.png" width="672">
Basing clonotyping only on homology of the VDJ cdr3 region, the number
of cells in expanded clones increased to 36.</p>
</div>
<div class="section level2">
<h2 id="clonotyping-by-v-and-j-gene-matching">2.2 Clonotyping by V and J gene matching<a class="anchor" aria-label="anchor" href="#clonotyping-by-v-and-j-gene-matching"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/VDJ_clonotype.html">VDJ_clonotype</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  VDJ.VJ.1chain <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  clone.strategy <span class="op">=</span> <span class="st">"VDJJ.VJJ"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_donut.html">VDJ_clonal_donut</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                          counts.to.use <span class="op">=</span> <span class="st">"clonotype_id_VDJJ.VJJ"</span><span class="op">)</span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="clonotyping_files/figure-html/unnamed-chunk-4-1.png" width="672">
In this small dataset, this yields to a reduction in expanded cells.</p>
</div>
<div class="section level2">
<h2 id="clonotyping-by-exact-cdr3-matching">2.3 Clonotyping by exact cdr3 matching<a class="anchor" aria-label="anchor" href="#clonotyping-by-exact-cdr3-matching"></a>
</h2>
<p>The function also allows to clonotype by exact sequence matching. The
results of this can vary from 10x default clonotyping due to the
handling</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/VDJ_clonotype.html">VDJ_clonotype</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  VDJ.VJ.1chain <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  clone.strategy <span class="op">=</span> <span class="st">"cdr3.aa"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_donut.html">VDJ_clonal_donut</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">yermanos2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                          counts.to.use <span class="op">=</span> <span class="st">"clonotype_id_cdr3.aa"</span><span class="op">)</span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="clonotyping_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="hierarchical-clonotyping">3. Hierarchical clonotyping<a class="anchor" aria-label="anchor" href="#hierarchical-clonotyping"></a>
</h2>
<p>In the examples above, the argument “hierarchical” was left at its
default: “none”. Here we explore the two other options: “single.chains”
and “double.and.single.chains”</p>
<p>To illustrate this, we make use of a larger VDJ dataset: T cells from
the CNS of EAE mice For more information on this dataset see
(doi.org/10.1101/2022.02.07.479381 Fig.6)</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlatypusDB_fetch.html">PlatypusDB_fetch</a></span><span class="op">(</span>PlatypusDB.links <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"kreiner2021a//VDJmatrix"</span><span class="op">)</span>, </span>
<span>                                     load.to.enviroment <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "kreiner2021a__VDJmatrix"</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Checking clonal expansion from 10x default clonotyping</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_donut.html">VDJ_clonal_donut</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                          counts.to.use <span class="op">=</span> <span class="st">"clonotype_id_10x"</span> <span class="co">#Column name of clonotype ids to plot</span></span>
<span>                          <span class="op">)</span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="clonotyping_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="filtering-out-cells-with-aberrant-chain-numbers">3.1 Filtering out cells with aberrant chain numbers<a class="anchor" aria-label="anchor" href="#filtering-out-cells-with-aberrant-chain-numbers"></a>
</h2>
<p>It is common practice to disregard cells with aberrant chain numbers.
The function offers this feature without adding a new clonotyping
strategy.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kreiner2021a__VDJmatrix_FILTERED</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/VDJ_clonotype.html">VDJ_clonotype</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  VDJ.VJ.1chain <span class="op">=</span> <span class="cn">T</span>, <span class="co">#Filtering aberrant cells</span></span>
<span>  clone.strategy <span class="op">=</span> <span class="st">"10x.default"</span><span class="op">)</span> <span class="co">#Leaving clonotyping strategy at default </span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Cells before filtering"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Cells before filtering"</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4811</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Cells after filtering"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Cells after filtering"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">kreiner2021a__VDJmatrix_FILTERED</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3531</span></span></code></pre>
<p>In this case a large number of cells was removed. As this does reduce
sample size for comparisons, we recommend to keep such cell.</p>
</div>
<div class="section level2">
<h2 id="hierarchical-single-chains">3.2 hierarchical = “single.chains”<a class="anchor" aria-label="anchor" href="#hierarchical-single-chains"></a>
</h2>
<p>This setting will integrate cells with only a single chain into
existing clonotypes defined by 1VDJ 1VJ chain cells</p>
<p>We clonotype again by cd3.aa sequence and visualize this merging of
clontypes using the flexible VDJ_clonal_expansion function</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/VDJ_clonotype.html">VDJ_clonotype</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  VDJ.VJ.1chain <span class="op">=</span> <span class="cn">F</span>, <span class="co">#Not filtering aberrant cells</span></span>
<span>  clone.strategy <span class="op">=</span> <span class="st">"cdr3.aa"</span>,</span>
<span>  hierarchical <span class="op">=</span> <span class="st">"single.chains"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Moving the new clonotype id column into the standard slot. </span></span>
<span><span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_id</span> <span class="op">&lt;-</span> <span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_id_cdr3.aa</span></span>
<span></span>
<span><span class="co">#Plotting clonal expansion and coloring bars by the number of VJ chains of each included cell</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_expansion.html">VDJ_clonal_expansion</a></span><span class="op">(</span><span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, color.by <span class="op">=</span> <span class="st">"Nr_of_VJ_chains"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="clonotyping_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<p>The green sections indicate that cells with only a single VDJ chain
have been integrated into existing high confidence clonotypes. Cells
with 2VJ chains on the other hand were grouped seperately. We suggest
this plot as a quality control.</p>
</div>
<div class="section level2">
<h2 id="hierarchical-double-and-single-chains">3.3 hierarchical = “double.and.single.chains”<a class="anchor" aria-label="anchor" href="#hierarchical-double-and-single-chains"></a>
</h2>
<p>This setting will integrate cells as discussed in 3.2. Further it
will join 1VDJ and 1VJ cells into clonotypes of 1VDJ 2VJ cells (deemed
“hub clonotypes”). Give that 1VDJ 2VJ chains can arise from mRNA
contamination but can also be biological, a filtering parameter helps to
set a cutoff against contamination.</p>
<p>We clonotype again by cd3.aa sequence and visualize this merging of
clontypes using the flexible VDJ_clonal_expansion function</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/VDJ_clonotype.html">VDJ_clonotype</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  VDJ.VJ.1chain <span class="op">=</span> <span class="cn">F</span>, <span class="co">#Not filtering aberrant cells</span></span>
<span>  clone.strategy <span class="op">=</span> <span class="st">"cdr3.aa"</span>,</span>
<span>  hierarchical <span class="op">=</span> <span class="st">"double.and.single.chains"</span>,</span>
<span>  triple.chain.count.threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co">#This requires the same 1VDJ 2VJ sequence combination to appear in at least 2 cells for it to be considered a hub clonotype. </span></span>
<span></span>
<span><span class="co">#Moving the new clonotype id column into the standard slot. </span></span>
<span><span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_id</span> <span class="op">&lt;-</span> <span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_id_cdr3.aa</span></span>
<span></span>
<span><span class="co">#Plotting clonal expansion and coloring bars by the number of VJ chains of each included cell</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_expansion.html">VDJ_clonal_expansion</a></span><span class="op">(</span><span class="va">kreiner2021a__VDJmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, color.by <span class="op">=</span> <span class="st">"Nr_of_VJ_chains"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="clonotyping_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p>As we can appreciate, cells with 1VDJ 2VJ chains were no used as hub
clones and integrated. Here, this increases the number of cells in the
second most expanded clonotype.</p>
</div>
<div class="section level2">
<h2 id="conclusion">4 Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Following the multitude of available clonotyping and filtering
strategies, Platypus aims to provide maximum flexibility to the user. We
recommend exploring different clonotyping strategies and integration
methods for aberrant cells, so to choose setting which are most suitable
for the current dataset and biological query.</p>
</div>
<div class="section level2">
<h2 id="version-information">5 Version information<a class="anchor" aria-label="anchor" href="#version-information"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23 ucrt)</span></span>
<span><span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span><span class="co">## Running under: Windows 10 x64 (build 19044)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] LC_COLLATE=German_Germany.utf8  LC_CTYPE=German_Germany.utf8   </span></span>
<span><span class="co">## [3] LC_MONETARY=German_Germany.utf8 LC_NUMERIC=C                   </span></span>
<span><span class="co">## [5] LC_TIME=German_Germany.utf8    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] forcats_0.5.1   stringr_1.4.0   purrr_0.3.4     readr_2.1.2    </span></span>
<span><span class="co">##  [5] tidyr_1.2.0     tibble_3.1.8    ggplot2_3.3.6   tidyverse_1.3.2</span></span>
<span><span class="co">##  [9] Platypus_3.4.1  dplyr_1.0.9    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] nlme_3.1-157        fs_1.5.2            ggtree_3.4.1       </span></span>
<span><span class="co">##  [4] lubridate_1.8.0     httr_1.4.3          rprojroot_2.0.3    </span></span>
<span><span class="co">##  [7] tools_4.2.1         backports_1.4.1     bslib_0.4.0        </span></span>
<span><span class="co">## [10] utf8_1.2.2          R6_2.5.1            DBI_1.1.3          </span></span>
<span><span class="co">## [13] lazyeval_0.2.2      colorspace_2.0-3    withr_2.5.0        </span></span>
<span><span class="co">## [16] tidyselect_1.1.2    curl_4.3.2          compiler_4.2.1     </span></span>
<span><span class="co">## [19] textshaping_0.3.6   cli_3.3.0           rvest_1.0.2        </span></span>
<span><span class="co">## [22] xml2_1.3.3          desc_1.4.1          labeling_0.4.2     </span></span>
<span><span class="co">## [25] sass_0.4.2          scales_1.2.0        pkgdown_2.0.6      </span></span>
<span><span class="co">## [28] systemfonts_1.0.4   digest_0.6.29       yulab.utils_0.0.5  </span></span>
<span><span class="co">## [31] rmarkdown_2.14      stringdist_0.9.8    pkgconfig_2.0.3    </span></span>
<span><span class="co">## [34] htmltools_0.5.3     highr_0.9           dbplyr_2.2.1       </span></span>
<span><span class="co">## [37] fastmap_1.1.0       rlang_1.0.4         readxl_1.4.0       </span></span>
<span><span class="co">## [40] rstudioapi_0.13     gridGraphics_0.5-1  jquerylib_0.1.4    </span></span>
<span><span class="co">## [43] generics_0.1.3      farver_2.1.1        jsonlite_1.8.0     </span></span>
<span><span class="co">## [46] googlesheets4_1.0.0 magrittr_2.0.3      ggplotify_0.1.0    </span></span>
<span><span class="co">## [49] patchwork_1.1.1     Rcpp_1.0.9          munsell_0.5.0      </span></span>
<span><span class="co">## [52] fansi_1.0.3         ape_5.6-2           lifecycle_1.0.1    </span></span>
<span><span class="co">## [55] stringi_1.7.8       yaml_2.3.5          grid_4.2.1         </span></span>
<span><span class="co">## [58] parallel_4.2.1      crayon_1.5.1        lattice_0.20-45    </span></span>
<span><span class="co">## [61] haven_2.5.0         cowplot_1.1.1       hms_1.1.1          </span></span>
<span><span class="co">## [64] knitr_1.39          pillar_1.8.0        reprex_2.0.1       </span></span>
<span><span class="co">## [67] glue_1.6.2          evaluate_0.15       ggfun_0.0.6        </span></span>
<span><span class="co">## [70] modelr_0.1.8        vctrs_0.4.1         treeio_1.20.1      </span></span>
<span><span class="co">## [73] tzdb_0.3.0          cellranger_1.1.0    gtable_0.3.0       </span></span>
<span><span class="co">## [76] assertthat_0.2.1    cachem_1.0.6        xfun_0.31          </span></span>
<span><span class="co">## [79] broom_1.0.0         tidytree_0.3.9      ragg_1.2.2         </span></span>
<span><span class="co">## [82] googledrive_2.0.0   gargle_1.2.0        aplot_0.1.6        </span></span>
<span><span class="co">## [85] memoise_2.0.1       ellipsis_0.3.2</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alexander Yermanos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
