<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Bulk Repertoires • Platypus</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Bulk Repertoires">
<meta property="og:description" content="Platypus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/quickstart.html">Get started</a>
</li>
<li>
  <a href="../vignette_index.html">Vignettes</a>
</li>
<li>
  <a href="../PlatypusDB_index.html">PlatypusDB</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://immunesim.readthedocs.io/en/latest/" class="external-link">ImmuneSIM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../cute_monotremes.html">Cute Monotremes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/alexyermanos" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/alexyermanos/Platypus" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Bulk Repertoires</h1>
                        <h4 data-toc-skip class="author">Alexander
Yermanos, Lorenzo Guerci, Kostas Stiklioraitis</h4>
            
            <h4 data-toc-skip class="date">2022-08-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alexyermanos/Platypus/tree/mastervignettes/Platypus_Bulk_Vignette.Rmd" class="external-link"><code>vignettes/Platypus_Bulk_Vignette.Rmd</code></a></small>
      <div class="hidden name"><code>Platypus_Bulk_Vignette.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The Platypus family of packages are meant to provide potential
pipelines and examples relevant to the broad field of computational
immunology. In this vignette, we introduce and provide examples for
using the package to analyze T- and B- cell receptor repertoire bulk
sequencing data outputs coming from MiXCR (Bolotin DA, et al., Nat
Methods. 2015;12(5):380-1) or molecular amplification fingerprinting
(MAF) (Khan TA, et al., Sci Adv. 2016;2(3):e1501371). The core set of
functions can be found at <a href="https://github.com/alexyermanos/Platypus" class="external-link uri">https://github.com/alexyermanos/Platypus</a> and examples of
use can be found in the publications <a href="https://doi.org/10.1093/nargab/lqab023" class="external-link uri">https://doi.org/10.1093/nargab/lqab023</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Platypus</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="converting-mafmixcr-bulk-sequencing-outputs-to-a-vgm-object">2. Converting MAF/MiXCR Bulk Sequencing Outputs to a VGM Object<a class="anchor" aria-label="anchor" href="#converting-mafmixcr-bulk-sequencing-outputs-to-a-vgm-object"></a>
</h2>
<p>Comprehensive details about the parameters of the VDJ_bulk_to_vgm()
function can be found in the respective function documentation. Note
that the input.type function parameter defaults to “MIXCR”; if working
with MAF outputs, the input.type parameter should be set to “MAF”.</p>
<div class="section level3">
<h3 id="running-the-vdj_bulk_to_vgm-function">2.1 Running The VDJ_bulk_to_vgm Function<a class="anchor" aria-label="anchor" href="#running-the-vdj_bulk_to_vgm-function"></a>
</h3>
<p>The user should save the local paths to the bulk sequencing output
.txt files in a list. This list will then serve as the primary input to
the VDJ_bulk_to_vgm function.</p>
<p>The VDJ_bulk_to_vgm function call is commented out throughout this
tutorial, as all RData output objects were saved separately in the
PlatybusDB data base, from which they can be loaded directly into R.
Therefore, this tutorial can be followed without downloading any sample
data on your local device.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a list containing the local paths to bulk sequencing output .txt </span></span>
<span><span class="co"># files processed with MiXCR: </span></span>
<span></span>
<span><span class="va">VDJ.bulk.out.directory.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">VDJ.bulk.out.directory.list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/PC_14VDJ.txt"</span><span class="op">)</span></span>
<span><span class="va">VDJ.bulk.out.directory.list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/PC_15VDJ.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a VGM object by running the VDJ_bulk_to_vgm() function:</span></span>
<span></span>
<span><span class="co"># vgm_bulk &lt;- VDJ_bulk_to_vgm(VDJ.bulk.out.directory.list = VDJ.bulk.out.directory.list,</span></span>
<span><span class="co">#            input.type = 'MIXCR', # if working with MAF outputs, the input.type </span></span>
<span><span class="co">#            # parameter should be set to "MAF"</span></span>
<span><span class="co">#            integrate.MIXCR.output = TRUE, </span></span>
<span><span class="co">#            group.id = c(1,2),</span></span>
<span><span class="co">#            cell.type = "Bcells",</span></span>
<span><span class="co">#            batches = c(1,2),</span></span>
<span><span class="co">#            vgm.expanded = TRUE)</span></span>
<span></span>
<span><span class="co"># integrate.MIXCR.output: argument which specifies whether MiXCR-derived </span></span>
<span><span class="co"># additional columns should be included in the VGM output.</span></span>
<span></span>
<span><span class="co"># vgm.expanded: argument for including vgm[[9]] in the output list.</span></span>
<span><span class="co"># Recommended to be left as default (TRUE) as some downstream functions</span></span>
<span><span class="co"># can only take vgm[[9]] and not vgm[[1]]; however, if the bulk datasets are </span></span>
<span><span class="co"># exceedingly large, it may be necessary to set vgm.expanded to FALSE. </span></span>
<span></span>
<span><span class="co"># clone.strategy: argument for specifying the clonotyping strategy. </span></span>
<span><span class="co"># MiXCR inputs have the clonotypes already assigned, therefore, users may wish </span></span>
<span><span class="co"># to only specify the clone.strategy when re-clonotyping is to be performed. </span></span>
<span><span class="co"># Possible options for clone.strategy are "cdr3.aa", "VDJJ.VJJ", </span></span>
<span><span class="co"># "VDJJ.VJJ.cdr3length". </span></span>
<span><span class="co"># For MAF inputs, it is necessary to set this argument to the preferred option. </span></span>
<span></span>
<span><span class="co"># Load the vgm_bulk object stored in PlatypusDB:</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://storage.googleapis.com/platypus_private/stiklioraitis2022a_vgm_bulk.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Please note that Sample IDs seen in the analyses below are assigned
based on the user-defined input order of the local paths to the
VDJ.bulk.out.directory.list list.</p>
</div>
<div class="section level3">
<h3 id="vdj_bulk_to_vgm-function-output">2.2 VDJ_bulk_to_vgm Function Output<a class="anchor" aria-label="anchor" href="#vdj_bulk_to_vgm-function-output"></a>
</h3>
<p>The output of the VDJ_bulk_to_vgm function is a list of two data
frames. vgm_bulk[[1]] is an aggregated data frame in which each line
corresponds to a different clonotype with the associated cell count,
whereas vgm_bulk[[9]] contains “per-cell” information (e.g., if we have
clonotype0 with 8000 cells, vgm_bulk[[9]] will contain 8000 rows for
each cell of clonotype0, whereas vgm_bulk[[1]] will only contain one row
per clonotype). The inclusion of both vgm_bulk[[1]] and vgm_bulk[[9]]
was done to ensure compatibility with downstream Platypus functions that
were originally designed for single cell sequence data analysis. Please
refer to the sections below for examples where either vgm_bulk[[1]] or
vgm_bulk[[9]] should be provided to the individual functions.</p>
<p>We can briefly look at the structure of both vgm_bulk[[1]] and
vgm_bulk[[9]]:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># inspecting the columns included in vgm_bulk[[1]]:</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "barcode"                 "sample_id"              </span></span>
<span><span class="co">##  [3] "group_id"                "clonotype_id_10x"       </span></span>
<span><span class="co">##  [5] "celltype"                "Nr_of_VDJ_chains"       </span></span>
<span><span class="co">##  [7] "Nr_of_VJ_chains"         "VDJ_cdr3s_aa"           </span></span>
<span><span class="co">##  [9] "VJ_cdr3s_aa"             "VDJ_cdr3s_nt"           </span></span>
<span><span class="co">## [11] "VJ_cdr3s_nt"             "VDJ_chain_contig"       </span></span>
<span><span class="co">## [13] "VJ_chain_contig"         "VDJ_chain"              </span></span>
<span><span class="co">## [15] "VJ_chain"                "VDJ_vgene"              </span></span>
<span><span class="co">## [17] "VJ_vgene"                "VDJ_dgene"              </span></span>
<span><span class="co">## [19] "VDJ_jgene"               "VJ_jgene"               </span></span>
<span><span class="co">## [21] "VDJ_cgene"               "VJ_cgene"               </span></span>
<span><span class="co">## [23] "VDJ_sequence_nt_raw"     "VJ_sequence_nt_raw"     </span></span>
<span><span class="co">## [25] "VDJ_sequence_nt_trimmed" "VJ_sequence_nt_trimmed" </span></span>
<span><span class="co">## [27] "VDJ_sequence_aa"         "VJ_sequence_aa"         </span></span>
<span><span class="co">## [29] "VDJ_trimmed_ref"         "VJ_trimmed_ref"         </span></span>
<span><span class="co">## [31] "VDJ_raw_consensus_id"    "VJ_raw_consensus_id"    </span></span>
<span><span class="co">## [33] "orig_barcode"            "clonotype_frequency"    </span></span>
<span><span class="co">## [35] "specifity"               "affinity"               </span></span>
<span><span class="co">## [37] "GEX_available"           "orig.ident"             </span></span>
<span><span class="co">## [39] "orig_barcode_GEX"        "seurat_clusters"        </span></span>
<span><span class="co">## [41] "PC_1"                    "PC_2"                   </span></span>
<span><span class="co">## [43] "UMAP_1"                  "UMAP_2"                 </span></span>
<span><span class="co">## [45] "tSNE_1"                  "tSNE_2"                 </span></span>
<span><span class="co">## [47] "batches"                 "clonotype_id"           </span></span>
<span><span class="co">## [49] "VDJ_nSeqFR1"             "VDJ_nSeqCDR1"           </span></span>
<span><span class="co">## [51] "VDJ_nSeqFR2"             "VDJ_nSeqCDR2"           </span></span>
<span><span class="co">## [53] "VDJ_nSeqFR3"             "VJ_nSeqCDR3"            </span></span>
<span><span class="co">## [55] "VDJ_nSeqFR4"             "VDJ_aaSeqFR1"           </span></span>
<span><span class="co">## [57] "VDJ_aaSeqCDR1"           "VDJ_aaSeqFR2"           </span></span>
<span><span class="co">## [59] "VDJ_aaSeqCDR2"           "VDJ_aaSeqFR3"           </span></span>
<span><span class="co">## [61] "VDJ_aaSeqCDR3"           "VDJ_aaSeqFR4"           </span></span>
<span><span class="co">## [63] "VDJ_SHM"                 "VJ_nSeqFR1"             </span></span>
<span><span class="co">## [65] "VJ_nSeqCDR1"             "VJ_nSeqFR2"             </span></span>
<span><span class="co">## [67] "VJ_nSeqCDR2"             "VJ_nSeqFR3"             </span></span>
<span><span class="co">## [69] "VJ_nSeqCDR3.1"           "VJ_nSeqFR4"             </span></span>
<span><span class="co">## [71] "VJ_aaSeqFR1"             "VJ_aaSeqCDR1"           </span></span>
<span><span class="co">## [73] "VJ_aaSeqFR2"             "VJ_aaSeqCDR2"           </span></span>
<span><span class="co">## [75] "VJ_aaSeqFR3"             "VJ_aaSeqCDR3"           </span></span>
<span><span class="co">## [77] "VJ_aaSeqFR4"             "VJ_SHM"</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># verifying that all columns in vgm_bulk[[1]] and vgm_bulk[[9]] match:</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>Note that if we were to re-run the VDJ_bulk_to_vgm function with the
integrate.MIXCR.output argument set to FALSE, only columns 1 to 48 are
returned:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># vgm_bulk_condensed &lt;- VDJ_bulk_to_vgm(VDJ.bulk.out.directory.list = VDJ.bulk.out.directory.list,</span></span>
<span><span class="co">#            input.type = 'MIXCR', # if working with MAF outputs, </span></span>
<span><span class="co">#            # the input.type parameter should be set to "MAF"</span></span>
<span><span class="co">#            integrate.MIXCR.output = FALSE, </span></span>
<span><span class="co">#            group.id = c(1,2),</span></span>
<span><span class="co">#            cell.type = "Bcells",</span></span>
<span><span class="co">#            batches = c(1,2),</span></span>
<span><span class="co">#            vgm.expanded = TRUE)</span></span>
<span></span>
<span><span class="co"># Load the vgm_bulk_condensed object stored in PlatybusDB:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://storage.googleapis.com/platypus_private/stiklioraitis2022a_vgm_bulk_condensed.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">vgm_bulk_condensed</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "barcode"                 "sample_id"              </span></span>
<span><span class="co">##  [3] "group_id"                "clonotype_id_10x"       </span></span>
<span><span class="co">##  [5] "celltype"                "Nr_of_VDJ_chains"       </span></span>
<span><span class="co">##  [7] "Nr_of_VJ_chains"         "VDJ_cdr3s_aa"           </span></span>
<span><span class="co">##  [9] "VJ_cdr3s_aa"             "VDJ_cdr3s_nt"           </span></span>
<span><span class="co">## [11] "VJ_cdr3s_nt"             "VDJ_chain_contig"       </span></span>
<span><span class="co">## [13] "VJ_chain_contig"         "VDJ_chain"              </span></span>
<span><span class="co">## [15] "VJ_chain"                "VDJ_vgene"              </span></span>
<span><span class="co">## [17] "VJ_vgene"                "VDJ_dgene"              </span></span>
<span><span class="co">## [19] "VDJ_jgene"               "VJ_jgene"               </span></span>
<span><span class="co">## [21] "VDJ_cgene"               "VJ_cgene"               </span></span>
<span><span class="co">## [23] "VDJ_sequence_nt_raw"     "VJ_sequence_nt_raw"     </span></span>
<span><span class="co">## [25] "VDJ_sequence_nt_trimmed" "VJ_sequence_nt_trimmed" </span></span>
<span><span class="co">## [27] "VDJ_sequence_aa"         "VJ_sequence_aa"         </span></span>
<span><span class="co">## [29] "VDJ_trimmed_ref"         "VJ_trimmed_ref"         </span></span>
<span><span class="co">## [31] "VDJ_raw_consensus_id"    "VJ_raw_consensus_id"    </span></span>
<span><span class="co">## [33] "orig_barcode"            "clonotype_frequency"    </span></span>
<span><span class="co">## [35] "specifity"               "affinity"               </span></span>
<span><span class="co">## [37] "GEX_available"           "orig.ident"             </span></span>
<span><span class="co">## [39] "orig_barcode_GEX"        "seurat_clusters"        </span></span>
<span><span class="co">## [41] "PC_1"                    "PC_2"                   </span></span>
<span><span class="co">## [43] "UMAP_1"                  "UMAP_2"                 </span></span>
<span><span class="co">## [45] "tSNE_1"                  "tSNE_2"                 </span></span>
<span><span class="co">## [47] "batches"                 "clonotype_id"</span></span></code></pre>
<p>Comparing the number of rows, we see that vgm_bulk[[9]] is an
expanded version of vgm_bulk[[1]]:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"vgm_bulk[[9]] contains "</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="st">" rows"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "vgm_bulk[[9]] contains 123690 rows"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"vgm_bulk[[1]] contains "</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="st">" rows"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "vgm_bulk[[1]] contains 96667 rows"</span></span></code></pre>
<p>And we can briefly inspect some of the information contained in the
VGMs:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For example, the first 3 most frequent clonotypes from Sample 1 : </span></span>
<span></span>
<span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"s1"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span><span class="fu">select</span><span class="op">(</span><span class="va">clonotype_id</span>, <span class="va">clonotype_frequency</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   clonotype_id clonotype_frequency</span></span>
<span><span class="co">## 1   clonotype0                 488</span></span>
<span><span class="co">## 2   clonotype1                 375</span></span>
<span><span class="co">## 3   clonotype2                 365</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># have the following CDR3 amino acid sequences:</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VDJ_cdr3s_aa</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CTGYYYAMDYW"   "CARHGGFWYFDVW" "CASYGNHWYFDVW"</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># and we can quantify the number of Somatic Hypermutations (SHMs) in their VDJ:</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VDJ_SHM</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2 3 1</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># or VJ regions:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VJ_SHM</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] NA NA NA</span></span></code></pre>
<p>Please note that with bulk repertoires, only sequence information for
the heavy (VDJ) or light (VJ) chains will be available (i.e., we don’t
instances of having VDJ + VJ information for a single clonotype).
Therefore, in the example above, the light chain (VJ) SHMs are
quantified as “NA”, as the dataset we’ve used in this tutorial only had
sequence information concerning B cell heavy chains.</p>
<p>Additionally, we can have a look at the at the unique VDJ constant
region genes represented in our data set:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VDJ_cgene</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "IGHM"   "IGHG1"  "IGHG2C" "IGHG2B" "IGHG3"  ""       "IGHE"   "IGHA"</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Notice the empty ("") element - these are cells for which the constant </span></span>
<span><span class="co"># region gene could not be assigned</span></span></code></pre></div>
<p>Of note: the VDJ_bulk_to_vgm function has an argument
best.match.only, which can be set to FALSE if one wishes to return all
matching V,J,D, and C genes in the output (i.e., not only those with the
highest confidence score):</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># vgm_bulk_all_matches &lt;- VDJ_bulk_to_vgm(VDJ.bulk.out.directory.list = VDJ.bulk.out.directory.list,</span></span>
<span><span class="co">#            input.type = 'MIXCR', # if working with MAF outputs, </span></span>
<span><span class="co">#            # the input.type parameter should be set to "MAF"</span></span>
<span><span class="co">#            integrate.MIXCR.output = TRUE, </span></span>
<span><span class="co">#            group.id = c(1,2),</span></span>
<span><span class="co">#            cell.type = "Bcells",</span></span>
<span><span class="co">#            batches = c(1,2),</span></span>
<span><span class="co">#            vgm.expanded = TRUE,</span></span>
<span><span class="co">#            best.match.only = FALSE)</span></span>
<span></span>
<span><span class="co"># Load the vgm_bulk_all_matches object stored in PlatypusDB:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://storage.googleapis.com/platypus_private/stiklioraitis2022a_vgm_bulk_all_matches.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">vgm_bulk_all_matches</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VDJ_cgene</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "IGHM"                      "IGHG1"                    </span></span>
<span><span class="co">##  [3] "IGHG2C"                    "IGHM;IGHG2C"              </span></span>
<span><span class="co">##  [5] "IGHG2C;IGHM"               "IGHG2C;IGHG1"             </span></span>
<span><span class="co">##  [7] "IGHG1;IGHM"                "IGHG1;IGHG2C"             </span></span>
<span><span class="co">##  [9] "IGHG2B"                    "IGHG1;IGHG2B"             </span></span>
<span><span class="co">## [11] "IGHM;IGHG2B"               "IGHG2C;IGHG3"             </span></span>
<span><span class="co">## [13] "IGHG2B;IGHG1"              "IGHG3"                    </span></span>
<span><span class="co">## [15] "IGHM;IGHG1"                "IGHG2B;IGHM;IGHG2C"       </span></span>
<span><span class="co">## [17] "IGHG2C;IGHG2B"             ""                         </span></span>
<span><span class="co">## [19] "IGHG3;IGHG2C"              "IGHG1;IGHG3"              </span></span>
<span><span class="co">## [21] "IGHG2B;IGHG2C"             "IGHG1;IGHM;IGHG2C"        </span></span>
<span><span class="co">## [23] "IGHG3;IGHM"                "IGHG3;IGHG1"              </span></span>
<span><span class="co">## [25] "IGHG1;IGHG2C;IGHG3"        "IGHG2B;IGHG3"             </span></span>
<span><span class="co">## [27] "IGHG3;IGHG2B"              "IGHG3;IGHG2B;IGHG2C"      </span></span>
<span><span class="co">## [29] "IGHG2C;IGHG1;IGHG3"        "IGHG2C;IGHG3;IGHG1"       </span></span>
<span><span class="co">## [31] "IGHG1;IGHG2B;IGHG3"        "IGHG1;IGHG2B;IGHG2C"      </span></span>
<span><span class="co">## [33] "IGHG1;IGHG2B;IGHG2C;IGHG3" "IGHG2B;IGHG2C;IGHG1"      </span></span>
<span><span class="co">## [35] "IGHG2C;IGHG1;IGHG2B"       "IGHE"                     </span></span>
<span><span class="co">## [37] "IGHG1;IGHG3;IGHG2C"        "IGHA"                     </span></span>
<span><span class="co">## [39] "IGHG3;IGHG2C;IGHG1"        "IGHG1;IGHG2C;IGHG2B"      </span></span>
<span><span class="co">## [41] "IGHG3;IGHG1;IGHG2C"        "IGHG2B;IGHG2C;IGHG3"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="vdj-repertoire-analysis">3. VDJ Repertoire Analysis<a class="anchor" aria-label="anchor" href="#vdj-repertoire-analysis"></a>
</h2>
<div class="section level3">
<h3 id="changing-the-clonotyping-strategy">3.1. Changing the Clonotyping Strategy<a class="anchor" aria-label="anchor" href="#changing-the-clonotyping-strategy"></a>
</h3>
<p>The clone.strategy argument of the VDJ_bulk_to_vgm function can be
used to perform (re)clonotyping. As MiXCR inputs come with clonotypes
already assigned, the user may only choose to use the clone.strategy
argument when reclonotyping is to be performed. When working with MAF
inputs, the user must set clone.strategy to their preferred option
(either “cdr3.aa”, “VDJJ.VJJ”, or “VDJJ.VJJ.cdr3length”).</p>
<p>The MiXCR dataset supplied and used for the rest of this tutorial was
clonotyped based on the VDJ region. In the example below, we demonstrate
how the clone.strategy may be used to re-clonotype the MiXCR input based
on the CDR3 amino acid sequence.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform reclonotyping on cdr3.aa with the clone.strategy argument</span></span>
<span></span>
<span><span class="co"># vgm_bulk_reclonotyped_cdr3aa &lt;- VDJ_bulk_to_vgm(VDJ.bulk.out.directory.list = VDJ.bulk.out.directory.list,</span></span>
<span><span class="co">#            input.type = 'MIXCR', # if working with MAF outputs, </span></span>
<span><span class="co">#            # the input.type parameter should be set to "MAF"</span></span>
<span><span class="co">#            integrate.MIXCR.output = TRUE, </span></span>
<span><span class="co">#            group.id = c(1,2),</span></span>
<span><span class="co">#            cell.type = "Bcells",</span></span>
<span><span class="co">#            batches = c(1,2),</span></span>
<span><span class="co">#            vgm.expanded = TRUE,</span></span>
<span><span class="co">#            clone.strategy = "cdr3.aa")</span></span>
<span></span>
<span><span class="co"># Load the vgm_bulk_reclonotyped_cdr3aa object stored in PlatypusDB:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://storage.googleapis.com/platypus_private/stiklioraitis2022a_vgm_bulk_reclonotyped_cdr3aa.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vgm_bulk</span> <span class="op">&lt;-</span> <span class="va">vgm_bulk_reclonotyped_cdr3aa</span></span></code></pre></div>
<p>After re-clonotyping on the CDR3 a.a. region, we’d expect to see a
reduced number of total clones as opposed to when clonotyping on the VDJ
region is performed. We can verify this with the code below:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"The original VDJ-clonotyped dataset contains a total of "</span>, </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"s1"</span>, <span class="op">]</span><span class="op">$</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span>, </span>
<span><span class="st">" clonotypes"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "The original VDJ-clonotyped dataset contains a total of 3939 clonotypes"</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"After reclonotyping on the CDR3 a.a. the dataset contains a total of "</span>, </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">vgm_bulk_reclonotyped_cdr3aa</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">[</span><span class="va">vgm_bulk_reclonotyped_cdr3aa</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"s1"</span>, <span class="op">]</span><span class="op">$</span><span class="va">clonotype_id</span><span class="op">)</span><span class="op">)</span>, </span>
<span><span class="st">" clonotypes"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "After reclonotyping on the CDR3 a.a. the dataset contains a total of 3939 clonotypes"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="somatic-hypermutations-shms">3.2. Somatic Hypermutations (SHMs)<a class="anchor" aria-label="anchor" href="#somatic-hypermutations-shms"></a>
</h3>
<p>The VDJ_plot_SHM function of Platypus can be used to create a boxplot
that will help us to visualize the distribution of somatic
hypermutations in each sample. Note that to run this function,
vgm_bulk[[1]] should be used.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SHM_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_plot_SHM.html">VDJ_plot_SHM</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, group.by <span class="op">=</span> <span class="st">"sample_id"</span>, </span>
<span>                        quantile.label <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span></span>
<span><span class="va">SHM_plt</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="visualizing-clonal-frequencies-using-donut-plots">3.3. Visualizing Clonal Frequencies Using Donut Plots<a class="anchor" aria-label="anchor" href="#visualizing-clonal-frequencies-using-donut-plots"></a>
</h3>
<p>To get a basic overview of clonal expansion, the VDJ_clonal_donut
function can be used to generate donut plots for each sample, where the
black section of the plot shows cells in unexpanded clones, while the
remaining sections indicate the expanded clones by their frequency.
vgm_bulk[[1]] should be used as input for the function.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">donut_plts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_donut.html">VDJ_clonal_donut</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, </span>
<span>non.expanded.color <span class="op">=</span> <span class="st">"black"</span>, counts.to.use <span class="op">=</span> <span class="st">"clonotype_id_10x"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using column clonotype_id_10x for counting clones</span></span></code></pre>
<pre><code><span><span class="co">## Clones: Expanded: 2297 / 58.31%; 1 cell 1642 / 41.69%; total: 3939</span></span></code></pre>
<pre><code><span><span class="co">## Cells: Expanded: 62833 / 97.45%; 1 cell 1642 / 2.55%; Total: 64475</span></span></code></pre>
<pre><code><span><span class="co">## Clones: Expanded: 1717 / 55.48%; 1 cell 1378 / 44.52%; total: 3095</span></span></code></pre>
<pre><code><span><span class="co">## Cells: Expanded: 30814 / 95.72%; 1 cell 1378 / 4.28%; Total: 32192</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">donut_plts</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>For large bulk data sets with many samples, generating all donut
plots may take a while. Indexing can be used to generate only select
plots for the chosen sample IDs (e.g., call donut_plts[1] to generate
the plot for s1 only).</p>
</div>
<div class="section level3">
<h3 id="calculating-common-repertoire-diversity-metrics">3.4. Calculating Common Repertoire Diversity Metrics<a class="anchor" aria-label="anchor" href="#calculating-common-repertoire-diversity-metrics"></a>
</h3>
<p>Diversity metrics, such as Shannon Evenness or the Jaccard index, can
be calculated and displayed graphically for any column in the
vgm_bulk[[1]] using the VDJ_diversity function:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate Shannon Evenness for the heavy chain CDR3:</span></span>
<span></span>
<span><span class="va">diversity_plt_Shannon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_diversity.html">VDJ_diversity</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                         feature.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_cdr3s_aa"</span><span class="op">)</span>, </span>
<span>                         grouping.column <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>                         metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"shannonevenness"</span><span class="op">)</span>, </span>
<span>                         subsample.to.same.n <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">diversity_plt_Shannon</span></span>
<span></span>
<span></span>
<span><span class="co"># Calculate the Jaccard Index for the heavy chain CDR3:</span></span>
<span></span>
<span><span class="va">diversity_plt_Jaccard</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_diversity.html">VDJ_diversity</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                         feature.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_cdr3s_aa"</span><span class="op">)</span>,</span>
<span>                         grouping.column <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>                         metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span><span class="op">)</span>, </span>
<span>                         subsample.to.same.n <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">diversity_plt_Jaccard</span></span></code></pre></div>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-15-1.png" width="672"><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-15-2.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="examining-isotype-distribution">3.5. Examining Isotype Distribution<a class="anchor" aria-label="anchor" href="#examining-isotype-distribution"></a>
</h3>
<p>We can examine the clonal expansion for a defined number of clones by
using the the VDJ_clonal_expansion function, where we set the number of
clones to show, and can color the bar-plot by the desired column to
examine the distribution of a selected parameter in the vgm (e.g.,
setting the color.by parameter to “isotype” will generate bars colored
by the respective IgH chain; in principle, any other column from the vgm
can be used).</p>
<p>Only vgm_bulk[[9]] should be used as the input to the the
VDJ_clonal_expansion function.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting the isotype distribution of the top 30 clones:</span></span>
<span></span>
<span><span class="va">clonal_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_expansion.html">VDJ_clonal_expansion</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span>, celltype <span class="op">=</span> <span class="st">"Bcells"</span>,</span>
<span>                                   clones <span class="op">=</span> <span class="st">"30"</span>, group.by <span class="op">=</span> <span class="st">"sample_id"</span>, </span>
<span>                                   color.by <span class="op">=</span> <span class="st">"isotype"</span>, </span>
<span>                                   isotypes.to.plot <span class="op">=</span> <span class="st">"all"</span>, </span>
<span>                                   treat.incomplete.clones <span class="op">=</span> <span class="st">"exclude"</span>, </span>
<span>                                   treat.incomplete.cells <span class="op">=</span> <span class="st">"proportional"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">clonal_out</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-16-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="sequence-similarity-networks">3.6. Sequence Similarity Networks<a class="anchor" aria-label="anchor" href="#sequence-similarity-networks"></a>
</h3>
<p>The VDJ_network function creates a sequence similarity network
between repertoires or within a repertoire by connecting those clones
with sequence similarity. Setting the per.sample argument to false
indicates that one network for multiple repertoires should be
produced.</p>
<p>Networks with a high number of sequences may take a very long time to
be displayed. Alternatively, the user may subsample the dataset to
reduce the runtime. In the example below, only the first 750 clones from
Sample 1 are used to generate the sequence similarity network.</p>
<p>Note that the is.bulk argument must be set to True to run the
VDJ_network function for bulk samples:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">S1</span> <span class="op">&lt;-</span> <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"s1"</span><span class="op">)</span> </span>
<span><span class="va">sample200_S1</span> <span class="op">&lt;-</span> <span class="fu">slice_sample</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">S1</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="va">network</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_network.html">VDJ_network</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">sample200_S1</span>, per.sample <span class="op">=</span> <span class="cn">T</span>,distance.cutoff <span class="op">=</span> <span class="fl">8</span>, </span>
<span>                  platypus.version <span class="op">=</span> <span class="st">"v3"</span>, hcdr3.only <span class="op">=</span> <span class="cn">T</span>, is.bulk <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Sample order: s1</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/plot.igraph.html" class="external-link">plot.igraph</a></span><span class="op">(</span><span class="va">network</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,vertex.label<span class="op">=</span><span class="cn">NA</span>,vertex.size<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span></code></pre></div>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="germline-gene-usage">3.7. Germline Gene Usage<a class="anchor" aria-label="anchor" href="#germline-gene-usage"></a>
</h3>
<p>Platypus allows for a separate analysis of V gene usage for Heavy and
Light Chains. With the VDJ_Vgene_usage_barplot function, the user may
plot the most frequently used IgH or IgK/L V genes. By default, this
function only provides visualizations for the HC V genes, but can also
provide for the LC if the LC.Vgene argument is set to TRUE. The user can
also select the number of most used genes to be depicted.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_Vgene_usage_barplot.html">VDJ_Vgene_usage_barplot</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, HC.gene.number <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                               LC.Vgene <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                               platypus.version <span class="op">=</span> <span class="st">"v3"</span>, </span>
<span>                               is.bulk <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Sample order: s1 ; s2</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barplot</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-18-1.png" width="672">
We can also generate stacked barplots showing the fraction of the most
frequently used IgH and IgK/L Vgenes using the
VDJ_Vgene_usage_stacked_barplot function:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stacked_barplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_Vgene_usage_stacked_barplot.html">VDJ_Vgene_usage_stacked_barplot</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                                   HC.gene.number <span class="op">=</span> <span class="fl">4</span>, </span>
<span>                                                   LC.Vgene <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                                   platypus.version <span class="op">=</span> <span class="st">"v3"</span>, </span>
<span>                                                   is.bulk <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Sample order: s1 ; s2</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stacked_barplot</span></span></code></pre></div>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-19-1.png" width="672">
Additionally, a circular visualization of how V and J genes are combined
throughout the repertoire can be generated using the VDJ_VJ_usage_circos
function of Platypus.</p>
<p>To run VDJ_VJ_usage_circos, set A.or.B = “A” (if we’re working with
heavy chains, and “B” if we’re working with light chains), and must set
filter1H1L = “F” (as in all cases, we’ll only have either light or heavy
information from bulk data). The c.threshold arguments is used to filter
out clonotypes with lower than the user-selected frequency. This can be
useful for bulk repertoire analysis, as the majority of clonotypes are
often not highly-expanded.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">circos_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_VJ_usage_circos.html">VDJ_VJ_usage_circos</a></span><span class="op">(</span>VGM <span class="op">=</span> <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                  platypus.version <span class="op">=</span> <span class="st">"v3"</span>, filter1H1L <span class="op">=</span> <span class="st">"F"</span>, </span>
<span>                                  c.threshold <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-20-1.png" width="672"><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-20-2.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="cdr3-sequence-similarity">3.8. CDR3 Sequence Similarity<a class="anchor" aria-label="anchor" href="#cdr3-sequence-similarity"></a>
</h3>
<p>Using the VDJ_logoplot_vector function, we can also look for any
specific heavy-chain (HC) and/or light-chain (LC) CDR3 amino acid
patterns arising across the different clones. This function produces a
logoplot of the CDR3 region, where the user can select the length of the
CDR3 regionts to be plotted using the length_cdr3 argument (if
length_cdr3 argument is set to auto, the most frequently appearing
length in the vector will be used).</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pasted_CDR3s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VDJ_cdr3s_aa</span>, <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VJ_cdr3s_aa</span><span class="op">)</span></span>
<span></span>
<span><span class="va">CDR3_logoplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_logoplot_vector.html">VDJ_logoplot_vector</a></span><span class="op">(</span>cdr3.vector <span class="op">=</span> <span class="va">pasted_CDR3s</span>, </span>
<span>                                     seq_type <span class="op">=</span> <span class="st">"aa"</span>, length_cdr3 <span class="op">=</span> <span class="st">"auto"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Returning logoplot based on 21595 sequences of a total 96667 input sequences</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">CDR3_logoplot</span><span class="op">)</span></span></code></pre></div>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="repertoire-overlap">3.9. Repertoire Overlap<a class="anchor" aria-label="anchor" href="#repertoire-overlap"></a>
</h3>
<p>The VDJ_overlap_heatmap function can be used to test the overlap of
selected features (e.g., CDR3s amino acid sequences, VDJ V genes)
between different samples:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cdr3_overlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_overlap_heatmap.html">VDJ_overlap_heatmap</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm_bulk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                    feature.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_cdr3s_aa"</span><span class="op">)</span>, </span>
<span>                                    grouping.column <span class="op">=</span> <span class="st">"sample_id"</span>, </span>
<span>                                    axis.label.size <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                                    pvalues.label.size <span class="op">=</span> <span class="fl">12</span>, </span>
<span>                                    add.barcode.table <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                                    plot.type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cdr3_overlap</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># Table of overlapping items</span></span></code></pre></div>
<pre><code><span><span class="co">##   sample.names.1. sample.names.2. overlap items.overlapping overlap_lab</span></span>
<span><span class="co">## 1              s1              s2       0                             0</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cdr3_overlap</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Platypus_Bulk_Vignette_files/figure-html/unnamed-chunk-22-1.png" width="672"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alexander Yermanos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
