<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Platypus Quick Start • Platypus</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Platypus Quick Start">
<meta property="og:description" content="Platypus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/quickstart.html">Get started</a>
</li>
<li>
  <a href="../vignette_index.html">Vignettes</a>
</li>
<li>
  <a href="../PlatypusDB_index.html">PlatypusDB</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://immunesim.readthedocs.io/en/latest/" class="external-link">ImmuneSIM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../cute_monotremes.html">Cute Monotremes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/alexyermanos" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/KreineVic" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/alexyermanos/Platypus" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Platypus Quick Start</h1>
                        <h4 data-toc-skip class="author">Platypus
Team</h4>
            
            <h4 data-toc-skip class="date">2022-07-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alexyermanos/Platypus/tree/mastervignettes/quickstart.Rmd" class="external-link"><code>vignettes/quickstart.Rmd</code></a></small>
      <div class="hidden name"><code>quickstart.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The Platypus family of packages are meant to provide potential
pipelines and examples relevant to the broad field of computational
immunology. While the majority of functions are currently tailored to
single-cell immune repertoire sequencing, we will continue to building
the package in a modular structure that allows integration to other
bioinformatic methods. Currently, the package can be used to separately
analyze gene expression (GEX) or immune receptor repertoire (VDJ)
sequencing data, in addition to integrating the two data sets to combine
phenotypic features with repertoire analysis. The package is designed to
primarily analyze the output from 10x genomics cellranger (output from
count for GEX and VDJ for enriched immune receptor libraries). The
functions could work with other barcode-based scSeq technologies
assuming the input columns are added correctly. The gene expression
analysis relies heavily upon Seurat, a commonly used R package for
single-cell sequencing (scSeq). The core set of functions can be found
at <a href="https://github.com/alexyermanos/Platypus" class="external-link uri">https://github.com/alexyermanos/Platypus</a> and examples of
use can be found in the publications <a href="https://doi.org/10.1093/nargab/lqab023" class="external-link uri">https://doi.org/10.1093/nargab/lqab023</a> and <a href="https://alexyermanos.github.io/Platypus/PlatypusDB_index.html" class="external-link uri">https://alexyermanos.github.io/Platypus/PlatypusDB_index.html</a></p>
<p>Stay tuned for updates <a href="https://twitter.com/AlexYermanos" class="external-link uri">https://twitter.com/AlexYermanos</a></p>
</div>
<div class="section level2">
<h2 id="platypus-version-3-v3">2. Platypus Version 3 (v3)<a class="anchor" aria-label="anchor" href="#platypus-version-3-v3"></a>
</h2>
<p>Following the original NARGAB publication of Platypus, we have
rebuilt the package to entirely revolve around the VDJ_GEX_matrix
function (vgm for short). This function integrates both repertoire and
transcriptome information (if desired) and serves as the input to all
secondary functions in future iterations of the package. The advantage
of this is having all repertoire and transcriptome information at a
per-cell level. This was originally required due to significant changes
to the clonotyping strategies in the default 10x genomics VDJ pipeline.
While the format of this object will be detailed below, the core of the
function creates a list in R where the first element corresponds to the
VDJ information at the per-cell level and the second element corresponds
to a Seurat object. The VGM can then be supplied to all other functions
that will assume this format by default. All functions should now be
updated to take the VGM structure (referred to as “v3” of Platypus) as
input in the “platypus.version” argument. “platypus.version” can be set
to either “v2” for backcompatibility found in the original manucsript or
“v3” for newer and future functions. Some new functions are only
compatible with “v3”.</p>
<div class="section level3">
<h3 id="citing-platypus">2.1 Citing Platypus<a class="anchor" aria-label="anchor" href="#citing-platypus"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>Yermanos A et al. Platypus<span class="sc">:</span> an open<span class="sc">-</span>access software <span class="cf">for</span> integrating lymphocyte single<span class="sc">-</span>cell immune repertoires with transcriptomes. NAR Genomics <span class="fu">Bioinforma</span> (<span class="dv">2021</span>) <span class="dv">3</span><span class="sc">:</span> doi<span class="sc">:</span><span class="fl">10.1093</span><span class="sc">/</span>NARGAB<span class="sc">/</span>LQAB023</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="sc">**</span><span class="er">*</span> DB BioRxiv Citation coming soon <span class="sc">**</span><span class="er">*</span> </span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="platypusdb">3. PlatypusDB<a class="anchor" aria-label="anchor" href="#platypusdb"></a>
</h2>
<p>We understand the pain that locating and utilizing public data can
inflict on aspiring and established scientists alike. We have therefore
created a public database that is integrated within the structure of the
Platypus package. This means that single-cell immune repertoire
sequencing data can be directly loaded into the R session and even
integrated with locally stored data. All examples in these vignettes
highlight the strength of this feature by primarily using public
datasets within our database. We highly encourage others to contribute
to this database by contacting use with the output files from your
cellranger runs (hopefully with at least v6 of cellranger and using the
standard references). Sharing and organizing public data can help save
the world :P</p>
<p>Locating, downloading, and realigning public data takes us A LOT of
time. In the case you use these aligned files for your own publications
we would highly appreciate you to cite the PlatypusDB manuscript ***
Citation coming soon ***.</p>
</div>
<div class="section level2">
<h2 id="installation">4. Installation…<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="from-cran">4.1. … from CRAN<a class="anchor" aria-label="anchor" href="#from-cran"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Platypus"</span><span class="op">)</span>

<span class="co">#Installing main imports</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Seurat"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Tidyverse"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Biostrings"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"jsonlite"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"seqinr"</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Platypus</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="or-from-github">4.2. …or from Github<a class="anchor" aria-label="anchor" href="#or-from-github"></a>
</h3>
<p>We are adding new features constantly. Recent changes may not yet be
reflected in the CRAN version. All Github versions are
developmental!</p>
<p>Individual R functions can additionally be found on the github in the
folder “R”. These can similarly be downloaded and loaded into the R
environment in case not all functions are desired. These functions are
actively updated and may include more features than the in tar.gz or
Github version file.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### Removing any previous versions of the package</span>
<span class="co">#First can ensure that there is no previous version installed locally</span>
<span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:Platypus"</span>, unload<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/remove.packages.html" class="external-link">remove.packages</a></span><span class="op">(</span><span class="st">"Platypus"</span><span class="op">)</span>

<span class="co">### Downloading and installing Platypus</span>
<span class="co"># First we need to download the most recent version from the master branch at https://github.com/alexyermanos/Platypus we can install the package using the following command. </span>
<span class="co"># WARNING: This needs to be replaced with your own directory where the downloaded package is found</span>
<span class="co"># For MacOS users it may look like this</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"~/Downloads/Platypus_3.3.3.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span>, type<span class="op">=</span><span class="st">"source"</span><span class="op">)</span>
<span class="co"># For windows it will likely look something like this. </span>
<span class="co"># WARNING: You will need to replace 'YourPCName' with your user name for the windows account in the directory. </span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"C:/Users/YourPCName/Downloads/Platypus_3.1.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span>, type<span class="op">=</span><span class="st">"source"</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Platypus</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="extracting-and-integrating-repertoire-data-with-vdj_gex_matrix-or-vgm-platypus-v3">5. Extracting and integrating repertoire data with VDJ_GEX_matrix or
VGM (Platypus v3)<a class="anchor" aria-label="anchor" href="#extracting-and-integrating-repertoire-data-with-vdj_gex_matrix-or-vgm-platypus-v3"></a>
</h2>
<p>Comprehensive details about the parameters of the VDJ_GEX_matrix()
function can be found in the respective function documentation and in
the dedicated vignette (VGM overview)</p>
<p>As an exemplary dataset we use GEX data from murine B and T cells and
VDJ data from B cells in the aged CNS. This small dataset contains VDJ
and GEX libraries from the central nervous system of four murine
samples. The corresponding publication can be found here: <a href="https://doi.org/10.1098/rspb.2020.2793" class="external-link uri">https://doi.org/10.1098/rspb.2020.2793</a></p>
<div class="section level3">
<h3 id="process-local-datasets">5.1. Process local datasets<a class="anchor" aria-label="anchor" href="#process-local-datasets"></a>
</h3>
<p>For local datasets, the paths to Cellranger output directories are
needed. VDJ and GEX are processed simulatneously using the
VDJ_GEX_matrix function, any input of the two can be omitted if not
necessary.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Creating a list with local paths to cellranger directories</span>

<span class="va">VDJ.out.directory.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> <span class="co">### Set directory to the outs folder of cellranger vdj</span>
<span class="va">VDJ.out.directory.list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/yermanos2021b__VDJ_RAW/Aged.CNS.pool.3m.Bcell.S1"</span><span class="op">)</span>
<span class="va">VDJ.out.directory.list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/yermanos2021b__VDJ_RAW/Aged.CNS.pool.12m.Bcell.S2"</span><span class="op">)</span>
<span class="va">VDJ.out.directory.list</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/yermanos2021b__VDJ_RAW/Aged.CNS.pool.18m.Bcell.S3"</span><span class="op">)</span>
<span class="va">VDJ.out.directory.list</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/yermanos2021b__VDJ_RAW/Aged.CNS.single.18m.Bcell.S4"</span><span class="op">)</span>
<span class="va">GEX.out.directory.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> <span class="co">### Set directory to the /outs folder of cellranger count</span>
<span class="va">GEX.out.directory.list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/yermanos2021b__GEX_RAW/Aged.CNS.pool.3m.Bcell.S1/"</span><span class="op">)</span>
<span class="va">GEX.out.directory.list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/yermanos2021b__GEX_RAW/Aged.CNS.pool.12m.Bcell.S2/"</span><span class="op">)</span>
<span class="va">GEX.out.directory.list</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/yermanos2021b__GEX_RAW/Aged.CNS.pool.18m.Bcell.S3/"</span><span class="op">)</span>
<span class="va">GEX.out.directory.list</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Downloads/yermanos2021b__GEX_RAW/Aged.CNS.single.18m.Bcell.S4/"</span><span class="op">)</span>

<span class="co">#Running the VDJ_GEX_matrix function</span>

<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>,
                               GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                               GEX.integrate <span class="op">=</span> <span class="cn">T</span>,
                               VDJ.combine <span class="op">=</span> <span class="cn">T</span>,
                               integrate.GEX.to.VDJ <span class="op">=</span> <span class="cn">T</span>,
                               integrate.VDJ.to.GEX <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>

<span class="co">#For output format see 5.3</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="process-raw-datasets-from-platypusdb">5.2. Process raw datasets from PlatypusDB<a class="anchor" aria-label="anchor" href="#process-raw-datasets-from-platypusdb"></a>
</h3>
<p>PlatypusDB stores both raw and processed data for each study.
Processed data can be loaded directly as a VGM object. Raw data can be
downloaded and then processed with the VDJ_GEX_matrix function.</p>
<p>These features are usful for integration of local and DB datasets.
For this please refer to the PlatypusDB vignette</p>
<p>Here we load samples from PlatypusDB, and process these. The output
format is identical to that of section 5.1</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Downloading PlatypusDB raw data in a list format</span>
<span class="co">#For structure of PlatypusDB links, please refer to the PlatypusDB vignette </span>
<span class="va">yermanos2021_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlatypusDB_fetch.html">PlatypusDB_fetch</a></span><span class="op">(</span>PlatypusDB.links <span class="op">=</span> 
                                     <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"yermanos2021b/ALL/ALL"</span><span class="op">)</span>, 
                                     load.to.enviroment <span class="op">=</span> <span class="cn">F</span>,
                                     load.to.list <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="co">#Running the VDJ_GEX_matrix function</span>

<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>Data.in <span class="op">=</span> <span class="va">yermanos2021_raw</span>,
                      GEX.integrate <span class="op">=</span> <span class="cn">T</span>,
                      VDJ.combine <span class="op">=</span> <span class="cn">T</span>,
                      integrate.GEX.to.VDJ <span class="op">=</span> <span class="cn">T</span>,
                      integrate.VDJ.to.GEX <span class="op">=</span> <span class="cn">T</span>,
                      verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
<span class="co">## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span>
<span class="co">## This message will be shown once per session</span></code></pre>
<p>We can briefly look at the datastructure</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "barcode"             "sample_id"           "group_id"           </span>
<span class="co">## [4] "clonotype_id_10x"    "clonotype_id"        "clonotype_frequency"</span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## By setting integrate.GEX.to.VDJ and integrate.VDJ.to.GEX to T, VDJ and GEX information will be found in vgm[[1]] and vgm[[2]] objects.</span>
<span class="co"># For example, the seurat-determined cluster is attached to each cell in the VDJ library by</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;</span>
<span class="co">## Levels: 0 1 2 3 4 5 6 7 8 9 10 11</span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># which corresponds to cells with the following VDJ_cdr3 </span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VDJ_cdr3s_aa</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "CARIGYAMDYW"        "CARDFTTVVARGYFDVW"  "CARGITTVVAYYYAMDYW"</span>
<span class="co">## [4] "CTTGPYDYYAMDYW"     "CARPHDYDGVDYW"      "CARRYYSNYAWFAYW"</span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># an NA indicates that the cell barcode in the VDJ library was not detected in the GEX object (or was filtered out, depending on mitochondrial gene limits, etc)</span></code></pre></div>
<p>Importantly, all samples are integrated in one vgm. Basic stats can
be retrieved from the VDJ_GEX_stats output which was run as part of the
VGM function (get_VDJ_stats = T)</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">##                                                                                  Repertoir path</span>
<span class="co">## VDJ.stats     N:/reddy/Shared_Folders/AA_AY/DBmanuscript_Aged_CNS/VDJ/Aged.CNS.pool.3m.Bcell.S1</span>
<span class="co">## VDJ.stats1   N:/reddy/Shared_Folders/AA_AY/DBmanuscript_Aged_CNS/VDJ/Aged.CNS.pool.12m.Bcell.S2</span>
<span class="co">## VDJ.stats2   N:/reddy/Shared_Folders/AA_AY/DBmanuscript_Aged_CNS/VDJ/Aged.CNS.pool.18m.Bcell.S3</span>
<span class="co">## VDJ.stats3 N:/reddy/Shared_Folders/AA_AY/DBmanuscript_Aged_CNS/VDJ/Aged.CNS.single.18m.Bcell.S4</span>
<span class="co">##            Sample name Nr unique barcodes Nr barcodes is_cell Nr cells 1VDJ 1VJ</span>
<span class="co">## VDJ.stats           s1                 78                  78                61</span>
<span class="co">## VDJ.stats1          s2                160                 160               122</span>
<span class="co">## VDJ.stats2          s3                208                 208               160</span>
<span class="co">## VDJ.stats3          s4                499                 499               372</span>
<span class="co">##            Nr cells 1VDJ 0VJ Nr cells 0VDJ 1VJ Nr cells 2 or more VDJ 1VJ</span>
<span class="co">## VDJ.stats                  8                 1                          0</span>
<span class="co">## VDJ.stats1                13                15                          1</span>
<span class="co">## VDJ.stats2                 7                16                          4</span>
<span class="co">## VDJ.stats3                34                60                          8</span>
<span class="co">##            Nr cells 1VDJ 2 or more VJ Nr cells 2 or more VDJ 2 or more VJ</span>
<span class="co">## VDJ.stats                           7                                   1</span>
<span class="co">## VDJ.stats1                          7                                   2</span>
<span class="co">## VDJ.stats2                         18                                   3</span>
<span class="co">## VDJ.stats3                         22                                   3</span>
<span class="co">##            Nr cells full_length Nr cells productive Nr cells high_confidence</span>
<span class="co">## VDJ.stats                    78                  78                       78</span>
<span class="co">## VDJ.stats1                  160                 160                      160</span>
<span class="co">## VDJ.stats2                  208                 208                      208</span>
<span class="co">## VDJ.stats3                  499                 499                      499</span>
<span class="co">##            Nr cells all true Nr cells all true and 1VDJ 1VJ Nr clonotypes</span>
<span class="co">## VDJ.stats                 78                             61            73</span>
<span class="co">## VDJ.stats1               160                            122           129</span>
<span class="co">## VDJ.stats2               208                            160           200</span>
<span class="co">## VDJ.stats3               499                            372           349</span>
<span class="co">##            Nr clonotypes 1VDJ 1VJ Nr clonotypes &lt; 1VDJ 1VJ</span>
<span class="co">## VDJ.stats                      58                        8</span>
<span class="co">## VDJ.stats1                     94                       25</span>
<span class="co">## VDJ.stats2                    154                       21</span>
<span class="co">## VDJ.stats3                    248                       77</span>
<span class="co">##            Nr clonotypes &gt; 1VDJ 1VJ % Nr unique barcodes % Nr barcodes is_cell</span>
<span class="co">## VDJ.stats                         7                  100                   100</span>
<span class="co">## VDJ.stats1                       10                  100                   100</span>
<span class="co">## VDJ.stats2                       25                  100                   100</span>
<span class="co">## VDJ.stats3                       24                  100                   100</span>
<span class="co">##            % Nr cells 1VDJ 1VJ % Nr cells 1VDJ 0VJ % Nr cells 0VDJ 1VJ</span>
<span class="co">## VDJ.stats                 78.2                10.3                 1.3</span>
<span class="co">## VDJ.stats1                76.2                 8.1                 9.4</span>
<span class="co">## VDJ.stats2                76.9                 3.4                 7.7</span>
<span class="co">## VDJ.stats3                74.5                 6.8                  12</span>
<span class="co">##            % Nr cells 2 or more VDJ 1VJ % Nr cells 1VDJ 2 or more VJ</span>
<span class="co">## VDJ.stats                             0                            9</span>
<span class="co">## VDJ.stats1                          0.6                          4.4</span>
<span class="co">## VDJ.stats2                          1.9                          8.7</span>
<span class="co">## VDJ.stats3                          1.6                          4.4</span>
<span class="co">##            % Nr cells 2 or more VDJ 2 or more VJ % Nr cells full_length</span>
<span class="co">## VDJ.stats                                    1.3                    100</span>
<span class="co">## VDJ.stats1                                   1.2                    100</span>
<span class="co">## VDJ.stats2                                   1.4                    100</span>
<span class="co">## VDJ.stats3                                   0.6                    100</span>
<span class="co">##            % Nr cells productive % Nr cells high_confidence % Nr cells all true</span>
<span class="co">## VDJ.stats                    100                        100                 100</span>
<span class="co">## VDJ.stats1                   100                        100                 100</span>
<span class="co">## VDJ.stats2                   100                        100                 100</span>
<span class="co">## VDJ.stats3                   100                        100                 100</span>
<span class="co">##            % Nr cells all true and 1VDJ 1VJ % Nr clonotypes</span>
<span class="co">## VDJ.stats                              78.2             100</span>
<span class="co">## VDJ.stats1                             76.2             100</span>
<span class="co">## VDJ.stats2                             76.9             100</span>
<span class="co">## VDJ.stats3                             74.5             100</span>
<span class="co">##            % Nr clonotypes 1VDJ 1VJ % Nr clonotypes &lt; 1VDJ 1VJ</span>
<span class="co">## VDJ.stats                      79.5                         11</span>
<span class="co">## VDJ.stats1                     72.9                       19.4</span>
<span class="co">## VDJ.stats2                       77                       10.5</span>
<span class="co">## VDJ.stats3                     71.1                       22.1</span>
<span class="co">##            % Nr clonotypes &gt; 1VDJ 1VJ Estimated.Number.of.Cells</span>
<span class="co">## VDJ.stats                         9.6                        78</span>
<span class="co">## VDJ.stats1                        7.8                       160</span>
<span class="co">## VDJ.stats2                       12.5                       208</span>
<span class="co">## VDJ.stats3                        6.9                       499</span>
<span class="co">##            Mean.Read.Pairs.per.Cell</span>
<span class="co">## VDJ.stats                   272,337</span>
<span class="co">## VDJ.stats1                  114,292</span>
<span class="co">## VDJ.stats2                   99,848</span>
<span class="co">## VDJ.stats3                   50,928</span>
<span class="co">##            Number.of.Cells.With.Productive.V.J.Spanning.Pair</span>
<span class="co">## VDJ.stats                                                 69</span>
<span class="co">## VDJ.stats1                                               132</span>
<span class="co">## VDJ.stats2                                               185</span>
<span class="co">## VDJ.stats3                                               405</span>
<span class="co">##            Number.of.Read.Pairs Valid.Barcodes Q30.Bases.in.Barcode</span>
<span class="co">## VDJ.stats            21,242,313          96.5%                91.7%</span>
<span class="co">## VDJ.stats1           18,286,818          96.7%                91.9%</span>
<span class="co">## VDJ.stats2           20,768,413          95.3%                91.7%</span>
<span class="co">## VDJ.stats3           25,413,451          96.1%                91.7%</span>
<span class="co">##            Q30.Bases.in.RNA.Read.1 Q30.Bases.in.RNA.Read.2 Q30.Bases.in.UMI</span>
<span class="co">## VDJ.stats                    73.3%                   79.8%            91.0%</span>
<span class="co">## VDJ.stats1                   73.2%                   79.8%            91.1%</span>
<span class="co">## VDJ.stats2                   73.1%                   79.2%            91.0%</span>
<span class="co">## VDJ.stats3                   73.3%                   79.5%            91.0%</span>
<span class="co">##            Reads.Mapped.to.Any.V.D.J.Gene Reads.Mapped.to.IGH</span>
<span class="co">## VDJ.stats                           82.6%               26.3%</span>
<span class="co">## VDJ.stats1                          82.0%               30.4%</span>
<span class="co">## VDJ.stats2                          80.0%               20.5%</span>
<span class="co">## VDJ.stats3                          83.2%               24.6%</span>
<span class="co">##            Reads.Mapped.to.IGK Reads.Mapped.to.IGL</span>
<span class="co">## VDJ.stats                29.4%               27.0%</span>
<span class="co">## VDJ.stats1               30.3%               21.3%</span>
<span class="co">## VDJ.stats2               38.0%               21.5%</span>
<span class="co">## VDJ.stats3               40.6%               18.1%</span>
<span class="co">##            Mean.Used.Read.Pairs.per.Cell Fraction.Reads.in.Cells</span>
<span class="co">## VDJ.stats                         36,548                   79.0%</span>
<span class="co">## VDJ.stats1                        29,068                   74.6%</span>
<span class="co">## VDJ.stats2                        33,816                   84.6%</span>
<span class="co">## VDJ.stats3                        18,887                   55.1%</span>
<span class="co">##            Median.IGH.UMIs.per.Cell Median.IGK.UMIs.per.Cell</span>
<span class="co">## VDJ.stats                        11                       30</span>
<span class="co">## VDJ.stats1                        7                       12</span>
<span class="co">## VDJ.stats2                       10                       31</span>
<span class="co">## VDJ.stats3                        4                       10</span>
<span class="co">##            Median.IGL.UMIs.per.Cell Cells.With.Productive.V.J.Spanning.Pair</span>
<span class="co">## VDJ.stats                        47                                   88.5%</span>
<span class="co">## VDJ.stats1                        8                                   82.5%</span>
<span class="co">## VDJ.stats2                       32                                   88.9%</span>
<span class="co">## VDJ.stats3                      6.5                                   81.2%</span>
<span class="co">##            Cells.With.Productive.V.J.Spanning..IGK..IGH..Pair</span>
<span class="co">## VDJ.stats                                               80.8%</span>
<span class="co">## VDJ.stats1                                              71.2%</span>
<span class="co">## VDJ.stats2                                              82.2%</span>
<span class="co">## VDJ.stats3                                              76.0%</span>
<span class="co">##            Cells.With.Productive.V.J.Spanning..IGL..IGH..Pair</span>
<span class="co">## VDJ.stats                                                9.0%</span>
<span class="co">## VDJ.stats1                                              11.9%</span>
<span class="co">## VDJ.stats2                                               6.7%</span>
<span class="co">## VDJ.stats3                                               5.8%</span>
<span class="co">##            Paired.Clonotype.Diversity Cells.With.IGH.Contig</span>
<span class="co">## VDJ.stats                       69.14                 98.7%</span>
<span class="co">## VDJ.stats1                         80                 93.8%</span>
<span class="co">## VDJ.stats2                     191.43                 96.2%</span>
<span class="co">## VDJ.stats3                     159.31                 90.6%</span>
<span class="co">##            Cells.With.IGK.Contig Cells.With.IGL.Contig</span>
<span class="co">## VDJ.stats                  91.0%                 38.5%</span>
<span class="co">## VDJ.stats1                 82.5%                 38.1%</span>
<span class="co">## VDJ.stats2                 92.3%                 41.8%</span>
<span class="co">## VDJ.stats3                 89.0%                 29.5%</span>
<span class="co">##            Cells.With.CDR3.annotated.IGH.Contig</span>
<span class="co">## VDJ.stats                                 98.7%</span>
<span class="co">## VDJ.stats1                                91.9%</span>
<span class="co">## VDJ.stats2                                93.3%</span>
<span class="co">## VDJ.stats3                                88.6%</span>
<span class="co">##            Cells.With.CDR3.annotated.IGK.Contig</span>
<span class="co">## VDJ.stats                                 82.1%</span>
<span class="co">## VDJ.stats1                                78.8%</span>
<span class="co">## VDJ.stats2                                88.9%</span>
<span class="co">## VDJ.stats3                                86.8%</span>
<span class="co">##            Cells.With.CDR3.annotated.IGL.Contig</span>
<span class="co">## VDJ.stats                                 11.5%</span>
<span class="co">## VDJ.stats1                                13.8%</span>
<span class="co">## VDJ.stats2                                 8.2%</span>
<span class="co">## VDJ.stats3                                 7.2%</span>
<span class="co">##            Cells.With.V.J.Spanning.IGH.Contig</span>
<span class="co">## VDJ.stats                               98.7%</span>
<span class="co">## VDJ.stats1                              93.1%</span>
<span class="co">## VDJ.stats2                              95.7%</span>
<span class="co">## VDJ.stats3                              88.8%</span>
<span class="co">##            Cells.With.V.J.Spanning.IGK.Contig</span>
<span class="co">## VDJ.stats                               89.7%</span>
<span class="co">## VDJ.stats1                              81.2%</span>
<span class="co">## VDJ.stats2                              92.3%</span>
<span class="co">## VDJ.stats3                              88.6%</span>
<span class="co">##            Cells.With.V.J.Spanning.IGL.Contig Cells.With.Productive.IGH.Contig</span>
<span class="co">## VDJ.stats                               16.7%                            98.7%</span>
<span class="co">## VDJ.stats1                              17.5%                            90.6%</span>
<span class="co">## VDJ.stats2                              10.1%                            92.3%</span>
<span class="co">## VDJ.stats3                              10.4%                            88.0%</span>
<span class="co">##            Cells.With.Productive.IGK.Contig Cells.With.Productive.IGL.Contig</span>
<span class="co">## VDJ.stats                             80.8%                            10.3%</span>
<span class="co">## VDJ.stats1                            78.8%                            13.8%</span>
<span class="co">## VDJ.stats2                            88.5%                             8.2%</span>
<span class="co">## VDJ.stats3                            86.6%                             7.2%</span>
<span class="co">##            rep_id Estimated.Number.of.Cells Mean.Reads.per.Cell</span>
<span class="co">## VDJ.stats       1                       303             819,292</span>
<span class="co">## VDJ.stats1      2                       261             919,694</span>
<span class="co">## VDJ.stats2      3                       945             413,917</span>
<span class="co">## VDJ.stats3      4                       595             369,552</span>
<span class="co">##            Median.Genes.per.Cell Number.of.Reads Valid.Barcodes</span>
<span class="co">## VDJ.stats                  1,199     248,245,523          83.2%</span>
<span class="co">## VDJ.stats1                   970     240,040,317          82.8%</span>
<span class="co">## VDJ.stats2                 1,338     391,152,436          85.3%</span>
<span class="co">## VDJ.stats3                   766     219,883,523          84.2%</span>
<span class="co">##            Sequencing.Saturation Q30.Bases.in.Barcode Q30.Bases.in.RNA.Read</span>
<span class="co">## VDJ.stats                  99.1%                97.7%                 92.3%</span>
<span class="co">## VDJ.stats1                 98.8%                97.8%                 89.9%</span>
<span class="co">## VDJ.stats2                 98.2%                97.8%                 93.1%</span>
<span class="co">## VDJ.stats3                 97.9%                97.7%                 92.3%</span>
<span class="co">##            Q30.Bases.in.UMI Reads.Mapped.to.Genome</span>
<span class="co">## VDJ.stats             97.5%                  90.7%</span>
<span class="co">## VDJ.stats1            97.6%                  83.1%</span>
<span class="co">## VDJ.stats2            97.6%                  92.2%</span>
<span class="co">## VDJ.stats3            97.5%                  91.1%</span>
<span class="co">##            Reads.Mapped.Confidently.to.Genome</span>
<span class="co">## VDJ.stats                               85.4%</span>
<span class="co">## VDJ.stats1                              77.3%</span>
<span class="co">## VDJ.stats2                              83.2%</span>
<span class="co">## VDJ.stats3                              85.7%</span>
<span class="co">##            Reads.Mapped.Confidently.to.Intergenic.Regions</span>
<span class="co">## VDJ.stats                                           10.7%</span>
<span class="co">## VDJ.stats1                                          15.0%</span>
<span class="co">## VDJ.stats2                                          11.3%</span>
<span class="co">## VDJ.stats3                                          15.8%</span>
<span class="co">##            Reads.Mapped.Confidently.to.Intronic.Regions</span>
<span class="co">## VDJ.stats                                          9.7%</span>
<span class="co">## VDJ.stats1                                        10.2%</span>
<span class="co">## VDJ.stats2                                        10.5%</span>
<span class="co">## VDJ.stats3                                        11.1%</span>
<span class="co">##            Reads.Mapped.Confidently.to.Exonic.Regions</span>
<span class="co">## VDJ.stats                                       65.0%</span>
<span class="co">## VDJ.stats1                                      52.0%</span>
<span class="co">## VDJ.stats2                                      61.4%</span>
<span class="co">## VDJ.stats3                                      58.7%</span>
<span class="co">##            Reads.Mapped.Confidently.to.Transcriptome</span>
<span class="co">## VDJ.stats                                      61.5%</span>
<span class="co">## VDJ.stats1                                     49.1%</span>
<span class="co">## VDJ.stats2                                     59.7%</span>
<span class="co">## VDJ.stats3                                     54.9%</span>
<span class="co">##            Reads.Mapped.Antisense.to.Gene Fraction.Reads.in.Cells</span>
<span class="co">## VDJ.stats                           12.7%                   84.2%</span>
<span class="co">## VDJ.stats1                          12.9%                   59.7%</span>
<span class="co">## VDJ.stats2                          11.7%                   88.8%</span>
<span class="co">## VDJ.stats3                          14.6%                   51.9%</span>
<span class="co">##            Total.Genes.Detected Median.UMI.Counts.per.Cell rep_id</span>
<span class="co">## VDJ.stats                14,652                      2,865      1</span>
<span class="co">## VDJ.stats1               14,033                      1,966      2</span>
<span class="co">## VDJ.stats2               16,905                      3,151      3</span>
<span class="co">## VDJ.stats3               14,629                      1,536      4</span></code></pre>
<p>Moreover, given the many parameters of the VGM function is may be
useful to check back on runtime parameters. These, as well as session
info are also saved in the vgm object</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co">#Runtime params</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co">#session info</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="gene-expression-analysis-platypus-v3">6. Gene expression analysis (Platypus v3)<a class="anchor" aria-label="anchor" href="#gene-expression-analysis-platypus-v3"></a>
</h2>
<p>The vgm object now contains all information needed for both V(D)J and
gene expression analysis. In the examples below we first focus on gene
expression</p>
<p>We can visualize the 2D plots for each sample individually. By
default, UMAP, PCA, and TSNE embeddings are included in the object. Here
it is important to supply the seurat object in vgm[[2]]. For the
parameter group.by we supply “sample_id” to color points by the entry in
their sample id column</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># In Platypus version 2, the output from GEX_automate was used as input to other GEX functions. These functions are still compatible with v3 if the vgm[[2]] seurat object is supplied as input.</span>
<span class="co"># For example, the following function can be used to calculate the DE genes for each cluster, as before. </span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,reduction <span class="op">=</span> <span class="st">"pca"</span>, group.by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,reduction <span class="op">=</span> <span class="st">"tsne"</span>, group.by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span>
<span class="co">#alternatively plot each sample separately</span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,reduction <span class="op">=</span> <span class="st">"umap"</span>, split.by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span>
<span class="co">#this also works with any other column of vgm[[2]]@meta.data</span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,reduction <span class="op">=</span> <span class="st">"umap"</span>, split.by <span class="op">=</span> <span class="st">"group_id"</span><span class="op">)</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-9-1.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-9-2.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-9-3.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-9-4.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-9-5.png" width="672"></p>
<p>We can observe that the majority of clusters have cells from both
samples, suggesting a similar distribution of transcriptional properties
between the two samples. We can also plot the cluster membership for
each of the distinct sample using the GEX_proportions_barplot
function.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/GEX_proportions_barplot.html">GEX_proportions_barplot</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, stacked.plot <span class="op">=</span> <span class="cn">T</span>, source.group <span class="op">=</span> <span class="st">"sample_id"</span>, target.group <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span>
<span class="co">#This function is very flexible and can be used to plot proportions of cells from and of any groups. For this use the source.group and target.group parameters to specify metadata columns.</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<p>Finally, we can also look at common B cell markers.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,reduction <span class="op">=</span> <span class="st">"umap"</span>,features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD19"</span>,<span class="st">"PTPRC"</span>, <span class="st">"EBF1"</span>, <span class="st">"H2-K1"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#To easily scout through genes in the dataset use:</span>
<span class="co">#View(as.data.frame(rownames(vgm[[2]])))</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<p>Platypus also allows us to assign cell type/state identity to
different clusters by using GEX_phenotype. This function takes the
Seurat object as input and uses canonical markers to easily match the
clustering to known cell types. The user also has the possibility to use
a custom list of markers and their associated cell types/states. While
the default cell states are quite basic, we find that the strength of
the function is that a user can label cells with custon phenotypes as
demonstrated below. For more information please see the functions
documentation.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#using defaults</span>
<span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_phenotype.html">GEX_phenotype</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, default <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="co">#custom criteria</span>
<span class="co">#vgm[[2]] &lt;- GEX_phenotype(vgm[[2]], default = F,cell.state.markers=c("CD8A+;CCL5+;CD44+;IL7R-;CD19-","CD8A+;CCL5-;CD44+;IL7R+;CD19-"),cell.state.names=c("EffectorCD8","MemoryCD8"))</span></code></pre></div>
<p>The resulting Seurat object now contains a “cell.state” column which
can be used for annotation in the DimPlot function of the Seurat
package.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"cell.state"</span><span class="op">)</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<div class="section level3">
<h3 id="differential-gene-expression-analysis">6.1. Differential Gene Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-gene-expression-analysis"></a>
</h3>
<p>After scaling, normalizing, and clustering the cells from the GEX
libraries we can now investigate which genes are differentially
expressed between either clusters or samples. First, we can investigate
the genes that define each of the clusters by using the GEX output of
the VGM function. Depending on the size of the dataset and the number of
cells this function can be quite slow. The output of this function is a
list in which each element contains the differentially expressed genes
for a given cluster. For example, the first element of the list will
correspond to a dataframe describing the genes for cluster0 that we
previously observed on the UMAP. This list object will correspond to the
length of the number of clusters that were previously calculated, and
has the same format as the FindMarkers function from Seurat. The pct.1
will correspond to the percentage of cells expressing the gene in the
cluster of interest, and the pct.2 to the percentage of cells in all
other clusters.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_expression_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_cluster_genes.html">GEX_cluster_genes</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gene_expression_cluster</span><span class="op">)</span> <span class="co"># length of 12, corresponding to 12 clusters</span></code></pre></div>
<pre><code><span class="co">## [1] 12</span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span><span class="op">)</span> <span class="co"># length of 12</span></code></pre></div>
<pre><code><span class="co">## [1] 12</span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gene_expression_cluster</span>,<span class="va">nrow</span><span class="op">)</span><span class="op">)</span> <span class="co">#Nr of differentially expressed genes per cluster</span></code></pre></div>
<pre><code><span class="co">##  [1]  711  506  459 1463  459  819 1021  937  871 1211 1061  956</span></code></pre>
<p>We can now look at some of the genes associated with cluster0 in the
previously displayed umap</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_expression_cluster</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##               p_val avg_logFC pct.1 pct.2     p_val_adj SYMBOL cluster</span>
<span class="co">## NKG7   0.000000e+00  3.311587 0.980 0.102  0.000000e+00   NKG7       0</span>
<span class="co">## CCL5  1.934113e-292  3.462946 0.970 0.110 6.127462e-288   CCL5       0</span>
<span class="co">## CTSW  2.215096e-254  2.329515 0.865 0.079 7.017646e-250   CTSW       0</span>
<span class="co">## CD8A  2.800031e-251  2.591763 0.801 0.058 8.870778e-247   CD8A       0</span>
<span class="co">## CD8B1 9.033911e-238  2.626240 0.835 0.080 2.862033e-233  CD8B1       0</span>
<span class="co">## THY1  3.170291e-232  2.295433 0.930 0.146 1.004380e-227   THY1       0</span></code></pre>
<p>Raw tables are informative, but not visually appealing.
Differentially expressed genes may be plotted in different modes using
Platypus.</p>
<p>It is also possible to create a heatmap displaying differentially
expressed genes for each cluster. This can be customized to sub-sample
cells in case certain clusters are too large for visualization purposes.
Additionally, the user can determine the number of genes to display for
each cluster based on the n.genes.per.cluster argument. The function
GEX_cluster_genes_heatmap can be used to produce a ggplot object, based
on the DoHeatmap from Seurat.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">agedCNS_heatmap_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_cluster_genes_heatmap.html">GEX_cluster_genes_heatmap</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, GEX_cluster_genes.output <span class="op">=</span> <span class="va">gene_expression_cluster</span>,n.genes.per.cluster <span class="op">=</span> <span class="fl">3</span>,max.cell <span class="op">=</span> <span class="fl">30</span>,metric <span class="op">=</span> <span class="st">"avg_logFC"</span>, platypus.version <span class="op">=</span> <span class="st">"v3"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">agedCNS_heatmap_clusters</span><span class="op">)</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-16-1.png" width="672"></p>
<p>After plotting the ggplot object we can clearly see genes enriched in
various clusters - mainly indicated by the diagonal.</p>
<p>Further, volcano plots may be used to show and annotate a larger
number of differentially expressed genes</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">agedCNS_heatmap_volcano</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_volcano.html">GEX_volcano</a></span><span class="op">(</span>DEGs.input <span class="op">=</span> <span class="va">gene_expression_cluster</span>, input.type <span class="op">=</span> <span class="st">"cluster.genes"</span>, RP.MT.filter <span class="op">=</span> <span class="cn">T</span>, color.p.threshold <span class="op">=</span> <span class="fl">0.01</span>, n.label.up <span class="op">=</span> <span class="fl">10</span>, n.label.down <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "Top 10 and bottom 10 genes will be labeled"</span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">agedCNS_heatmap_volcano</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co">#genes specific to cluster 0</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<p>We can extract and test for differential expressed genes between the
two samples (or between other subgroups) by using the
GEX_DEgenes_persample function.</p>
<p>This function allows us to also create a heatmap displaying the top
most up- or downregulated genes for each cluster based on log fold
change (avg_logFC) or p value (adj_p_value). Additionally, the user can
determine the number of up- and downreulated genes to be displayed for
each sample. In this case, the output returns a list where the first
element contains the dataframe with the differntial expression
information and the second element contains the heatmap displaying the
most up-/downregulated genes.<br>
If more than two samples are to be examined please refer to
GEX_pairwise_degs two chunks below</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DE_genes_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_DEgenes.html">GEX_DEgenes</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,min.pct <span class="op">=</span> <span class="fl">.25</span>, grouping.column <span class="op">=</span> <span class="st">"sample_id"</span>,group1 <span class="op">=</span> <span class="st">"s1"</span>, group2 <span class="op">=</span> <span class="st">"s2"</span>,return.plot <span class="op">=</span> <span class="st">"volcano"</span>,up.genes <span class="op">=</span> <span class="fl">5</span>,down.genes <span class="op">=</span> <span class="fl">5</span>,logFC <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="co">#This function is flexible and takes any column name as grouping.column to allow easy exploration of differences between custom groups</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DE_genes_samples</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">##              p_val avg_logFC pct.1 pct.2    p_val_adj SYMBOL signif_pval</span>
<span class="co">## RGS10 1.564907e-13  1.216417 0.436 0.138 4.957783e-09  RGS10        TRUE</span>
<span class="co">## CTSD  3.522094e-12  1.866324 0.568 0.297 1.115835e-07   CTSD        TRUE</span>
<span class="co">## C1QC  7.536629e-11  1.552118 0.375 0.134 2.387680e-06   C1QC        TRUE</span>
<span class="co">## CST3  2.673577e-10  1.840032 0.679 0.455 8.470159e-06   CST3        TRUE</span>
<span class="co">## HPGDS 3.129809e-10  1.120999 0.287 0.077 9.915547e-06  HPGDS        TRUE</span>
<span class="co">## C1QA  5.307456e-10  1.487025 0.372 0.138 1.681455e-05   C1QA        TRUE</span>
<span class="co">##       signif_logfc color_threshold</span>
<span class="co">## RGS10         TRUE            TRUE</span>
<span class="co">## CTSD          TRUE            TRUE</span>
<span class="co">## C1QC          TRUE            TRUE</span>
<span class="co">## CST3          TRUE            TRUE</span>
<span class="co">## HPGDS         TRUE            TRUE</span>
<span class="co">## C1QA          TRUE            TRUE</span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DE_genes_samples</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> </code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<p>The same function can also return DEGs between seurat clusters or any
2 groups of cells Here DEGs between cluster 1 and 3 are calculated</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DE_genes_cl1_vs_3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_DEgenes.html">GEX_DEgenes</a></span><span class="op">(</span>GEX<span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,min.pct <span class="op">=</span> <span class="fl">.25</span>, grouping.column <span class="op">=</span> <span class="st">"seurat_clusters"</span>,group1 <span class="op">=</span> <span class="st">"1"</span>, group2 <span class="op">=</span> <span class="st">"3"</span>,return.plot <span class="op">=</span> <span class="st">"heatmap"</span>,up.genes <span class="op">=</span> <span class="fl">10</span>,down.genes <span class="op">=</span> <span class="fl">10</span>,logFC <span class="op">=</span> <span class="cn">F</span>, size.top.colorbar <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "CD74"   "H2-AA"  "CD79A"  "H2-EB1" "H2-AB1" "EBF1"   "TMSB10" "H3F3A" </span>
<span class="co">##  [9] "FAU"    "LY6D"   "C1QC"   "P2RY12" "CX3CR1" "CSF1R"  "C1QA"   "C1QB"  </span>
<span class="co">## [17] "SPARC"  "FCER1G" "OLFML3" "GPR34"</span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#head(DE_genes_cl1_vs_3[[1]]) </span>
<span class="va">DE_genes_cl1_vs_3</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> </code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-19-1.png" width="672"></p>
<p>To gain a complete overview of DEGs between groups or clusters the
GEX_pairwise_degs function is used This function calculates DEGs for
every possible pairwise comparison between groups of the selected
column. It is thereby recommended to use this function with a maximum
number of 12 groups/clusters/samples This function automatically saves
plots where top label.n.top.genes genes are labeled by their p value.
Additionally, genes of special interest that should be labeled
irregardless of their p value can be supplied to genes.to.label</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#DE_clusters_all &lt;- GEX_pairwise_DEGs(GEX = vgm[[2]], group.by = "seurat_clusters", min.pct = 0.25, RP.MT.filter = T, label.n.top.genes = 10, genes.to.label = c("CD74", "EBF1"), save.plot = F)</span></code></pre></div>
<p>Now that cluster defining genes have been examined, a helper plotting
function allows to visualize selected genes.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dottile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_dottile_plot.html">GEX_dottile_plot</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD19"</span>, <span class="st">"CD74"</span>,<span class="st">"SDC1"</span>, <span class="st">"EBF1"</span>,<span class="st">"PTPRC"</span>,<span class="st">"CD93"</span>,<span class="st">"CD38"</span>,<span class="st">"CD24A"</span>,<span class="st">"CD34"</span>,<span class="st">"CD1D1"</span>,<span class="st">"CR2"</span>,<span class="st">"MS4A1"</span>,<span class="st">"CXCR5"</span>,<span class="st">"SELL"</span>,<span class="st">"CD40"</span>,<span class="st">"CD83"</span>,<span class="st">"H2-AB1"</span>,<span class="st">"H2-EB1"</span>,<span class="st">"CD27"</span>,<span class="st">"POU2AF1"</span>,<span class="st">"NT5E"</span>,<span class="st">"FAS"</span>,<span class="st">"PDCD1LG2"</span>,<span class="st">"PRDM1"</span>,<span class="st">"ITGAM"</span>,<span class="st">"IL10"</span>,<span class="st">"IL12A"</span>,<span class="st">"HAVCR2"</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>, threshold.to.plot <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> 
<span class="co">#threshold.to.plot specifies how many % of cells have to express a gene to show a dot.</span>
<span class="va">dottile</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, plot.subtitle <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## Warning: Removed 194 rows containing missing values (geom_point).</span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#For visualisation purposes in the RMD format</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
<p>To examine patterns of co-expression two functions allow to generate
both overview and single-cell resolution plots.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#overview</span>
<span class="va">coexpression_dotmap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_coexpression_coefficient.html">GEX_coexpression_coefficient</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD19"</span>, <span class="st">"CD74"</span>,<span class="st">"SDC1"</span>, <span class="st">"EBF1"</span>,<span class="st">"PTPRC"</span>,<span class="st">"CD93"</span>,<span class="st">"CD38"</span>,<span class="st">"CD24A"</span>,<span class="st">"CD34"</span><span class="op">)</span>, plot.dotmap <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Calculating coexpression for 9 genes with 2064 cells</span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coexpression_dotmap</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>
<span class="co">#detail of two selected genes</span>
<span class="fu"><a href="../reference/GEX_scatter_coexpression.html">GEX_scatter_coexpression</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, gene.1 <span class="op">=</span> <span class="st">"CD19"</span>, gene.2 <span class="op">=</span> <span class="st">"EBF1"</span>, color.theme <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span>
<span class="co">#to save use: </span>
<span class="co">#ggsave(last_plot(), filename = "Coexpression_scatter.png")</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-22-1.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-22-2.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="go-term-analysis">6.2. GO Term analysis<a class="anchor" aria-label="anchor" href="#go-term-analysis"></a>
</h3>
<p>Earlier we mentioned how we can match the unbiased clustering to
known cell types using canonical markers. Platypus also allows us to run
a GO or KEGG term analysis in order to obtain information on the most
significant GO/KEGG terms and their visualizations using the GEX_GOterm
function. (To do this, we first need to organize the top genes that
define each Seurat cluster and convert them into a single dataframe.
This is done using the GEX_topN_DE_genes_per_cluster function). This
function does require internet connection.</p>
<p>To return plots as PDF directly to the current working directory set
go.plot = T</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ontology_agedCNS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_GOterm.html">GEX_GOterm</a></span><span class="op">(</span>GEX.cluster.genes.output <span class="op">=</span> <span class="va">gene_expression_cluster</span>, topNgenes <span class="op">=</span> <span class="fl">10</span>, go.plots <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ontology_agedCNS</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co">#Cluster 0 </span></code></pre></div>
<p>We can additionally extract the top N genes per cluster directly
(with filtering) using the following function:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top_10_genes_per_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_topN_DE_genes_per_cluster.html">GEX_topN_DE_genes_per_cluster</a></span><span class="op">(</span>GEX_cluster_genes.output <span class="op">=</span> <span class="va">gene_expression_cluster</span>, n.genes <span class="op">=</span> <span class="fl">10</span>, by_FC <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">top_10_genes_per_cluster</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##               p_val avg_logFC pct.1 pct.2    p_val_adj SYMBOL cluster</span>
<span class="co">## CD74   8.127208e-94 -6.762404 0.076 0.598 2.574781e-89   CD74       0</span>
<span class="co">## CST3   1.123493e-19 -5.047920 0.351 0.481 3.559339e-15   CST3       0</span>
<span class="co">## H2-AA  6.924181e-75 -4.816353 0.016 0.479 2.193650e-70  H2-AA       0</span>
<span class="co">## CD79A  4.947812e-91 -4.666191 0.008 0.540 1.567516e-86  CD79A       0</span>
<span class="co">## H2-EB1 1.379589e-68 -4.276945 0.010 0.444 4.370677e-64 H2-EB1       0</span>
<span class="co">## EBF1   4.670730e-89 -4.164532 0.004 0.529 1.479734e-84   EBF1       0</span></code></pre>
</div>
<div class="section level3">
<h3 id="gsea-term-analysis">6.3. GSEA Term analysis<a class="anchor" aria-label="anchor" href="#gsea-term-analysis"></a>
</h3>
<p>We can also perform a Gene Set Enrichment Analysis (GSEA) using the
GEX_GSEA function. For this, the user needs to provide the path to a gmt
file containing the gene sets, which can be downloaded for example from
MSigDB. (<a href="http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp" class="external-link uri">http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp</a>
C7: immunologic signature gene sets) For instance we can perform GSEA
for cluster0 and look at the most significant pathways.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gsea_EAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_GSEA.html">GEX_GSEA</a></span><span class="op">(</span>GEX.cluster.genes.output <span class="op">=</span> <span class="va">gene_expression_cluster</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, MT.Rb.filter <span class="op">=</span> <span class="cn">T</span>, path.to.pathways <span class="op">=</span> <span class="st">"~/Downloads/c7.all.v7.4.symbols.gmt"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="vdj-repertoire-anaylsis">7. VDJ Repertoire anaylsis<a class="anchor" aria-label="anchor" href="#vdj-repertoire-anaylsis"></a>
</h2>
<div class="section level3">
<h3 id="checking-vgm-output">7.1. Checking VGM output<a class="anchor" aria-label="anchor" href="#checking-vgm-output"></a>
</h3>
<p>Now we can analyze the VDJ repertoire data before integrating VDJ and
GEX information. This may be useful if only VDJ libraries have been
sequenced without the accompanying gene expression data. We first start
by examining the structure of vgm[[1]]</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "barcode"                 "sample_id"              </span>
<span class="co">##  [3] "group_id"                "clonotype_id_10x"       </span>
<span class="co">##  [5] "clonotype_id"            "clonotype_frequency"    </span>
<span class="co">##  [7] "celltype"                "Nr_of_VDJ_chains"       </span>
<span class="co">##  [9] "Nr_of_VJ_chains"         "VDJ_cdr3s_aa"           </span>
<span class="co">## [11] "VJ_cdr3s_aa"             "VDJ_cdr3s_nt"           </span>
<span class="co">## [13] "VJ_cdr3s_nt"             "VDJ_umis"               </span>
<span class="co">## [15] "VJ_umis"                 "VDJ_chain_contig"       </span>
<span class="co">## [17] "VJ_chain_contig"         "VDJ_chain"              </span>
<span class="co">## [19] "VJ_chain"                "VDJ_vgene"              </span>
<span class="co">## [21] "VJ_vgene"                "VDJ_dgene"              </span>
<span class="co">## [23] "VDJ_jgene"               "VJ_jgene"               </span>
<span class="co">## [25] "VDJ_cgene"               "VJ_cgene"               </span>
<span class="co">## [27] "VDJ_sequence_nt_raw"     "VJ_sequence_nt_raw"     </span>
<span class="co">## [29] "VDJ_sequence_nt_trimmed" "VJ_sequence_nt_trimmed" </span>
<span class="co">## [31] "VDJ_sequence_aa"         "VJ_sequence_aa"         </span>
<span class="co">## [33] "VDJ_raw_ref"             "VJ_raw_ref"             </span>
<span class="co">## [35] "VDJ_trimmed_ref"         "VJ_trimmed_ref"         </span>
<span class="co">## [37] "VDJ_raw_consensus_id"    "VJ_raw_consensus_id"    </span>
<span class="co">## [39] "orig_barcode"            "specifity"              </span>
<span class="co">## [41] "affinity"                "GEX_available"          </span>
<span class="co">## [43] "orig.ident"              "seurat_clusters"        </span>
<span class="co">## [45] "PC_1"                    "PC_2"                   </span>
<span class="co">## [47] "UMAP_1"                  "UMAP_2"                 </span>
<span class="co">## [49] "tSNE_1"                  "tSNE_2"                 </span>
<span class="co">## [51] "batch_id"</span></code></pre>
<p>If GEX data has been integrated, this dataframe also contains
dimensional reduction coordinates and cluster id. If trim.and.align was
set to TRUE, the columns VDJ_sequence_nt_trimmed to VJ_trimmed_ref
contain aligned and reference sequences. The clonotype_frequency column
takes input from the clonotyping output of cellranger, which is saved
int the clonotype_id_10x column.</p>
<p>This object contains both cells with less than 2 V(D)J chains as well
as with more than 2. To show how this is done the VDJ_cgene column is
taken as an example</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VDJ_cgene</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "IGHM"        "IGHD"        ""            "IGHM;IGHM"   "IGHG2B"     </span>
<span class="co">## [6] "IGHM;IGHD"   "IGHG1"       "IGHA"        "IGHM;IGHG2C"</span></code></pre>
<p>Apart from classical isotypes, we can see an empty (““) element. This
is for cells, for which only a VJ chain is available. Secondly we can
see 3 examples of cells containing more than one VDJ chain. These are
delimited by”;“. Importantly, this format is maintained throughout all
columns.</p>
<p>Both cells with less and more than 2 chains can be cumbersome for
analysis. Platypus V3 functions generally support such cells. To check
their quantity and, if needed, exclude these, the following line can be
used</p>
<p>For advanced filtering and selection options concerning cells with
aberrant number of chains, please refer to the detailed VDJ_GEX_matrix
vignette.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">" Cell count by number of VDJ chains"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  Cell count by number of VDJ chains</span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Nr_of_VDJ_chains</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##   0   1   2 </span>
<span class="co">##  92 831  22</span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n Cell count by number of VJ chains"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  Cell count by number of VJ chains</span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Nr_of_VJ_chains</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##   0   1   2 </span>
<span class="co">##  62 820  63</span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Subset the VGM matrix to only include cells with 1 VDJ and 1 VJ chain</span>
<span class="co">#vgm[[1]] &lt;- subset(vgm[[1]], Nr_of_VDJ_chains == 1 &amp; Nr_of_VJ_chains == 1)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="changing-the-clonotype-strategy">7.2. Changing the clonotype strategy<a class="anchor" aria-label="anchor" href="#changing-the-clonotype-strategy"></a>
</h3>
<p>Often paired nucleotide CDRH3 and CDRL3 clonotyping may not be the
best strategy given somatic hypermutation may occur in the CDR3 region.
Therefore, there could be highly similar clones that likely bind the
same antigen that are officially part of different clonal families. To
address this, we have added a function that allows for various heuristic
clonotyping strategies. This involves clonotyping by identical amino
acid CDRH3 + CDRL3 seuqence, identical germline usage, or seqeunce
homology requirements. This function works both for V3 and V2 of
platypus. The version needs to be specified. For V3 (vgm)
platypus.version a new clonotyping function was added to improve the
existing one. For the full documentation see VDJ_clonotype().</p>
<p>Three new parameters have been added therein: 1. global.clonotype If
set to TRUE, clonotypes will be generated across all samples of the vgm.
This may be useful if more than one sample has been taken from the same
animal (e.g. spleen and bone marrow). Moreover this can provide an
alternative approach to investigate clone sharing across donors or mice.
After global clonotyping, the resulting clonotype_id can be used as a
grouping variable and summary statistics can be generated per clone
(including its presence in multiple repertoires.) A key update as part
of the hierarchical clonotyping strategy is the same-sample-criterion.
This is explained below. 2. hierarchical For cells with aberrant numbers
of VDJ and VJ chains: the functions implements a hierarchical
clonotyping strategy for such cells. In brief, clonotypes are determined
based on all cells with exactly 1VDJ and 1VJ chain. Other cells are then
“joined in” in post. This follows three steps 1. hierarchical = “none”
Create clonotypes from 1 VDJ 1 VJ cells. 2. hierarchical =
“single.chains” Cells with a single chain (e.g. 0VDJ and 1VJ) are merged
into existing clonotypes containing the same or a homologous VJ chain,
depending on the clonotyping strategy. Here the same-sample-criterion is
applied when using global clonotyping. If multiple matching clonotypes
have been found for that cell then ones within the same sample_id are
prioritized. Further if multiple matches exist, the cell is merged into
the existing clone with the highest frequency 2. hierarchical =
“double.and.single.chains” A combination of 1VDJ and 2VJ chain will
first be checked for frequency in the dataset (by exact nucleotide CDR3
matching). Only if this count exceeds the triple.chain.count.threshold,
the clone is used as a “hub clone”. This protects from merging
clonotypes on the basis of rare doublets. Existing clonotypes are then
merged into the 1 VDJ 2 VJ clonotypes as they match with the assumption
that e.g. a cell with 1 VDJ 1 VJ is part of that same clonotype, but
missing a VJ chain due to stochastical sampling</p>
<p>In the example that follows, we use VDJ_clonotype to group cells into
clones based on identical CDRH3 + CDRL3 amino acid sequence. We will
compare this to the case in which we group B cells by using the same
germline genes (both heavy chain and light chain).</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonotype.html">VDJ_clonotype</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, clone.strategy <span class="op">=</span> <span class="st">"cdr3.aa"</span>, global.clonotype <span class="op">=</span> <span class="cn">F</span>, VDJ.VJ.1chain <span class="op">=</span> <span class="cn">F</span>, hierarchical <span class="op">=</span> <span class="st">"single.chains"</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## Clonotyping strategy: cdr3.aa</span></code></pre>
<pre><code><span class="co">## Filtered out 22 cells containing more than one VDJ AND VJ chain or two VDJ chains, as these likely correspond to doublets</span></code></pre>
<pre><code><span class="co">## Processing sample 1 of 4</span></code></pre>
<pre><code><span class="co">## Attempting to merge in 9 aberrant cells</span></code></pre>
<pre><code><span class="co">## Processing sample 2 of 4</span></code></pre>
<pre><code><span class="co">## Attempting to merge in 28 aberrant cells</span></code></pre>
<pre><code><span class="co">## Processing sample 3 of 4</span></code></pre>
<pre><code><span class="co">## Attempting to merge in 23 aberrant cells</span></code></pre>
<pre><code><span class="co">## Processing sample 4 of 4</span></code></pre>
<pre><code><span class="co">## Attempting to merge in 94 aberrant cells</span></code></pre>
<pre><code><span class="co">## Backing up 10x default clonotyping in columns clonotype_id_10x and clonotype_frequency_10x before updating clonotype_id and clonotype_frequency columns</span></code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">" Nr and distribution of clonotypes using exact CDR3.aa matching \n"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  Nr and distribution of clonotypes using exact CDR3.aa matching</span></code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_id_cdr3.aa</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 321</span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_frequency_cdr3.aa</span><span class="op">)</span><span class="op">)</span> <span class="co">#Check distribution of clonotypes with identical CDR3 aa sequences</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##   1   2   3   4   5   6   7   8   9  12  24 </span>
<span class="co">## 593 118  69  24  15  36   7  16   9  12  24</span></code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonotype.html">VDJ_clonotype</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, clone.strategy <span class="op">=</span> <span class="st">"VDJJ.VJJ"</span>, global.clonotype <span class="op">=</span> <span class="cn">F</span>, VDJ.VJ.1chain <span class="op">=</span> <span class="cn">F</span>, hierarchical <span class="op">=</span> <span class="st">"double.and.single.chains"</span>, triple.chain.count.threshold <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## Clonotyping strategy: VDJJ.VJJ</span></code></pre>
<pre><code><span class="co">## Filtered out 0 cells containing more than one VDJ AND VJ chain or two VDJ chains, as these likely correspond to doublets</span></code></pre>
<pre><code><span class="co">## Found 0 exact matching clones with 3 chains and a frequency of at least 4. These will be used as high confidence clonotypes.</span></code></pre>
<pre><code><span class="co">## Processing sample 1 of 4</span></code></pre>
<pre><code><span class="co">## Attempting to merge in 9 aberrant cells</span></code></pre>
<pre><code><span class="co">## Processing sample 2 of 4</span></code></pre>
<pre><code><span class="co">## Attempting to merge in 28 aberrant cells</span></code></pre>
<pre><code><span class="co">## Processing sample 3 of 4</span></code></pre>
<pre><code><span class="co">## Attempting to merge in 23 aberrant cells</span></code></pre>
<pre><code><span class="co">## Processing sample 4 of 4</span></code></pre>
<pre><code><span class="co">## Attempting to merge in 94 aberrant cells</span></code></pre>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n Nr and distribution of clonotypes using germline gene matching \n "</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  Nr and distribution of clonotypes using germline gene matching </span>
<span class="co">## </span></code></pre>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_id_VDJJ.VJJ</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 295</span></code></pre>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_frequency_VDJJ.VJJ</span><span class="op">)</span><span class="op">)</span> <span class="co">#Check distribution of clonotypes with identical germline genes</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##   1   2   3   4   5   6   7   8  10  12  24 </span>
<span class="co">## 534 148  84  32  30  12  21  16  10  12  24</span></code></pre>
<p>Columns with new clonotype information are added to the VGM so to
allow for multiple different strategies to be compared easily. For most
Platypus downstream functions, the columns default columns
“clonotype_id” and “clonotype_frequency” are used. The VDJ_clonotype
function sets overwrites these columns, so that the last clonotyping run
is the “active” one and used by downstream functions.</p>
<p>To manually change the active clonotype, use these lines by adapting
the column names to use as input</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_id</span> <span class="op">&lt;-</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_id_cdr3.aa</span>
<span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_frequency</span> <span class="op">&lt;-</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_frequency_cdr3.aa</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extracting-full-length-sequences-from-the-vdjregion">7.3. Extracting full-length sequences from the VDJRegion<a class="anchor" aria-label="anchor" href="#extracting-full-length-sequences-from-the-vdjregion"></a>
</h3>
<p>At the start of this vignette the VGM function was run with
trim.and.align = F. Here we run the function again, but this time with
trim.and.align = T. This does take significantly longer</p>
<p>To accelerate runtime, parallel computing options are available. If
running a Windows machine parallel.processing should be set to
“parlapply” This does require the package Parallel and its dependencies.
The additional parameter numcores allows to set the numbers of cores to
use. This is important when running the function on a cluster. If on a
MAC or Linux machine, set parallel.processing to “mclapply”</p>
<p>Reference sequences are optained from the cellranger output and
thereby the 10x Genomics reference. Further trimming is made based on
annotations by cellranger (feature$region_type). Alignment is done via
Biostrings::pairwiseAlignment and the alignment with maximum score is
returned. If alignments are not complete, the vgm parameters
gap.opening.cost and gap.extension.cost can be modified</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>Data.in <span class="op">=</span> <span class="va">yermanos2021_raw</span>, <span class="co">#For local dataset input see section 5.1</span>
                               GEX.integrate <span class="op">=</span> <span class="cn">T</span>,
                               VDJ.combine <span class="op">=</span> <span class="cn">T</span>,
                               integrate.GEX.to.VDJ <span class="op">=</span> <span class="cn">T</span>,
                               integrate.VDJ.to.GEX <span class="op">=</span> <span class="cn">T</span>, 
                               trim.and.align <span class="op">=</span> <span class="cn">T</span>, <span class="co">#Set this to TRUE for full length sequence recovery</span>
                               gap.opening.cost <span class="op">=</span> <span class="fl">10</span>,
                               gap.extension.cost <span class="op">=</span> <span class="fl">4</span>, <span class="co">#Tweak to optimize alignments if neccessary</span>
                               parallel.processing <span class="op">=</span> <span class="st">"parlapply"</span><span class="op">)</span> </code></pre></div>
<p>No trimmed and reference sequence columns should be filled.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##               barcode sample_id group_id clonotype_id_10x clonotype_id</span>
<span class="co">## 1 s1_AAACGGGGTTTAGGAA        s1        1       clonotype7   clonotype7</span>
<span class="co">##   clonotype_frequency celltype Nr_of_VDJ_chains Nr_of_VJ_chains VDJ_cdr3s_aa</span>
<span class="co">## 1                   1   B cell                1               1  CARIGYAMDYW</span>
<span class="co">##   VJ_cdr3s_aa                      VDJ_cdr3s_nt</span>
<span class="co">## 1 CQQGNTLPPTF TGTGCTCGAATAGGATATGCTATGGACTACTGG</span>
<span class="co">##                         VJ_cdr3s_nt VDJ_umis VJ_umis</span>
<span class="co">## 1 TGCCAACAGGGTAATACGCTTCCTCCGACGTTC        2       8</span>
<span class="co">##              VDJ_chain_contig             VJ_chain_contig VDJ_chain VJ_chain</span>
<span class="co">## 1 AAACGGGGTTTAGGAA-1_contig_2 AAACGGGGTTTAGGAA-1_contig_1       IGH      IGK</span>
<span class="co">##   VDJ_vgene  VJ_vgene VDJ_dgene VDJ_jgene VJ_jgene VDJ_cgene VJ_cgene</span>
<span class="co">## 1   IGHV8-8 IGKV10-96               IGHJ4    IGKJ1      IGHM     IGKC</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            VDJ_sequence_nt_raw</span>
<span class="co">## 1 TGGGAAGTGTGCAGCCATGGGCAGGCTTACTTCTTCATTCCTGTTACTGATTGTCCCTGCATATGTCCTGTCCCAGGTTACTCTGAAAGAGTCTGGCCCTGGGATATTGCAGCCCTCCCAGACCCTCAGTCTGACTTGTTCTTTCTCTGGGTTTTCACTGAGCACTTTTGGTATGGGTGTAGGCTGGATTCGTCAGCCTTCAGGGAAGGGTCTGGAGTGGCTGGCACACATTTGGTGGGATGATGATAAGTACTATAACCCAGCCCTGAAGAGTCGGCTCACAATCTCCAAGGATACCTCCAAAAACCAGGTATTCCTCAAGATCGCCAATGTGGACACTGCAGATACTGCCACATACTACTGTGCTCGAATAGGATATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        VJ_sequence_nt_raw</span>
<span class="co">## 1 TTGGGGCATTGTAATTGAAGTCAAGACTCAGCCTGGACATGATGTCCTCTGCTCAGTTCCTTGGTCTCCTGTTGCTCTGTTTTCAAGGTACCAGATGTGATATCCAGATGACACAGACTACATCCTCCCTGTCTGCCTCTCTGGGAGACAGAGTCACCATCAGTTGCAGGGCAAGTCAGGACATTAGCAATTATTTAAACTGGTATCAGCAGAAACCAGATGGAACTGTTAAACTCCTGATCTACTACACATCAAGATTACACTCAGGAGTCCCATCAAGGTTCAGTGGCAGTGGGTCTGGAACAGATTATTCTCTCACCATTAGCAACCTGGAGCAAGAAGATATTGCCACTTACTTTTGCCAACAGGGTAATACGCTTCCTCCGACGTTCGGTGGAGGCACCAAGCTGGAAATCAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                    VDJ_sequence_nt_trimmed</span>
<span class="co">## 1 ATGGGCAGGCTTACTTCTTCATTCCTGTTACTGATTGTCCCTGCATATGTCCTGTCCCAGGTTACTCTGAAAGAGTCTGGCCCTGGGATATTGCAGCCCTCCCAGACCCTCAGTCTGACTTGTTCTTTCTCTGGGTTTTCACTGAGCACTTTTGGTATGGGTGTAGGCTGGATTCGTCAGCCTTCAGGGAAGGGTCTGGAGTGGCTGGCACACATTTGGTGGGATGATGATAAGTACTATAACCCAGCCCTGAAGAGTCGGCTCACAATCTCCAAGGATACCTCCAAAAACCAGGTATTCCTCAAGATCGCCAATGTGGACACTGCAGATACTGCCACATACTACTGTGCTCGAATAGGATATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCA</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                          VJ_sequence_nt_trimmed</span>
<span class="co">## 1 ATGATGTCCTCTGCTCAGTTCCTTGGTCTCCTGTTGCTCTGTTTTCAAGGTACCAGATGTGATATCCAGATGACACAGACTACATCCTCCCTGTCTGCCTCTCTGGGAGACAGAGTCACCATCAGTTGCAGGGCAAGTCAGGACATTAGCAATTATTTAAACTGGTATCAGCAGAAACCAGATGGAACTGTTAAACTCCTGATCTACTACACATCAAGATTACACTCAGGAGTCCCATCAAGGTTCAGTGGCAGTGGGTCTGGAACAGATTATTCTCTCACCATTAGCAACCTGGAGCAAGAAGATATTGCCACTTACTTTTGCCAACAGGGTAATACGCTTCCTCCGACGTTCGGTGGAGGCACCAAGCTGGAAATCAAA</span>
<span class="co">##                                                                                                                            VDJ_sequence_aa</span>
<span class="co">## 1 MGRLTSSFLLLIVPAYVLSQVTLKESGPGILQPSQTLSLTCSFSGFSLSTFGMGVGWIRQPSGKGLEWLAHIWWDDDKYYNPALKSRLTISKDTSKNQVFLKIANVDTADTATYYCARIGYAMDYWGQGTSVTVSS</span>
<span class="co">##                                                                                                                    VJ_sequence_aa</span>
<span class="co">## 1 MMSSAQFLGLLLLCFQGTRCDIQMTQTTSSLSASLGDRVTISCRASQDISNYLNWYQQKPDGTVKLLIYYTSRLHSGVPSRFSGSGSGTDYSLTISNLEQEDIATYFCQQGNTLPPTFGGGTKLEIK</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        VDJ_raw_ref</span>
<span class="co">## 1 ATGGGCAGGCTTACTTCTTCATTCCTGTTACTGATTGTCCCTGCATATGTCCTGTCCCAGGTTACTCTGAAAGAGTCTGGCCCTGGGATATTGCAGCCCTCCCAGACCCTCAGTCTGACTTGTTCTTTCTCTGGGTTTTCACTGAGCACTTTTGGTATGGGTGTAGGCTGGATTCGTCAGCCTTCAGGGAAGGGTCTGGAGTGGCTGGCACACATTTGGTGGGATGATGATAAGTACTATAACCCAGCCCTGAAGAGTCGGCTCACAATCTCCAAGGATACCTCCAAAAACCAGGTATTCCTCAAGATCGCCAATGTGGACACTGCAGATACTGCCACATACTACTGTGCTCGAATAGATTACTATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCAGGTGTTGCTGTCTCCCAAGAGCATCCTTGAAGGTTCAGATGAATACCTGGTATGCAAAATCCACTACGGAGGCAAAAACAAAGATCTGCATGTGCCCATTCCAGCTGTCGCAGAGATGAACCCCAATGTAAATGTGTTCGTCCCACCACGGGATGGCTTCTCTGGCCCTGCACCACGCAAGTCTAAACTCATCTGCGAGGCCACGAACTTCACTCCAAAACCGATCACAGTATCCTGGCTAAAGGATGGGAAGCTCGTGGAATCTGGCTTCACCACAGATCCGGTGACCATCGAGAACAAAGGATCCACACCCCAAACCTACAAGGTCATAAGCACACTTACCATCTCTGAAATCGACTGGCTGAACCTGAATGTGTACACCTGCCGTGTGGATCACAGGGGTCTCACCTTCTTGAAGAACGTGTCCTCCACATGTGCTGCCAGTCCCTCCACAGACATCCTAACCTTCACCATCCCCCCCTCCTTTGCCGACATCTTCCTCAGCAAGTCCGCTAACCTGACCTGTCTGGTCTCAAACCTGGCAACCTATGAAACCCTGAATATCTCCTGGGCTTCTCAAAGTGGTGAACCACTGGAAACCAAAATTAAAATCATGGAAAGCCATCCCAATGGCACCTTCAGTGCTAAGGGTGTGGCTAGTGTTTGTGTGGAAGACTGGAATAACAGGAAGGAATTTGTGTGTACTGTGACTCACAGGGATCTGCCTTCACCACAGAAGAAATTCATCTCAAAACCCAATGAGGTGCACAAACATCCACCTGCTGTGTACCTGCTGCCACCAGCTCGTGAGCAACTGAACCTGAGGGAGTCAGCCACAGTCACCTGCCTGGTGAAGGGCTTCTCTCCTGCAGACATCAGTGTGCAGTGGCTTCAGAGAGGGCAACTCTTGCCCCAAGAGAAGTATGTGACCAGTGCCCCGATGCCAGAGCCTGGGGCCCCAGGCTTCTACTTTACCCACAGCATCCTGACTGTGACAGAGGAGGAATGGAACTCCGGAGAGACCTATACCTGTGTTGTAGGCCACGAGGCCCTGCCACACCTGGTGACCGAGAGGACCGTGGACAAGTCCACTGGTAAACCCACACTGTACAATGTCTCCCTGATCATGTCTGACACAGGCGGCACCTGCTAT</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          VJ_raw_ref</span>
<span class="co">## 1 ATGATGTCCTCTGCTCAGTTCCTTGGTCTCCTGTTGCTCTGTTTTCAAGGTACCAGATGTGATATCCAGATGACACAGACTACATCCTCCCTGTCTGCCTCTCTGGGAGACAGAGTCACCATCAGTTGCAGGGCAAGTCAGGACATTAGCAATTATTTAAACTGGTATCAGCAGAAACCAGATGGAACTGTTAAACTCCTGATCTACTACACATCAAGATTACACTCAGGAGTCCCATCAAGGTTCAGTGGCAGTGGGTCTGGAACAGATTATTCTCTCACCATTAGCAACCTGGAGCAAGAAGATATTGCCACTTACTTTTGCCAACAGGGTAATACGCTTCCTCCGTGGACGTTCGGTGGAGGCACCAAGCTGGAAATCAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTCTTGAACAACTTCTACCCCAAAGACATCAATGTCAAGTGGAAGATTGATGGCAGTGAACGACAAAATGGCGTCCTGAACAGTTGGACTGATCAGGACAGCAAAGACAGCACCTACAGCATGAGCAGCACCCTCACGTTGACCAAGGACGAGTATGAACGACATAACAGCTATACCTGTGAGGCCACTCACAAGACATCAACTTCACCCATTGTCAAGAGCTTCAACAGGAATGAGTGT</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                               VDJ_trimmed_ref</span>
<span class="co">## 1 ATGGGCAGGCTTACTTCTTCATTCCTGTTACTGATTGTCCCTGCATATGTCCTGTCCCAGGTTACTCTGAAAGAGTCTGGCCCTGGGATATTGCAGCCCTCCCAGACCCTCAGTCTGACTTGTTCTTTCTCTGGGTTTTCACTGAGCACTTTTGGTATGGGTGTAGGCTGGATTCGTCAGCCTTCAGGGAAGGGTCTGGAGTGGCTGGCACACATTTGGTGGGATGATGATAAGTACTATAACCCAGCCCTGAAGAGTCGGCTCACAATCTCCAAGGATACCTCCAAAAACCAGGTATTCCTCAAGATCGCCAATGTGGACACTGCAGATACTGCCACATACTACTGTGCTCGAATAGATTACTATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCA</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                     VJ_trimmed_ref</span>
<span class="co">## 1 ATGATGTCCTCTGCTCAGTTCCTTGGTCTCCTGTTGCTCTGTTTTCAAGGTACCAGATGTGATATCCAGATGACACAGACTACATCCTCCCTGTCTGCCTCTCTGGGAGACAGAGTCACCATCAGTTGCAGGGCAAGTCAGGACATTAGCAATTATTTAAACTGGTATCAGCAGAAACCAGATGGAACTGTTAAACTCCTGATCTACTACACATCAAGATTACACTCAGGAGTCCCATCAAGGTTCAGTGGCAGTGGGTCTGGAACAGATTATTCTCTCACCATTAGCAACCTGGAGCAAGAAGATATTGCCACTTACTTTTGCCAACAGGGTAATACGCTTCCTCCGTGGACGTTCGGTGGAGGCACCAAGCTGGAAATCAAA</span>
<span class="co">##      VDJ_raw_consensus_id     VJ_raw_consensus_id     orig_barcode specifity</span>
<span class="co">## 1 clonotype7_concat_ref_1 clonotype7_concat_ref_2 AAACGGGGTTTAGGAA        NA</span>
<span class="co">##   affinity GEX_available orig.ident seurat_clusters PC_1 PC_2 UMAP_1 UMAP_2</span>
<span class="co">## 1       NA         FALSE       &lt;NA&gt;            &lt;NA&gt;   NA   NA     NA     NA</span>
<span class="co">##   tSNE_1 tSNE_2</span>
<span class="co">## 1     NA     NA</span>
<span class="co">##                                                                            batch_id</span>
<span class="co">## 1 N:/reddy/Shared_Folders/AA_AY/DBmanuscript_Aged_CNS/GEX/Aged.CNS.pool.3m.Bcell.S1</span></code></pre>
<p>A different way to get germline reference sequences is by using
MIXCR. MIXCR can be called directly from R on both Windows and MAC
machines. Given that output files have to be read in, it is important to
set the working directory correctly and (Essential for Windows users)
have a version of MIXCR available in that working directory.</p>
<p>Moreover to quantify the number of somatic variants or to extract
full-length sequences for expression, it is often useful to have the
nucleotide sequence from framework region 1 (FR1) to framework region 4
(FR4). Using the call_MIXCR function, the full-length VDJRegion
sequences can be added to the clonal information and easily extracted
thereafter. This function works on UNIX/Mac and furthermore requires
that mixcr is already downloaded locally (with license agreement).</p>
<p>FOR MAC/UNIX users</p>
<p>One needs to supply the directory to the executable in the call_MIXCR
function as below. Either “mmu” or “hsa” for mouse and human,
respectively. Again, the format is similar to the input, in that the
outer list corresponds to the individual repertoire and the inner list
is a dataframe with various information, including the full-length VDJ
sequences (e.g., VDJ.AA.LC and VDJ.AA.HC for the light and heavy chain
amino acid sequence). One will notice that the germline sequence is
still very long (e.g., in the example below the “full_HC_germline”
length is over 600 nucleotides). This will be filled in using the
separate function VDJ_extract_germline.</p>
<p>For MAC users</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### WARNING: You will need to download MiXCR and change the mixcr.directory to the folder which contains the mixcr.jar executable</span>

<span class="co">#VDJ_mixcr_out &lt;- VDJ_call_MIXCR(VDJ = vgm[[1]], mixcr.directory = "~/Downloads/Directory-with-MiXcr/",species = "mmu", platypus.version = "v3", operating.system = "Darwin", simplify = T)</span>

<span class="co">#set simplify to T to append only a selected columns of the MIXCR output to the vgm matrix</span></code></pre></div>
<p>FOR Windows users The mixcr.jar executable needs to be in the current
working directory!</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#check working directory</span>
<span class="co">#getwd()</span>
<span class="co">### WARNING: You will need to download MiXCR and have it in your current working directory</span>
<span class="va">VDJ_mixcr_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_call_MIXCR.html">VDJ_call_MIXCR</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, mixcr.directory <span class="op">=</span> <span class="st">"Is set automatically to current working directory"</span>,species <span class="op">=</span> <span class="st">"mmu"</span>, platypus.version <span class="op">=</span> <span class="st">"v3"</span>, operating.system <span class="op">=</span> <span class="st">"Windows"</span>, simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="co">#set simplify to T to append only a selected columns of the MIXCR output to the vgm matrix</span>
<span class="co">#set to False to save as separate object with the complete MIXCR output as in this case</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plotting-shms">7.4. Plotting SHMs<a class="anchor" aria-label="anchor" href="#plotting-shms"></a>
</h3>
<p>Directly from the MIXCR output we can plot the frequency of somatic
hypermutations and run basic tests of significance between groups</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/VDJ_plot_SHM.html">VDJ_plot_SHM</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">VDJ_mixcr_out</span>, group.by <span class="op">=</span> <span class="st">"sample_id"</span>, quantile.label <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [[1]]</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## [[2]]</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## [[3]]</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## [[4]]</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## [[5]]</span></code></pre>
<pre><code><span class="co">## Warning: ggrepel: 1 unlabeled data points (too many overlaps). Consider</span>
<span class="co">## increasing max.overlaps</span></code></pre>
<p><img src="quickstart_files/figure-html/unnamed-chunk-34-1.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-34-2.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-34-3.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-34-4.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-34-5.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="visualizing-clonal-frequencies-v3-only">7.5. Visualizing clonal frequencies (V3 only)<a class="anchor" aria-label="anchor" href="#visualizing-clonal-frequencies-v3-only"></a>
</h3>
<p>For a basic view of clonal expansion the VDJ_clonal_donut produces
circular plots per sample. Label position and size should only be
adjusted after deciding on a plot export format. This function uses the
clonotype_id column as input. If VDJ_clonotyping function was used, its
result are stored in that column. If not, the default 10x clonotyping is
stored there. To retrieve default 10x clonotyping, use the column
clonotyping_id_10x</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">donuts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_donut.html">VDJ_clonal_donut</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, expanded.colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey50"</span>, <span class="st">"grey65"</span>, <span class="st">"grey80"</span><span class="op">)</span>, non.expanded.color <span class="op">=</span> <span class="st">"black"</span>, counts.to.use <span class="op">=</span> <span class="st">"freq_column"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning in VDJ_clonal_donut(VDJ = vgm[[1]], expanded.colors = c("grey50", :</span>
<span class="co">## Column name for counts.to.use was not found in VDJ. Defaulting to</span>
<span class="co">## 'clonotype_id_10x'</span></code></pre>
<pre><code><span class="co">## Using column clonotype_id for counting clones</span></code></pre>
<pre><code><span class="co">## Clones: Expanded: 5 / 6.85%; 1 cell 68 / 93.15%; total: 73</span></code></pre>
<pre><code><span class="co">## Cells: Expanded: 10 / 12.82%; 1 cell 68 / 87.18%; Total: 78</span></code></pre>
<pre><code><span class="co">## Clones: Expanded: 14 / 10.85%; 1 cell 115 / 89.15%; total: 129</span></code></pre>
<pre><code><span class="co">## Cells: Expanded: 45 / 28.12%; 1 cell 115 / 71.88%; Total: 160</span></code></pre>
<pre><code><span class="co">## Clones: Expanded: 7 / 3.5%; 1 cell 193 / 96.5%; total: 200</span></code></pre>
<pre><code><span class="co">## Cells: Expanded: 15 / 7.21%; 1 cell 193 / 92.79%; Total: 208</span></code></pre>
<pre><code><span class="co">## Clones: Expanded: 64 / 18.34%; 1 cell 285 / 81.66%; total: 349</span></code></pre>
<pre><code><span class="co">## Cells: Expanded: 214 / 42.89%; 1 cell 285 / 57.11%; Total: 499</span></code></pre>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Counts to use = "freq_column" uses the counts in the clonotype_frequency column. Counts to use = "vgm" simply counts the rows of a given clonotype in the VGM table (these counts may differ if cells have been filtered out due to overlapping barcodes or if another clonotyping strategy was used)</span>
<span class="va">donuts</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-35-1.png" width="672"></p>
<p>The black section shows cells in unexpanded clones. The rest shows
expanded clones by their frequency. The middle label shows the total
number of clones (cells)</p>
</div>
<div class="section level3">
<h3 id="calculating-common-repertoire-diversity-metrics">7.6. Calculating common repertoire diversity metrics<a class="anchor" aria-label="anchor" href="#calculating-common-repertoire-diversity-metrics"></a>
</h3>
<p>This requires the Vegan package. We can calculate the diversity for
any column(s) in the vgm. If more than one column is provided, the
content will be pasted together before calculating diversity metrics</p>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Shannon Evenness for the VDJ chain CDR3</span>
<span class="va">diversity_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_diversity.html">VDJ_diversity</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,feature.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_cdr3s_aa"</span><span class="op">)</span>,grouping.column <span class="op">=</span> <span class="st">"sample_id"</span>,metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"shannonevenness"</span><span class="op">)</span>, subsample.to.same.n <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="va">diversity_plot</span>
<span class="co">#Gini-Simpson index for pasted VDJ and VJ chain CDR3s</span>
<span class="va">diversity_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_diversity.html">VDJ_diversity</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,feature.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_cdr3s_aa"</span>, <span class="st">"VJ_cdr3s_aa"</span><span class="op">)</span>,grouping.column <span class="op">=</span> <span class="st">"sample_id"</span>,metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ginisimpson"</span><span class="op">)</span>, subsample.to.same.n <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="va">diversity_plot</span>
<span class="co">#exact values can be retrived as </span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">diversity_plot</span><span class="op">$</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 4 x 4</span></span>
<span class="co">##   group metric metric_name        colors </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> s1    0.015<span style="text-decoration: underline;">1</span> Gini-Simpson index #FF0000</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> s2    0.025<span style="text-decoration: underline;">5</span> Gini-Simpson index #80FF00</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> s3    0.015<span style="text-decoration: underline;">1</span> Gini-Simpson index #00FFFF</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> s4    0.020<span style="text-decoration: underline;">8</span> Gini-Simpson index #8000FF</span></code></pre>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Jaccard index between repertoires of the two samples</span>
<span class="va">diversity_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_diversity.html">VDJ_diversity</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,feature.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_cgene"</span><span class="op">)</span>,grouping.column <span class="op">=</span> <span class="st">"sample_id"</span>,metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span><span class="op">)</span>, subsample.to.same.n <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="va">diversity_plot</span> </code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-36-1.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-36-2.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-36-3.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="examining-isotypes">7.7. Examining isotypes<a class="anchor" aria-label="anchor" href="#examining-isotypes"></a>
</h3>
<p>Additional VDJ functions include plotting the clonal expansion for
each clone. This can be performed by the VDJ_clonal_expansion function
and setting the number of clones to show.</p>
<p>The function has two operating modes: Set color.by to “isotype” to
exctract and color bars by isotype. Set subtypes to TRUE to view IgG
subtypes. Set color.by to any other column in vgm[[1]] to examine the
distribution of other parameters (This will be used during VDJ - GEX
integration)</p>
<p>The following code extracts and plots the isotype distribution of the
top 30 clones. We can see a clear IgM isotype majority of the top four
clones in the first patient when using the default clonotyping strategy.
We can additionally use the new clonotyping strategies to compare how
changing the clonal defintion impacts the clonal expansion profiles. We
simply supply the output from VDJ_clonotype</p>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clonal_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_expansion.html">VDJ_clonal_expansion</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,celltype <span class="op">=</span> <span class="st">"Bcells"</span>,clones <span class="op">=</span> <span class="st">"30"</span>, group.by <span class="op">=</span> <span class="st">"sample_id"</span>, color.by <span class="op">=</span> <span class="st">"isotype"</span>, isotypes.to.plot <span class="op">=</span> <span class="st">"all"</span>, treat.incomplete.clones <span class="op">=</span> <span class="st">"exclude"</span>, treat.incomplete.cells <span class="op">=</span> <span class="st">"proportional"</span><span class="op">)</span>
<span class="co">#group by specifies how many separate plots should be generated. If vgm contains global clonotype information this can be set to "none"</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">clonal_out</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co">#Plot sample 1</span>
<span class="co">#print(clonal_out[[2]][[1]]) #Formatted data for plot sample 1</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-37-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="sequence-similarity-networks">7.8. Sequence similarity networks<a class="anchor" aria-label="anchor" href="#sequence-similarity-networks"></a>
</h3>
<p>Other functions are specifically tailored to repertoire analysis -
such as VDJ_network, which creates a sequence similarity network between
repertoires or within a repertoire by connecting those clones with
sequence similarity. This function relies upon igraph to visually
display and construct the graph - which means that networks with high
number of sequences will not display easily. In the following example we
are using a small dataset. In case of a bigger dataset one may subsample
by using the sample.n function on vgm[[1]]. Setting the per.mouse
argument to false indicates that one network for multiple repertoires
should be produced.</p>
<p>This function is quite specialized and we are working on building a
better networks module. But hopefully the code can be used as a template
for your own analyses.</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#subsampled_VGM &lt;- dplyr::sample_n(vgm[[1]], 60)</span>
<span class="va">cl_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_network.html">VDJ_network</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,per.sample <span class="op">=</span> <span class="cn">T</span>,distance.cutoff <span class="op">=</span> <span class="fl">8</span>, platypus.version <span class="op">=</span> <span class="st">"v3"</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">{</span>
<span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/plot.igraph.html" class="external-link">plot.igraph</a></span><span class="op">(</span><span class="va">cl_network</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,vertex.label<span class="op">=</span><span class="cn">NA</span>,vertex.size<span class="op">=</span><span class="fl">7</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#For a plot including clonotype frequencies</span>
<span class="va">cl_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_network.html">VDJ_network</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,per.sample <span class="op">=</span> <span class="cn">F</span>,distance.cutoff <span class="op">=</span> <span class="fl">8</span>, platypus.version <span class="op">=</span> <span class="st">"v3"</span><span class="op">)</span>
<span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/plot.igraph.html" class="external-link">plot.igraph</a></span><span class="op">(</span><span class="va">cl_network</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>,vertex.label<span class="op">=</span><span class="cn">NA</span>,vertex.size<span class="op">=</span><span class="fl">3</span><span class="op">+</span><span class="op">(</span><span class="fl">0.95</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">cl_network</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_frequency</span><span class="op">)</span><span class="op">)</span>,vertex.color<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-38-1.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-38-2.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-38-3.png" width="672"></p>
<p>For more details see the documentation of the VDJ_network function,
but essentially information such as clonal frequency and which sample
(here still indicated by the “sample_id” column) are stored in the
second element of the output list. Here we can see only a few clones
that are showing connections (produced by edges between those with a
distance of 8 amino acids or less between heavy and light chain paired
CDR3 sequence homology)</p>
</div>
<div class="section level3">
<h3 id="germline-gene-usage-heatmaps">7.9. Germline gene usage heatmaps<a class="anchor" aria-label="anchor" href="#germline-gene-usage-heatmaps"></a>
</h3>
<p>It is also possible to produce heatmaps of the germline gene usage in
the context of heavy chain V gene and light chain V gene. The output of
the VDJ_Vgene_usage function is a matrix for each repertoire
corresponding to the order specified by VDJ_analyze. The outer list
corresponds to the sample and the inner list corresponds to a matrix,
where the rows correspond to the heavy chain V genes and the columns
correspond to the light chains of the V genes. Therefore the
output[[1]][i,j] corresponds to the number of clones using the
combination of IGH-Vgene[i] and IGK/L-Vgene[j].</p>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#First calculate adjacency matrix for V gene usage</span>
<span class="va">Vgene_usage</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_Vgene_usage.html">VDJ_Vgene_usage</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, platypus.version <span class="op">=</span> <span class="st">"v3"</span><span class="op">)</span>
<span class="co">#library(pheatmap)</span>
<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">Vgene_usage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,show_rownames <span class="op">=</span> <span class="cn">F</span>,show_colnames <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">Vgene_usage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "matrix" "array"</span></code></pre>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Vgene_usage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "IGHV8-8"  "IGHV1-9"  "IGHV1-26" "IGHV14-1" "IGHV5-17" "IGHV5-16"</span></code></pre>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Vgene_usage</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "IGKV10-96" "IGKV6-32"  "IGKV19-93" "IGKV8-24"  "IGKV1-110" "IGKV4-55"</span></code></pre>
<p><img src="quickstart_files/figure-html/unnamed-chunk-39-1.png" width="672"></p>
<p>This can then be easily plotted as a heatmap to observe patterns
between repertoires or can be used to calculate V gene correlation using
the “pheatmap” package.</p>
<p>Platypus also allows a separate analysis of V gene usage for HC and
LC. The VDJ_Vgene_usage_barplot allows the user to plot most frequently
used IgH or IgK/L V genes. By default, this function only provides
visualizations for the HC V genes, but can also provide for the LC if
LC. Vgene is set to TRUE. The User can also select the number of most
used genes to be depicted.</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Vgene_usage_barplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_Vgene_usage_barplot.html">VDJ_Vgene_usage_barplot</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, HC.gene.number <span class="op">=</span> <span class="fl">10</span>, LC.Vgene <span class="op">=</span> <span class="cn">T</span>, LC.gene.number <span class="op">=</span> <span class="fl">10</span>, platypus.version <span class="op">=</span> <span class="st">"v3"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Sample order: s1 ; s2 ; s3 ; s4</span></code></pre>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Vgene_usage_barplot</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#VDJ chains</span>
<span class="va">Vgene_usage_stackedbarplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_Vgene_usage_stacked_barplot.html">VDJ_Vgene_usage_stacked_barplot</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, LC.Vgene <span class="op">=</span> <span class="cn">F</span>,HC.gene.number <span class="op">=</span> <span class="fl">10</span>, Fraction.HC <span class="op">=</span> <span class="fl">1</span>, platypus.version <span class="op">=</span> <span class="st">"v3"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Sample order: s1 ; s2 ; s3 ; s4</span></code></pre>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Vgene_usage_stackedbarplot</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-40-1.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-40-2.png" width="672"></p>
<p>Furthermore, we can also produce a circular visualization of how V
and J genes are combined throughout the repertoire. In the example that
follows we use VDJ_VJ_usage_circos to look at the V gene with the
corresponding J gene for each expanded clonotype.</p>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#VDJ and VJ V and J gene pairing</span>
<span class="va">vj_circos_bcells</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_VJ_usage_circos.html">VDJ_VJ_usage_circos</a></span><span class="op">(</span>VGM <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, c.threshold <span class="op">=</span> <span class="fl">5</span>,label.threshold<span class="op">=</span><span class="fl">5</span>,cell.level <span class="op">=</span> <span class="cn">T</span>, platypus.version <span class="op">=</span> <span class="st">"v3"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Chosen clonotype column:  clonotype_id_10x</span></code></pre>
<pre><code><span class="co">## Chosen clonotype.frequency column:  clonotype_frequency_10x</span></code></pre>
<pre><code><span class="co">## Chosen clonotype.frequency column:  clonotype_frequency</span></code></pre>
<pre><code><span class="co">## Processing sample 1</span></code></pre>
<pre><code><span class="co">## Processing sample 2</span></code></pre>
<div class="sourceCode" id="cb147"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#VDJ and VJ pairing </span>
<span class="va">HL_circos_bcells</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_alpha_beta_Vgene_circos.html">VDJ_alpha_beta_Vgene_circos</a></span><span class="op">(</span>VGM <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, c.threshold <span class="op">=</span> <span class="fl">5</span>,label.threshold<span class="op">=</span><span class="fl">5</span>,cell.level <span class="op">=</span> <span class="cn">T</span>, V.or.J<span class="op">=</span> <span class="st">"both"</span>, platypus.version <span class="op">=</span> <span class="st">"v3"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Chosen clonotype.frequency column:  clonotype_frequency</span></code></pre>
<pre><code><span class="co">## [1] "s2"</span>
<span class="co">## [1] "s4"</span></code></pre>
<pre><code><span class="co">## Plotting for sample s2</span></code></pre>
<pre><code><span class="co">## Plotting for sample s4</span></code></pre>
<p><img src="quickstart_files/figure-html/unnamed-chunk-41-1.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-41-2.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-41-3.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-41-4.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="assessing-cdr3-sequence-similarity">7.10. Assessing CDR3 sequence similarity<a class="anchor" aria-label="anchor" href="#assessing-cdr3-sequence-similarity"></a>
</h3>
<p>Finally, one can also look at any specific HC and LC CDR3 amino acid
patterns arising across the different clones. Using the
VDJ_logoplot_vector function the user can plot a logoplot of the CDR3
region od a certain length, specifyed by the length_cdr3 argument. For
instance, the logoplot below corresponds to all those CDR3 aminoacid
sequences of length 25</p>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasted_CDR3s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VDJ_cdr3s_aa</span>, <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VJ_cdr3s_aa</span><span class="op">)</span>
<span class="va">agedCNS_CDR3_logoplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_logoplot_vector.html">VDJ_logoplot_vector</a></span><span class="op">(</span>cdr3.vector <span class="op">=</span> <span class="va">pasted_CDR3s</span>, seq_type <span class="op">=</span> <span class="st">"aa"</span>, length_cdr3 <span class="op">=</span> <span class="st">"auto"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Returning logoplot based on 106 sequences of a total 945 input sequences</span></code></pre>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">agedCNS_CDR3_logoplot</span><span class="op">)</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-42-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="assessing-nr-of-variants-per-clone">7.11. Assessing Nr of variants per clone<a class="anchor" aria-label="anchor" href="#assessing-nr-of-variants-per-clone"></a>
</h3>
<p>Depending on the clonotype strategy, within a clone several different
variants of CDR3s or full length sequences may be contained. To count
variants, the function VDJ_variants_per_clone is used. It returns a
table per sample_id with stats about variants</p>
<div class="sourceCode" id="cb155"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">variants_agedCNS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_variants_per_clone.html">VDJ_variants_per_clone</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, variants.of <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_cdr3s_aa"</span>, <span class="st">"VJ_cdr3s_aa"</span><span class="op">)</span>, clonotypes.col <span class="op">=</span> <span class="st">"clonotype_id_10x"</span>, split.by <span class="op">=</span> <span class="st">"sample_id"</span>, stringDist.method <span class="op">=</span> <span class="st">"levenshtein"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">variants_agedCNS</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##   split_group clonotype_id n sequences n variants mean variants distance</span>
<span class="co">## 1          s1   clonotype7           1          1                     NA</span>
<span class="co">## 2          s1  clonotype58           1          1                     NA</span>
<span class="co">## 3          s1  clonotype65           1          1                     NA</span>
<span class="co">## 4          s1  clonotype43           1          1                     NA</span>
<span class="co">## 5          s1  clonotype33           1          1                     NA</span>
<span class="co">## 6          s1  clonotype49           1          1                     NA</span>
<span class="co">##   max variants distance min variants distance         most frequent variant</span>
<span class="co">## 1                    NA                    NA       CARIGYAMDYW/CQQGNTLPPTF</span>
<span class="co">## 2                    NA                    NA CARDFTTVVARGYFDVW/CQQDYSSPWTF</span>
<span class="co">## 3                    NA                    NA CARGITTVVAYYYAMDYW/CLQYDNLYTF</span>
<span class="co">## 4                    NA                    NA    CTTGPYDYYAMDYW/CQQHYSTPYTF</span>
<span class="co">## 5                    NA                    NA    CARPHDYDGVDYW/CSQSTHVPPWTF</span>
<span class="co">## 6                    NA                    NA   CARRYYSNYAWFAYW/CQQWSSYPYTF</span>
<span class="co">##   n VDJ cgene variants most frequent VDJ cgene</span>
<span class="co">## 1                    1                    IGHM</span>
<span class="co">## 2                    1                    IGHD</span>
<span class="co">## 3                    1                    IGHD</span>
<span class="co">## 4                    1                    IGHD</span>
<span class="co">## 5                    1                    IGHM</span>
<span class="co">## 6                    1                    IGHM</span></code></pre>
<div class="sourceCode" id="cb157"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#set split.by to "none" if clonotyping was conducted across all samples</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="searching-for-overlap-between-repertoires">7.12. Searching for overlap between repertoires<a class="anchor" aria-label="anchor" href="#searching-for-overlap-between-repertoires"></a>
</h3>
<p>Public clones or sequences are sequences that are shared between
individuals. The VDJ_overlap_heatmap function can quantify this overlap
and return a list of overlapping items</p>
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#overlap of VDJ V genes</span>
<span class="va">VDJv_overlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_overlap_heatmap.html">VDJ_overlap_heatmap</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,feature.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_vgene"</span><span class="op">)</span> ,grouping.column <span class="op">=</span> <span class="st">"sample_id"</span>, axis.label.size <span class="op">=</span> <span class="fl">20</span>, pvalues.label.size <span class="op">=</span> <span class="fl">12</span>, add.barcode.table <span class="op">=</span> <span class="cn">T</span>, plot.type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">VDJv_overlap</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co">#summary of overlap</span></code></pre></div>
<pre><code><span class="co">##   V1 V2 overlap</span>
<span class="co">## 1 s1 s2      30</span>
<span class="co">## 2 s1 s3      33</span>
<span class="co">## 3 s1 s4      38</span>
<span class="co">## 4 s2 s3      40</span>
<span class="co">## 5 s2 s4      47</span>
<span class="co">## 6 s3 s4      53</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                    items.overlapping</span>
<span class="co">## 1                                                                                                                                                                                                         IGHV8-8;IGHV1-9;IGHV1-26;IGHV5-17;IGHV5-16;IGHV1-54;IGHV1-72;IGHV1-64;IGHV5-9-1;IGHV1-19;IGHV1-18;IGHV8-5;IGHV3-6;IGHV1-52;IGHV1-80;IGHV1-74;IGHV1-39;IGHV1-22;IGHV10-1;IGHV1-62-2;IGHV9-3;IGHV1-53;IGHV1-69;IGHV9-1;IGHV1-76;IGHV1-50;IGHV1-81;IGHV1-12;IGHV1-82;IGHV1-85</span>
<span class="co">## 2                                                                                                                                                                               IGHV8-8;IGHV1-9;IGHV1-26;IGHV5-17;IGHV5-16;IGHV1-54;IGHV1-72;IGHV1-64;IGHV5-9-1;IGHV1-71;IGHV1-19;IGHV1-18;IGHV1-59;IGHV8-5;IGHV3-6;IGHV1-52;IGHV1-80;IGHV1-74;IGHV1-39;IGHV10-1;IGHV1-62-2;IGHV9-3;IGHV2-5;IGHV1-53;IGHV1-69;IGHV1-76;IGHV1-81;IGHV1-12;IGHV1-66;IGHV14-2;IGHV1-4;IGHV1-82;IGHV1-85</span>
<span class="co">## 3                                                                                                                                   IGHV8-8;IGHV1-9;IGHV1-26;IGHV14-1;IGHV5-17;IGHV5-16;IGHV1-54;IGHV1-72;IGHV1-64;IGHV5-9-1;IGHV2-3;IGHV1-71;IGHV1-19;IGHV1-18;IGHV1-75;IGHV1-59;IGHV4-1;IGHV3-6;IGHV1-52;IGHV1-80;IGHV1-74;IGHV1-39;IGHV1-22;IGHV10-1;IGHV1-62-2;IGHV9-3;IGHV2-5;IGHV1-53;IGHV1-69;IGHV9-1;IGHV1-76;IGHV1-50;IGHV1-81;IGHV1-12;IGHV1-66;IGHV14-2;IGHV1-82;IGHV1-85</span>
<span class="co">## 4                                                                                                                   IGHV5-9-1;IGHV5-16;IGHV9-3;IGHV1-12;IGHV1-39;IGHV14-3;IGHV1-9;IGHV1-82;IGHV1-26;IGHV10-3;IGHV7-3;IGHV1-19;IGHV1-74;IGHV1-76;IGHV3-6;IGHV1-53;IGHV1-69;IGHV1-85;IGHV1-81;IGHV1-18;IGHV1-54;IGHV8-8;IGHV1-15;IGHV2-9-1;IGHV1-64;IGHV1-52;IGHV5-12;IGHV10-1;IGHV1-78;IGHV5-4;IGHV6-3;IGHV6-6;IGHV5-17;IGHV1-72;IGHV1-55;IGHV1-80;IGHV8-5;IGHV5-9;IGHV2-2;IGHV1-62-2</span>
<span class="co">## 5                                                     IGHV5-9-1;IGHV5-16;IGHV9-3;IGHV1-50;IGHV1-12;IGHV1-39;IGHV14-3;IGHV1-2;IGHV8-12;IGHV1-9;IGHV1-82;IGHV1-58;IGHV1-36;IGHV1-26;IGHV10-3;IGHV7-3;IGHV1-19;IGHV1-74;IGHV1-76;IGHV3-6;IGHV1-53;IGHV1-69;IGHV1-85;IGHV1-81;IGHV1-18;IGHV1-54;IGHV8-8;IGHV1-15;IGHV2-9-1;IGHV1-22;IGHV1-64;IGHV1-52;IGHV1-42;IGHV5-12;IGHV10-1;IGHV1-78;IGHV5-4;IGHV6-3;IGHV6-6;IGHV5-17;IGHV9-1;IGHV1-72;IGHV1-55;IGHV1-80;IGHV2-2;IGHV2-6;IGHV1-62-2</span>
<span class="co">## 6 IGHV1-55;IGHV1-9;IGHV1-71;IGHV1-64;IGHV1-53;IGHV1-82;IGHV7-1;IGHV1-18;IGHV5-4;IGHV1-52;IGHV1-26;IGHV1-7;IGHV5-9-1;IGHV2-9-1;IGHV5-12;IGHV13-2;IGHV1-72;IGHV1-69;IGHV1-19;IGHV10-3;IGHV1-81;IGHV1-80;IGHV6-3;IGHV1-5;IGHV1-15;IGHV6-6;IGHV1-84;IGHV5-16;IGHV9-3;IGHV1-78;IGHV1-66;IGHV2-5;IGHV3-6;IGHV1-12;IGHV1-39;IGHV7-3;IGHV1-62-2;IGHV1-74;IGHV10-1;IGHV1-59;IGHV8-8;IGHV1-77;IGHV15-2;IGHV14-3;IGHV14-2;IGHV5-17;IGHV1-61;IGHV3-1;IGHV1-54;IGHV2-2;IGHV1-63;IGHV1-76;IGHV1-85</span>
<span class="co">##   overlap_lab</span>
<span class="co">## 1          30</span>
<span class="co">## 2          33</span>
<span class="co">## 3          38</span>
<span class="co">## 4          40</span>
<span class="co">## 5          47</span>
<span class="co">## 6          53</span></code></pre>
<div class="sourceCode" id="cb160"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">VDJv_overlap</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co">#Table of overlapping items</span></code></pre></div>
<pre><code><span class="co">##   overlapping_items Freq in s1</span>
<span class="co">## 1           IGHV8-8          1</span>
<span class="co">## 2           IGHV1-9          3</span>
<span class="co">## 3          IGHV1-26         12</span>
<span class="co">## 4          IGHV5-17          1</span>
<span class="co">## 5          IGHV5-16          2</span>
<span class="co">## 6          IGHV1-54          1</span>
<span class="co">##                                                                                                                                                                                                                                       Barcodes s1</span>
<span class="co">## 1                                                                                                                                                                                                                             s1_AAACGGGGTTTAGGAA</span>
<span class="co">## 2                                                                                                                                                                                     s1_AAGGCAGTCTCTGCTG;s1_AGAGTGGAGGAGTCTG;s1_GATCGCGAGTACGATA</span>
<span class="co">## 3 s1_ACAGCTACAGTCGTGC;s1_CATATGGTCTTTAGTC;s1_CCTCAGTGTACTTAGC;s1_CGAGCCACAATGTAAG;s1_CGATGGCAGGGTATCG;s1_GAAACTCCAAGCTGAG;s1_GACGCGTTCAGGCGAA;s1_GATGAGGCAGCGTTCG;s1_GGCCGATAGTTAAGTG;s1_TCAGGATTCACAGGCC;s1_TGCCCATAGTTAACGA;s1_TTTACTGCAGCAGTTT</span>
<span class="co">## 4                                                                                                                                                                                                                             s1_AGAGCTTGTCATGCAT</span>
<span class="co">## 5                                                                                                                                                                                                         s1_AGCAGCCTCAAACCGT;s1_GTCAAGTAGACTTGAA</span>
<span class="co">## 6                                                                                                                                                                                                                             s1_AGCGTCGGTTTCGCTC</span>
<span class="co">##   Freq in s2</span>
<span class="co">## 1          2</span>
<span class="co">## 2          7</span>
<span class="co">## 3          8</span>
<span class="co">## 4          2</span>
<span class="co">## 5          2</span>
<span class="co">## 6          1</span>
<span class="co">##                                                                                                                                                       Barcodes s2</span>
<span class="co">## 1                                                                                                                         s2_CATATTCAGCTGAACG;s2_GCATGTATCTGCGGCA</span>
<span class="co">## 2                     s2_ACGAGGATCGCATGGC;s2_AGAGCGAAGCCACCTG;s2_AGCAGCCTCTGTACGA;s2_CGATGTATCCTATGTT;s2_CGTAGCGAGTCCCACG;s2_CTAGAGTGTAAATGAC;s2_CTAGCCTGTGACTACT</span>
<span class="co">## 3 s2_ACGTCAAAGCGCTTAT;s2_CTACACCAGCCCTAAT;s2_GCACATAAGGATATAC;s2_GGGACCTTCGGACAAG;s2_GTGCATACACCTATCC;s2_TCTCTAAGTATATGGA;s2_TGCCCATTCGGCTTGG;s2_TTGGCAATCTGGCGTG</span>
<span class="co">## 4                                                                                                                         s2_CTGCTGTTCGTTTGCC;s2_GTCGGGTCACGACGAA</span>
<span class="co">## 5                                                                                                                         s2_AAAGTAGTCACCCGAG;s2_CAACCTCGTGCGCTTG</span>
<span class="co">## 6                                                                                                                                             s2_CAGTAACGTGGTACAG</span>
<span class="co">##   Freq in s3</span>
<span class="co">## 1          4</span>
<span class="co">## 2          8</span>
<span class="co">## 3          9</span>
<span class="co">## 4          1</span>
<span class="co">## 5          2</span>
<span class="co">## 6          1</span>
<span class="co">##                                                                                                                                                                           Barcodes s3</span>
<span class="co">## 1                                                                                                     s3_CTAGAGTGTAGCCTCG;s3_GCCTCTATCTCTTATG;s3_GGGTTGCCATACGCCG;s3_TGGGAAGAGGACTGGT</span>
<span class="co">## 2                     s3_AAAGATGTCTTATCTG;s3_AGTGAGGTCATTTGGG;s3_ATGAGGGCACTACAGT;s3_CAGCGACCAAGTAATG;s3_CTACATTTCTTGTCAT;s3_CTCGGAGGTCTAGCCG;s3_CTGAAGTCACGAGAGT;s3_TGTTCCGGTAGGGTAC</span>
<span class="co">## 3 s3_ACATCAGAGTGGTCCC;s3_AGAGCTTGTACATGTC;s3_AGCCTAATCGATGAGG;s3_CATGACAGTCCGTTAA;s3_CCTTCGAGTATGCTTG;s3_CTCGGAGAGAAACCTA;s3_CTGCTGTGTTACAGAA;s3_GGGATGAAGGTGCAAC;s3_TGGCCAGAGACGCAAC</span>
<span class="co">## 4                                                                                                                                                                 s3_GAGGTGAAGGCATGGT</span>
<span class="co">## 5                                                                                                                                             s3_AGGCCGTCAGTACACT;s3_TCGTAGATCTTATCTG</span>
<span class="co">## 6                                                                                                                                                                 s3_TGACTAGGTAGTACCT</span>
<span class="co">##   Freq in s4</span>
<span class="co">## 1          4</span>
<span class="co">## 2         38</span>
<span class="co">## 3         24</span>
<span class="co">## 4         17</span>
<span class="co">## 5          6</span>
<span class="co">## 6          9</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Barcodes s4</span>
<span class="co">## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         s4_CGAATGTCAGTTTACG;s4_CGGACGTGTCAAAGAT;s4_TCTCTAATCGGCGCAT;s4_TTAGGACCAGCGTCCA</span>
<span class="co">## 2 s4_AACCATGGTCTGGAGA;s4_AAGGAGCGTATATGGA;s4_AAGGCAGCAAGCCCAC;s4_ACAGCCGCACACATGT;s4_ACGATGTGTCGCTTTC;s4_ACGCAGCGTGCCTGGT;s4_AGAATAGCAACTGCGC;s4_AGCTCCTTCATTGCCC;s4_ATGTGTGTCCTTCAAT;s4_CAAGGCCAGTGACTCT;s4_CACACTCCAAGCCCAC;s4_CAGCAGCGTTCCGGCA;s4_CCGTTCAAGTTAACGA;s4_CCTAAAGAGTACATGA;s4_CCTCAGTAGAGGTACC;s4_CGGAGTCAGGTGGGTT;s4_CGTAGGCAGAGTAATC;s4_CTGAAACAGGCAGGTT;s4_CTGTGCTCACGTAAGG;s4_GAATAAGTCCTCAACC;s4_GACGTGCTCACTTCAT;s4_GATGCTAAGCGTTTAC;s4_GCGAGAATCGCGTAGC;s4_GGACATTGTCCAACTA;s4_GGAGCAATCAGGCGAA;s4_GTCAAGTCAACGCACC;s4_GTCACGGCAGTATCTG;s4_GTGTGCGTCTGGCGAC;s4_TAAACCGAGACTTTCG;s4_TACCTTACAAAGGTGC;s4_TACTCATGTAACGTTC;s4_TAGGCATGTAATTGGA;s4_TCGCGAGTCGTACCGG;s4_TGATTTCAGATCCCGC;s4_TGATTTCTCGGCCGAT;s4_TGGCCAGAGCTAACAA;s4_TTCTCAACAAGCGCTC;s4_TTGGAACTCTTCGGTC</span>
<span class="co">## 3                                                                                                                                                                                                                                                                                         s4_AGGTCATAGGGCTTCC;s4_CAAGATCTCCGGGTGT;s4_CACATAGTCGCCAGCA;s4_CGAGAAGTCAACGAAA;s4_CGGGTCATCAGGCCCA;s4_CGGTTAACATTCTTAC;s4_CGTCAGGCAGATCTGT;s4_CTACGTCGTACTTGAC;s4_CTCATTATCAGTCCCT;s4_GACACGCCACTGCCAG;s4_GACTACAGTACATGTC;s4_GGAATAAGTTCACCTC;s4_GGGCATCGTAGCGTAG;s4_GGGTCTGTCCGCTGTT;s4_GGTGTTAGTAACGCGA;s4_GTGGGTCTCGGATGGA;s4_GTTCATTGTTAAGATG;s4_GTTTCTAGTATATGAG;s4_TACTTACGTGCCTGGT;s4_TCACGAATCTAACTGG;s4_TCCCGATTCTCTGTCG;s4_TGTTCCGGTTGCCTCT;s4_TTGGAACGTGTGAAAT;s4_TTTATGCGTACCGTTA</span>
<span class="co">## 4                                                                                                                                                                                                                                                                                                                                                                                                                                     s4_AAATGCCTCTGGGCCA;s4_ACACCAACAGTCAGCC;s4_ACTTACTTCTCAAGTG;s4_ATCATGGGTTCGGCAC;s4_CACACTCAGATCCCAT;s4_CACAGTAAGATGTAAC;s4_CAGCTAAAGGGTGTTG;s4_CCTAAAGCAATGTTGC;s4_CCTACCACAATCACAC;s4_CTGTTTAGTCGCTTCT;s4_GCACATAAGATCCTGT;s4_GGCGACTAGGTGCTTT;s4_GGCTCGACACCAGCAC;s4_GGGAGATCAGTTTACG;s4_TACCTTACACGCGAAA;s4_TCCCGATCACGTAAGG;s4_TGACTAGTCCTCATTA</span>
<span class="co">## 5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 s4_CGAACATAGATCCCAT;s4_CGGGTCAGTGCCTTGG;s4_CTACGTCCACGGCGTT;s4_GTAACGTTCAATCTCT;s4_GTCGGGTAGACTAGGC;s4_TAAGAGACATCGGTTA</span>
<span class="co">## 6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     s4_AACTCAGTCGAGAACG;s4_ACGATACGTAGGCATG;s4_AGCAGCCTCCTGCCAT;s4_ATCGAGTTCTCTGTCG;s4_ATTGGTGTCATTGCGA;s4_CCGTACTTCCGGGTGT;s4_CGTTAGAGTTGCCTCT;s4_TCATTTGGTCATGCCG;s4_TGACGGCCATTGGGCC</span></code></pre>
<div class="sourceCode" id="cb162"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#overlap of clones</span>
<span class="va">clone_overlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_overlap_heatmap.html">VDJ_overlap_heatmap</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,feature.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_cdr3s_aa"</span>,<span class="st">"VJ_cdr3s_aa"</span><span class="op">)</span> ,grouping.column <span class="op">=</span> <span class="st">"sample_id"</span>, axis.label.size <span class="op">=</span> <span class="fl">20</span>, pvalues.label.size <span class="op">=</span> <span class="fl">12</span>, add.barcode.table <span class="op">=</span> <span class="cn">T</span>, plot.type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">clone_overlap</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">#Pheatmap plots will function only with length(unique(grouping.column)) &gt; 4</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-44-1.png" width="672">
Given the small size of the two repertoires, it is unsurprising that no
clones are shared</p>
</div>
</div>
<div class="section level2">
<h2 id="integrating-repertoire-and-gene-expression">8. Integrating repertoire and gene expression<a class="anchor" aria-label="anchor" href="#integrating-repertoire-and-gene-expression"></a>
</h2>
<p>The strength of the current 5’ sequencing protocols are that the gene
expression (GEX) and repertoire (VDJ) libraries are extracted from the
same sample, which can then be linked back to demonstrate that a given T
cell has a certain gene expression pattern and also a certain T cell
receptor sequence. The following functions are meant to integrate these
two pieces of information.</p>
<div class="section level3">
<h3 id="integrating-transcriptional-clusters-to-the-vdj-objects">8.1. Integrating transcriptional clusters to the VDJ objects<a class="anchor" aria-label="anchor" href="#integrating-transcriptional-clusters-to-the-vdj-objects"></a>
</h3>
<p>One thing we may ask is how similar the B or T cells in a given
clonal family are on the transcriptional level. The basic work of
matching barcodes from V(D)J sequencing to those from GEX sequencing was
already done in the VGM function. This part is therefore dedicated to
functions that help explore and exploite this data</p>
<p>First we get basic stats on the overlap between GEX and VDJ. In this
dataset, GEX data contains both B and T cells while VDJ data contains
only BCR sequences.</p>
<div class="sourceCode" id="cb163"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Nr of cells for which VDJ info is available</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 945</span></code></pre>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Nr of cells for which GEX info is available</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 2064</span></code></pre>
<div class="sourceCode" id="cb167"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#VDJ sequences for which GEX is available</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">VDJ_available</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 303</span></code></pre>
<div class="sourceCode" id="cb169"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#We can also plot this </span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"VDJ_available"</span>, shuffle <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-45-1.png" width="672"></p>
<p>If additional columns have been added to VDJ (e.g. by clonotyping),
these can be transferred to GEX via the VGM_integrate function</p>
<div class="sourceCode" id="cb170"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VGM_integrate.html">VGM_integrate</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Cells in GEX: 2064; cells in VDJ: 945</span>
<span class="co">##  Overlap: 303</span></code></pre>
<pre><code><span class="co">## Integrating all VDJ columns which are not in GEX to GEX</span></code></pre>
<pre><code><span class="co">## Added columns to GEX: clonotype_id; clonotype_frequency</span></code></pre>
</div>
<div class="section level3">
<h3 id="relating-clonal-expansion-to-transcriptional-cluster-membership">8.2. Relating clonal expansion to transcriptional cluster
membership<a class="anchor" aria-label="anchor" href="#relating-clonal-expansion-to-transcriptional-cluster-membership"></a>
</h3>
<p>The VDJ_clonal_expansion function has been used above to plot isotype
information. Here its extended functionality is used to color clones by
their transcriptional cluster</p>
<div class="sourceCode" id="cb174"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clonal_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_expansion.html">VDJ_clonal_expansion</a></span><span class="op">(</span>VDJ <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,clones <span class="op">=</span> <span class="st">"30"</span>, group.by <span class="op">=</span> <span class="st">"none"</span>, color.by <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span>
<span class="co">#group by specifies how many separate plots should be generated. If vgm contains global clonotype information this can be set to "none"</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">clonal_out</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-47-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="visualizing-clones-on-the-2-dimensional-landscape">8.3. Visualizing clones on the 2 dimensional landscape<a class="anchor" aria-label="anchor" href="#visualizing-clones-on-the-2-dimensional-landscape"></a>
</h3>
<p>To better visualize this, the cells of any clone can also be
highlighted on a dimensional reduction</p>
<p>Because the legend of these plots can get quite large, we can split
the plot and the legend and draw them separately. For this the gridExtra
and cowplot package is required</p>
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#here we overlay the top 5 clones</span>
<span class="va">plot_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_overlay_clones.html">VDJ_GEX_overlay_clones</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, n.clones <span class="op">=</span> <span class="fl">5</span>, by.sample <span class="op">=</span> <span class="cn">F</span>, ncol.facet <span class="op">=</span> <span class="fl">1</span>, split.plot.and.legend <span class="op">=</span> <span class="cn">F</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Scale for 'colour' is already present. Adding another scale for 'colour',</span>
<span class="co">## which will replace the existing scale.</span></code></pre>
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot_out</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co"># the plot</span>

<span class="co">#We can also plot any clone of interest by adding a column were TRUE values select which clones are to be plotted</span>
<span class="va">interesting_clones</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clonotype7"</span>, <span class="st">"clonotype42"</span><span class="op">)</span>
<span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">clones_to_plot</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span>
<span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">clones_to_plot</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">clonotype_id_10x</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">interesting_clones</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span>
<span class="va">plot_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_overlay_clones.html">VDJ_GEX_overlay_clones</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, clones.to.plot <span class="op">=</span> <span class="st">"clones_to_plot"</span>, by.sample <span class="op">=</span> <span class="cn">F</span>, ncol.facet <span class="op">=</span> <span class="fl">1</span>, split.plot.and.legend <span class="op">=</span> <span class="cn">T</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Scale for 'colour' is already present. Adding another scale for 'colour',</span>
<span class="co">## which will replace the existing scale.</span></code></pre>
<div class="sourceCode" id="cb179"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot_out</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co"># the plot</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">plot_out</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># the legend. Alternatively use gridExtra::grid.arrange(plot_out[[2]])</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-48-1.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-48-2.png" width="672"><img src="quickstart_files/figure-html/unnamed-chunk-48-3.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="specific-gene-expression-information-on-the-clonal-level">8.4. Specific gene expression information on the clonal level<a class="anchor" aria-label="anchor" href="#specific-gene-expression-information-on-the-clonal-level"></a>
</h3>
<p>We previously integrated the GEX information into the format of the
VDJ output. However, we may want to ask how the gene expression looks
for certain clonotypes (e.g., how many of the cells in the top clone are
expressing markers of activated B cells). This is done by adding an
extra column to the VDJ and GEX objects containing binary information on
expansion.</p>
<div class="sourceCode" id="cb180"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VGM_expanded_clones.html">VGM_expanded_clones</a></span><span class="op">(</span><span class="va">vgm</span>, add.to.VDJ <span class="op">=</span> <span class="cn">T</span>, add.to.GEX <span class="op">=</span> <span class="cn">T</span>, expansion.threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Added new columns to VDJ: expanded_clonotype_frequency</span></code></pre>
<pre><code><span class="co">## Added new columns to GEX: expanded_clonotype_frequency</span></code></pre>
<div class="sourceCode" id="cb183"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#We can use the dottile function to look at selected genes</span>
<span class="fu"><a href="../reference/GEX_dottile_plot.html">GEX_dottile_plot</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD19"</span>, <span class="st">"CD74"</span>,<span class="st">"IL2RA"</span>, <span class="st">"CD27"</span>,<span class="st">"CD80"</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"expanded_clonotype_frequency"</span>, threshold.to.plot <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: Removed 2 rows containing missing values (geom_point).</span></code></pre>
<p><img src="quickstart_files/figure-html/unnamed-chunk-49-1.png" width="672"></p>
<p>For a more unbiased view, DEGs can be calculated between expanded and
non-expanded clones</p>
<div class="sourceCode" id="cb185"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DE_byexpansion</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_DEgenes.html">GEX_DEgenes</a></span><span class="op">(</span>GEX <span class="op">=</span> <span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span>, group1 <span class="op">=</span> <span class="st">"TRUE"</span>, group2 <span class="op">=</span> <span class="st">"FALSE"</span>, grouping.column <span class="op">=</span> <span class="st">"expanded_clonotype_frequency"</span>, return.plot <span class="op">=</span> <span class="st">"volcano"</span>, logFC <span class="op">=</span> <span class="cn">F</span>, label.n.top.genes <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DE_byexpansion</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##               p_val  avg_logFC pct.1 pct.2 p_val_adj SYMBOL signif_pval</span>
<span class="co">## EDF1   0.0001018582 -0.6945464 0.200 0.489         1   EDF1       FALSE</span>
<span class="co">## POLR1D 0.0002441627 -0.6863222 0.214 0.489         1 POLR1D       FALSE</span>
<span class="co">## TRA2B  0.0002955248 -0.5573743 0.243 0.498         1  TRA2B       FALSE</span>
<span class="co">## HSPE1  0.0003391985 -0.5913888 0.329 0.579         1  HSPE1       FALSE</span>
<span class="co">## RBX1   0.0003399626 -0.6474882 0.286 0.554         1   RBX1       FALSE</span>
<span class="co">## PSMA4  0.0006829527 -0.4248722 0.114 0.335         1  PSMA4       FALSE</span>
<span class="co">##        signif_logfc color_threshold</span>
<span class="co">## EDF1           TRUE           FALSE</span>
<span class="co">## POLR1D         TRUE           FALSE</span>
<span class="co">## TRA2B          TRUE           FALSE</span>
<span class="co">## HSPE1          TRUE           FALSE</span>
<span class="co">## RBX1           TRUE           FALSE</span>
<span class="co">## PSMA4          TRUE           FALSE</span></code></pre>
<div class="sourceCode" id="cb187"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DE_byexpansion</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-50-1.png" width="672"></p>
</div>
</div>
<div class="section level2">
<h2 id="version-information">9 Version information<a class="anchor" aria-label="anchor" href="#version-information"></a>
</h2>
<div class="sourceCode" id="cb188"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.0.5 (2021-03-31)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 19044)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252   </span>
<span class="co">## [3] LC_MONETARY=German_Germany.1252 LC_NUMERIC=C                   </span>
<span class="co">## [5] LC_TIME=German_Germany.1252    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] Platypus_3.4.0     SeuratObject_4.0.4 Seurat_4.1.0       forcats_0.5.1     </span>
<span class="co">##  [5] stringr_1.4.0      purrr_0.3.4        readr_2.1.2        tidyr_1.1.3       </span>
<span class="co">##  [9] tibble_3.1.2       ggplot2_3.3.5      tidyverse_1.3.1    dplyr_1.0.7       </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] utf8_1.2.1            reticulate_1.20       tidyselect_1.1.1     </span>
<span class="co">##   [4] RSQLite_2.2.10        AnnotationDbi_1.52.0  htmlwidgets_1.5.4    </span>
<span class="co">##   [7] grid_4.0.5            Rtsne_0.15            munsell_0.5.0        </span>
<span class="co">##  [10] codetools_0.2-18      ragg_1.2.1            ica_1.0-2            </span>
<span class="co">##  [13] future_1.24.0         miniUI_0.1.1.1        withr_2.4.3          </span>
<span class="co">##  [16] colorspace_2.0-2      Biobase_2.50.0        highr_0.9            </span>
<span class="co">##  [19] knitr_1.37            rstudioapi_0.13       stats4_4.0.5         </span>
<span class="co">##  [22] ROCR_1.0-11           tensor_1.5            listenv_0.8.0        </span>
<span class="co">##  [25] labeling_0.4.2        polyclip_1.10-0       bit64_4.0.5          </span>
<span class="co">##  [28] farver_2.1.0          pheatmap_1.0.12       rprojroot_2.0.2      </span>
<span class="co">##  [31] parallelly_1.30.0     vctrs_0.3.8           generics_0.1.2       </span>
<span class="co">##  [34] xfun_0.27             ggseqlogo_0.1         R6_2.5.1             </span>
<span class="co">##  [37] doParallel_1.0.17     spatstat.utils_2.2-0  cachem_1.0.6         </span>
<span class="co">##  [40] assertthat_0.2.1      promises_1.2.0.1      scales_1.1.1         </span>
<span class="co">##  [43] gtable_0.3.0          org.Mm.eg.db_3.12.0   globals_0.14.0       </span>
<span class="co">##  [46] goftest_1.2-2         rlang_0.4.10          systemfonts_1.0.4    </span>
<span class="co">##  [49] GlobalOptions_0.1.2   splines_4.0.5         lazyeval_0.2.2       </span>
<span class="co">##  [52] spatstat.geom_2.2-0   broom_0.7.12          yaml_2.2.1           </span>
<span class="co">##  [55] reshape2_1.4.4        abind_1.4-5           modelr_0.1.8         </span>
<span class="co">##  [58] backports_1.4.1       httpuv_1.6.1          useful_1.2.6         </span>
<span class="co">##  [61] tools_4.0.5           ellipsis_0.3.2        spatstat.core_2.2-0  </span>
<span class="co">##  [64] jquerylib_0.1.4       RColorBrewer_1.1-2    BiocGenerics_0.36.1  </span>
<span class="co">##  [67] ggridges_0.5.3        Rcpp_1.0.7            plyr_1.8.6           </span>
<span class="co">##  [70] zlibbioc_1.36.0       rpart_4.1-15          deldir_0.2-10        </span>
<span class="co">##  [73] pbapply_1.5-0         cowplot_1.1.1         S4Vectors_0.28.1     </span>
<span class="co">##  [76] zoo_1.8-9             haven_2.4.3           ggrepel_0.9.1        </span>
<span class="co">##  [79] cluster_2.1.2         fs_1.5.2              magrittr_2.0.1       </span>
<span class="co">##  [82] data.table_1.14.0     RSpectra_0.16-0       scattermore_0.7      </span>
<span class="co">##  [85] circlize_0.4.14       lmtest_0.9-38         reprex_2.0.1         </span>
<span class="co">##  [88] RANN_2.6.1            fitdistrplus_1.1-6    matrixStats_0.59.0   </span>
<span class="co">##  [91] hms_1.1.1             patchwork_1.1.1       mime_0.11            </span>
<span class="co">##  [94] evaluate_0.14         xtable_1.8-4          readxl_1.3.1         </span>
<span class="co">##  [97] shape_1.4.6           IRanges_2.24.1        gridExtra_2.3        </span>
<span class="co">## [100] compiler_4.0.5        KernSmooth_2.23-20    crayon_1.5.0         </span>
<span class="co">## [103] htmltools_0.5.2       mgcv_1.8-36           later_1.2.0          </span>
<span class="co">## [106] tzdb_0.2.0            lubridate_1.8.0       DBI_1.1.2            </span>
<span class="co">## [109] dbplyr_2.1.1          MASS_7.3-54           Matrix_1.3-4         </span>
<span class="co">## [112] ade4_1.7-18           permute_0.9-7         cli_3.1.1            </span>
<span class="co">## [115] parallel_4.0.5        igraph_1.2.6          pkgconfig_2.0.3      </span>
<span class="co">## [118] pkgdown_2.0.2         plotly_4.10.0         spatstat.sparse_2.0-0</span>
<span class="co">## [121] xml2_1.3.3            foreach_1.5.2         bslib_0.3.1          </span>
<span class="co">## [124] XVector_0.30.0        stringdist_0.9.8      rvest_1.0.2          </span>
<span class="co">## [127] digest_0.6.27         sctransform_0.3.3     RcppAnnoy_0.0.18     </span>
<span class="co">## [130] vegan_2.5-7           Biostrings_2.58.0     spatstat.data_2.1-2  </span>
<span class="co">## [133] rmarkdown_2.11        cellranger_1.1.0      leiden_0.3.9         </span>
<span class="co">## [136] uwot_0.1.10           shiny_1.7.1           lifecycle_1.0.1      </span>
<span class="co">## [139] nlme_3.1-152          jsonlite_1.7.2        seqinr_4.2-8         </span>
<span class="co">## [142] desc_1.4.0            viridisLite_0.4.0     limma_3.46.0         </span>
<span class="co">## [145] fansi_0.5.0           pillar_1.7.0          lattice_0.20-44      </span>
<span class="co">## [148] fastmap_1.1.0         httr_1.4.2            survival_3.2-11      </span>
<span class="co">## [151] GO.db_3.12.1          glue_1.4.2            png_0.1-7            </span>
<span class="co">## [154] iterators_1.0.14      bit_4.0.4             stringi_1.7.4        </span>
<span class="co">## [157] sass_0.4.0            blob_1.2.2            textshaping_0.3.6    </span>
<span class="co">## [160] memoise_2.0.1         irlba_2.3.3           future.apply_1.8.1</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alexander Yermanos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
