<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Platypus VDJ_GEX_matrix • Platypus</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Platypus VDJ_GEX_matrix">
<meta property="og:description" content="Platypus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/vignette_index.html">Get started</a>
</li>
<li>
  <a href="../vignette_index.html">Vignettes</a>
</li>
<li>
  <a href="../PlatypusDB_index.html">PlatypusDB</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://immunesim.readthedocs.io/en/latest/" class="external-link">ImmuneSIM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../cute_monotremes.html">Cute Monotremes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/alexyermanos" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/KreineVic" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/alexyermanos/Platypus" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Platypus VDJ_GEX_matrix</h1>
                        <h4 data-toc-skip class="author">Alexander
Yermanos, Victor Kreiner, Andreas Agrafiotis</h4>
            
            <h4 data-toc-skip class="date">2022-03-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alexyermanos/Platypus/tree/mastervignettes/vgm_overview.Rmd" class="external-link"><code>vignettes/vgm_overview.Rmd</code></a></small>
      <div class="hidden name"><code>vgm_overview.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="vgm-introduction">1. VGM Introduction<a class="anchor" aria-label="anchor" href="#vgm-introduction"></a>
</h2>
<p>The VGM is the central data object of the current iteration of
Platypus that is produced by the VDJ_GEX_matrix function. The main idea
was to create a single object that contains all relevant immune
information that can be supplied to all other functions in our package.
This additionally provides the benefit that in the case of custom data
formats (e.g. pre-existing Seurat objects or non 10x single-cell
repertoire information), the user can adapt the necessary column names
and still use the downstream functions of Platypus.</p>
<p>For downstream examples of how the VGM interacts with other Platypus
functions, please refer to the Platypus Quickstart vignette. This
vignette will describe how we create the VGM from the output of 10x
genomics cellranger and how this can be modulated by the various
fucntion arguments.</p>
<p>As the function processes both GEX and VDJ, this vignette is devided
into 3 parts: 2. General settings 3.1 Gene expression (GEX) 3.2. Feature
Barcodes (FB) 4. Immune receptor repertoire (VDJ)</p>
<p>Most examples use the yermanos2021a dataset from PlatypusDB, which
contains B and T cells (GEX + VDJ) and is also featured in the
quickstart vignette due to the low number of cells. For more
information, please refer to the corresponding publication: <a href="https://doi.org/10.1098/rspb.2020.2793" class="external-link uri">https://doi.org/10.1098/rspb.2020.2793</a></p>
</div>
<div class="section level2">
<h2 id="general-information">2. General Information<a class="anchor" aria-label="anchor" href="#general-information"></a>
</h2>
<div class="section level3">
<h3 id="input-formats">2.1 Input formats<a class="anchor" aria-label="anchor" href="#input-formats"></a>
</h3>
<p>The VGM takes three different input formats that will be covered in
the respective sections: 1. Local Paths to cellranger output files
(covered below) 2. Data.in list input from either PlatypusDB_fetch() or
PlatypusDB_load_from_disk() (covered in the PlatypusDB vignette) This is
for the case where a user would like to download raw PlatypusDB datasets
and integrate these with local data. 3. A processed Seurat object as
Seurat.in (Covered in section 3.1.4) This is for the case that a user
would like to use an existing seurat object, which may be desired when
using custom normalization and integration methods for GEX data.</p>
<p>Local paths should be provided to cellranger directories which… …for
GEX: Corresponds to the “outs” folder from cellranger count function.
Under default parameters, the directory supplied as input to the VGM
function should contain the filtered_feature_annotations folder from 10x
cellranger count. …for VDJ: Corresponds to the “outs” folder from the
cellranger vdj function. Under default parameters, the directory
supplied as input to the VGM function should contain files such as
clonotypes.csv 10x cellranger vdj.</p>
<p>Below is an example of a basic run. The user will need to change the
input directory to their own local output files from cellranger.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Creating a list with local paths to cellranger directories</span>

<span class="va">VDJ.out.directory.list</span> <span class="op">&lt;-</span>  
        <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"C:/Users/PlatypusDB/yermanos2021b__VDJ_RAW/Aged.CNS.pool.3m.Bcell.S1"</span>,                     <span class="st">"C:/Users/PlatypusDB/yermanos2021b__VDJ_RAW/Aged.CNS.pool.12m.Bcell.S2"</span><span class="op">)</span>
<span class="va">GEX.out.directory.list</span> <span class="op">&lt;-</span> 
        <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"C:/Users/PlatypusDB/yermanos2021b__GEX_RAW/Aged.CNS.pool.3m.Bcell.S1"</span>,                     <span class="st">"C:/Users/PlatypusDB/yermanos2021b__GEX_RAW/Aged.CNS.pool.12m.Bcell.S2"</span><span class="op">)</span>


<span class="co">#Running the VDJ_GEX_matrix function</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>,
                      GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="co">#For more detailed runtime messages</span></code></pre></div>
<pre><code><span class="co">## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
<span class="co">## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span>
<span class="co">## This message will be shown once per session</span></code></pre>
<p>The function can also be run with just GEX or VDJ input. The output
object will have the same format as if both GEX and VDJ folders were
supplied as input to ensure compatibility with all downstream functions.
If this is needed, please provide only the desired input</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Only VDJ run</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span><span class="op">)</span> 

<span class="co">#Only GEX run</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span><span class="op">)</span> </code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="output-format">2.2 Output format<a class="anchor" aria-label="anchor" href="#output-format"></a>
</h2>
<p>Irregardless of the input, the output format stays the same. This
allows compatibility with all other downstream Platypus as well as
custom functions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "VDJ"            "GEX"            "VDJ.GEX.stats"  "Running params"</span>
<span class="co">## [5] "sessionInfo"</span></code></pre>
<p>The VGM output is a list of 5 elements as seen above. Certain
downstream functions may add additional list elements but will always
maintain the first 5.</p>
<p>VDJ is a data.frame with standard column output</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                 barcode sample_id group_id clonotype_id_10x celltype</span>
<span class="co">## 1 s1_AAACGGGGTTTAGGAA-1        s1        1       clonotype7   B cell</span>
<span class="co">## 2 s1_AAGGCAGTCTCTGCTG-1        s1        1      clonotype58   B cell</span>
<span class="co">## 3 s1_ACAGCTACAGTCGTGC-1        s1        1      clonotype65   B cell</span>
<span class="co">## 4 s1_ACTTGTTGTACTTAGC-1        s1        1      clonotype43   B cell</span>
<span class="co">## 5 s1_AGAGCTTGTCATGCAT-1        s1        1      clonotype33   B cell</span>
<span class="co">## 6 s1_AGAGTGGAGGAGTCTG-1        s1        1      clonotype49   B cell</span>
<span class="co">##   Nr_of_VDJ_chains Nr_of_VJ_chains       VDJ_cdr3s_aa  VJ_cdr3s_aa</span>
<span class="co">## 1                1               1        CARIGYAMDYW  CQQGNTLPPTF</span>
<span class="co">## 2                1               1  CARDFTTVVARGYFDVW  CQQDYSSPWTF</span>
<span class="co">## 3                1               1 CARGITTVVAYYYAMDYW   CLQYDNLYTF</span>
<span class="co">## 4                1               1     CTTGPYDYYAMDYW  CQQHYSTPYTF</span>
<span class="co">## 5                1               1      CARPHDYDGVDYW CSQSTHVPPWTF</span>
<span class="co">## 6                1               1    CARRYYSNYAWFAYW  CQQWSSYPYTF</span>
<span class="co">##                                             VDJ_cdr3s_nt</span>
<span class="co">## 1                      TGTGCTCGAATAGGATATGCTATGGACTACTGG</span>
<span class="co">## 2    TGTGCAAGAGACTTTACTACGGTAGTAGCCCGGGGGTACTTCGATGTCTGG</span>
<span class="co">## 3 TGTGCAAGAGGGATTACTACGGTAGTAGCTTATTACTATGCTATGGACTACTGG</span>
<span class="co">## 4             TGTACTACGGGCCCTTATGATTACTATGCTATGGACTACTGG</span>
<span class="co">## 5                TGTGCAAGGCCCCATGATTACGACGGAGTTGACTACTGG</span>
<span class="co">## 6          TGTGCAAGACGCTACTATAGTAACTACGCCTGGTTTGCTTACTGG</span>
<span class="co">##                            VJ_cdr3s_nt            VDJ_chain_contig</span>
<span class="co">## 1    TGCCAACAGGGTAATACGCTTCCTCCGACGTTC AAACGGGGTTTAGGAA-1_contig_2</span>
<span class="co">## 2    TGTCAGCAGGATTATAGCTCTCCGTGGACGTTC AAGGCAGTCTCTGCTG-1_contig_2</span>
<span class="co">## 3       TGTCTACAGTATGATAATCTGTACACGTTC ACAGCTACAGTCGTGC-1_contig_1</span>
<span class="co">## 4    TGTCAGCAACATTATAGCACTCCGTACACGTTC ACTTGTTGTACTTAGC-1_contig_1</span>
<span class="co">## 5 TGCTCTCAAAGTACACATGTTCCTCCGTGGACGTTC AGAGCTTGTCATGCAT-1_contig_2</span>
<span class="co">## 6    TGCCAGCAGTGGAGTAGTTACCCGTACACGTTC AGAGTGGAGGAGTCTG-1_contig_2</span>
<span class="co">##               VJ_chain_contig VDJ_chain VJ_chain VDJ_vgene  VJ_vgene VDJ_dgene</span>
<span class="co">## 1 AAACGGGGTTTAGGAA-1_contig_1       IGH      IGK   IGHV8-8 IGKV10-96          </span>
<span class="co">## 2 AAGGCAGTCTCTGCTG-1_contig_1       IGH      IGK   IGHV1-9  IGKV6-32          </span>
<span class="co">## 3 ACAGCTACAGTCGTGC-1_contig_2       IGH      IGK  IGHV1-26 IGKV19-93          </span>
<span class="co">## 4 ACTTGTTGTACTTAGC-1_contig_2       IGH      IGK  IGHV14-1  IGKV8-24          </span>
<span class="co">## 5 AGAGCTTGTCATGCAT-1_contig_1       IGH      IGK  IGHV5-17 IGKV1-110          </span>
<span class="co">## 6 AGAGTGGAGGAGTCTG-1_contig_1       IGH      IGK   IGHV1-9  IGKV4-55   IGHD2-5</span>
<span class="co">##   VDJ_jgene VJ_jgene VDJ_cgene VJ_cgene</span>
<span class="co">## 1     IGHJ4    IGKJ1      IGHM     IGKC</span>
<span class="co">## 2     IGHJ1    IGKJ1      IGHD     IGKC</span>
<span class="co">## 3     IGHJ4    IGKJ2      IGHD     IGKC</span>
<span class="co">## 4     IGHJ4    IGKJ2      IGHD     IGKC</span>
<span class="co">## 5     IGHJ2    IGKJ1      IGHM     IGKC</span>
<span class="co">## 6     IGHJ3    IGKJ2      IGHM     IGKC</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 VDJ_sequence_nt_raw</span>
<span class="co">## 1                                                                                                                                                      TGGGAAGTGTGCAGCCATGGGCAGGCTTACTTCTTCATTCCTGTTACTGATTGTCCCTGCATATGTCCTGTCCCAGGTTACTCTGAAAGAGTCTGGCCCTGGGATATTGCAGCCCTCCCAGACCCTCAGTCTGACTTGTTCTTTCTCTGGGTTTTCACTGAGCACTTTTGGTATGGGTGTAGGCTGGATTCGTCAGCCTTCAGGGAAGGGTCTGGAGTGGCTGGCACACATTTGGTGGGATGATGATAAGTACTATAACCCAGCCCTGAAGAGTCGGCTCACAATCTCCAAGGATACCTCCAAAAACCAGGTATTCCTCAAGATCGCCAATGTGGACACTGCAGATACTGCCACATACTACTGTGCTCGAATAGGATATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 2                                                                                                           TTGGGGAGTGTCCTCTCCAAAGTCCTTGAACATAGACTCTAACCATGGAATGGACCTGGGTCTTTCTCTTCCTCCTGTCAGTAACTGCAGGTGTCCACTCCCAGGTTCAGCTGCAGCAGTCTGGAGCTGAGCTGATGAAGCCTGGGGCCTCAGTGAAGCTTTCCTGCAAGGCTACTGGCTACACATTCACTGGCTACTGGATAGAGTGGGTAAAGCAGAGGCCTGGACATGGCCTTGAGTGGATTGGAGAGATTTTACCTGGAAGTGGTAGTACTAACTACAATGAGAAGTTCAAGGGCAAGGCCACATTCACTGCAGATACATCCTCCAACACAGCCTACATGCAACTCAGCAGCCTGACAACTGAGGACTCTGCCATCTATTACTGTGCAAGAGACTTTACTACGGTAGTAGCCCGGGGGTACTTCGATGTCTGGGGCACAGGGACCACGGTCACCGTCTCCTCAGGTAATGAAAAGGGACCTGACATGTTCCTCCTCTCAGAGTGCAAAGCCCCAGAGGAAAATGAAAAGATAAACCTGGGCTGTTTAGTAATTGGAAGTCAGCCACTGAAAATCAGCTGGGAGCCAAAGAAGTCAAGTATAGTTGAACATGTCTTCCCCTCTGAAATGAGAAATGGCAATTATACAATGGTCCTCCAGGTCACTGTGCTGGCCTC</span>
<span class="co">## 3                                                                                              GGGAACATATGTACAATGTCCTCACCACAGACACTGAACACACTGACTCTAACCATGGGATGGAGCTGGATCTTTCTCTTTCTCCTGTCAGGAACTGCAGGTGTCCTCTCTGAGGTCCAGCTGCAACAATCTGGACCTGAGCTGGTGAAGCCTGGGGCTTCAGTGAAGATATCCTGTAAGGCTTCTGGATACACGTTCACTGACTACTACATGAACTGGGTGAAGCAGAGCCATGGAAAGAGCCTTGAGTGGATTGGAGATATTAATCCTAACAATGGTGGTACTAGCTACAACCAGAAGTTCAAGGGCAAGGCCACATTGACTGTAGACAAGTCCTCCAGCACAGCCTACATGGAGCTCCGCAGCCTGACATCTGAGGACTCTGCAGTCTATTACTGTGCAAGAGGGATTACTACGGTAGTAGCTTATTACTATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCAGGTAATGAAAAGGGACCTGACATGTTCCTCCTCTCAGAGTGCAAAGCCCCAGAGGAAAATGAAAAGATAAACCTGGGCTGTTTAGTAATTGGAAGTCAGCCACTGAAAATCAGCTGGGAGCCAAAGAAGTCAAGTATAGTTGAACATGTCTTCCCCTCTGAAATGAGAAATGGCAATTATACAATGGTCCTCCAGGTCACTGTGCTGGCCTC</span>
<span class="co">## 4 GGGACTCAACTTCCTTCTTCTCCAGCCAGAATGTCCTTATGTAAGAAAGATCCTGTATGCAAATCATGTGAGACTGTGATGATTAATATAGGGATATCCACACCAAACATCATATGAGCCCTGTCTTCTCTACAGCCACTGAATCTCAAGATCCTTACAATGAAATGCAGCTGGGTCATCTTCTTCCTGATGGCAGTGGTTACAGGGGTCAATTCAGAGGTTCAGCTGCAGCAGTCTGGGGCAGAGCTTGTGAGGCCAGGGGCCTCAGTCAAGTTGTCCTGCACAGCTTCTGGCTTCAACATTAAAGACTACTATATGCACTGGGTGAAGCAGAGGCCTGAACAGGGCCTGGAGTGGATTGGAAGGATTGATCCTGAGGATGGTGATACTGAATATGCCCCGAAGTTCCAGGGCAAGGCCACTATGACTGCAGACACATCCTCCAACACAGCCTACCTGCAGCTCAGCAGCCTGACATCTGAGGACACTGCCGTCTATTACTGTACTACGGGCCCTTATGATTACTATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCAGGTAATGAAAAGGGACCTGACATGTTCCTCCTCTCAGAGTGCAAAGCCCCAGAGGAAAATGAAAAGATAAACCTGGGCTGTTTAGTAATTGGAAGTCAGCCACTGAAAATCAGCTGGGAGCCAAAGAAGTCAAGTATAGTTGAACATGTCTTCCCCTCTGAAATGAGAAATGGCAATTATACAATGGTCCTCCAGGTCACTGTGCTGGCCTC</span>
<span class="co">## 5                                                                                                             TGGATTCCCAGGTCCTCACATTCAGTGATCAGCACTGAACACAGACCACTCACCATGGACTCCAGGCTCAATTTAGTTTTCCTTGTCCTTATTTTAAAAGGTGTCCAGTGTGAGGTGCAGCTGGTGGAGTCTGGGGGAGGCTTAGTGAAGCCTGGAGGGTCCCTGAAACTCTCCTGTGCAGCCTCTGGATTCACTTTCAGTGACTATGGAATGCACTGGGTTCGTCAGGCTCCAGAGAAGGGGCTGGAGTGGGTTGCATACATTAGTAGTGGCAGTAGTACCATCTACTATGCAGACACAGTGAAGGGCCGATTCACCATCTCCAGAGACAATGCCAAGAACACCCTGTTCCTGCAAATGACCAGTCTGAGGTCTGAGGACACGGCCATGTATTACTGTGCAAGGCCCCATGATTACGACGGAGTTGACTACTGGGGCCAAGGCACCACTCTCACAGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 6                                                                                                       TGGGGAGCATATGATCAGTGTCCTCTCCAAAGTCCTTGAACATAGACTCTAACCATGGAATGGACCTGGGTCTTTCTCTTCCTCCTGTCAGTAACTGCAGGTGTCCACTCCCAGGTTCAGCTGCAGCAGTCTGGAGCTGAGCTGATGAAGCCTGGGGCCTCAGTGAAGCTTTCCTGCAAGGCTACTGGCTACACATTCACTGGCTACTGGATAGAGTGGGTAAAGCAGAGGCCTGGACATGGCCTTGAGTGGATTGGAGAGATTTTACCTGGAAGTGGTAGTACTAACTACAATGAGAAGTTCAAGGGCAAGGCCACATTCACTGCAGATACATCCTCCAACACAGCCTACATGCAACTCAGCAGCCTGACAACTGAGGACTCTGCCATCTATTACTGTGCAAGACGCTACTATAGTAACTACGCCTGGTTTGCTTACTGGGGCCAAGGGACTCTGGTCACTGTCTCTGCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        VJ_sequence_nt_raw</span>
<span class="co">## 1                                                                                 TTGGGGCATTGTAATTGAAGTCAAGACTCAGCCTGGACATGATGTCCTCTGCTCAGTTCCTTGGTCTCCTGTTGCTCTGTTTTCAAGGTACCAGATGTGATATCCAGATGACACAGACTACATCCTCCCTGTCTGCCTCTCTGGGAGACAGAGTCACCATCAGTTGCAGGGCAAGTCAGGACATTAGCAATTATTTAAACTGGTATCAGCAGAAACCAGATGGAACTGTTAAACTCCTGATCTACTACACATCAAGATTACACTCAGGAGTCCCATCAAGGTTCAGTGGCAGTGGGTCTGGAACAGATTATTCTCTCACCATTAGCAACCTGGAGCAAGAAGATATTGCCACTTACTTTTGCCAACAGGGTAATACGCTTCCTCCGACGTTCGGTGGAGGCACCAAGCTGGAAATCAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">## 2                                                                                                    GGCAGGCAAGGGCATCAAGATGAAGTCACAGACCCAGGTCTTCGTATTTCTACTGCTCTGTGTGTCTGGTGCTCATGGGAGTATTGTGATGACCCAGACTCCCAAATTCCTGCTTGTATCAGCAGGAGACAGGGTTACCATAACCTGCAAGGCCAGTCAGAGTGTGAGTAATGATGTAGCTTGGTACCAACAGAAGCCAGGGCAGTCTCCTAAACTGCTGATATACTATGCATCCAATCGCTACACTGGAGTCCCTGATCGCTTCACTGGCAGTGGATATGGGACGGATTTCACTTTCACCATCAGCACTGTGCAGGCTGAAGACCTGGCAGTTTATTTCTGTCAGCAGGATTATAGCTCTCCGTGGACGTTCGGTGGAGGCACCAAGCTGGAAATCAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">## 3                                                                                  TTATTTGGGGAGTCATTCTTGGTCAGGAGACGTTGTAGAAATGAGACCGTCTATTCAGTTCCTGGGGCTCTTGTTGTTCTGGCTTCATGGTGCTCAGTGTGACATCCAGATGACACAGTCTCCATCCTCACTGTCTGCATCTCTGGGAGGCAAAGTCACCATCACTTGCAAGGCAAGCCAAGACATTAACAAGTATATAGCTTGGTACCAACACAAGCCTGGAAAAGGTCCTAGGCTGCTCATACATTACACATCTACATTACAGCCAGGCATCCCATCAAGGTTCAGTGGAAGTGGGTCTGGGAGAGATTATTCCTTCAGCATCAGCAACCTGGAGCCTGAAGATATTGCAACTTATTATTGTCTACAGTATGATAATCTGTACACGTTCGGAGGGGGGACCAAGCTGGAAATAAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">## 4 TTCCGATCTACTTGTTGTACTTAGCCTGTGTTCACATTCTTATTTGGGGAGTGTTGCTGGTGTCCAGGCATGATGAGCATCAGACAGGCTGGGCAGCAAGATGGAATCACAGACCCAGGTCCTCATGTTTCTTCTGCTCTGGGTATCTGGTGCCTGTGCAGACATTGTGATGACACAGTCTCCATCCTCCCTGGCTATGTCAGTAGGACAGAAGGTCACTATGAGCTGCAAGTCCAGTCAGAGCCTTTTAAATAGTAGCAATCAAAAGAACTATTTGGCCTGGTACCAGCAGAAACCAGGACAGTCTCCTAAACTTCTGGTATACTTTGCATCCACTAGGGAATCTGGGGTCCCTGATCGCTTCATAGGCAGTGGATCTGGGACAGATTTCACTCTTACCATCAGCAGTGTGCAGGCTGAAGACCTGGCAGATTACTTCTGTCAGCAACATTATAGCACTCCGTACACGTTCGGAGGGGGGACCAAGCTGGAAATAAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">## 5                   AGAGCTTGTCATGCATTTTGGGCCTATTTCTTTTTTGGGGACTGATCAGTCTCCTCAGGCTGTCTCCTCAGGTTGCCTCCTCAAAATGAAGTTGCCTGTTAGGCTGTTGGTGCTGATGTTCTGGATTCCTGCTTCCAGCAGTGATGTTGTGATGACCCAAACTCCACTCTCCCTGCCTGTCAGTCTTGGAGATCAAGCCTCCATCTCTTGCAGATCTAGTCAGAGCCTTGTACACAGTAATGGAAACACCTATTTACATTGGTACCTGCAGAAGCCAGGCCAGTCTCCAAAGCTCCTGATCTACAAAGTTTCCAACCGATTTTCTGGGGTCCCAGACAGGTTCAGTGGCAGTGGATCAGGGACAGATTTCACACTCAAGATCAGCAGAGTGGAGGCTGAGGATCTGGGAGTTTATTTCTGCTCTCAAAGTACACATGTTCCTCCGTGGACGTTCGGTGGAGGCACCAAGCTGGAAATCAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">## 6                                                             TTGGGGACTTATGAGAATAGTAGTAATTAGCTAGGGACCAAAGTTCAAAGACAAAATGGATTTTCAAGTGCAGATTTTCAGCTTCCTGCTAATCAGTGCCTCAGTCATACTGTCCAGAGGACAAATTGTTCTCACCCAGTCTCCAGCAATCATGTCTGCATCTCCAGGGGAGAAGGTCACCATGACCTGCAGTGCCAGCTCAAGTGTAAGTTACATGTACTGGTACCAGCAGAAGCCAGGATCCTCCCCCAGACTCCTGATTTATGACACATCCAACCTGGCTTCTGGAGTCCCTGTTCGCTTCAGTGGCAGTGGGTCTGGGACCTCTTACTCTCTCACAATCAGCCGAATGGAGGCTGAAGATGCTGCCACTTATTACTGCCAGCAGTGGAGTAGTTACCCGTACACGTTCGGAGGGGGGACCAAGCTGGAAATAAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">##   VDJ_sequence_nt_trimmed VJ_sequence_nt_trimmed VDJ_sequence_aa VJ_sequence_aa</span>
<span class="co">## 1                                                                              </span>
<span class="co">## 2                                                                              </span>
<span class="co">## 3                                                                              </span>
<span class="co">## 4                                                                              </span>
<span class="co">## 5                                                                              </span>
<span class="co">## 6                                                                              </span>
<span class="co">##   VDJ_trimmed_ref VJ_trimmed_ref     VDJ_raw_consensus_id</span>
<span class="co">## 1                                 clonotype7_concat_ref_1</span>
<span class="co">## 2                                clonotype58_concat_ref_1</span>
<span class="co">## 3                                clonotype65_concat_ref_1</span>
<span class="co">## 4                                clonotype43_concat_ref_1</span>
<span class="co">## 5                                clonotype33_concat_ref_1</span>
<span class="co">## 6                                clonotype49_concat_ref_1</span>
<span class="co">##        VJ_raw_consensus_id     orig_barcode clonotype_frequency specifity</span>
<span class="co">## 1  clonotype7_concat_ref_2 AAACGGGGTTTAGGAA                   1        NA</span>
<span class="co">## 2 clonotype58_concat_ref_2 AAGGCAGTCTCTGCTG                   1        NA</span>
<span class="co">## 3 clonotype65_concat_ref_2 ACAGCTACAGTCGTGC                   1        NA</span>
<span class="co">## 4 clonotype43_concat_ref_2 ACTTGTTGTACTTAGC                   1        NA</span>
<span class="co">## 5 clonotype33_concat_ref_2 AGAGCTTGTCATGCAT                   1        NA</span>
<span class="co">## 6 clonotype49_concat_ref_2 AGAGTGGAGGAGTCTG                   1        NA</span>
<span class="co">##   affinity GEX_available    orig.ident orig_barcode_GEX seurat_clusters</span>
<span class="co">## 1       NA         FALSE          &lt;NA&gt;             &lt;NA&gt;            &lt;NA&gt;</span>
<span class="co">## 2       NA          TRUE SeuratProject AAGGCAGTCTCTGCTG               5</span>
<span class="co">## 3       NA          TRUE SeuratProject ACAGCTACAGTCGTGC               0</span>
<span class="co">## 4       NA          TRUE SeuratProject ACTTGTTGTACTTAGC               0</span>
<span class="co">## 5       NA          TRUE SeuratProject AGAGCTTGTCATGCAT               0</span>
<span class="co">## 6       NA          TRUE SeuratProject AGAGTGGAGGAGTCTG               0</span>
<span class="co">##        PC_1      PC_2    UMAP_1    UMAP_2   tSNE_1     tSNE_2     batches</span>
<span class="co">## 1        NA        NA        NA        NA       NA         NA Unspecified</span>
<span class="co">## 2 -6.257892 -2.522445 -4.813108 0.6205567 2.188638  -3.411929 Unspecified</span>
<span class="co">## 3 -6.375512 -5.164284 -6.385532 1.9866081 6.511564  -7.350143 Unspecified</span>
<span class="co">## 4 -7.476580 -4.585612 -7.730432 0.4828904 2.640636 -11.282818 Unspecified</span>
<span class="co">## 5 -6.710098 -6.034031 -6.045865 3.3426221 9.943012  -6.504101 Unspecified</span>
<span class="co">## 6 -6.667460 -2.615907 -5.995199 0.3335993 1.924303  -6.548645 Unspecified</span>
<span class="co">##   clonotype_id</span>
<span class="co">## 1   clonotype7</span>
<span class="co">## 2  clonotype58</span>
<span class="co">## 3  clonotype65</span>
<span class="co">## 4  clonotype43</span>
<span class="co">## 5  clonotype33</span>
<span class="co">## 6  clonotype49</span></code></pre>
<p>The GEX object vgm[[2]] is a Seurat object. Metadata can be accessed
as shown below. Depending on integration parameters (Section 2.3), the
GEX object can also contain information from the immune receptor VDJ
data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "orig.ident"              "nCount_RNA"             </span>
<span class="co">##  [3] "nFeature_RNA"            "orig_barcode"           </span>
<span class="co">##  [5] "VDJ_available"           "sample_id"              </span>
<span class="co">##  [7] "group_id"                "percent.mt"             </span>
<span class="co">##  [9] "RNA_snn_res.0.5"         "seurat_clusters"        </span>
<span class="co">## [11] "clonotype_id_10x"        "celltype"               </span>
<span class="co">## [13] "Nr_of_VDJ_chains"        "Nr_of_VJ_chains"        </span>
<span class="co">## [15] "VDJ_cdr3s_aa"            "VJ_cdr3s_aa"            </span>
<span class="co">## [17] "VDJ_cdr3s_nt"            "VJ_cdr3s_nt"            </span>
<span class="co">## [19] "VDJ_chain_contig"        "VJ_chain_contig"        </span>
<span class="co">## [21] "VDJ_chain"               "VJ_chain"               </span>
<span class="co">## [23] "VDJ_vgene"               "VJ_vgene"               </span>
<span class="co">## [25] "VDJ_dgene"               "VDJ_jgene"              </span>
<span class="co">## [27] "VJ_jgene"                "VDJ_cgene"              </span>
<span class="co">## [29] "VJ_cgene"                "VDJ_sequence_nt_raw"    </span>
<span class="co">## [31] "VJ_sequence_nt_raw"      "VDJ_sequence_nt_trimmed"</span>
<span class="co">## [33] "VJ_sequence_nt_trimmed"  "VDJ_sequence_aa"        </span>
<span class="co">## [35] "VJ_sequence_aa"          "VDJ_trimmed_ref"        </span>
<span class="co">## [37] "VJ_trimmed_ref"          "VDJ_raw_consensus_id"   </span>
<span class="co">## [39] "VJ_raw_consensus_id"     "clonotype_frequency"    </span>
<span class="co">## [41] "specifity"               "affinity"               </span>
<span class="co">## [43] "orig_barcode_GEX"        "PC_1"                   </span>
<span class="co">## [45] "PC_2"                    "UMAP_1"                 </span>
<span class="co">## [47] "UMAP_2"                  "tSNE_1"                 </span>
<span class="co">## [49] "tSNE_2"                  "batches"</span></code></pre>
<p><img src="vgm_overview_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
<p>VDJ.GEX.stats is a table containing statistics about the processed
datasets. This is useful for QC. Many of the values in this dataframe
are imported from the metrics.csv tables provided by Cellranger. In case
these tables are not available, the output will contain NA values.</p>
<p>The generation of this table can be toggled by setting get.VDJ.stats
= F</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "Repertoir path"                                    </span>
<span class="co">##  [2] "Sample name"                                       </span>
<span class="co">##  [3] "Nr unique barcodes"                                </span>
<span class="co">##  [4] "Nr barcodes is_cell"                               </span>
<span class="co">##  [5] "Nr cells 1VDJ 1VJ"                                 </span>
<span class="co">##  [6] "Nr cells 1VDJ 0VJ"                                 </span>
<span class="co">##  [7] "Nr cells 0VDJ 1VJ"                                 </span>
<span class="co">##  [8] "Nr cells 2 or more VDJ 1VJ"                        </span>
<span class="co">##  [9] "Nr cells 1VDJ 2 or more VJ"                        </span>
<span class="co">## [10] "Nr cells 2 or more VDJ 2 or more VJ"               </span>
<span class="co">## [11] "Nr cells full_length"                              </span>
<span class="co">## [12] "Nr cells productive"                               </span>
<span class="co">## [13] "Nr cells high_confidence"                          </span>
<span class="co">## [14] "Nr cells all true"                                 </span>
<span class="co">## [15] "Nr cells all true and 1VDJ 1VJ"                    </span>
<span class="co">## [16] "Nr clonotypes"                                     </span>
<span class="co">## [17] "Nr clonotypes 1VDJ 1VJ"                            </span>
<span class="co">## [18] "Nr clonotypes &lt; 1VDJ 1VJ"                          </span>
<span class="co">## [19] "Nr clonotypes &gt; 1VDJ 1VJ"                          </span>
<span class="co">## [20] "% Nr unique barcodes"                              </span>
<span class="co">## [21] "% Nr barcodes is_cell"                             </span>
<span class="co">## [22] "% Nr cells 1VDJ 1VJ"                               </span>
<span class="co">## [23] "% Nr cells 1VDJ 0VJ"                               </span>
<span class="co">## [24] "% Nr cells 0VDJ 1VJ"                               </span>
<span class="co">## [25] "% Nr cells 2 or more VDJ 1VJ"                      </span>
<span class="co">## [26] "% Nr cells 1VDJ 2 or more VJ"                      </span>
<span class="co">## [27] "% Nr cells 2 or more VDJ 2 or more VJ"             </span>
<span class="co">## [28] "% Nr cells full_length"                            </span>
<span class="co">## [29] "% Nr cells productive"                             </span>
<span class="co">## [30] "% Nr cells high_confidence"                        </span>
<span class="co">## [31] "% Nr cells all true"                               </span>
<span class="co">## [32] "% Nr cells all true and 1VDJ 1VJ"                  </span>
<span class="co">## [33] "% Nr clonotypes"                                   </span>
<span class="co">## [34] "% Nr clonotypes 1VDJ 1VJ"                          </span>
<span class="co">## [35] "% Nr clonotypes &lt; 1VDJ 1VJ"                        </span>
<span class="co">## [36] "% Nr clonotypes &gt; 1VDJ 1VJ"                        </span>
<span class="co">## [37] "Estimated.Number.of.Cells"                         </span>
<span class="co">## [38] "Mean.Read.Pairs.per.Cell"                          </span>
<span class="co">## [39] "Number.of.Cells.With.Productive.V.J.Spanning.Pair" </span>
<span class="co">## [40] "Number.of.Read.Pairs"                              </span>
<span class="co">## [41] "Valid.Barcodes"                                    </span>
<span class="co">## [42] "Q30.Bases.in.Barcode"                              </span>
<span class="co">## [43] "Q30.Bases.in.RNA.Read.1"                           </span>
<span class="co">## [44] "Q30.Bases.in.RNA.Read.2"                           </span>
<span class="co">## [45] "Q30.Bases.in.UMI"                                  </span>
<span class="co">## [46] "Reads.Mapped.to.Any.V.D.J.Gene"                    </span>
<span class="co">## [47] "Reads.Mapped.to.IGH"                               </span>
<span class="co">## [48] "Reads.Mapped.to.IGK"                               </span>
<span class="co">## [49] "Reads.Mapped.to.IGL"                               </span>
<span class="co">## [50] "Mean.Used.Read.Pairs.per.Cell"                     </span>
<span class="co">## [51] "Fraction.Reads.in.Cells"                           </span>
<span class="co">## [52] "Median.IGH.UMIs.per.Cell"                          </span>
<span class="co">## [53] "Median.IGK.UMIs.per.Cell"                          </span>
<span class="co">## [54] "Median.IGL.UMIs.per.Cell"                          </span>
<span class="co">## [55] "Cells.With.Productive.V.J.Spanning.Pair"           </span>
<span class="co">## [56] "Cells.With.Productive.V.J.Spanning..IGK..IGH..Pair"</span>
<span class="co">## [57] "Cells.With.Productive.V.J.Spanning..IGL..IGH..Pair"</span>
<span class="co">## [58] "Paired.Clonotype.Diversity"                        </span>
<span class="co">## [59] "Cells.With.IGH.Contig"                             </span>
<span class="co">## [60] "Cells.With.IGK.Contig"                             </span>
<span class="co">## [61] "Cells.With.IGL.Contig"                             </span>
<span class="co">## [62] "Cells.With.CDR3.annotated.IGH.Contig"              </span>
<span class="co">## [63] "Cells.With.CDR3.annotated.IGK.Contig"              </span>
<span class="co">## [64] "Cells.With.CDR3.annotated.IGL.Contig"              </span>
<span class="co">## [65] "Cells.With.V.J.Spanning.IGH.Contig"                </span>
<span class="co">## [66] "Cells.With.V.J.Spanning.IGK.Contig"                </span>
<span class="co">## [67] "Cells.With.V.J.Spanning.IGL.Contig"                </span>
<span class="co">## [68] "Cells.With.Productive.IGH.Contig"                  </span>
<span class="co">## [69] "Cells.With.Productive.IGK.Contig"                  </span>
<span class="co">## [70] "Cells.With.Productive.IGL.Contig"                  </span>
<span class="co">## [71] "rep_id"                                            </span>
<span class="co">## [72] "Estimated.Number.of.Cells"                         </span>
<span class="co">## [73] "Mean.Reads.per.Cell"                               </span>
<span class="co">## [74] "Median.Genes.per.Cell"                             </span>
<span class="co">## [75] "Number.of.Reads"                                   </span>
<span class="co">## [76] "Valid.Barcodes"                                    </span>
<span class="co">## [77] "Sequencing.Saturation"                             </span>
<span class="co">## [78] "Q30.Bases.in.Barcode"                              </span>
<span class="co">## [79] "Q30.Bases.in.RNA.Read"                             </span>
<span class="co">## [80] "Q30.Bases.in.UMI"                                  </span>
<span class="co">## [81] "Reads.Mapped.to.Genome"                            </span>
<span class="co">## [82] "Reads.Mapped.Confidently.to.Genome"                </span>
<span class="co">## [83] "Reads.Mapped.Confidently.to.Intergenic.Regions"    </span>
<span class="co">## [84] "Reads.Mapped.Confidently.to.Intronic.Regions"      </span>
<span class="co">## [85] "Reads.Mapped.Confidently.to.Exonic.Regions"        </span>
<span class="co">## [86] "Reads.Mapped.Confidently.to.Transcriptome"         </span>
<span class="co">## [87] "Reads.Mapped.Antisense.to.Gene"                    </span>
<span class="co">## [88] "Fraction.Reads.in.Cells"                           </span>
<span class="co">## [89] "Total.Genes.Detected"                              </span>
<span class="co">## [90] "Median.UMI.Counts.per.Cell"                        </span>
<span class="co">## [91] "rep_id"</span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>,
                      GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      get.VDJ.stats <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="co">#Turn off VDJ stats </span></code></pre></div>
<p>The VGM also stores the parameter used during function call in case a
user saves a VGM but deletes/overwrites the initial code….although this
should not happen…right?</p>
<p>When the VGM is called under default parameters, the function input
arguments can be located in the fourth list element.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##                                                                                                                                sample.path.vdj </span>
<span class="co">## "C:/Users/PlatypusDB/yermanos2021b__VDJ_RAW/Aged.CNS.pool.3m.Bcell.S1 ; C:/Users/PlatypusDB/yermanos2021b__VDJ_RAW/Aged.CNS.pool.12m.Bcell.S2" </span>
<span class="co">##                                                                                                                              samples.paths.GEX </span>
<span class="co">## "C:/Users/PlatypusDB/yermanos2021b__GEX_RAW/Aged.CNS.pool.3m.Bcell.S1 ; C:/Users/PlatypusDB/yermanos2021b__GEX_RAW/Aged.CNS.pool.12m.Bcell.S2" </span>
<span class="co">##                                                                                                                          FB.out.directory.list </span>
<span class="co">##                                                                                                                                         "none" </span>
<span class="co">##                                                                                                                                    VDJ.combine </span>
<span class="co">##                                                                                                                                         "TRUE" </span>
<span class="co">##                                                                                                                                  GEX.integrate </span>
<span class="co">##                                                                                                                                         "TRUE" </span>
<span class="co">##                                                                                                                           integrate.GEX.to.VDJ </span>
<span class="co">##                                                                                                                                         "TRUE" </span>
<span class="co">##                                                                                                                           integrate.VDJ.to.GEX </span>
<span class="co">##                                                                                                                                         "TRUE" </span>
<span class="co">##                                                                                                                         exclude.GEX.not.in.VDJ </span>
<span class="co">##                                                                                                                                        "FALSE" </span>
<span class="co">##                                                                                                                filter.overlapping.barcodes.GEX </span>
<span class="co">##                                                                                                                                         "TRUE" </span>
<span class="co">##                                                                                                                filter.overlapping.barcodes.VDJ </span>
<span class="co">##                                                                                                                                         "TRUE" </span>
<span class="co">##                                                                                                                  exclude.on.cell.state.markers </span>
<span class="co">##                                                                                                                                         "none" </span>
<span class="co">##                                                                                                                                  get.VDJ.stats </span>
<span class="co">##                                                                                                                                         "TRUE" </span>
<span class="co">##                                                                                                                                       numcores </span>
<span class="co">##                                                                                                                                            "1" </span>
<span class="co">##                                                                                                                                 trim.and.align </span>
<span class="co">##                                                                                                                                        "FALSE" </span>
<span class="co">##                                                                                                              select.excess.chains.by.umi.count </span>
<span class="co">##                                                                                                                                        "FALSE" </span>
<span class="co">##                                                                                                        excess.chain.confidence.count.threshold </span>
<span class="co">##                                                                                                                                         "1000" </span>
<span class="co">##                                                                                                                              gap.opening.cost, </span>
<span class="co">##                                                                                                                                           "10" </span>
<span class="co">##                                                                                                                             gap.extension.cost </span>
<span class="co">##                                                                                                                                            "4" </span>
<span class="co">##                                                                                                                            parallel.processing </span>
<span class="co">##                                                                                                                                         "none" </span>
<span class="co">##                                                                                                                             integration.method </span>
<span class="co">##                                                                                                                                   "scale.data" </span>
<span class="co">##                                                                                                                                VDJ.gene.filter </span>
<span class="co">##                                                                                                                                         "TRUE" </span>
<span class="co">##                                                                                                                                    mito.filter </span>
<span class="co">##                                                                                                                                           "20" </span>
<span class="co">##                                                                                                                              norm.scale.factor </span>
<span class="co">##                                                                                                                                        "10000" </span>
<span class="co">##                                                                                                                                  n.feature.rna </span>
<span class="co">##                                                                                                                                            "0" </span>
<span class="co">##                                                                                                                                n.count.rna.min </span>
<span class="co">##                                                                                                                                            "0" </span>
<span class="co">##                                                                                                                                n.count.rna.max </span>
<span class="co">##                                                                                                                                          "Inf" </span>
<span class="co">##                                                                                                                            n.variable.features </span>
<span class="co">##                                                                                                                                         "2000" </span>
<span class="co">##                                                                                                                             cluster.resolution </span>
<span class="co">##                                                                                                                                          "0.5" </span>
<span class="co">##                                                                                                                                   neighbor.dim </span>
<span class="co">##                                                                                                                         "1;2;3;4;5;6;7;8;9;10" </span>
<span class="co">##                                                                                                                                        mds.dim </span>
<span class="co">##                                                                                                                         "1;2;3;4;5;6;7;8;9;10" </span>
<span class="co">##                                                                                                                             subsample.barcodes </span>
<span class="co">##                                                                                                                                        "FALSE" </span>
<span class="co">##                                                                                                                                       group.id </span>
<span class="co">##                                                                                                                                          "1;2" </span>
<span class="co">##                                                                                                                             FB.count.threshold </span>
<span class="co">##                                                                                                                                           "10" </span>
<span class="co">##                                                                                                                             FB.ratio.threshold </span>
<span class="co">##                                                                                                                                            "2"</span></code></pre>
<p>The fifth element of the VGM contains the utils::sessionInfo() output
to record the versions of R and accompanying packages used during the
VGM creation.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "sessionInfo"</span></code></pre>
</div>
<div class="section level2">
<h2 id="parameters-for-vdj-gex-integration">2.3 Parameters for VDJ GEX integration<a class="anchor" aria-label="anchor" href="#parameters-for-vdj-gex-integration"></a>
</h2>
<p>A key feature of Platypus is the direct pairing of VDJ and GEX data.
This is currently achieved by VGM combining the relevant data and
metadata from VDJ (vgm[[1]]) and GEX (vgm[[2]]) objects using the cell
barcode and sample_id information.<br>
Several parameters control this integration:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Running the VDJ_GEX_matrix function</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>,
                      GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      VDJ.combine <span class="op">=</span> <span class="cn">T</span>, <span class="co">#Whether to combine all samples into one VDJ dataframe (is highly recommended)</span>
                      GEX.integrate <span class="op">=</span> <span class="cn">T</span>, <span class="co">#Whether to integrate all GEX samples. For integration methods see GEX section.  </span>
                      integrate.GEX.to.VDJ <span class="op">=</span> <span class="cn">T</span>, <span class="co">#Whether to copy GEX metadata into VDJdataframe </span>
                      integrate.VDJ.to.GEX <span class="op">=</span> <span class="cn">T</span>, <span class="co">#Whether to copy VDJ data into GEX</span>
                      exclude.GEX.not.in.VDJ <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="co">#Whether to exclude cells in GEX, for which no VDJ data is available. Set this to TRUE if you only want gene expression information for those cells with immune receptor sequences.</span></code></pre></div>
<p>In some cases, cell barcodes may not be unique across samples. This
may occur by pure chance, barcode hopping during library construction
and sequencing, or due to low diversity of barcodes during capture. The
VGM deals with this in two ways. Firstly, a sample-id prefix is appended
to every barcode.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">barcode</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## [1] "s1_AAACGGGGTTTAGGAA-1"</span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## [1] "s1_AAAGATGAGTCCGGTC-1"</span></code></pre>
<p>Second, the duplicated barcodes can be filtered out to prevent the
emergence of unlikely public clones. This is necessary, for example, if
a public clone is discovered in two distinct VDJ samples and have the
exact same cell barcode. Public clones containing identical cell
barcodes are highly unlikely given the massive potential barcode space.
If this filtering is set to TRUE, the function will prompt a callback
with the number of filtered cells.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Running the VDJ_GEX_matrix function</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>,
                      GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      filter.overlapping.barcodes.GEX <span class="op">=</span> <span class="cn">T</span>,
                      filter.overlapping.barcodes.VDJ <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </code></pre></div>
</div>
<div class="section level2">
<h2 id="gex">3.1 GEX<a class="anchor" aria-label="anchor" href="#gex"></a>
</h2>
<p>The VGM function attempts to simplify the standard GEX processing and
integration functions common to Seurat and Harmony packages, given the
majority of immunological studies use this pipeline. Although we are not
making the statement that all gene expression datasets should be
processed using identical parameters, we find that this function
simplifies the standard copy-paste from the Seurat website.</p>
<div class="section level3">
<h3 id="gex-qc-filtering-parameters">3.1.1 GEX QC filtering parameters<a class="anchor" aria-label="anchor" href="#gex-qc-filtering-parameters"></a>
</h3>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Running the VDJ_GEX_matrix function</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>, 
                      VDJ.gene.filter <span class="op">=</span> <span class="cn">T</span>, <span class="co">#Remove all VDJ genes from GEX before clustering.</span>
                      mito.filter <span class="op">=</span> <span class="fl">20</span>, <span class="co">#Remove all cells with a higher % of reads mapped to Mitochondiral genes. Data gathered via Seurat::PercentageFeatureSet(., pattern = "^MT-")</span>
                      n.count.rna.min <span class="op">=</span> <span class="fl">0</span>, <span class="co">#Remove all cells with a total RNA count below this</span>
                      n.count.rna.max <span class="op">=</span> <span class="cn">Inf</span>,  <span class="co">#Remove all cells with a total RNA count above this</span>
                      n.feature.rna <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="co">#Remove all cells with a gene count lower than this</span></code></pre></div>
<p>The default settings are meant to be inclusive, so to not impose
filtering to on the dataset which is not directly apparent to the
user.</p>
</div>
<div class="section level3">
<h3 id="gex-filtering-by-phenotype">3.1.2 GEX filtering by phenotype<a class="anchor" aria-label="anchor" href="#gex-filtering-by-phenotype"></a>
</h3>
<p>The VGM also offers the option to filter cells based on their gene
expression profiles. Removing unwanted cells before clustering can
result in more accurate conclusions concerning the cell types of
interest. For example, it is likely better to filter out non B and T
cells if integrating and analyzing repertoire features such as clonal
expansion.</p>
<p>The input format for the exclude.on.cell.state.markers argument is
the same as to the GEX_phenotype function. In the example below, we
filter out all CD14 positive cells as well as all CD3 epsilon and gamma
double negative cells.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Running the VDJ_GEX_matrix function</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>, 
                      exclude.on.cell.state.markers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14+"</span>, <span class="st">"CD3E-;CD3G-"</span><span class="op">)</span><span class="op">)</span> <span class="co">#Remove all cells with a total RNA count above this</span></code></pre></div>
<pre><code><span class="co">## Setting GEX directory to provided path/filtered_feature_bc_matrix </span>
<span class="co">## [1] "2022-03-15 16:04:32 CET"</span></code></pre>
<pre><code><span class="co">## Warning in simpleLoess(y, x, w, span, degree = degree, parametric =</span>
<span class="co">## parametric, : pseudoinverse used at -2.091</span></code></pre>
<pre><code><span class="co">## Warning in simpleLoess(y, x, w, span, degree = degree, parametric =</span>
<span class="co">## parametric, : neighborhood radius 0.49627</span></code></pre>
<pre><code><span class="co">## Warning in simpleLoess(y, x, w, span, degree = degree, parametric =</span>
<span class="co">## parametric, : reciprocal condition number 2.8445e-015</span></code></pre>
<pre><code><span class="co">## Warning in simpleLoess(y, x, w, span, degree = degree, parametric =</span>
<span class="co">## parametric, : There are other near singularities as well. 0.090619</span></code></pre>
<pre><code><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="co">## </span>
<span class="co">## Number of nodes: 118</span>
<span class="co">## Number of edges: 3533</span>
<span class="co">## </span>
<span class="co">## Running Louvain algorithm...</span>
<span class="co">## Maximum modularity in 10 random starts: 0.6521</span>
<span class="co">## Number of communities: 3</span>
<span class="co">## Elapsed time: 0 seconds</span>
<span class="co">## [1] "2022-03-15 16:04:36 CET"</span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Plotting this confirms the filtering </span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14"</span>,<span class="st">"CD3E"</span>,<span class="st">"CD3G"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning in Seurat::FeaturePlot(vgm[[2]], c("CD14", "CD3E", "CD3G")): All cells</span>
<span class="co">## have the same value (0) of CD14.</span></code></pre>
<p><img src="vgm_overview_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>As a workflow we first recommend clustering all cells and running a
differential gene expression analysis by cluster using the
GEX_cluster_genes() function, which reveals the cluster-defining genes
of each cluster. In case that unwanted cells are present (e.g. a
contamination of Neutrophils in a B cell dataset), the cluster
signatures can be used to identify the best genes for filtering via the
initial VGM call.</p>
</div>
<div class="section level3">
<h3 id="gex-integration-methods">3.1.3 GEX integration methods<a class="anchor" aria-label="anchor" href="#gex-integration-methods"></a>
</h3>
<p>The VGM offers four methods of GEX dataset integration if
GEX.integrate is set to TRUE:</p>
<p>1.”scale.data” integration is based on Seurat logNormalize followed
by the ScaleData function based on found variable features (set using
n.variable.features)</p>
<ol start="2" style="list-style-type: decimal">
<li><p>“anchors” is an integration method which uses similar cell states
to align to datasets. Extensive documentation is provided here: <a href="https://satijalab.org/seurat/articles/integration_introduction.html" class="external-link uri">https://satijalab.org/seurat/articles/integration_introduction.html</a></p></li>
<li><p>“sct” employs the SCTransform function from Seurat</p></li>
<li><p>“harmony” uses the Bioconductor package Harmony. This may be
better for larger datasets given the runtime and required memory of
anchors and sct.</p></li>
</ol>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Running the VDJ_GEX_matrix function</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      GEX.integrate <span class="op">=</span> <span class="cn">T</span>, 
                      integration.method <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span> <span class="co">#Default</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"harmony"</span><span class="op">)</span>

<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      GEX.integrate <span class="op">=</span> <span class="cn">T</span>, 
                      integration.method <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gex-clustering-optimisation-parameters">3.1.3 GEX clustering optimisation parameters<a class="anchor" aria-label="anchor" href="#gex-clustering-optimisation-parameters"></a>
</h3>
<p>Indipendently of integration method, GEX data is scaled and variable
features and PCA dimensions are used to calculate low dimensional
embedding Three parameters control these:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Running the VDJ_GEX_matrix function</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      GEX.integrate <span class="op">=</span> <span class="cn">T</span>, 
                      integration.method <span class="op">=</span> <span class="st">"scale.data"</span>, <span class="co">#Default</span>
                      norm.scale.factor <span class="op">=</span> <span class="fl">10000</span>, <span class="co">#passed to Seurat::NormalizeData</span>
                      n.variable.features <span class="op">=</span> <span class="fl">2000</span>, <span class="co">#Passed to Seurat::FindVariableFeatures</span>
                      cluster.resolution <span class="op">=</span> <span class="fl">0.5</span>, <span class="co">#Passed to Seurat::FindClusters</span>
                      neighbor.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>, <span class="co">#Passed to Seurat::FindNeighbors</span>
                      mds.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="co">#Passed to Seurat::RunTSNE and Seurat::RunUMAP</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="input-a-processed-seurat-object">3.1.4 Input a processed Seurat object<a class="anchor" aria-label="anchor" href="#input-a-processed-seurat-object"></a>
</h3>
<p>The VGM offers simple and flexible GEX processing, but remains one of
many options for GEX processing. We therefore made it possible to use
the VDJ processing and VDJ-GEX integration capabilities of the VGM
function with an already processed Seurat object.</p>
<p>For a Seurat object to be compatible as input, it must contain two
metadata columns: 1. sample_id with sample ids from s1,s2,s3 to sn of
character or factor class. These must be in the same order as
VDJ.out.directory list elements. 2. group_id of character class.</p>
<p>This input Seurat object will not be processed concerning
normalisation or dimensional embeddings. Nonetheless the following
filtering operations are still available, as shown below</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Running the VDJ_GEX_matrix function</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>vgm <span class="ot">&lt;-</span> <span class="fu">VDJ_GEX_matrix</span>(<span class="at">Seurat.in =</span> preprocessed_GEX, </span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">exclude.GEX.not.in.VDJ =</span> F, </span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">integrate.GEX.to.VDJ =</span> T, </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">integrate.VDJ.to.GEX =</span> T, </span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">filter.overlapping.barcodes.GEX =</span> T, </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">exclude.on.cell.state.markers =</span> <span class="fu">c</span>(<span class="st">"CD3E+"</span>)</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>                      <span class="at">GEX.integrate =</span> T) </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="feature-barcode-import">3.2.1 Feature Barcode import<a class="anchor" aria-label="anchor" href="#feature-barcode-import"></a>
</h3>
<p>Platypus supports feature barcode technology (also referred to as
hashing barcodes), but at this time does not support CITE-seq data.
(This is on the Platypus-Team To-Do list)</p>
<p>Feature Barcode (FB) data may be imported in two different modes,
depending on Cellranger processing proceedures.</p>
<ol style="list-style-type: decimal">
<li><p>FB data processed indipendently of GEX data via Cellranger count
will yield an output folder structure which is identical to GEX. These
output directories can be fed into the VGM using the
FB.out.directory.list</p></li>
<li><p>FB data can also be processed in combination with GEX using
Cellranger multi and aggr. This yields a single folder structure with
both GEX and FB matrices contained within the same output files. In this
case, the function will determine the input type (GEX of FB) of each
matrix based on the numbers of Features. Any matrix below 100 features
is regarded FB and every matrix above that as GEX.</p></li>
</ol>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">FB.out.directory.list</span> <span class="op">&lt;-</span>  
        <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"~path_to_CellrangerCount_outs_directory"</span>,                   
             <span class="st">"~path_to_CellrangerCount_outs_directory"</span><span class="op">)</span>

<span class="co">#Running with separate FB input</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      FB.out.directory.list <span class="op">=</span> <span class="va">FB.out.directory.list</span><span class="op">)</span> 

<span class="co">#Running with FB GEX combined input</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.cellranger.aggr.out.directory.list</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="feature-barcode-assignment">3.2.2 Feature Barcode assignment<a class="anchor" aria-label="anchor" href="#feature-barcode-assignment"></a>
</h3>
<p>The concept of feature barcodes lies in the attribution of a certain
sample or group barcode to a cell given the number of sequenced counts
of that barcode. While this does work well in most cases, FB data can be
noisy and a subset of cells may be difficult to confidently attribute to
a single barcode. The VGM assignes FBs to cells by two criteria:</p>
<ol style="list-style-type: decimal">
<li><p>FB.count.threshold determines how many counts for any FB are
neccessary to be considered. This defaults to 10. For example, in case a
cell has counts &lt; 10 for all FBs, no single FB will be assigned
(Function returns “Not assignable”).</p></li>
<li><p>FB.ratio.threshold determines the minimum ratio between the most
frequent and second most frequent FB for the most frequent to be
confidently assigned. This defaults to 2. For clarity we can consider
the following example</p></li>
</ol>
<table class="table">
<tr>
<th>
barcode
</th>
<th>
FB-1
</th>
<th>
FB-2
</th>
<th>
FB-3
</th>
</tr>
<tr>
<td>
Cell 1
</td>
<th>
3
</th>
<th>
4
</th>
<th>
9
</th>
</tr>
<tr>
<td>
Cell 2
</td>
<th>
1
</th>
<th>
32
</th>
<th>
43
</th>
</tr>
<tr>
<td>
Cell 3
</td>
<th>
100
</th>
<th>
1
</th>
<th>
13
</th>
</tr>
</table>
<p>For Cell 1: 9/4 &gt; 2, so FB-3 meets the FB.ratio.threshold. But: 9
&lt; 10 so FB-3 does not meet the FB.count.threshold. For this cell the
function returns “Not assignable”</p>
<p>For Cell 2: 43 and 32 &gt; 10 so both FB-2 and FB-3 meet the
FB.count.threshold. But: 43/32 &lt; 2 so FB-3 does not meet the
FB.ratio.threshold. Again the function returns “Not assignable”</p>
<p>For Cell 3: 100 &gt; 10 and 100/32 &gt; 2 so FB-1 meets both
criteria. The function returns “FB-1”</p>
<p>Tweaking these parameters can help to make barcode assignments more
inclusive, but also more susceptible to false assignments.</p>
<p>As a QC, we recommend verifying that variability in FB coverage
across libraries and samples is consistent and that FB assignments match
with expected numbers from e.g. pre-sorting by FACS.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Running with separate FB input</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      FB.out.directory.list <span class="op">=</span> ,
                      FB.count.threshold <span class="op">=</span> <span class="fl">10</span>, 
                      FB.ratio.threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> 

<span class="co">#Running with FB GEX combined input</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.cellranger.aggr.out.directory.list</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="feature-barcode-filtering">3.2.3 Feature Barcode filtering<a class="anchor" aria-label="anchor" href="#feature-barcode-filtering"></a>
</h3>
<p>In many cases, hashing barcodes are being combined with CITE-seq or
other surface barcodes. For processing FB barcodes, all other barcodes
need to be filtered out. For this the VGM allows excluding Feature
barcodes by their names and a regex expression. In the example below we
are filtering out all FBs that have “CITE” or “TetTCR” in their
name.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Running with separate FB input</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.out.directory.list</span>,
                      FB.out.directory.list <span class="op">=</span> , 
                      FB.exclude.pattern <span class="op">=</span> <span class="st">"(CITE|TetTCR)"</span><span class="op">)</span> 

<span class="co">#Running with FB GEX combined input</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>GEX.out.directory.list <span class="op">=</span> <span class="va">GEX.cellranger.aggr.out.directory.list</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="vdj">4.1. VDJ<a class="anchor" aria-label="anchor" href="#vdj"></a>
</h2>
<p>The VGM allows to reformat and merge several dataframes from
Cellranger outputs and additionally return aligned and trimmed receptor
sequences.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Basic run with VDJ only</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## [1] "2022-03-15 16:04:37 CET"</span>
<span class="co">## [1] "2022-03-15 16:04:37 CET"</span>
<span class="co">## [1] "2022-03-15 16:04:38 CET"</span>
<span class="co">## [1] "2022-03-15 16:04:38 CET"</span></code></pre>
<div class="section level3">
<h3 id="accomodating-cells-with-aberrant-chain-numbers">4.1.1 Accomodating cells with aberrant chain numbers<a class="anchor" aria-label="anchor" href="#accomodating-cells-with-aberrant-chain-numbers"></a>
</h3>
<p>Due to stochastical sampling, inter-cellular mRNA and biological
peculiarities, a single cell barcode may be associated with one, two or
more TCR or BCR contigs. A classical cell contains 1VDJ and 1VJ chain.
Cells with only one chain are frequent, while chains with more than 2
are rare.</p>
<p>The VGM function and format is fully compatible with any combination
of chains, without the need for cell filtering. Fields attributed to a
missing chain will contain and empty string (““). In fields which
contain information about 2 chains (e.g. VDJ_cdr3_aa in a cell with 2VDJ
chains) different chains are separated by”;”</p>
<p>For filtering purposes two numeric columns containing the number of
chains are included</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">Nr_of_VJ_chains</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                  barcode sample_id group_id clonotype_id_10x celltype</span>
<span class="co">## 21 s1_CCGTACTGTCCGACGT-1        s1        1      clonotype18   B cell</span>
<span class="co">## 24 s1_CGAGCCACAATGTAAG-1        s1        1      clonotype44   B cell</span>
<span class="co">## 32 s1_GAAACTCAGGAATCGC-1        s1        1      clonotype35   B cell</span>
<span class="co">## 53 s1_GTACTTTAGAGGGATA-1        s1        1      clonotype11   B cell</span>
<span class="co">## 54 s1_GTAGGCCGTCCCGACA-1        s1        1      clonotype17   B cell</span>
<span class="co">## 55 s1_GTAGTCACAGTAAGAT-1        s1        1      clonotype20   B cell</span>
<span class="co">##    Nr_of_VDJ_chains Nr_of_VJ_chains    VDJ_cdr3s_aa VJ_cdr3s_aa</span>
<span class="co">## 21                1               0   CARRNHPYYFDYW            </span>
<span class="co">## 24                1               0 CARETAQVPYYFDYW            </span>
<span class="co">## 32                1               0  CAIGHYYGSSSDVW            </span>
<span class="co">## 53                1               0     CALYGSSYDYW            </span>
<span class="co">## 54                1               0    CVNGIYYYFDYW            </span>
<span class="co">## 55                1               0    CARDSSGWFAYW            </span>
<span class="co">##                                     VDJ_cdr3s_nt VJ_cdr3s_nt</span>
<span class="co">## 21       TGTGCAAGACGGAACCACCCCTACTACTTTGACTACTGG            </span>
<span class="co">## 24 TGTGCAAGAGAGACAGCTCAGGTTCCGTACTACTTTGACTACTGG            </span>
<span class="co">## 32    TGTGCAATAGGGCATTACTACGGTAGTAGCTCCGATGTCTGG            </span>
<span class="co">## 53             TGTGCCCTCTACGGTAGTAGCTACGACTACTGG            </span>
<span class="co">## 54          TGTGTAAATGGGATTTATTACTACTTTGACTACTGG            </span>
<span class="co">## 55          TGTGCAAGAGACAGCTCAGGCTGGTTTGCTTACTGG            </span>
<span class="co">##               VDJ_chain_contig VJ_chain_contig VDJ_chain VJ_chain VDJ_vgene</span>
<span class="co">## 21 CCGTACTGTCCGACGT-1_contig_1                       IGH            IGHV4-1</span>
<span class="co">## 24 CGAGCCACAATGTAAG-1_contig_1                       IGH           IGHV1-26</span>
<span class="co">## 32 GAAACTCAGGAATCGC-1_contig_1                       IGH           IGHV1-74</span>
<span class="co">## 53 GTACTTTAGAGGGATA-1_contig_1                       IGH            IGHV2-3</span>
<span class="co">## 54 GTAGGCCGTCCCGACA-1_contig_1                       IGH            IGHV9-1</span>
<span class="co">## 55 GTAGTCACAGTAAGAT-1_contig_1                       IGH           IGHV1-39</span>
<span class="co">##    VJ_vgene VDJ_dgene VDJ_jgene VJ_jgene VDJ_cgene VJ_cgene</span>
<span class="co">## 21                        IGHJ2               IGHM         </span>
<span class="co">## 24            IGHD3-2     IGHJ2               IGHM         </span>
<span class="co">## 32            IGHD1-1     IGHJ1               IGHM         </span>
<span class="co">## 53                        IGHJ2               IGHM         </span>
<span class="co">## 54                        IGHJ2               IGHM         </span>
<span class="co">## 55            IGHD3-2     IGHJ3               IGHM         </span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   VDJ_sequence_nt_raw</span>
<span class="co">## 21                                       GAAGCAAAGGGGATCAGCCCGAGATTCTCATTCAGTGATCAACACTGAACACACATCCCTTACCATGGATTTTGGGCTGATTTTTTTTATTGTTGCTCTTTTAAAAGGGGTCCAGTGTGAGGTGAAGCTTCTCCAGTCTGGAGGTGGCCTGGTGCAGCCTGGAGGATCCCTGAAACTCTCCTGTGCAGCCTCAGGAATCGATTTTAGTAGATACTGGATGAGTTGGGTTCGGCGGGCTCCAGGGAAAGGACTAGAATGGATTGGAGAAATTAATCCAGATAGCAGTACAATAAACTATGCACCATCTCTAAAGGATAAATTCATCATCTCCAGAGACAACGCCAAAAATACGCTGTACCTGCAAATGAGCAAAGTGAGATCTGAGGACACAGCCCTTTATTACTGTGCAAGACGGAACCACCCCTACTACTTTGACTACTGGGGCCAAGGCACCACTCTCACAGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 24                         TTGGGGACCCCTGAAAACAACATATGTACAATGTCCTCACCACAGACACTGAACACACTGACTCTAACCATGGGATGGAGCTGGATCTTTCTCTTTCTCCTGTCAGGAACTGCAGGTGTCCTCTCTGAGGTCCAGCTGCAACAATCTGGACCTGAGCTGGTGAAGCCTGGGGCTTCAGTGAAGATATCCTGTAAGGCTTCTGGATACACGTTCACTGACTACTACATGAACTGGGTGAAGCAGAGCCATGGAAAGAGCCTTGAGTGGATTGGAGATATTAATCCTAACAATGGTGGTACTAGCTACAACCAGAAGTTCAAGGGCAAGGCCACATTGACTGTAGACAAGTCCTCCAGCACAGCCTACATGGAGCTCCGCAGCCTGACATCTGAGGACTCTGCAGTCTATTACTGTGCAAGAGAGACAGCTCAGGTTCCGTACTACTTTGACTACTGGGGCCAAGGCACCACTCTCACAGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 32 GTTACTCTGGAATCGCGTCAGACGTGTTTCTTTTTTGGGGAGAAAAACATGAGATCACTGTTCTCTCTACAGTTACTGAGCACACAGGACCTCACCATGAGATGGAGCTGTATCATCCTCTTCTTGGTAGCAACAGCTACAGGTGTCCACTCCCAGGTCCAACTGCAGCAGCCTGGGGCTGAACTGGTGAAGCCTGGGGCTTCAGTGAAGGTGTCCTGCAAGGCTTCTGGCTACACCTTCACCAGCTACTGGATGCACTGGGTGAAGCAGAGGCCTGGCCAAGGCCTTGAGTGGATTGGAAGGATTCATCCTTCTGATAGTGATACTAACTACAATCAAAAGTTCAAGGGCAAGGCCACATTGACTGTAGACAAATCCTCCAGCACAGCCTACATGCAGCTCAGCAGCCTGACATCTGAGGACTCTGCGGTCTATTACTGTGCAATAGGGCATTACTACGGTAGTAGCTCCGATGTCTGGGGCACAGGGACCACGGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 53                                                         CTAAAGGGGTTCTTATCTGGGGATCCTCTTCTCATAGAGCCTCCATCAGACCATGGCTGTCCTGGCACTGCTCCTCTGCCTGGTGACATTCCCAAGCTGTGTCCTGTCCCAGGTGCAGCTGAAGGAGTCAGGACCTGGCCTGGTGGCGCCCTCACAGAGCCTGTCCATCACATGCACTGTCTCAGGGTTCTCATTAACCAGCTATGGTGTAAGCTGGGTTCGCCAGCCTCCAGGAAAGGGTCTGGAGTGGCTGGGAGTAATATGGGGTGACGGGAGCACAAATTATCATTCAGCTCTCATATCCAGACTGAGCATCAGCAAGGATAACTCCAAGAGCCAAGTTTTCTTAAAACTGAACAGTCTGCAAACTGATGACACAGCCACGTACTACTGTGCCCTCTACGGTAGTAGCTACGACTACTGGGGCCAAGGCACCACTCTCACAGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 54                                          TTTTGGGGAAGGGAGTGACCAGTTAGTCTTAAGGCACCACTGAGCCCAAGTCTTAGACATCATGGATTGGGTGTGGACCTTGCTATTCCTGATAGCAGCTGCCCAAAGTGCCCAAGCACAGATCCAGTTGGTGCAGTCTGGACCTGAGCTGAAGAAGCCTGGAGAGACAGTCAAGATCTCCTGCAAGGCTTCTGGGTATACCTTCACAGAATATCCAATGCACTGGGTGAAGCAGGCTCCAGGAAAGGGTTTCAAGTGGATGGGCATGATATACACCGACACTGGAGAGCCAACATATGCTGAAGAGTTCAAGGGACGGTTTGCCTTCTCTTTGGAGACCTCTGCCAGCACTGCCTATTTGCAGATCAACAACCTCAAAAATGAGGACACGGCTACATATTTCTGTGTAAATGGGATTTATTACTACTTTGACTACTGGGGCCAAGGCACCACTCTCACAGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 55                                                                        ACTTATTTGGGGGAAGACACTGACTCTAACCATGGGATGGAGCTGGATCTTTCTCTTCCTCCTCTCAGGAACTGCAGGTGTCCACTCTGAGTTCCAGCTGCAGCAGTCTGGACCTGAGCTGGTGAAGCCTGGCGCTTCAGTGAAGATATCCTGCAAGGCTTCTGGTTACTCATTCACTGACTACAACATGAACTGGGTGAAGCAGAGCAATGGAAAGAGCCTTGAGTGGATTGGAGTAATTAATCCTAACTATGGTACTACTAGCTACAATCAGAAGTTCAAGGGCAAGGCCACATTGACTGTAGACCAATCTTCCAGCACAGCCTACATGCAGCTCAACAGCCTGACATCTGAGGACTCTGCAGTCTATTACTGTGCAAGAGACAGCTCAGGCTGGTTTGCTTACTGGGGCCAAGGGACTCTGGTCACTGTCTCTGCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">##    VJ_sequence_nt_raw VDJ_sequence_nt_trimmed VJ_sequence_nt_trimmed</span>
<span class="co">## 21                                                                  </span>
<span class="co">## 24                                                                  </span>
<span class="co">## 32                                                                  </span>
<span class="co">## 53                                                                  </span>
<span class="co">## 54                                                                  </span>
<span class="co">## 55                                                                  </span>
<span class="co">##    VDJ_sequence_aa VJ_sequence_aa VDJ_trimmed_ref VJ_trimmed_ref</span>
<span class="co">## 21                                                              </span>
<span class="co">## 24                                                              </span>
<span class="co">## 32                                                              </span>
<span class="co">## 53                                                              </span>
<span class="co">## 54                                                              </span>
<span class="co">## 55                                                              </span>
<span class="co">##        VDJ_raw_consensus_id VJ_raw_consensus_id     orig_barcode</span>
<span class="co">## 21 clonotype18_concat_ref_1                     CCGTACTGTCCGACGT</span>
<span class="co">## 24 clonotype44_concat_ref_1                     CGAGCCACAATGTAAG</span>
<span class="co">## 32 clonotype35_concat_ref_1                     GAAACTCAGGAATCGC</span>
<span class="co">## 53 clonotype11_concat_ref_1                     GTACTTTAGAGGGATA</span>
<span class="co">## 54 clonotype17_concat_ref_1                     GTAGGCCGTCCCGACA</span>
<span class="co">## 55 clonotype20_concat_ref_1                     GTAGTCACAGTAAGAT</span>
<span class="co">##    clonotype_frequency specifity affinity     batches clonotype_id</span>
<span class="co">## 21                   1        NA       NA Unspecified  clonotype18</span>
<span class="co">## 24                   1        NA       NA Unspecified  clonotype44</span>
<span class="co">## 32                   1        NA       NA Unspecified  clonotype35</span>
<span class="co">## 53                   1        NA       NA Unspecified  clonotype11</span>
<span class="co">## 54                   1        NA       NA Unspecified  clonotype17</span>
<span class="co">## 55                   1        NA       NA Unspecified  clonotype20</span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">Nr_of_VJ_chains</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                  barcode sample_id group_id clonotype_id_10x celltype</span>
<span class="co">## 16 s1_CACCTTGCAAACGCGA-1        s1        1      clonotype28   B cell</span>
<span class="co">## 25 s1_CGATGGCAGGGTATCG-1        s1        1       clonotype2   B cell</span>
<span class="co">## 30 s1_CTGATAGGTTCTGTTT-1        s1        1      clonotype57   B cell</span>
<span class="co">## 42 s1_GCTGCTTTCTCTGAGA-1        s1        1      clonotype22   B cell</span>
<span class="co">## 57 s1_GTCACGGCACGCATCG-1        s1        1      clonotype60   B cell</span>
<span class="co">## 67 s1_TGACTAGTCGCCTGTT-1        s1        1      clonotype52   B cell</span>
<span class="co">##    Nr_of_VDJ_chains Nr_of_VJ_chains                VDJ_cdr3s_aa</span>
<span class="co">## 16                1               2               CARGAPNWYFDVW</span>
<span class="co">## 25                1               2                 CARWFAWFAYW</span>
<span class="co">## 30                1               2           CAKNYGSSYSYWYFDVW</span>
<span class="co">## 42                2               2 CTGDYAMDYW;CTLITTVVAKDAMDYW</span>
<span class="co">## 57                1               2           CARRPYYSNSHYAMDYW</span>
<span class="co">## 67                1               2            CARDYYGSSLYYFDYW</span>
<span class="co">##                    VJ_cdr3s_aa</span>
<span class="co">## 16     CALWYSTHYVF;CWQGTHFPQTF</span>
<span class="co">## 25     CQNDYSYPLTF;CQQYSSYPYTF</span>
<span class="co">## 30       CKQSYNLYTF;CQQSNSWLTF</span>
<span class="co">## 42      CQQWSSNPLTF;CWQGTHFPPF</span>
<span class="co">## 57 CALWYSNHLVF;CGVGDTIKEQFVYVF</span>
<span class="co">## 67      CQQYWSTRTF;CLQYDNLLYTF</span>
<span class="co">##                                                                       VDJ_cdr3s_nt</span>
<span class="co">## 16                                         TGTGCAAGAGGGGCTCCCAACTGGTACTTCGATGTCTGG</span>
<span class="co">## 25                                               TGTGCAAGATGGTTTGCCTGGTTTGCTTACTGG</span>
<span class="co">## 30                             TGTGCAAAAAACTATGGTAGTAGCTACAGCTACTGGTACTTCGATGTCTGG</span>
<span class="co">## 42 TGTACCGGGGATTACGCTATGGACTACTGG;TGTACTCTCATTACTACGGTAGTAGCCAAGGATGCTATGGACTACTGG</span>
<span class="co">## 57                             TGTGCAAGAAGGCCCTACTATAGTAACTCCCACTATGCTATGGACTACTGG</span>
<span class="co">## 67                                TGTGCTAGAGATTACTACGGTAGTAGCTTGTACTACTTTGACTACTGG</span>
<span class="co">##                                                                        VJ_cdr3s_nt</span>
<span class="co">## 16             TGTGCTCTATGGTACAGCACCCATTATGTTTTC;TGCTGGCAAGGTACACATTTTCCTCAGACGTTC</span>
<span class="co">## 25             TGTCAGAATGATTATAGTTATCCGCTCACGTTC;TGTCAGCAATATAGCAGCTATCCGTACACGTTC</span>
<span class="co">## 30                   TGCAAGCAATCTTATAATCTGTACACGTTC;TGTCAACAGAGTAACAGCTGGCTCACGTTC</span>
<span class="co">## 42                TGCCAGCAGTGGAGTAGTAACCCGCTCACGTTC;TGCTGGCAAGGTACACATTTTCCTCCGTTC</span>
<span class="co">## 57 TGTGCTCTATGGTACAGCAACCATTTGGTGTTC;TGTGGTGTGGGTGATACAATTAAGGAACAATTTGTGTATGTTTTC</span>
<span class="co">## 67                TGTCAACAGTATTGGAGTACTCGGACGTTC;TGTCTACAGTATGATAATCTTCTGTACACGTTC</span>
<span class="co">##                                           VDJ_chain_contig</span>
<span class="co">## 16                             CACCTTGCAAACGCGA-1_contig_1</span>
<span class="co">## 25                             CGATGGCAGGGTATCG-1_contig_3</span>
<span class="co">## 30                             CTGATAGGTTCTGTTT-1_contig_3</span>
<span class="co">## 42 GCTGCTTTCTCTGAGA-1_contig_2;GCTGCTTTCTCTGAGA-1_contig_3</span>
<span class="co">## 57                             GTCACGGCACGCATCG-1_contig_1</span>
<span class="co">## 67                             TGACTAGTCGCCTGTT-1_contig_3</span>
<span class="co">##                                            VJ_chain_contig VDJ_chain VJ_chain</span>
<span class="co">## 16 CACCTTGCAAACGCGA-1_contig_2;CACCTTGCAAACGCGA-1_contig_3       IGH  IGL;IGK</span>
<span class="co">## 25 CGATGGCAGGGTATCG-1_contig_1;CGATGGCAGGGTATCG-1_contig_2       IGH  IGK;IGK</span>
<span class="co">## 30 CTGATAGGTTCTGTTT-1_contig_1;CTGATAGGTTCTGTTT-1_contig_2       IGH  IGK;IGK</span>
<span class="co">## 42 GCTGCTTTCTCTGAGA-1_contig_1;GCTGCTTTCTCTGAGA-1_contig_4   IGH;IGH  IGK;IGK</span>
<span class="co">## 57 GTCACGGCACGCATCG-1_contig_2;GTCACGGCACGCATCG-1_contig_3       IGH  IGL;IGL</span>
<span class="co">## 67 TGACTAGTCGCCTGTT-1_contig_1;TGACTAGTCGCCTGTT-1_contig_2       IGH  IGK;IGK</span>
<span class="co">##           VDJ_vgene            VJ_vgene VDJ_dgene   VDJ_jgene    VJ_jgene</span>
<span class="co">## 16         IGHV1-18     IGLV2;IGKV1-135   IGHD4-1       IGHJ1 IGLJ2;IGKJ1</span>
<span class="co">## 25         IGHV1-26   IGKV8-19;IGKV6-23                 IGHJ3 IGKJ5;IGKJ2</span>
<span class="co">## 30         IGHV1-80   IGKV8-21;IGKV5-43                 IGHJ1 IGKJ2;IGKJ5</span>
<span class="co">## 42 IGHV6-6;IGHV14-4  IGKV4-59;IGKV1-135           IGHJ4;IGHJ4 IGKJ5;IGKJ2</span>
<span class="co">## 57         IGHV1-76         IGLV1;IGLV3   IGHD2-5       IGHJ4 IGLJ1;IGLJ2</span>
<span class="co">## 67         IGHV14-2 IGKV13-84;IGKV19-93                 IGHJ2 IGKJ1;IGKJ2</span>
<span class="co">##    VDJ_cgene    VJ_cgene</span>
<span class="co">## 16      IGHM  IGLC2;IGKC</span>
<span class="co">## 25      IGHM   IGKC;IGKC</span>
<span class="co">## 30      IGHM   IGKC;IGKC</span>
<span class="co">## 42 IGHM;IGHM   IGKC;IGKC</span>
<span class="co">## 57      IGHM IGLC1;IGLC2</span>
<span class="co">## 67      IGHM   IGKC;IGKC</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      VDJ_sequence_nt_raw</span>
<span class="co">## 16                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             TTTGGGGAACATATGTCCAATGTCCTCTCCACAGGCACTGAACACACTGACTCTAACCATGGGATGGAGCTGGATCTTTCTCCTCTTCCTGTCAGGAACTGCAGGTGTCCTCTCTGAGGTCCAGCTGCAACAGTCTGGACCTGAGCTGGTGAAGCCTGGGGCTTCAGTGAAGATACCCTGCAAGGCTTCTGGATACACATTCACTGACTACAACATGGACTGGGTGAAGCAGAGCCATGGAAAGAGCCTTGAGTGGATTGGAGATATTAATCCTAACAATGGTGGTACTATCTACAACCAGAAGTTCAAGGGCAAGGCCACATTGACTGTAGACAAGTCCTCCAGCACAGCCTACATGGAGCTCCGCAGCCTGACATCTGAGGACACTGCAGTCTATTACTGTGCAAGAGGGGCTCCCAACTGGTACTTCGATGTCTGGGGCACAGGGACCACGGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 25                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             TTATGGGGAATGTCCTCACCACAGACACTGAACACACTGACTCTAACCATGGGATGGAGCTGGATCTTTCTCTTTCTCCTGTCAGGAACTGCAGGTGTCCTCTCTGAGGTCCAGCTGCAACAATCTGGACCTGAGCTGGTGAAGCCTGGGGCTTCAGTGAAGATATCCTGTAAGGCTTCTGGATACACGTTCACTGACTACTACATGAACTGGGTGAAGCAGAGCCATGGAAAGAGCCTTGAGTGGATTGGAGATATTAATCCTAACAATGGTGGTACTAGCTACAACCAGAAGTTCAAGGGCAAGGCCACATTGACTGTAGACAAGTCCTCCAGCACAGCCTACATGGAGCTCCGCAGCCTGACATCTGAGGACTCTGCAGTCTATTACTGTGCAAGATGGTTTGCCTGGTTTGCTTACTGGGGCCAAGGGACTCTGGTCACTGTCTCTGCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 30                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          TGGGGACAGTCCCTGAACACACTGACTCTAACCATGGAATGGCCTTTGATCTTTCTCTTCCTCCTGTCAGGAACTGCAGGTGTCCAATCCCAGGTTCAGCTGCAGCAGTCTGGGGCTGAGCTGGTGAAGCCTGGGGCCTCAGTGAAGATTTCCTGCAAAGCTTCTGGCTACGCATTCAGTAGCTACTGGATGAACTGGGTGAAGCAGAGGCCTGGAAAGGGTCTTGAGTGGATTGGACAGATTTATCCTGGAGATGGTGATACTAACTACAACGGAAAGTTCAAGGGCAAGGCCACACTGACTGCAGACAAATCCTCCAGCACAGCCTACATGCAGCTCAGCAGCCTGACCTCTGAGGACTCTGCGGTCTATTTCTGTGCAAAAAACTATGGTAGTAGCTACAGCTACTGGTACTTCGATGTCTGGGGCACAGGGACCACGGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 42 TGGGGGACAGATGCACAAACCTGGACTCACAAGTTTTTCTCTTCAGTGACAAACACAGACATAGAACATTCACGATGTACTTGGGACTGAACTGTGTATTCATAGTTTTTCTCTTAAAAGGTGTCCAGAGTGAAGTGAAGCTTGAGGAGTCTGGAGGAGGCTTGGTGCAACCTGGAGGATCCATGAAACTCTCTTGTGCTGCCTCTGGATTCACTTTTAGTGACGCCTGGATGGACTGGGTCCGCCAGTCTCCAGAGAAGGGGCTTGAGTGGGTTGCTGAAATTAGAAACAAAGCTAATAATCATGCAACATACTATGCTGAGTCTGTGAAAGGGAGGTTCACCATCTCAAGAGATGATTCCAAAAGTAGTGTCTACCTGCAAATGAACAGCTTAAGAGCTGAAGACACTGGCATTTATTACTGTACCGGGGATTACGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA;TTATGGGGATGAACACTGTTTTCTCTACAGTCACTGAATCTCAATGTCCTTACAATGAAATGCAGCTGGGTCATCTTCTTCCTGATGGCAGTGGTTATAGGGGTCAATTCAGAGGTTCAGCTGCAGCAGTCTGGGGCTGAGCTTGTGAGGCCAGGGGCCTCAGTCAAGTTGTCCTGCACAGCTTCTGGCTTTAACATTAAAGACGACTATATGCACTGGGTGAAGCAGAGGCCTGAACAGGGCCTGGAGTGGATTGGATGGATTGATCCTGAGAATGGTGATACTGAATATGCCTCGAAGTTCCAGGGCAAGGCCACTATAACAGCAGACACATCCTCCAACACAGCCTACCTGCAGCTCAGCAGCCTGACATCTGAGGACACTGCCGTCTATTACTGTACTCTCATTACTACGGTAGTAGCCAAGGATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 57                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                TCTGTCACGGCACGCATCGTTTTATCCGTTTACTTTTATGGGGATGATCAGTGTCCTCTCTACACAGTCCCTGACGACACTGATTCTAACCATGGGATGGAGCTGGATCTTTTTCTTCCTCCTGTCAGGAACTGCAGGTGTCCACTGTCAGGTCCAGCTGAAGCAGTCTGGGGCTGAGCTGGTGAGGCCTGGGGCTTCAGTGAAGCTGTCCTGCAAGGCTTCTGGCTACACTTTCACTGACTACTATATAAACTGGGTGAAGCAGAGGCCTGGACAGGGACTTGAGTGGATTGCAAGGATTTATCCTGGAAGTGGTAATACTTACTACAATGAGAAGTTCAAGGGCAAGGCCACACTGACTGCAGAAAAATCCTCCAGCACTGCCTACATGCAGCTCAGCAGCCTGACATCTGAGGACTCTGCTGTCTATTTCTGTGCAAGAAGGCCCTACTATAGTAACTCCCACTATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">## 67                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       TATTTGGGGATGAACCCTGTCTTCTCTACAGCCACTGAATCTCAAGGTCCTTACAATGAAATGCAGCTGGATCATCTTCTTCCTGATGGCAGTGGTTACAGGGGTCAATTCAGAGGTTCAGCTGCAGCAGTCTGGGGCAGAGCTTGTGAAGCCAGGGGCCTCAGTCAAGTTGTCCTGCACAGCTTCTGGCTTCAACATTAAAGACTACTATATGCACTGGGTGAAGCAGAGGACTGAACAGGGCCTGGAGTGGATTGGAAGGATTGATCCTGAGGATGGTGAAACTAAATATGCCCCGAAATTCCAGGGCAAGGCCACTATAACAGCAGACACATCCTCCAACACAGCCTACCTGCAGCTCAGCAGCCTGACATCTGAGGACACTGCCGTCTATTACTGTGCTAGAGATTACTACGGTAGTAGCTTGTACTACTTTGACTACTGGGGCCAAGGCACCACTCTCACAGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            VJ_sequence_nt_raw</span>
<span class="co">## 16                             ACTCACCTTGCAAACGCGATCAACAGTTGTATCTTTTATGGGGGACCAATATTGAAAATAATAGACTTGGTTTGTGAATTATGGCCTGGACTTCACTTATACTCTCTCTCCTGGCTCTCTGCTCAGGAGCCAGTTCCCAGGCTGTTGTGACTCAGGAATCTGCACTCACCACATCACCTGGTGGAACAGTCATACTCACTTGTCGCTCAAGTACTGGGGCTGTTACAACTAGTAACTATGCCAACTGGGTCCAAGAAAAACCAGATCATTTATTCACTGGTCTAATAGGTGGTACCAGCAACCGAGCTCCAGGTGTTCCTGTCAGATTCTCAGGCTCCCTGATTGGAGACAAGGCTGCCCTCACCATCACAGGGGCACAGACTGAGGATGATGCAATGTATTTCTGTGCTCTATGGTACAGCACCCATTATGTTTTCGGCGGTGGAACCAAGGTCACTGTCCTAGGTCAGCCCAAGTCCACTCCCACTCTCACCGTGTTTCCACCTTCCTCTGAGGAGCTCAAGGAAAACAAAGCCACACTGGTGTGTCTGATTTCCAACTTTTCCCCGAGTGGTGTGACAGTGGCCTG;ACTGATCACTCTCCTATGTTCATTTCCTCAAAATGATGAGTCCTGCCCAGTTCCTGTTTCTGTTAGTGCTCTGGATTCGGGAAACCAACGGTGATGTTGTGATGACCCAGACTCCACTCACTTTGTCGGTTACCATTGGACAACCAGCCTCCATCTCTTGCAAGTCAAGTCAGAGCCTCTTAGATAGTGATGGAAAGACATATTTGAATTGGTTGTTACAGAGGCCAGGCCAGTCTCCAAAGCGCCTAATCTATCTGGTGTCTAAACTGGACTCTGGAGTCCCTGACAGGTTCACTGGCAGTGGATCAGGGACAGATTTCACACTGAAAATCAGCAGAGTGGAGGCTGAGGATTTGGGAGTTTATTATTGCTGGCAAGGTACACATTTTCCTCAGACGTTCGGTGGAGGCACCAAGCTGGAAATCAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">## 25                                                                                                               TTTATGGGGACATCTGAAAGGCAGGTGGAGCAAGATGGAATCACAGACTCAGGTCCTCATGTCCCTGCTGTTCTGGGTATCTGGTACCTGTGGGGACATTGTGATGACACAGTCTCCATCCTCCCTGACTGTGACAGCAGGAGAGAAGGTCACTATGAGCTGCAAGTCCAGTCAGAGTCTGTTAAACAGTGGAAATCAAAAGAACTACTTGACCTGGTACCAGCAGAAACCAGGGCAGCCTCCTAAACTGTTGATCTACTGGGCATCCACTAGGGAATCTGGGGTCCCTGATCGCTTCACAGGCAGTGGATCTGGAACAGATTTCACTCTCACCATCAGCAGTGTGCAGGCTGAAGACCTGGCAGTTTATTACTGTCAGAATGATTATAGTTATCCGCTCACGTTCGGTGCTGGGACCAAGCTGGAGCTGAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC;TTATGGGGAAATACATCAGACCAGCATGGGCATCAAGATGGAGACACATTCTCAGGTCTTTGTATACATGTTGCTGTGGTTGTCTGGTGTTGAAGGAGACATTGTGATGACCCAGTCTCACAAATTCATGTCCACATCAGTAGGAGACAGGGTCAGCATCACCTGCAAGGCCAGTCAGGATGTGGGTACTGCTGTAGCCTGGTATCAACAGAAACCAGGGCAATCTCCTAAACTACTGATTTACTGGGCATCCACCCGGCACACTGGAGTCCCTGATCGCTTCACAGGCAGTGGATCTGGGACAGATTTCACTCTCACCATTAGCAATGTGCAGTCTGAAGACTTGGCAGATTATTTCTGTCAGCAATATAGCAGCTATCCGTACACGTTCGGAGGGGGGACCAAGCTGGAAATAAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">## 30                                                                                                                                        AGACAGGCAGTGGGAGCAAGATGGATTCACAGGCCCAGGTTCTTATATTGCTGCTGCTATGGGTATCTGGTACCTGTGGGGACATTGTGATGTCACAGTCTCCATCCTCCCTGGCTGTGTCAGCAGGAGAGAAGGTCACTATGAGCTGCAAATCCAGTCAGAGTCTGCTCAACAGTAGAACCCGAAAGAACTACTTGGCTTGGTACCAGCAGAAACCAGGGCAGTCTCCTAAACTGCTGATCTACTGGGCATCCACTAGGGAATCTGGGGTCCCTGATCGCTTCACAGGCAGTGGATCTGGGACAGATTTCACTCTCACCATCAGCAGTGTGCAGGCTGAAGACCTGGCAGTTTATTACTGCAAGCAATCTTATAATCTGTACACGTTCGGAGGGGGGACCAAGCTGGAAATAAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC;ATGAGCCACACAAACTCAGGGAAAGCTCGAAGATGGTTTTCACACCTCAGATACTTGGACTTATGCTTTTTTGGATTTCAGCCTCCAGAGGTGATATTGTGCTAACTCAGTCTCCAGCCACCCTGTCTGTGACTCCAGGAGATAGCGTCAGTCTTTCCTGCAGGGCCAGCCAAAGTATTAGCAACAACCTACACTGGTATCAACAAAAATCACATGAGTCTCCAAGGCTTCTCATCAAGTATGCTTCCCAGTCCATCTCTGGGATCCCCTCCAGGTTCAGTGGCAGTGGATCAGGGACAGATTTCACTCTCAGTATCAACAGTGTGGAGACTGAAGATTTTGGAATGTATTTCTGTCAACAGAGTAACAGCTGGCTCACGTTCGGTGCTGGGACCAAGCTGGAGCTGAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">## 42                                                                                           TTATGGGGAAAGTACTTATGAGAATAGCAGTAATTAGCTAGGGACCAAAATTCAAAGACAAAATGGATTTTCAAGTGCAGATTTTCAGCTTCCTGCTAATCAGTGCCTCAGTCATAATATCCAGAGGACAAATTGTTCTCACCCAGTCTCCAGCAATCATGTCTGCATCTCCAGGGGAGAAGGTCACCATGACCTGCAGTGCCAGCTCAAGTGTAAGTTACATGCACTGGTACCAGCAGAAGTCAGGCACCTCCCCCAAAAGATGGATTTATGACACATCCAAACTGGCTTCTGGAGTCCCTGCTCGCTTCAGTGGCAGTGGGTCTGGGACCTCTTACTCTCTCACAATCAGCAGCATGGAGGCTGAAGATGCTGCCACTTATTACTGCCAGCAGTGGAGTAGTAACCCGCTCACGTTCGGTGCTGGGACCAAGCTGGAGCTGAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC;ACTGATCACTCTCCTATGTTCATTTCCTCAAAATGATGAGTCCTGCCCAGTTCCTGTTTCTGTTAGTGCTCTGGATTCGGGAAACCAACGGTGATGTTGTGATGACCCAGACTCCACTCACTTTGTCGGTTACCATTGGACAACCAGCCTCCATCTCTTGCAAGTCAAGTCAGAGCCTCTTAGATAGTGATGGAAAGACATATTTGAATTGGTTGTTACAGAGGCCAGGCCAGTCTCCAAAGCGCCTAATCTATCTGGTGTCTAAACTGGACTCTGGAGTCCCTGACAGGTTCACTGGCAGTGGATCAGGGACAGATTTCACACTGAAAATCAGCAGAGTGGAGGCTGAGGATTTGGGAGTTTATTATTGCTGGCAAGGTACACATTTTCCTCCGTTCGGAGGGGGGACCAAGCTGGAAATAAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">## 57 GGGGACCAATATTGAAAAGAATAGACCTGGTTTGTGAATTATGGCCTGGATTTCACTTATACTCTCTCTCCTGGCTCTCAGCTCAGGGGCCATTTCCCAGGCTGTTGTGACTCAGGAATCTGCACTCACCACATCACCTGGTGAAACAGTCACACTCACTTGTCGCTCAAGTACTGGGGCTGTTACAACTAGTAACTATGCCAACTGGGTCCAAGAAAAACCAGATCATTTATTCACTGGTCTAATAGGTGGTACCAACAACCGAGCTCCAGGTGTTCCTGCCAGATTCTCAGGCTCCCTGATTGGAGACAAGGCTGCCCTCACCATCACAGGGGCACAGACTGAGGATGAGGCAATATATTTCTGTGCTCTATGGTACAGCAACCATTTGGTGTTCGGTGGAGGAACCAAACTGACTGTCCTAGGCCAGCCCAAGTCTTCGCCATCAGTCACCCTGTTTCCACCTTCCTCTGAAGAGCTCGAGACTAACAAGGCCACACTGGTGTGTA;TCTGTCACGGCACGCATCGGACTATCAAATATCTTCTATGGGAGAGAGAACTACAACCTGTCTGTCTCAGCAGAGATCAGTAGTACCTGCATTATGGCCTGGACTCCTCTCTTCTTCTTCTTTGTTCTTCATTGCTCAGGTTCTTTCTCCCAACTTGTGCTCACTCAGTCATCTTCAGCCTCTTTCTCCCTGGGAGCCTCAGCAAAACTCACGTGCACCTTGAGTAGTCAGCACAGTACGTACACCATTGAATGGTATCAGCAACAGCCACTCAAGCCTCCTAAGTATGTGATGGAGCTTAAGAAAGATGGAAGCCACAGCACAGGTGATGGGATTCCTGATCGCTTCTCTGGATCCAGCTCTGGTGCTGATCGCTACCTTAGCATTTCCAACATCCAGCCTGAAGATGAAGCAATATACATCTGTGGTGTGGGTGATACAATTAAGGAACAATTTGTGTATGTTTTCGGCGGTGGAACCAAGGTCACTGTCCTAGGTCAGCCCAAGTCCACTCCCACTCTCACCGTGTTTCCACCTTCCTCTGAGGAGCTCAAGGAAAACAAAGCCACACTGGTGTGTCTGATTTCCAACTTTTCCCCGAGTGGTGTGACAGTGGCCTG</span>
<span class="co">## 67                                                                                                           TGGGGAATGTCAGGTCACAGCAGAAACATGAAGTTTCCTTCTCAACTTCTGCTCTTACTGCTGTTTGGAATCCCAGGCATGATATGTGACATCCAGATGACACAATCTTCATCCTCCTTTTCTGTATCTCTAGGAGACAGAGTCACCATTACTTGCAAGGCAAGTGAGGACATATATAATCGGTTAGCCTGGTATCAGCAGAAACCAGGAAATGCTCCTAGGCTCTTAATATCTGGTGCAACCAGTTTGGAAACTGGGGTTCCTTCAAGATTCAGTGGCAGTGGATCTGGAAAGGATTACACTCTCAGCATTACCAGTCTTCAGACTGAAGATGTTGCTACTTATTACTGTCAACAGTATTGGAGTACTCGGACGTTCGGTGGAGGCACCAAGCTGGAAATCAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC;TGACTAGTCGCCTGTTCGTACTTCGTTTTATTATTTGGGAGTCATTCTTGGTCAGGAGACGTTGTAGAAATGAGACCGTCTATTCAGTTCCTGGGGCTCTTGTTGTTCTGGCTTCATGGTGCTCAGTGTGACATCCAGATGACACAGTCTCCATCCTCACTGTCTGCATCTCTGGGAGGCAAAGTCACCATCACTTGCAAGGCAAGCCAAGACATTAACAAGTATATAGCTTGGTACCAACACAAGCCTGGAAAAGGTCCTAGGCTGCTCATACATTACACATCTACATTACAGCCAGGCATCCCATCAAGGTTCAGTGGAAGTGGGTCTGGGAGAGATTATTCCTTCAGCATCAGCAACCTGGAGCCTGAAGATATTGCAACTTATTATTGTCTACAGTATGATAATCTTCTGTACACGTTCGGAGGGGGGACCAAGCTGGAAATAAAACGGGCTGATGCTGCACCAACTGTATCCATCTTCCCACCATCCAGTGAGCAGTTAACATCTGGAGGTGCCTCAGTCGTGTGCTTC</span>
<span class="co">##    VDJ_sequence_nt_trimmed VJ_sequence_nt_trimmed VDJ_sequence_aa</span>
<span class="co">## 16                                                               </span>
<span class="co">## 25                                                               </span>
<span class="co">## 30                                                               </span>
<span class="co">## 42                                                               </span>
<span class="co">## 57                                                               </span>
<span class="co">## 67                                                               </span>
<span class="co">##    VJ_sequence_aa VDJ_trimmed_ref VJ_trimmed_ref     VDJ_raw_consensus_id</span>
<span class="co">## 16                                               clonotype28_concat_ref_1</span>
<span class="co">## 25                                                clonotype2_concat_ref_1</span>
<span class="co">## 30                                               clonotype57_concat_ref_1</span>
<span class="co">## 42                                               clonotype22_concat_ref_1</span>
<span class="co">## 57                                               clonotype60_concat_ref_1</span>
<span class="co">## 67                                               clonotype52_concat_ref_1</span>
<span class="co">##         VJ_raw_consensus_id     orig_barcode clonotype_frequency specifity</span>
<span class="co">## 16 clonotype28_concat_ref_3 CACCTTGCAAACGCGA                   1        NA</span>
<span class="co">## 25  clonotype2_concat_ref_2 CGATGGCAGGGTATCG                   2        NA</span>
<span class="co">## 30 clonotype57_concat_ref_2 CTGATAGGTTCTGTTT                   1        NA</span>
<span class="co">## 42 clonotype22_concat_ref_3 GCTGCTTTCTCTGAGA                   1        NA</span>
<span class="co">## 57 clonotype60_concat_ref_2 GTCACGGCACGCATCG                   1        NA</span>
<span class="co">## 67 clonotype52_concat_ref_3 TGACTAGTCGCCTGTT                   1        NA</span>
<span class="co">##    affinity     batches clonotype_id</span>
<span class="co">## 16       NA Unspecified  clonotype28</span>
<span class="co">## 25       NA Unspecified   clonotype2</span>
<span class="co">## 30       NA Unspecified  clonotype57</span>
<span class="co">## 42       NA Unspecified  clonotype22</span>
<span class="co">## 57       NA Unspecified  clonotype60</span>
<span class="co">## 67       NA Unspecified  clonotype52</span></code></pre>
</div>
<div class="section level3">
<h3 id="filtering-chains-by-umi-counts">4.1.2 Filtering chains by UMI counts<a class="anchor" aria-label="anchor" href="#filtering-chains-by-umi-counts"></a>
</h3>
<p>While filtering out all cells with aberrant chain numbers is often
used during processing of VDJ data, Platypus offers a format which can
accomodate and integrate these cells into analysis. (See
VDJ_clonotype_v3()). A third option was proposed by Zhang W et al. (Sci
Adv. 2021 10.1126/sciadv.abf5835): To choose between excess chains, by
the count of unique molecular identifier of each contig (UMIs). The VGM
function implements this strategy with two parameters: 1.
select.excess.chains.by.umi.count Is a boolean. Once set to TRUE the VGM
will filter excess chains based on UMI count</p>
<ol start="2" style="list-style-type: decimal">
<li>excess.chain.confidence.count.threshold is an integer that allows to
tweak filtering. It defaults to 1000 (any large value would achive the
same). To illustrate the filtering behavior we can consider the
following example with cells having 2 VJ chains</li>
</ol>
<p>select.excess.chains.by.umi.count = T</p>
<table class="table">
<tr>
<th>
barcode
</th>
<th>
Nr_of_VJ_chains
</th>
<th>
VJ_UMIs
</th>
</tr>
<tr>
<td>
Cell 1
</td>
<td>
2
</td>
<td>
1;1
</td>
</tr>
<tr>
<td>
Cell 2
</td>
<td>
2
</td>
<td>
1;5
</td>
</tr>
<tr>
<td>
Cell 3
</td>
<td>
2
</td>
<td>
3;3
</td>
</tr>
</table>
<p>FOR: excess.chain.confidence.count.threshold = 1000</p>
<p>Cell 1 -&gt; both chains are below the threshold and are therefore
subject to filtering. Given that both chains have the same UMI count, a
one contig is eliminated at random. Cell 2 -&gt; both chains are below
the threshold and subject to filtering. Chain 2 has the higher UMI and
is therefore kept Cell 3 -&gt; proceeds as for Cell 1</p>
<p>FOR: excess.chain.confidence.count.threshold = 3</p>
<p>Cell 1 -&gt; same as above Cell 2 -&gt; chain 1 is below threshold
and subject to filtering. Chain 2 is above threshold and therefore
considered a high confidence chain and not filtered Cell 3 -&gt; Both
chains are equal or above threshold, both are considered high confident
and no chain is filtered.</p>
<p>UMI distribution within VDJ datasets can vary. To optimize this
filtering parameter we recommend investigating UMI frequencies in cells
with double chains.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>,
                      select.excess.chains.by.umi.count <span class="op">=</span> <span class="cn">T</span>,
                      excess.chain.confidence.count.threshold <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span> </code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="vdj-trimming-and-aligning-and-translating">4.2 VDJ trimming and aligning and translating<a class="anchor" aria-label="anchor" href="#vdj-trimming-and-aligning-and-translating"></a>
</h2>
<p>It may be useful to obtain full-length sequences (e.g., for
phylogenetics or experimental expression and validation, see Pogson M.
et al. Nat Comm. 2016, 10.1038/ncomms12535) from single-cell sequencing
data. We have therefore added an option to recover both trimmed and
untrimmed sequences as determined by the cellranger vdj function. If the
user sets the trim.and.align parameter to TRUE, this will return
processed VDJ contig strings as detailed below:</p>
<ol style="list-style-type: decimal">
<li><p>Trimming: The raw nt contig sequences are trimmed to the start of
the V segment and the end of the J segment. (VDJ/VJ_sequence_nt_trimmed)
These trimmed sequences contain signal peptide information. In the case
that a user does not want this, we currently use MIXCR to realign and
extract the sequence that spans FR1 to FR4 for both heavy and light
chains. This can be done using the VDJ_call_MiXCR() function within
Platypus, although it does require the user to download mixcr
locally.</p></li>
<li><p>Translation: The trimmed contigs are translated to yield full
amino acid sequences. (VDJ/VJ_sequence_aa)</p></li>
<li><p>Realigning to reference: The untrimmed germline sequences (as
determined by 10x genomics) for each clonotype consensus sequence are
returned are returned when trim.and.align = T. This can be used on IMGT
or with MIXCR to further provide a shorter reference sequence that
covers just the FR1 to FR4. Furthermore the function aligns trimmed
contig sequences to reference contigs using
Biostrings::pairwiseAlignment with alignment = “local”; see optimization
parameters in section 4.1.2). This results in a reference sequence from
the start of the V segment to the end of the J segment.
(VDJ/VJ_trimmed_ref) Most of the time this trimmed germline sequence
will be out of frame given the CDR3 deletions and insertions. In the
case that a user wants an in-frame germline (e.g. for expression), the
CDR3 will likely need to be replace. We are currently developing a
pipeline that supplies CDR3s of those recovered sequences that are
closest to germline - so stay tuned :)</p></li>
</ol>
<p>The trimmed contigs are aligned to the sequences of the
concat_ref.fasta file and the reference is trimed accordingly. (This is
done using: Biostrings::pairwiseAlignment with alignment = “local”; see
optimization parameters in section 4.1.2)</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>, 
                      trim.and.align <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## [1] "2022-03-15 16:04:39 CET"</span>
<span class="co">## [1] "2022-03-15 16:04:39 CET"</span>
<span class="co">## [1] "2022-03-15 16:04:53 CET"</span>
<span class="co">## [1] "2022-03-15 16:05:14 CET"</span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VDJ_sequence_nt_raw"</span>, <span class="st">"VDJ_sequence_nt_trimmed"</span>, <span class="st">"VDJ_sequence_aa"</span>, <span class="st">"VDJ_trimmed_ref"</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            VDJ_sequence_nt_raw</span>
<span class="co">## 1 TGGGAAGTGTGCAGCCATGGGCAGGCTTACTTCTTCATTCCTGTTACTGATTGTCCCTGCATATGTCCTGTCCCAGGTTACTCTGAAAGAGTCTGGCCCTGGGATATTGCAGCCCTCCCAGACCCTCAGTCTGACTTGTTCTTTCTCTGGGTTTTCACTGAGCACTTTTGGTATGGGTGTAGGCTGGATTCGTCAGCCTTCAGGGAAGGGTCTGGAGTGGCTGGCACACATTTGGTGGGATGATGATAAGTACTATAACCCAGCCCTGAAGAGTCGGCTCACAATCTCCAAGGATACCTCCAAAAACCAGGTATTCCTCAAGATCGCCAATGTGGACACTGCAGATACTGCCACATACTACTGTGCTCGAATAGGATATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCAGAGAGTCAGTCCTTCCCAAATGTCTTCCCCCTCGTCTCCTGCGAGAGCCCCCTGTCTGATAAGAATCTGGTGGCCATGGGCTGCCTGGCCCGGGACTTCCTGCCCAGCACCATTTCCTTCACCTGGAACTACCAGAACAACACTGAAGTCATCCAGGGTATCAGAACCTTCCCAACACTGAGGACAGGGGGCAAGTACCTAGCCACCTCGCA</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                    VDJ_sequence_nt_trimmed</span>
<span class="co">## 1 ATGGGCAGGCTTACTTCTTCATTCCTGTTACTGATTGTCCCTGCATATGTCCTGTCCCAGGTTACTCTGAAAGAGTCTGGCCCTGGGATATTGCAGCCCTCCCAGACCCTCAGTCTGACTTGTTCTTTCTCTGGGTTTTCACTGAGCACTTTTGGTATGGGTGTAGGCTGGATTCGTCAGCCTTCAGGGAAGGGTCTGGAGTGGCTGGCACACATTTGGTGGGATGATGATAAGTACTATAACCCAGCCCTGAAGAGTCGGCTCACAATCTCCAAGGATACCTCCAAAAACCAGGTATTCCTCAAGATCGCCAATGTGGACACTGCAGATACTGCCACATACTACTGTGCTCGAATAGGATATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCA</span>
<span class="co">##                                                                                                                            VDJ_sequence_aa</span>
<span class="co">## 1 MGRLTSSFLLLIVPAYVLSQVTLKESGPGILQPSQTLSLTCSFSGFSLSTFGMGVGWIRQPSGKGLEWLAHIWWDDDKYYNPALKSRLTISKDTSKNQVFLKIANVDTADTATYYCARIGYAMDYWGQGTSVTVSS</span>
<span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                               VDJ_trimmed_ref</span>
<span class="co">## 1 ATGGGCAGGCTTACTTCTTCATTCCTGTTACTGATTGTCCCTGCATATGTCCTGTCCCAGGTTACTCTGAAAGAGTCTGGCCCTGGGATATTGCAGCCCTCCCAGACCCTCAGTCTGACTTGTTCTTTCTCTGGGTTTTCACTGAGCACTTTTGGTATGGGTGTAGGCTGGATTCGTCAGCCTTCAGGGAAGGGTCTGGAGTGGCTGGCACACATTTGGTGGGATGATGATAAGTACTATAACCCAGCCCTGAAGAGTCGGCTCACAATCTCCAAGGATACCTCCAAAAACCAGGTATTCCTCAAGATCGCCAATGTGGACACTGCAGATACTGCCACATACTACTGTGCTCGAATAGATTACTATGCTATGGACTACTGGGGTCAAGGAACCTCAGTCACCGTCTCCTCA</span></code></pre>
<div class="section level3">
<h3 id="optimizing-parameters-of-vdj-aligning">4.2.1 Optimizing parameters of VDJ aligning<a class="anchor" aria-label="anchor" href="#optimizing-parameters-of-vdj-aligning"></a>
</h3>
<p>A key difficulty of TCR and BCR sequences, is that CDR3-containing
junction regions almost never matches known references. This leaves
alignment gaps and can lead to the highest scoring alignment to not
contain the junction-following J segment. The VGM therefore allows to
tweak the alignment parameters gap.opening.cost and
gap.extension.cost</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>, 
                      trim.and.align <span class="op">=</span> <span class="cn">T</span>, 
                      gap.opening.cost <span class="op">=</span> <span class="fl">10</span>, 
                      gap.extension.cost <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> </code></pre></div>
</div>
<div class="section level3">
<h3 id="optimizing-alignment-runtime">4.2.2 Optimizing alignment runtime<a class="anchor" aria-label="anchor" href="#optimizing-alignment-runtime"></a>
</h3>
<p>To accelerate alignment the VGM allows for multicore processing via
the functions mclapply or parlapply depending on the operating system
from the Parallel package. Given initiation times of multicore
processes, we only recommend using it for datasets with &gt;2000
cells.</p>
<p>The parameter numcores can be used to set the number of cores used.
This defaults to all available cores, therefore, setting a limit when
running the function on a cluster is crucial.</p>
<p>By default parallel.processing is set to “none” and the function uses
standard lapply</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#For LINUX and WINDOWS users</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>, 
                      trim.and.align <span class="op">=</span> <span class="cn">T</span>, 
                      parallel.processing <span class="op">=</span> <span class="st">"parlapply"</span>,
                      numcores <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> 

<span class="co">#For MAC users</span>
<span class="va">vgm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_GEX_matrix.html">VDJ_GEX_matrix</a></span><span class="op">(</span>VDJ.out.directory.list <span class="op">=</span> <span class="va">VDJ.out.directory.list</span>, 
                      trim.and.align <span class="op">=</span> <span class="cn">T</span>, 
                      parallel.processing <span class="op">=</span> <span class="st">"mclapply"</span>,
                      numcores <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> </code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">5. Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>We hope that this comprehensive vignette of the VGM function enables
any user to employ the full functionality of this function. If you have
any issues, requests or ideas concerning existing or new features,
please reach out to us via Github or Email.</p>
</div>
<div class="section level2">
<h2 id="version-information">6. Version information<a class="anchor" aria-label="anchor" href="#version-information"></a>
</h2>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.0.5 (2021-03-31)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 19044)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252   </span>
<span class="co">## [3] LC_MONETARY=German_Germany.1252 LC_NUMERIC=C                   </span>
<span class="co">## [5] LC_TIME=German_Germany.1252    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] Platypus_3.3.3     SeuratObject_4.0.4 Seurat_4.1.0       forcats_0.5.1     </span>
<span class="co">##  [5] stringr_1.4.0      purrr_0.3.4        readr_2.1.2        tidyr_1.1.3       </span>
<span class="co">##  [9] tibble_3.1.2       ggplot2_3.3.5      tidyverse_1.3.1    dplyr_1.0.7       </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] readxl_1.3.1          backports_1.4.1       systemfonts_1.0.4    </span>
<span class="co">##   [4] plyr_1.8.6            igraph_1.2.6          lazyeval_0.2.2       </span>
<span class="co">##   [7] splines_4.0.5         listenv_0.8.0         scattermore_0.7      </span>
<span class="co">##  [10] digest_0.6.27         useful_1.2.6          htmltools_0.5.2      </span>
<span class="co">##  [13] fansi_0.5.0           magrittr_2.0.1        memoise_2.0.1        </span>
<span class="co">##  [16] tensor_1.5            cluster_2.1.2         ROCR_1.0-11          </span>
<span class="co">##  [19] tzdb_0.2.0            Biostrings_2.58.0     globals_0.14.0       </span>
<span class="co">##  [22] modelr_0.1.8          matrixStats_0.59.0    pkgdown_2.0.2        </span>
<span class="co">##  [25] spatstat.sparse_2.0-0 colorspace_2.0-2      rvest_1.0.2          </span>
<span class="co">##  [28] ggrepel_0.9.1         textshaping_0.3.6     haven_2.4.3          </span>
<span class="co">##  [31] xfun_0.27             crayon_1.5.0          jsonlite_1.7.2       </span>
<span class="co">##  [34] spatstat.data_2.1-2   survival_3.2-11       zoo_1.8-9            </span>
<span class="co">##  [37] glue_1.4.2            polyclip_1.10-0       gtable_0.3.0         </span>
<span class="co">##  [40] zlibbioc_1.36.0       XVector_0.30.0        seqinr_4.2-8         </span>
<span class="co">##  [43] leiden_0.3.9          future.apply_1.8.1    BiocGenerics_0.36.1  </span>
<span class="co">##  [46] abind_1.4-5           scales_1.1.1          DBI_1.1.2            </span>
<span class="co">##  [49] miniUI_0.1.1.1        Rcpp_1.0.7            viridisLite_0.4.0    </span>
<span class="co">##  [52] xtable_1.8-4          reticulate_1.20       spatstat.core_2.2-0  </span>
<span class="co">##  [55] stats4_4.0.5          htmlwidgets_1.5.4     httr_1.4.2           </span>
<span class="co">##  [58] RColorBrewer_1.1-2    ellipsis_0.3.2        ica_1.0-2            </span>
<span class="co">##  [61] farver_2.1.0          pkgconfig_2.0.3       uwot_0.1.10          </span>
<span class="co">##  [64] sass_0.4.0            dbplyr_2.1.1          deldir_0.2-10        </span>
<span class="co">##  [67] utf8_1.2.1            labeling_0.4.2        tidyselect_1.1.1     </span>
<span class="co">##  [70] rlang_0.4.10          reshape2_1.4.4        later_1.2.0          </span>
<span class="co">##  [73] munsell_0.5.0         cellranger_1.1.0      tools_4.0.5          </span>
<span class="co">##  [76] cachem_1.0.6          cli_3.1.1             generics_0.1.2       </span>
<span class="co">##  [79] ade4_1.7-18           broom_0.7.12          ggridges_0.5.3       </span>
<span class="co">##  [82] evaluate_0.14         fastmap_1.1.0         yaml_2.2.1           </span>
<span class="co">##  [85] ragg_1.2.1            goftest_1.2-2         knitr_1.37           </span>
<span class="co">##  [88] fs_1.5.2              fitdistrplus_1.1-6    RANN_2.6.1           </span>
<span class="co">##  [91] pbapply_1.5-0         future_1.23.0         nlme_3.1-152         </span>
<span class="co">##  [94] mime_0.11             xml2_1.3.3            compiler_4.0.5       </span>
<span class="co">##  [97] rstudioapi_0.13       plotly_4.10.0         png_0.1-7            </span>
<span class="co">## [100] spatstat.utils_2.2-0  reprex_2.0.1          bslib_0.3.1          </span>
<span class="co">## [103] stringi_1.7.4         highr_0.9             RSpectra_0.16-0      </span>
<span class="co">## [106] desc_1.4.0            lattice_0.20-44       Matrix_1.3-4         </span>
<span class="co">## [109] vctrs_0.3.8           pillar_1.7.0          lifecycle_1.0.1      </span>
<span class="co">## [112] spatstat.geom_2.2-0   lmtest_0.9-38         jquerylib_0.1.4      </span>
<span class="co">## [115] RcppAnnoy_0.0.18      data.table_1.14.0     cowplot_1.1.1        </span>
<span class="co">## [118] irlba_2.3.3           httpuv_1.6.1          patchwork_1.1.1      </span>
<span class="co">## [121] R6_2.5.1              promises_1.2.0.1      KernSmooth_2.23-20   </span>
<span class="co">## [124] gridExtra_2.3         IRanges_2.24.1        parallelly_1.30.0    </span>
<span class="co">## [127] codetools_0.2-18      MASS_7.3-54           assertthat_0.2.1     </span>
<span class="co">## [130] rprojroot_2.0.2       withr_2.4.3           sctransform_0.3.3    </span>
<span class="co">## [133] S4Vectors_0.28.1      mgcv_1.8-36           parallel_4.0.5       </span>
<span class="co">## [136] hms_1.1.1             grid_4.0.5            rpart_4.1-15         </span>
<span class="co">## [139] rmarkdown_2.11        Rtsne_0.15            shiny_1.7.1          </span>
<span class="co">## [142] lubridate_1.8.0</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alexander Yermanos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
