<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ Platypus</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette.html">FOR TESTING ONLY / Platypus vignette / FOR TESTING ONLY</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="automate_GEX.html">automate_GEX()</a></code> </p>
        </td>
        <td><p>Automates the transcriptional analysis of the gene expression libraries from cellranger. This function will integrate multiple samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="call_MIXCR.html">call_MIXCR()</a></code> </p>
        </td>
        <td><p>Extracts information on the VDJRegion level using MiXCR. This function assumes the user can run an executable instance of MiXCR and is elgible to use MiXCR as determined by license agreements. The VDJRegion corresponds to the recombined heavy and light chain loci starting from framework region 1 (FR1) and extending to frame work region 4 (FR4). This can be useful for extracting full-length sequences ready to clone and further calculating somatic hypermutation occurances.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_automate.html">GEX_automate()</a></code> </p>
        </td>
        <td><p>Automates the transcriptional analysis of the gene expression libraries from cellranger. This function will integrate multiple samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_clonotype.html">GEX_clonotype()</a></code> </p>
        </td>
        <td><p>Integrates VDJ and gene expression libraries by providing cluster membership seq_per_vdj object and the index of the cell in the Seurat RNA-seq object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_cluster_genes.html">GEX_cluster_genes()</a></code> </p>
        </td>
        <td><p>Extracts the differentially expressed genes between two samples. This function uses the FindMarkers function from the Seurat package. Further parameter control can be accomplished by calling the function directly on the output of automate_GEX or VDJ_GEX_matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_cluster_genes_heatmap.html">GEX_cluster_genes_heatmap()</a></code> </p>
        </td>
        <td><p>Produces a heatmap displaying the expression of the top genes that define each cluster in the Seurat object. The output heatmap is derived from DoHeatmap from Seurat and thereby can be edited using typical ggplot interactions. The number of genes per cluster and the nunber of cells to display can be specified by the user. Either the log fold change or the p value can be used to select the top n genes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_cluster_membership.html">GEX_cluster_membership()</a></code> </p>
        </td>
        <td><p>Plots the cluster membership for each of the distinct samples in the Seurat object from the automate_GEX function. The distinct samples are determined by "sample_id" field in the Seurat object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_coexpression_coefficient.html">GEX_coexpression_coefficient()</a></code> </p>
        </td>
        <td><p>Clonal frequency plot displaying clonal expansion for either T and B cells with Platypus v3 input.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_DEgenes.html">GEX_DEgenes()</a></code> </p>
        </td>
        <td><p>Extracts the differentially expressed genes between two groups of cells. These groups are defined as cells having either of two entries (group1, group2) in the grouping.column of the input Seurat object metadata This function uses the FindMarkers function from the Seurat package.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_DEgenes_persample.html">GEX_DEgenes_persample()</a></code> </p>
        </td>
        <td><p>!Only for Platypus version v2. For more flexibility and platypus v3 please refer to GEX_Degenes. Extracts the differentially expressed genes between two samples. This function uses the FindMarkers function from the Seurat package. Further parameter control can be accomplished by calling the function directly on the output of automate_GEX and further extracting sample information from the "sample_id" component of the Seurat object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_dottile_plot.html">GEX_dottile_plot()</a></code> </p>
        </td>
        <td><p>Outputs a dotplot for gene expression, where the color of each dot is scaled by the gene expression level and the size is scaled by the % of cells positive for the gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_GOterm.html">GEX_GOterm()</a></code> </p>
        </td>
        <td><p>Runs a GO term analysis on a submitted list of genes. Works with the output of GEX_topN_DE_genes_per_cluster or a custom list of genes to obtain GOterms.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_GSEA.html">GEX_GSEA()</a></code> </p>
        </td>
        <td><p>Conducts a Gene Set Enrichment Analysis (GSEA) on a set of genes submitted in a data frame with a metric each. Works with the output of GEX_genes_cluster or a custom data frame contianing the gene symbols either in a column "symbols" or as rownames and a metric for each gene. The name of the column containing the metric has to be declared via the input metric_colname.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_heatmap.html">GEX_heatmap()</a></code> </p>
        </td>
        <td><p>Produces a heatmap containing gene expression information at the clonotype level. The rows correspond to different genes that can either be determined by premade sets of B or T cell markers, or can be customly supplied by the user. The columns correspond to individual cells and the colors correspond to the different clonotype families.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_pairwise_DEGs.html">GEX_pairwise_DEGs()</a></code> </p>
        </td>
        <td><p>Produces and saves a list of volcano plots with each showing differentially expressed genes between pairs groups. If e.g. seurat_clusters used as group.by, a plot will be generated for every pairwise comparison of clusters. For large numbers of this may take longer to run. Only available for platypus v3</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_phenotype.html">GEX_phenotype()</a></code> </p>
        </td>
        <td><p>Integrates VDJ and gene expression libraries by providing cluster membership seq_per_vdj object and the index of the cell in the Seurat RNA-seq object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_phenotype_per_clone.html">GEX_phenotype_per_clone()</a></code> </p>
        </td>
        <td><p>Integrates VDJ and gene expression libraries by providing cluster membership seq_per_vdj object and the index of the cell in the Seurat RNA-seq object.
! For platypus.version == "v3" and VDJ_GEX_matrix output the function will iterate over entries in the sample_id column of the GEX by default.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_proportions_barplot.html">GEX_proportions_barplot()</a></code> </p>
        </td>
        <td><p>Plots proportions of a group of cells within a secondary group of cells. E.g. The proportions of samples in seurat clusters, or the proportions of samples in defined cell subtypes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_scatter_coexpression.html">GEX_scatter_coexpression()</a></code> </p>
        </td>
        <td><p>Clonal frequency plot displaying clonal expansion for either T and B cells with Platypus v3 input.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_topN_DE_genes_per_cluster.html">GEX_topN_DE_genes_per_cluster()</a></code> </p>
        </td>
        <td><p>Organizes the top N genes that define each Seurat cluster and converts them into a single dataframe. This can be useful for obtaining insight into cluster-specific phenotypes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_visualize_clones.html">GEX_visualize_clones()</a></code> </p>
        </td>
        <td><p>!Only for platypus version v2. For platypus v3 refer to: VDJ_GEX_overlay_clones() Visualize selected clonotypes on the tSNE or UMAP projection.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GEX_volcano.html">GEX_volcano()</a></code> </p>
        </td>
        <td><p>Plots a volcano plot from the output of the FindMarkers function from the Seurat package or the GEX_cluster_genes function alternatively.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_alpha_beta_Vgene_circos.html">VDJ_alpha_beta_Vgene_circos()</a></code> </p>
        </td>
        <td><p>Makes a Circos plot from the VDJ_analyze output. Connects the V-alpha with the corresponding V-beta gene for each clonotype.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_analyze.html">VDJ_analyze()</a></code> </p>
        </td>
        <td><p>Processes and organizes the repertoire sequening data from cellranger vdj and returns a list of dataframes, where each dataframe corresponds to an individual repertoire. The function will return split CDR3 sequences, germline gene information, filter out those clones with either incomplete information or doublets (multiple CDR3 sequences for a given chain). This function should be called once for desired integrated repertoire and transcriptome. For example, if there are 3 VDJ libraries and 3 GEX libraries and the goal is to analyze all three GEX libraries together (e.g. one UMAP/tSNE reduction) this then function should be called one time and the three VDJ directories should be provided as input to the single function call.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_call_MIXCR.html">VDJ_call_MIXCR()</a></code> </p>
        </td>
        <td><p>Extracts information on the VDJRegion level using MiXCR on WINDOWS, MAC and UNIX systems for input from both Platypus v2 (VDJ.per.clone) or v3 (Output of VDJ_GEX_matrix) This function assumes the user can run an executable instance of MiXCR and is elgible to use MiXCR as determined by license agreements. ! FOR WINDOWS USERS THE EXECUTABLE MIXCR.JAR HAS TO PRESENT IN THE CURRENT WORKING DIRECTORY ! The VDJRegion corresponds to the recombined heavy and light chain loci starting from framework region 1 (FR1) and extending to frame work region 4 (FR4). This can be useful for extracting full-length sequences ready to clone and further calculating somatic hypermutation occurrences.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_circos.html">VDJ_circos()</a></code> </p>
        </td>
        <td><p>Plots a Circos diagram from an adjacency matrix. Uses the Circlize chordDiagram function. Is called by VDJ_clonotype_clusters_circos(), VDJ_alpha_beta_Vgene_circos() and VDJ_VJ_usage_circos() functions or works on its own when supplied with an adjacency matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_clonal_donut.html">VDJ_clonal_donut()</a></code> </p>
        </td>
        <td><p>Generate circular plots of clonal expansion per repertoire directly from the VDJ matrix of the VDJ_GEX_matrix function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_clonal_expansion.html">VDJ_clonal_expansion()</a></code> </p>
        </td>
        <td><p>Clonal frequency plot displaying clonal expansion for either T and B cells with Platypus v3 input.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_clonal_lineages.html">VDJ_clonal_lineages()</a></code> </p>
        </td>
        <td><p>Organizes and extracts full-length sequences for clonal lineage inference. The output sequence can either contain the germline sequence as determined by cellranger or can just contain the sequences contained in each clonal family.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_clonotype.html">VDJ_clonotype()</a></code> </p>
        </td>
        <td><p>Returns a list of clonotype dataframes following additional clonotyping. This function works best following filtering to ensure that each clone only has one heavy chain and one light chain.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_clonotype_clusters_circos.html">VDJ_clonotype_clusters_circos()</a></code> </p>
        </td>
        <td><p>Makes a Circos plot from the VDJ_GEX_integrate output. Connects the clonotypes with the corresponding clusters.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_contigs_to_vgm.html">VDJ_contigs_to_vgm()</a></code> </p>
        </td>
        <td><p>Formats "VDJ_contigs_annotations.csv" files from cell ranger to match the VDJ_GEX_matrix output using only cells with 1VDJ and 1VJ chain</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_diversity.html">VDJ_diversity()</a></code> </p>
        </td>
        <td><p>Calculates and plots common diversity and overlap measures for repertoires and alike. Require the vegan package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_dublets.html">VDJ_dublets()</a></code> </p>
        </td>
        <td><p>Produces a matrix indicating either the number of cells or clones which contain multiple heavy or light chains (or alpha/beta in the case of T cells).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_extract_germline.html">VDJ_extract_germline()</a></code> </p>
        </td>
        <td><p>Extracts the full-length germline sequence as determined by cellranger. This function returns an object that now contains the reference germline for each of the clones. If multiple clones (as determined by cellranger) have been merged using the VDJ_clonotype function then these sequences may have distinct germline sequences despite being in the same clonal family (nested list). This is particularly possible when homology thresholds were used to determine the clonotypes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_GEX_clonal_lineage_clusters.html">VDJ_GEX_clonal_lineage_clusters()</a></code> </p>
        </td>
        <td><p>Integrates the transcriptional cluster information into the clonal lineages. This requires that automate_GEX, VDJ_clonal_lineages, and VDJ_GEX_integrate have already been ran. The transcriptional cluster will be added to the end of the Name for each sequence.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_GEX_expansion.html">VDJ_GEX_expansion()</a></code> </p>
        </td>
        <td><p>Integrates VDJ and gene expression libraries by providing cluster membership seq_per_vdj object. Output will plot which transcriptional cluster (GEX) that the cells of a given clonotype are found in.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_GEX_integrate.html">VDJ_GEX_integrate()</a></code> </p>
        </td>
        <td><p>Integrates VDJ and gene expression libraries by providing cluster membership seq_per_vdj object and the index of the cell in the Seurat RNA-seq object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_GEX_matrix.html">VDJ_GEX_matrix()</a></code> </p>
        </td>
        <td><p>Processes both raw VDJ and GEX Cellranger output to compile a single cell level table containing all available information for each cell.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_GEX_overlay_clones.html">VDJ_GEX_overlay_clones()</a></code> </p>
        </td>
        <td><p>Highlights the cells belonging to any number of top clonotypes or of specifically selected clonotypes from one or more samples or groups in a GEX dimensional reduction.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_GEX_stats.html">VDJ_GEX_stats()</a></code> </p>
        </td>
        <td><p>Gives stats on number and quality of reads.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_isotypes_per_clone.html">VDJ_isotypes_per_clone()</a></code> </p>
        </td>
        <td><p>Clonal frequency plot displaying the isotype usage of each clone. ! For platypus v3 use VDJ_clonal_expansion</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_logoplot.html">VDJ_logoplot()</a></code> </p>
        </td>
        <td><p>Plots a logoplot of the CDR3 aminoacid region ! For platypus v3 and more flexibility use VDJ_logoplot_vector</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_logoplot_vector.html">VDJ_logoplot_vector()</a></code> </p>
        </td>
        <td><p>Plots a logoplot of the CDR3 aminoacid region</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_network.html">VDJ_network()</a></code> </p>
        </td>
        <td><p>Creates a similarity network where clones with similar CDR3s are connected.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_overlap_heatmap.html">VDJ_overlap_heatmap()</a></code> </p>
        </td>
        <td><p>Yields overlap heatmap and datatable of features or combined features for different samples or groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_per_clone.html">VDJ_per_clone()</a></code> </p>
        </td>
        <td><p>VDJ_per_clone</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_plot_SHM.html">VDJ_plot_SHM()</a></code> </p>
        </td>
        <td><p>Plots for SHM based on MIXCR output generated using the VDJ_call_MIXCR function and appended to the VDJ.GEX.matrix.output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_reclonotype_list_arrange.html">VDJ_reclonotype_list_arrange()</a></code> </p>
        </td>
        <td><p>Organizes the top N genes that define each Seurat cluster and converts them into a single dataframe. This can be useful for obtaining insight into cluster-specific phenotypes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_tree.html">VDJ_tree()</a></code> </p>
        </td>
        <td><p>Produces neighbor joining phylogenetic trees from the output of VDJ_clonal_lineages</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_variants_per_clone.html">VDJ_variants_per_clone()</a></code> </p>
        </td>
        <td><p>Returns statistics and plots to examine diversity of any sequence or metadata item within clones on a by sample level or global level</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_Vgene_usage.html">VDJ_Vgene_usage()</a></code> </p>
        </td>
        <td><p>Produces a matrix counting the number of occurences for each VDJ and VJ Vgene combinations for each list enty in VDJ.clonotype.output or for each sample_id in VDJ.matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_Vgene_usage_barplot.html">VDJ_Vgene_usage_barplot()</a></code> </p>
        </td>
        <td><p>Produces a barplot with the most frequently used IgH and IgK/L Vgenes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_Vgene_usage_stacked_barplot.html">VDJ_Vgene_usage_stacked_barplot()</a></code> </p>
        </td>
        <td><p>Produces a stacked barplot with the fraction of the most frequently used IgH and IgK/L Vgenes. This function can be used in combination with the VDJ_Vgene_usage_barplot to vizualize V gene usage per sample and among samples.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="VDJ_VJ_usage_circos.html">VDJ_VJ_usage_circos()</a></code> </p>
        </td>
        <td><p>Makes a Circos plot from the VDJ_analyze output. Connects the V gene with the corresponding J gene for each clonotype.</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Alexander Yermanos.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


