<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ Platypus</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/quickstart.html">Get started</a>
</li>
<li>
  <a href="../vignette_index.html">Vignettes</a>
</li>
<li>
  <a href="../PlatypusDB_index.html">PlatypusDB</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://immunesim.readthedocs.io/en/latest/" class="external-link">ImmuneSIM</a>
    </li>
  </ul></li>
<li>
  <a href="../cute_monotremes.html">Cute Monotremes</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://twitter.com/alexyermanos" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/KreineVic" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/alexyermanos/Platypus" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="AbForests_AntibodyForest.html">AbForests_AntibodyForest()</a></code> </p>
        </td>
        <td><p>Infer and draw B cell evolutionary networks</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_CompareForests.html">AbForests_CompareForests()</a></code> </p>
        </td>
        <td><p>Comparison of distinct B cell repertoires</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_ConvertStructure.html">AbForests_ConvertStructure()</a></code> </p>
        </td>
        <td><p>Extract transcriptome/isotype information and B cell receptor sequences from single cell immune repertoire formatted as list of data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_CsvToDf.html">AbForests_CsvToDf()</a></code> </p>
        </td>
        <td><p>Convert list of csvs, to nested list of data.frames</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_ForestMetrics.html">AbForests_ForestMetrics()</a></code> </p>
        </td>
        <td><p>Calculate metrics for networks</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_PlotGraphs.html">AbForests_PlotGraphs()</a></code> </p>
        </td>
        <td><p>Plot igraph and ggplot objects</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_PlyloToMatrix.html">AbForests_PlyloToMatrix()</a></code> </p>
        </td>
        <td><p>Conversion of phylogenetic tree to distance matrix</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_RemoveNets.html">AbForests_RemoveNets()</a></code> </p>
        </td>
        <td><p>Filter sub-repertoires with less than N unique sequences or with less than C unique cells</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_SubRepertoiresByCells.html">AbForests_SubRepertoiresByCells()</a></code> </p>
        </td>
        <td><p>Split single cell immune repertoire into sub-repertoires by isotype based on number of B cells</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_SubRepertoiresByUniqueSeq.html">AbForests_SubRepertoiresByUniqueSeq()</a></code> </p>
        </td>
        <td><p>Split single cell immune repertoire into sub-repertoires by isotype based on number of unique sequences</p></td>
      </tr><tr><td>
          <p><code><a href="AbForests_UniqueAntibodyVariants.html">AbForests_UniqueAntibodyVariants()</a></code> </p>
        </td>
        <td><p>Count the number of unique antibody variants per clonal lineage</p></td>
      </tr><tr><td>
          <p><code><a href="AlphaFold_prediction.html">AlphaFold_prediction()</a></code> </p>
        </td>
        <td><p>Structure prediction of Mixcr wrapper output with Alpha Fold</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests-class.html">AntibodyForests-class</a></code> </p>
        </td>
        <td><p>Class used for AntibodyForests functions</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests.html">AntibodyForests()</a></code> </p>
        </td>
        <td><p>Infer B cell evolutionary networks and/or sequence similarity networks</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_communities.html">AntibodyForests_communities()</a></code> </p>
        </td>
        <td><p>Network clustering/community detection for the AntibodyForests similarity networks</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_dynamics.html">AntibodyForests_dynamics()</a></code> </p>
        </td>
        <td><p>Create a nested list of longitudinal AntibodyForests objects</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_embeddings.html">AntibodyForests_embeddings()</a></code> </p>
        </td>
        <td><p>Structural node embeddings for the AntibodyForests minimum spanning trees/ sequence similarity networks</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_expand_intermediates.html">AntibodyForests_expand_intermediates()</a></code> </p>
        </td>
        <td><p>Infer intermediate nodes in the minimum spanning trees/ sequences similiarity networks created by the AntibodyForests function</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_heterogeneous.html">AntibodyForests_heterogeneous()</a></code> </p>
        </td>
        <td><p>Bipartite sequence-cell networks in AntibodyForests</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_infer_ancestral.html">AntibodyForests_infer_ancestral()</a></code> </p>
        </td>
        <td><p>Creates phylogenetic trees, infers ancestral sequences, and converts the resulting trees into igraph objects.</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_join_trees.html">AntibodyForests_join_trees()</a></code> </p>
        </td>
        <td><p>Joins a list of trees/networks as AntibodyForests objects into a single AntibodyForests object</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_kernels.html">AntibodyForests_kernels()</a></code> </p>
        </td>
        <td><p>Graph kernel methods for graph structure/topology comparisons</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_label_propagation.html">AntibodyForests_label_propagation()</a></code> </p>
        </td>
        <td><p>Propagate label annotations/values on sparsely labeled networks as AntibodyForests objects.</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_metrics.html">AntibodyForests_metrics()</a></code> </p>
        </td>
        <td><p>Node metrics for the AntibodyForests sequence similarity networks and minimum spanning trees.</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_node_transitions.html">AntibodyForests_node_transitions()</a></code> </p>
        </td>
        <td><p>Calculates the node transitions frequencies for a given feature and an AntibodyForests object</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_overlap.html">AntibodyForests_overlap()</a></code> </p>
        </td>
        <td><p>Edge overlap heatmaps for a set of AntibodyForests sequence similarity networks or minimum spanning trees.</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_paths.html">AntibodyForests_paths()</a></code> </p>
        </td>
        <td><p>Calculates the longest/shortest paths from a node to a given node for the AntibodyForests minimum spanning trees / sequence similarity networks</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_phylo.html">AntibodyForests_phylo()</a></code> </p>
        </td>
        <td><p>Converts the igraph networks of a given AntibodyForests object into a given (useful to convert the minimum spanning trees into a phylogenetic tree)</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_plot.html">AntibodyForests_plot()</a></code> </p>
        </td>
        <td><p>Custom plots for trees/networks created with AntibodyForests</p></td>
      </tr><tr><td>
          <p><code><a href="AntibodyForests_plot_metrics.html">AntibodyForests_plot_metrics()</a></code> </p>
        </td>
        <td><p>Plots the resulting node metrics from the AntibodyForests_metrics function</p></td>
      </tr><tr><td>
          <p><code><a href="automate_GEX.html">automate_GEX()</a></code> </p>
        </td>
        <td><p>GEX processing wrapper in Platypus V2</p></td>
      </tr><tr><td>
          <p><code><a href="Bcell_sequences_example_tree.html">Bcell_sequences_example_tree</a></code> </p>
        </td>
        <td><p>Example csv file 1</p></td>
      </tr><tr><td>
          <p><code><a href="Bcell_tree_2.html">Bcell_tree_2</a></code> </p>
        </td>
        <td><p>Example csv file 2</p></td>
      </tr><tr><td>
          <p><code><a href="call_MIXCR.html">call_MIXCR()</a></code> </p>
        </td>
        <td><p>Calls MiXCR VDJ object of Platypus V2</p></td>
      </tr><tr><td>
          <p><code><a href="CellPhoneDB_analyse.html">CellPhoneDB_analyse()</a></code> </p>
        </td>
        <td><p>Cellphone DB utility</p></td>
      </tr><tr><td>
          <p><code><a href="class_switch_prob_hum.html">class_switch_prob_hum</a></code> </p>
        </td>
        <td><p>class_switch_prob_hum
The probability matrix of class switching for human b cells. The row names of the matrix are the isotypes the cell is switching from, the column names are the isotypes the cell is switching to. All B cells start from IGHM, and switch to one of the other isotypes or remain the same.</p></td>
      </tr><tr><td>
          <p><code><a href="class_switch_prob_mus.html">class_switch_prob_mus</a></code> </p>
        </td>
        <td><p>class_switch_prob_mus
The probability matrix of class switching for mouse b cells. The row names of the matrix are the isotypes the cell is switching from, the column names are the isotypes the cell is switching to. All B cells start from IGHM, and switch to one of the other isotypes or remain the same.</p></td>
      </tr><tr><td>
          <p><code><a href="clonofreq.isotype.data.html">clonofreq.isotype.data()</a></code> </p>
        </td>
        <td><p>Get information about the clonotype counts grouped by isotype.</p></td>
      </tr><tr><td>
          <p><code><a href="clonofreq.isotype.plot.html">clonofreq.isotype.plot()</a></code> </p>
        </td>
        <td><p>Get information about the clonotype counts grouped by isotype.</p></td>
      </tr><tr><td>
          <p><code><a href="clonofreq.html">clonofreq()</a></code> </p>
        </td>
        <td><p>Plot clonal frequency barplot of the outout simulated data</p></td>
      </tr><tr><td>
          <p><code><a href="clonofreq.trans.data.html">clonofreq.trans.data()</a></code> </p>
        </td>
        <td><p>Get information about the clonotype counts grouped by transcriptome state(cell type).</p></td>
      </tr><tr><td>
          <p><code><a href="clonofreq.trans.plot.html">clonofreq.trans.plot()</a></code> </p>
        </td>
        <td><p>Get information about the clonotype counts grouped by transcriptome state(cell type).</p></td>
      </tr><tr><td>
          <p><code><a href="cluster.id.igraph.html">cluster.id.igraph()</a></code> </p>
        </td>
        <td><p>Get clone network igraphs colored by seurat cluster id.</p></td>
      </tr><tr><td>
          <p><code><a href="colors.html">colors</a></code> </p>
        </td>
        <td><p>colors
A vector of characters specifying colors used in igraph phylogenetic tree. Default colors: "#66C2A5", "#FC8D62", "#8DA0CB", "#E78AC3" ,"#A6D854"</p></td>
      </tr><tr><td>
          <p><code><a href="dot_plot.html">dot_plot()</a></code> </p>
        </td>
        <td><p>Function to cutomise the Dot Plot of CellPhoneDB analysis results.</p></td>
      </tr><tr><td>
          <p><code><a href="Echidna_simulate_repertoire.html">Echidna_simulate_repertoire()</a></code> </p>
        </td>
        <td><p>Simulate immune repertoire and transcriptome data</p></td>
      </tr><tr><td>
          <p><code><a href="Echidna_vae_generate.html">Echidna_vae_generate()</a></code> </p>
        </td>
        <td><p>Simulate B or T cell receptor sequences by variational autoencodes(VAEs) trained with experimental data.</p></td>
      </tr><tr><td>
          <p><code><a href="get.avr.mut.data.html">get.avr.mut.data()</a></code> </p>
        </td>
        <td><p>Get information about somatic hypermutation in the simulation. This function return a barplot showing the average mutation.</p></td>
      </tr><tr><td>
          <p><code><a href="get.avr.mut.plot.html">get.avr.mut.plot()</a></code> </p>
        </td>
        <td><p>Get information about somatic hypermutation in the simulation. This function return a barplot showing the average mutation.</p></td>
      </tr><tr><td>
          <p><code><a href="get.barplot.errorbar.html">get.barplot.errorbar()</a></code> </p>
        </td>
        <td><p>Return a barplot of mean and standard error bar of certain value of each clone.</p></td>
      </tr><tr><td>
          <p><code><a href="get.elbow.html">get.elbow()</a></code> </p>
        </td>
        <td><p>Get the seurat object from simulated transciptome output.</p></td>
      </tr><tr><td>
          <p><code><a href="get.n.node.data.html">get.n.node.data()</a></code> </p>
        </td>
        <td><p>Get the number of unique variants in each clone in a vector. The output is the vector representing the numbers of unique variants.</p></td>
      </tr><tr><td>
          <p><code><a href="get.n.node.plot.html">get.n.node.plot()</a></code> </p>
        </td>
        <td><p>Get the number of unique variants in each clone in a vector and the barplot. The first item in the output is the vector representing the numbers of unique variants, the second item is the barplot.</p></td>
      </tr><tr><td>
          <p><code><a href="get.seq.distance.html">get.seq.distance()</a></code> </p>
        </td>
        <td><p>Computing sequence distance according to the number of unmatched bases.</p></td>
      </tr><tr><td>
          <p><code><a href="get.umap.html">get.umap()</a></code> </p>
        </td>
        <td><p>Further process the seurat object from simulated transciptome output and make UMAP ready for plotting.</p></td>
      </tr><tr><td>
          <p><code><a href="get.vgu.matrix.html">get.vgu.matrix()</a></code> </p>
        </td>
        <td><p>Get paired v gene heavy chain and light chain matrix on clonotype level. A v gene usage pheatmap can be obtain by p&lt;-pheatmap::pheatmap(vgu_matrix,show_colnames= T, main = "V Gene Usage"), where the vgu_matrix is the output of this function.</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_clonotype.html">GEX_clonotype()</a></code> </p>
        </td>
        <td><p>Platypus V2 GEX and VDJ integration for clonotypes</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_cluster_genes.html">GEX_cluster_genes()</a></code> </p>
        </td>
        <td><p>Differentially expressed genes between clusters or data subsets</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_cluster_genes_heatmap.html">GEX_cluster_genes_heatmap()</a></code> </p>
        </td>
        <td><p>Heatmap of cluster defining genes</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_cluster_membership.html">GEX_cluster_membership()</a></code> </p>
        </td>
        <td><p>Cluster membership plots by sample</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_coexpression_coefficient.html">GEX_coexpression_coefficient()</a></code> </p>
        </td>
        <td><p>Coexpression of selected genes</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_DEgenes.html">GEX_DEgenes()</a></code> </p>
        </td>
        <td><p>Wrapper for differential gene expression analysis and plotting</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_DEgenes_persample.html">GEX_DEgenes_persample()</a></code> </p>
        </td>
        <td><p>Platypus V2 Differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_dottile_plot.html">GEX_dottile_plot()</a></code> </p>
        </td>
        <td><p>GEX Dottile plots</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_gene_visualization.html">GEX_gene_visualization()</a></code> </p>
        </td>
        <td><p>Visualization of marker expression in a data set or of predefined genes (B cells, CD4 T cells and CD8 T cells).</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_GOterm.html">GEX_GOterm()</a></code> </p>
        </td>
        <td><p>GEX GO-Term analysis and plotting</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_GSEA.html">GEX_GSEA()</a></code> </p>
        </td>
        <td><p>GEX Gene Set Enrichment Analysis and plotting</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_heatmap.html">GEX_heatmap()</a></code> </p>
        </td>
        <td><p>Flexible GEX heatmap wrapper</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_lineage_trajectories.html">GEX_lineage_trajectories()</a></code> </p>
        </td>
        <td><p>This is a function to infer single cell trajectories and identifying lineage structures on clustered cells. Using the slingshot library</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_pairwise_DEGs.html">GEX_pairwise_DEGs()</a></code> </p>
        </td>
        <td><p>Wrapper for calculating pairwise differentially expressed genes</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_phenotype.html">GEX_phenotype()</a></code> </p>
        </td>
        <td><p>Assignment of cells to phenotypes based on selected markers</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_phenotype_per_clone.html">GEX_phenotype_per_clone()</a></code> </p>
        </td>
        <td><p>Plotting of GEX phenotype by VDJ clone</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_projecTILS.html">GEX_projecTILS()</a></code> </p>
        </td>
        <td><p>ProjectTILs tool utility</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_proportions_barplot.html">GEX_proportions_barplot()</a></code> </p>
        </td>
        <td><p>Plots proportions of a group of cells within a secondary group of cells. E.g. The proportions of samples in seurat clusters, or the proportions of samples in defined cell subtypes</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_pseudobulk.html">GEX_pseudobulk()</a></code> </p>
        </td>
        <td><p>Function that performs pseudo-bulking on the data (VGM input), according to criteria specified by the User, and uses the pseudo-bulked data to perform Differential Gene Expression (DGE) analysis.</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_pseudotime_trajectory_plot.html">GEX_pseudotime_trajectory_plot()</a></code> </p>
        </td>
        <td><p>This function plots pseudotime along the trajectories which have been constructed with the GEX_trajectories() function.</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_scatter_coexpression.html">GEX_scatter_coexpression()</a></code> </p>
        </td>
        <td><p>Scatter plot for coexpression of two selected genes</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_topN_DE_genes_per_cluster.html">GEX_topN_DE_genes_per_cluster()</a></code> </p>
        </td>
        <td><p>Platypus V2 GEX DE genes helper</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_trajectories.html">GEX_trajectories()</a></code> </p>
        </td>
        <td><p>This is a function which infers trajectories along ordered cells on dimensionality reduced data. It projects trajectrories on a dim. red. plot such as Umap. This uses Monocle3 or Monocle2.</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_visualize_clones.html">GEX_visualize_clones()</a></code> </p>
        </td>
        <td><p>Platypus V2 GEX and VDJ integration for visualizing clone clustering</p></td>
      </tr><tr><td>
          <p><code><a href="GEX_volcano.html">GEX_volcano()</a></code> </p>
        </td>
        <td><p>Flexible wrapper for GEX volcano plots</p></td>
      </tr><tr><td>
          <p><code><a href="hotspot_df.html">hotspot_df</a></code> </p>
        </td>
        <td><p>hotspot_df
Hotspot mutations taken from Yaari et al., Frontiers in Immunology, 2013.
This contains transition probabilities for all 5mer combinations based
on high throughput sequencing data. The transition probabilities are for
the middle nucleotide in each 5mer set. This can be customized by changing the genes
and sequences. Custom mutation hotspots can be supplied
by modifying this dataframe. Repeating particular hotspot entries allows
for the hotspot to mutate more than one time per SHM event.</p></td>
      </tr><tr><td>
          <p><code><a href="hum_b_h.html">hum_b_h</a></code> </p>
        </td>
        <td><p>hum_b_h</p></td>
      </tr><tr><td>
          <p><code><a href="hum_b_l.html">hum_b_l</a></code> </p>
        </td>
        <td><p>hum_b_l</p></td>
      </tr><tr><td>
          <p><code><a href="hum_t_h.html">hum_t_h</a></code> </p>
        </td>
        <td><p>hum_t_h</p></td>
      </tr><tr><td>
          <p><code><a href="hum_t_l.html">hum_t_l</a></code> </p>
        </td>
        <td><p>hum_t_l</p></td>
      </tr><tr><td>
          <p><code><a href="iso_SHM_prob.html">iso_SHM_prob</a></code> </p>
        </td>
        <td><p>iso_SHM_prob
A probability dataframe specifying SHM.nuc.prob for cells of different isotypes. The first column is the names of isotypes, while the second column is the SHM.nuc.prob of cell of that isotype. user can define different SHM.nuc.prob for isotypes.</p></td>
      </tr><tr><td>
          <p><code><a href="mus_b_h.html">mus_b_h</a></code> </p>
        </td>
        <td><p>mus_b_h</p></td>
      </tr><tr><td>
          <p><code><a href="mus_b_l.html">mus_b_l</a></code> </p>
        </td>
        <td><p>mus_b_l</p></td>
      </tr><tr><td>
          <p><code><a href="mus_b_trans.html">mus_b_trans</a></code> </p>
        </td>
        <td><p>mus_b_trans
A data frame contains mouse B cell average gene expression for multiple cell types, with the rows representing the gene names, column names representing the cell type names.
The original single cell sequencing data is retrieved from 10xgenomics and combined with experimental data
The expression level for different cell types are obtained by calculating the average expression after sorting the original data by markers as shown below.</p></td>
      </tr><tr><td>
          <p><code><a href="mus_t_h.html">mus_t_h</a></code> </p>
        </td>
        <td><p>mus_t_h</p></td>
      </tr><tr><td>
          <p><code><a href="mus_t_l.html">mus_t_l</a></code> </p>
        </td>
        <td><p>mus_t_l</p></td>
      </tr><tr><td>
          <p><code><a href="no.empty.node.html">no.empty.node()</a></code> </p>
        </td>
        <td><p>Get clone network igraphs without empty mode. Empty node represents the 'extincted' sequences, that are not in any living cell but once existed.</p></td>
      </tr><tr><td>
          <p><code><a href="one_spot_df.html">one_spot_df</a></code> </p>
        </td>
        <td><p>one_spot_df</p></td>
      </tr><tr><td>
          <p><code><a href="pheno_SHM_prob.html">pheno_SHM_prob</a></code> </p>
        </td>
        <td><p>pheno_SHM_prob
A probability dataframe specifying SHM.nuc.prob for cells of different phenotypes. The first column is the names of phenotypes, while the second column is the SHM.nuc.prob of cell of that phenotype. user can define different SHM.nuc.prob for phenotypes.</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusDB_AIRR_to_VGM.html">PlatypusDB_AIRR_to_VGM()</a></code> </p>
        </td>
        <td><p>AIRR to Platypus V3 VGM compatibility function</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusDB_fetch.html">PlatypusDB_fetch()</a></code> </p>
        </td>
        <td><p>Loads and saves RData objects from the PlatypusDB</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusDB_find_CDR3s.html">PlatypusDB_find_CDR3s()</a></code> </p>
        </td>
        <td><p>CDR3 query function for PlatypusDB</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusDB_list_projects.html">PlatypusDB_list_projects()</a></code> </p>
        </td>
        <td><p>Metadata download by project for PlatypusDB</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusDB_load_from_disk.html">PlatypusDB_load_from_disk()</a></code> </p>
        </td>
        <td><p>PlatypusDB utility for import of local datasets</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusDB_VGM_to_AIRR.html">PlatypusDB_VGM_to_AIRR()</a></code> </p>
        </td>
        <td><p>Platypus V3 VGM to AIRR compatibility function</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusML_balance.html">PlatypusML_balance()</a></code> </p>
        </td>
        <td><p>Secondary ML for crossvalidation</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusML_classification.html">PlatypusML_classification()</a></code> </p>
        </td>
        <td><p>Core ML for crossvalidation</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusML_feature_extraction_GEX.html">PlatypusML_feature_extraction_GEX()</a></code> </p>
        </td>
        <td><p>Extraction of features from GEX matrix of VGM</p></td>
      </tr><tr><td>
          <p><code><a href="PlatypusML_feature_extraction_VDJ.html">PlatypusML_feature_extraction_VDJ()</a></code> </p>
        </td>
        <td><p>Extraction of features from VDJ table of VGM</p></td>
      </tr><tr><td>
          <p><code><a href="select.top.clone.html">select.top.clone()</a></code> </p>
        </td>
        <td><p>Get the index of top ranking clones.</p></td>
      </tr><tr><td>
          <p><code><a href="small_vgm.html">small_vgm</a></code> </p>
        </td>
        <td><p>Small VDJ GEX matrix (VGM) for function testing purposes</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_celltype_plot.html">Spatial_celltype_plot()</a></code> </p>
        </td>
        <td><p>Plotting celltype assign to cell according to their phenotype on the spatial image.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_cluster.html">Spatial_cluster()</a></code> </p>
        </td>
        <td><p>Plotting clusters of cells by choosing between 10X Genomics clustering or reclustering the cells.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_density_plot.html">Spatial_density_plot()</a></code> </p>
        </td>
        <td><p>Plotting the contour density of selected cells or of all cells.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_evolution_of_clonotype_plot.html">Spatial_evolution_of_clonotype_plot()</a></code> </p>
        </td>
        <td><p>Plotting the phylogenetic network of a clonotype based on the somatic hypermutations of the immune repertoire sequences on a spatial image.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_marker_expression.html">Spatial_marker_expression()</a></code> </p>
        </td>
        <td><p>Plotting a gene of interest in selected cells on the spatial image.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_module_expression.html">Spatial_module_expression()</a></code> </p>
        </td>
        <td><p>Plotting the expression of a gene module on the spatial image with or without a threshold.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_nb_SHM_compare_to_germline_plot.html">Spatial_nb_SHM_compare_to_germline_plot()</a></code> </p>
        </td>
        <td><p>Plotting number of somatic hypermutation of clones compare to the germline sequence of the clonotype.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_scaling_parameters.html">Spatial_scaling_parameters()</a></code> </p>
        </td>
        <td><p>Scaling of the spatial parameters to be able to express the gene expression on the spatial image.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_selection_expanded_clonotypes.html">Spatial_selection_expanded_clonotypes()</a></code> </p>
        </td>
        <td><p>Selection of VGM[[1]]/VDJ data of the x more expanded clonotypes.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_selection_of_cells_on_image.html">Spatial_selection_of_cells_on_image()</a></code> </p>
        </td>
        <td><p>Allows to select an area on the spatial image and to isolate the cells expressed on this part and repeat this process several times.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_VDJ_assignment.html">Spatial_VDJ_assignment()</a></code> </p>
        </td>
        <td><p>Assign simulated immune repertoire sequences (BCR or TCR) simulated by Echidna to transcriptome and location in a spatial image in function of cell type.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_VDJ_plot.html">Spatial_VDJ_plot()</a></code> </p>
        </td>
        <td><p>Plotting immune repertoire data as clonotype or isotype for cells on a spatial image.</p></td>
      </tr><tr><td>
          <p><code><a href="Spatial_vgm_formation.html">Spatial_vgm_formation()</a></code> </p>
        </td>
        <td><p>Addition of the spatial information to the VGM matrix, output of VDJ_GEX_matrix()</p></td>
      </tr><tr><td>
          <p><code><a href="special_v.html">special_v</a></code> </p>
        </td>
        <td><p>special_v
a dataframe, of heavy and light chain v gene combination and their probability to be selected for expansion.</p></td>
      </tr><tr><td>
          <p><code><a href="trans_switch_prob_b.html">trans_switch_prob_b</a></code> </p>
        </td>
        <td><p>trans_switch_prob_b
The probability for B cell transcriptome states switching. The row names of the matrix are the cell states the cell is switching from, the column names are the cells states the cell is switching to.</p></td>
      </tr><tr><td>
          <p><code><a href="trans_switch_prob_t.html">trans_switch_prob_t</a></code> </p>
        </td>
        <td><p>trans_switch_prob_t
The probability for T cell transcriptome states switching. The row names of the matrix are the cell states the cell is switching from, the column names are the cells states the cell is switching to.</p></td>
      </tr><tr><td>
          <p><code><a href="umap.top.highlight.html">umap.top.highlight()</a></code> </p>
        </td>
        <td><p>Set idents for top abundant clones in Seurat object, get ready for highlight the top abundant clones in UMAP.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_abundances.html">VDJ_abundances()</a></code> </p>
        </td>
        <td><p>Calculate abundances/counts of specific features for a VDJ dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_alpha_beta_Vgene_circos.html">VDJ_alpha_beta_Vgene_circos()</a></code> </p>
        </td>
        <td><p>Produces a Circos plot from the VDJ_analyze output. Connects the V-alpha with the corresponding V-beta gene for each clonotype.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_analyze.html">VDJ_analyze()</a></code> </p>
        </td>
        <td><p>Platypus V2 VDJ processing wrapper.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_antigen_integrate.html">VDJ_antigen_integrate()</a></code> </p>
        </td>
        <td><p>Integrates antigen-specific information into the VDJ/VDJ.GEX.matrix[[1]] object</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_assemble_for_PnP.html">VDJ_assemble_for_PnP()</a></code> </p>
        </td>
        <td><p>Ab sequence assembly for recombinant PnP expression</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_bulk_to_vgm.html">VDJ_bulk_to_vgm()</a></code> </p>
        </td>
        <td><p>Utility function for bulk data to standard Platypus format conversion</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_call_enclone.html">VDJ_call_enclone()</a></code> </p>
        </td>
        <td><p>(Re)clonotype a VDJ object using cellranger's enclone tool</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_call_MIXCR.html">VDJ_call_MIXCR()</a></code> </p>
        </td>
        <td><p>MiXCR wrapper for Platypus V3 VDJ object</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_call_MIXCR_full.html">VDJ_call_MIXCR_full()</a></code> </p>
        </td>
        <td><p>MiXCR wrapper for Platypus V3 VDJ object. In addition to the VDJ_call_MIXCR function, the output also contains the concatenated sequences
from FR1 all the way to FR2 for both the VDJ and VJ.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_call_RECON.html">VDJ_call_RECON()</a></code> </p>
        </td>
        <td><p>Calls the Kaplinsky/RECON tool</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_circos.html">VDJ_circos()</a></code> </p>
        </td>
        <td><p>Plots a Circos diagram from an adjacency matrix. Uses the Circlize chordDiagram function. Is called by VDJ_clonotype_clusters_circos(), VDJ_alpha_beta_Vgene_circos() and VDJ_VJ_usage_circos() functions or works on its own when supplied with an adjacency matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_clonal_donut.html">VDJ_clonal_donut()</a></code> </p>
        </td>
        <td><p>Circular VDJ expansion plots</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_clonal_expansion.html">VDJ_clonal_expansion()</a></code> </p>
        </td>
        <td><p>Flexible wrapper for clonal expansion barplots by isotype, GEX cluster etc.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_clonal_expansion_abundances.html">VDJ_clonal_expansion_abundances()</a></code> </p>
        </td>
        <td><p>Wrapper function for VDJ_abundances to obtain ranked clonotype barplots</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_clonal_lineages.html">VDJ_clonal_lineages()</a></code> </p>
        </td>
        <td><p>Platypus V2 lineage utility</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_clonotype.html">VDJ_clonotype()</a></code> </p>
        </td>
        <td><p>Platypus V3 clonotyping wrapper</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_contigs_to_vgm.html">VDJ_contigs_to_vgm()</a></code> </p>
        </td>
        <td><p>Formats "VDJ_contigs_annotations.csv" files from cell ranger to match the VDJ_GEX_matrix output using only cells with 1VDJ and 1VJ chain</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_db_annotate.html">VDJ_db_annotate()</a></code> </p>
        </td>
        <td><p>Wrapper function of VDJ_antigen_integrate function</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_db_load.html">VDJ_db_load()</a></code> </p>
        </td>
        <td><p>Load and preprocess a list of antigen-specific databases</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_diversity.html">VDJ_diversity()</a></code> </p>
        </td>
        <td><p>Calculates and plots common diversity and overlap measures for repertoires and alike. Requires the vegan package</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_dublets.html">VDJ_dublets()</a></code> </p>
        </td>
        <td><p>Platypus V2 annotation utility</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_dynamics.html">VDJ_dynamics()</a></code> </p>
        </td>
        <td><p>Tracks a specific VDJ column across multiple samples/timepoints.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_enclone.html">VDJ_enclone()</a></code> </p>
        </td>
        <td><p>Updated clonotyping function based on implications for cells with different chain numbers than 1 VDJ 1 VJ chains.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_expand_aberrants.html">VDJ_expand_aberrants()</a></code> </p>
        </td>
        <td><p>Expand the aberrant cells in a VDJ dataframe by converting them into additional rows</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_extract_germline.html">VDJ_extract_germline()</a></code> </p>
        </td>
        <td><p>Platypus V2 utility for full germline sequence via MiXCR</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_get_public.html">VDJ_get_public()</a></code> </p>
        </td>
        <td><p>Function to get shared/public elements across multiple repertoires</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_GEX_clonal_lineage_clusters.html">VDJ_GEX_clonal_lineage_clusters()</a></code> </p>
        </td>
        <td><p>Platypus V2 lineage - GEX integration utility</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_GEX_clonotyme.html">VDJ_GEX_clonotyme()</a></code> </p>
        </td>
        <td><p>Pseudotime analysis for scRNA and repertoire sequencing datasets</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_GEX_clonotype_clusters_circos.html">VDJ_GEX_clonotype_clusters_circos()</a></code> </p>
        </td>
        <td><p>Makes a Circos plot from the VDJ_GEX_integrate output. Connects the clonotypes with the corresponding clusters.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_GEX_expansion.html">VDJ_GEX_expansion()</a></code> </p>
        </td>
        <td><p>Platypus V2 utility</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_GEX_integrate.html">VDJ_GEX_integrate()</a></code> </p>
        </td>
        <td><p>only Platypus v2 Integrates VDJ and gene expression libraries by providing cluster membership seq_per_vdj object and the index of the cell in the Seurat RNA-seq object.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_GEX_matrix.html">VDJ_GEX_matrix()</a></code> </p>
        </td>
        <td><p>VDJ GEX processing and integration wrapper</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_GEX_overlay_clones.html">VDJ_GEX_overlay_clones()</a></code> </p>
        </td>
        <td><p>Overlay clones on GEX projection</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_GEX_stats.html">VDJ_GEX_stats()</a></code> </p>
        </td>
        <td><p>Standalone VDJ and GEX statistics.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_isotypes_per_clone.html">VDJ_isotypes_per_clone()</a></code> </p>
        </td>
        <td><p>Platypus V2 clonal utility</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_kmers.html">VDJ_kmers()</a></code> </p>
        </td>
        <td><p>Calculates and plots kmers distributions and frequencies.</p></td>
      </tr><tr><td>
          <p><code><a href="vdj_length_prob.html">vdj_length_prob</a></code> </p>
        </td>
        <td><p>vdj_length_prob
A list dataframe specifying lengths and probabilities of bases deleted or inserted at each junction site of VDJ recombination event.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_logoplot_vector.html">VDJ_logoplot_vector()</a></code> </p>
        </td>
        <td><p>Flexible logoplot wrapper</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_network.html">VDJ_network()</a></code> </p>
        </td>
        <td><p>Similarity networks based on CDR3 regions</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_ordination.html">VDJ_ordination()</a></code> </p>
        </td>
        <td><p>Performs ordination/dimensionality reduction for a species incidence matrix, depending on the species selected in the feature.columns parameter.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_overlap_heatmap.html">VDJ_overlap_heatmap()</a></code> </p>
        </td>
        <td><p>Wrapper to determine and plot overlap between VDJ features across groups</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_per_clone.html">VDJ_per_clone()</a></code> </p>
        </td>
        <td><p>VDJ_per_clone</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_phylogenetic_trees.html">VDJ_phylogenetic_trees()</a></code> </p>
        </td>
        <td><p>Creates phylogenetic trees from a VDJ dataframe</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_phylogenetic_trees_plot.html">VDJ_phylogenetic_trees_plot()</a></code> </p>
        </td>
        <td><p>Phylogenetic tree plotting</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_plot_SHM.html">VDJ_plot_SHM()</a></code> </p>
        </td>
        <td><p>Plotting of somatic hypermutation counts</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_public.html">VDJ_public()</a></code> </p>
        </td>
        <td><p>Function to get shared/public elements across multiple repertoires</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_rarefaction.html">VDJ_rarefaction()</a></code> </p>
        </td>
        <td><p>Plots rarefaction curves for species denoted in the feature.columns parameter across groups determined by grouping.columns</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_reclonotype_list_arrange.html">VDJ_reclonotype_list_arrange()</a></code> </p>
        </td>
        <td><p>Platypus V2 dataframe utility</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_select_clonotypes.html">VDJ_select_clonotypes()</a></code> </p>
        </td>
        <td><p>Select clonotypes</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_structure_analysis.html">VDJ_structure_analysis()</a></code> </p>
        </td>
        <td><p>Analysis of antibody structures</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_tree.html">VDJ_tree()</a></code> </p>
        </td>
        <td><p>Platypus V2 phylogenetic trees.</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_variants_per_clone.html">VDJ_variants_per_clone()</a></code> </p>
        </td>
        <td><p>Wrapper for variant analysis by clone</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_Vgene_usage.html">VDJ_Vgene_usage()</a></code> </p>
        </td>
        <td><p>V(D)J gene usage stacked barplots</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_Vgene_usage_barplot.html">VDJ_Vgene_usage_barplot()</a></code> </p>
        </td>
        <td><p>V(D)J gene usage barplots</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_Vgene_usage_stacked_barplot.html">VDJ_Vgene_usage_stacked_barplot()</a></code> </p>
        </td>
        <td><p>V(D)J gene usage stacked barplots</p></td>
      </tr><tr><td>
          <p><code><a href="VDJ_VJ_usage_circos.html">VDJ_VJ_usage_circos()</a></code> </p>
        </td>
        <td><p>Makes a Circos plot from the VDJ_analyze output. Connects the V gene with the corresponding J gene for each clonotype.</p></td>
      </tr><tr><td>
          <p><code><a href="VGM_expanded_clones.html">VGM_expanded_clones()</a></code> </p>
        </td>
        <td><p>VDJ utility for T/F column for clonal expansion</p></td>
      </tr><tr><td>
          <p><code><a href="VGM_expand_featurebarcodes.html">VGM_expand_featurebarcodes()</a></code> </p>
        </td>
        <td><p>Utility for feature barcode assignment including clonal information</p></td>
      </tr><tr><td>
          <p><code><a href="VGM_integrate.html">VGM_integrate()</a></code> </p>
        </td>
        <td><p>Utility for VDJ GEX matrix to integrated VDJ and GEX objects after addition of data to either</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alexander Yermanos.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

