<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Replaces the original sample_id column of a vgm object with a pasted version of the original sample_id and the last digits of the feature barcode. — VGM_expand_featurebarcodes • Platypus</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Replaces the original sample_id column of a vgm object with a pasted version of the original sample_id and the last digits of the feature barcode. — VGM_expand_featurebarcodes" />
<meta property="og:description" content="The VGM_expand_featurebarcodes function function can be used to trace back the cell origin of each sample after using cell hashing for single-cell sequencing.
The original sample_id is stored in a new column called original_sample_id. Additionally, a second new column is created containing final barcode assignment information.
Those barcodes match the origin FB_assignment if by.majority.barcodes is set to FALSE (default). However, if this input parameter is set to TRUE, the majority barcode assignment in stored in this colum.
Note: The majority barcode of a cell is the feature barcode which is  most frequently assigned to the cells clonotype (10x default clonotype).
The majority barcode assignment can be used under the assumption that all cells which are assigned to the same clonotype (within one sample), originate from the same donor organ or at least the same donor depending on the experimental setup.
For example: The original sample_id of a cell is &quot;s1&quot;, the cell belongs to &quot;clonotype1&quot; and the feature barcode assigned to it is &quot;i1-TotalSeq-C0953&quot;. If by.majority.barcodes default (FALSE) is used, the resulting new sample_id would be &quot;s1_0953&quot;.
However, if majority barcode assignment is used AND &quot;i1-TotalSeq-C0953&quot; is not the most frequently occurring barcode in &quot;clonotype1&quot; but rather barcode &quot;i1-TotalSeq-C0951&quot;, the new sample_id would be &quot;s1_0951&quot;.
--&amp;gt; e.g., if 15 cells belong to clonotype1: 3 cells have no assigned barcode, 2 are assigned to &quot;i1-TotalSeq-C0953&quot; and 10 are assigned to &quot;i1-TotalSeq-C0951&quot; --&amp;gt; all 15 cells will have the new sample_id &quot;s1_0951&quot;." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/vignette_index.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_index.html">View all</a>
    </li>
    <li class="dropdown-header">First steps</li>
    <li>
      <a href="../articles/quickstart.html">Quickstart</a>
    </li>
    <li>
      <a href="../articles/platypus_database.html">Platypus Database</a>
    </li>
    <li>
      <a href="../articles/airr_compatibility.html">AIRR compatibility and TRUST4</a>
    </li>
    <li>
      <a href="../articles/vdj_gex_integration.html">VDJ-GEX integration</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">GEX analysis using Seurat</li>
    <li>
      <a href="../articles/quality_control.html">Quality control</a>
    </li>
    <li>
      <a href="../articles/feature_barcodes.html">Feature Barcodes</a>
    </li>
    <li>
      <a href="../articles/GEX_overview.html">GEX overview</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">VDJ analysis</li>
    <li>
      <a href="../articles/clonotyping.html">Clonotyping</a>
    </li>
    <li>
      <a href="../articles/clonal_expansion.html">Clonal expansion</a>
    </li>
    <li>
      <a href="../articles/germline_genes.html">Germline gene and isotype usage</a>
    </li>
    <li>
      <a href="../articles/sampling_diversity.html">Sampling and diversity analysis</a>
    </li>
    <li>
      <a href="../articles/somatic_hypermutations.html">Somatic hypermutations</a>
    </li>
    <li>
      <a href="../articles/TCR_look_up.html">TCR look up and specificity</a>
    </li>
    <li>
      <a href="../articles/mixcr_antibody_expression.html">MiXCR and Antibody expression</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Data simulation</li>
    <li>
      <a href="../articles/echidna.html">Echidna</a>
    </li>
  </ul>
</li>
<li>
  <a href="../PlatypusDB_index.html">PlatypusDB</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://immunesim.readthedocs.io/en/latest/">ImmuneSIM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../cute_monotremes.html">Cute Monotremes</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://twitter.com/alexyermanos">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/KreineVic">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/alexyermanos/Platypus">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Replaces the original sample_id column of a vgm object with a pasted version of the original sample_id and the last digits of the feature barcode.</h1>
    <small class="dont-index">Source: <a href='https://github.com/alexyermanos/Platypus/tree/masterR/VGM_expand_featurebarcodes.R'><code>R/VGM_expand_featurebarcodes.R</code></a></small>
    <div class="hidden name"><code>VGM_expand_featurebarcodes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The VGM_expand_featurebarcodes function function can be used to trace back the cell origin of each sample after using cell hashing for single-cell sequencing.</p>
<p>The original sample_id is stored in a new column called original_sample_id. Additionally, a second new column is created containing final barcode assignment information.
Those barcodes match the origin FB_assignment if by.majority.barcodes is set to FALSE (default). However, if this input parameter is set to TRUE, the majority barcode assignment in stored in this colum.</p>
<p>Note: The majority barcode of a cell is the feature barcode which is  most frequently assigned to the cells clonotype (10x default clonotype).
The majority barcode assignment can be used under the assumption that all cells which are assigned to the same clonotype (within one sample), originate from the same donor organ or at least the same donor depending on the experimental setup.</p>
<p>For example: The original sample_id of a cell is "s1", the cell belongs to "clonotype1" and the feature barcode assigned to it is "i1-TotalSeq-C0953". If by.majority.barcodes default (FALSE) is used, the resulting new sample_id would be "s1_0953".
However, if majority barcode assignment is used AND "i1-TotalSeq-C0953" is not the most frequently occurring barcode in "clonotype1" but rather barcode "i1-TotalSeq-C0951", the new sample_id would be "s1_0951".
--&gt; e.g., if 15 cells belong to clonotype1: 3 cells have no assigned barcode, 2 are assigned to "i1-TotalSeq-C0953" and 10 are assigned to "i1-TotalSeq-C0951" --&gt; all 15 cells will have the new sample_id "s1_0951".</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>VGM_expand_featurebarcodes</span><span class='op'>(</span>
  <span class='va'>vgm</span>,
  <span class='va'>by.majority.barcodes</span>,
  <span class='va'>integrate.in.gex</span>,
  <span class='va'>vdj.only</span>,
  <span class='va'>platypus.version</span>
<span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>vgm</th>
      <td><p>VGM output of VDJ_GEX_matrix function (Platypus V3)</p></td>
    </tr>
    <tr>
      <th>by.majority.barcodes</th>
      <td><p>Logical. Default is FALSE. Indicated whether strict barcode assignment or majority barcode assignment should be used to create the new sample_id. If TRUE, for each clonotype the most frequent feature barcode will be chosen and assigned to each cell, even if that cell itself does not have this particular barcode assigned.</p></td>
    </tr>
    <tr>
      <th>integrate.in.gex</th>
      <td><p>Logical. Default is FALSE. If TRUE, the newly created sample_id's are integrated into gex component as well. Not recommended if no further gex analysis is done due to much longer computational time.</p></td>
    </tr>
    <tr>
      <th>vdj.only</th>
      <td><p>Logical. Defines if only vdj information is provided as input. Default is set to FALSE. If set to TRUE a vdj dataframe has to be provided as input (vgm = vdj). Also, integrate.in.gex is automatically set to FALSE since no gex (vgm[[2]]) information is provided.</p></td>
    </tr>
    <tr>
      <th>platypus.version</th>
      <td><p>This function works with "v3" only, there is no need to set this parameter.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>This function returns a vgm with new sample_id's in case vdj.only is set to FALSE (default). If vdj.only is set to true only the vdj dataframe with new sample_id's is returned.
Note: If vdj.only is set to default (FALSE), VDJ information in the metadata of the GEX object is necessary. For this set integrate.VDJ.to.GEX to TRUE in the VDJ_GEX_matrix function</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'><span class='co'>#For Platypus version 3</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='co'># 1. If only vdj data (vgm[[1]]) and</span></span>
<span class='r-in'><span class='co'>#strict feature barcode assignment is used:</span></span>
<span class='r-in'><span class='va'>vgm_expanded_fb</span> <span class='op'>&lt;-</span> <span class='fu'>VGM_expand_featurebarcodes</span><span class='op'>(</span></span>
<span class='r-in'>vgm <span class='op'>=</span> <span class='va'>small_vgm</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>,</span>
<span class='r-in'>by.majority.barcodes <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span class='r-in'>integrate.in.gex<span class='op'>=</span><span class='cn'>FALSE</span>, vdj.only<span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span>
<span class='r-err co'><span class='r-pr'>#&gt;</span> <span class='error'>Error in if (short_fb == "able") {    short_fb &lt;- "not_assignable"}</span> Argument hat Länge 0</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'># 2. If whole vgm and strict fb assignment is used</span></span>
<span class='r-in'><span class='co'>#(gex and vdj - necessary if gene expression analysis</span></span>
<span class='r-in'><span class='co'># of sub-samples is desired):</span></span>
<span class='r-in'><span class='va'>vgm_expanded_fb</span> <span class='op'>&lt;-</span> <span class='fu'>VGM_expand_featurebarcodes</span><span class='op'>(</span></span>
<span class='r-in'>vgm <span class='op'>=</span> <span class='va'>small_vgm</span>,</span>
<span class='r-in'>by.majority.barcodes <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span class='r-in'>integrate.in.gex<span class='op'>=</span><span class='cn'>TRUE</span>, vdj.only<span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span>
<span class='r-err co'><span class='r-pr'>#&gt;</span> <span class='error'>Error in if (short_fb == "able") {    short_fb &lt;- "not_assignable"}</span> Argument hat Länge 0</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'># 3. If whole vgm and majority barcode assignement is used</span></span>
<span class='r-in'><span class='co'>#(gex and vdj) - necessary if gene expression analysis</span></span>
<span class='r-in'><span class='co'>#of sub-samples is desired):</span></span>
<span class='r-in'><span class='va'>vgm_expanded_fb</span> <span class='op'>&lt;-</span> <span class='fu'>VGM_expand_featurebarcodes</span><span class='op'>(</span>vgm <span class='op'>=</span> <span class='va'>small_vgm</span>,</span>
<span class='r-in'>by.majority.barcodes <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span class='r-in'>integrate.in.gex<span class='op'>=</span><span class='cn'>TRUE</span>, vdj.only<span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span>
<span class='r-err co'><span class='r-pr'>#&gt;</span> <span class='error'>Error in order(curr_feature_barcodes)</span> Argument 1 ist kein Vektor</span>
<span class='r-in'></span>
<span class='r-in'><span class='co'>#Note: Majority barcode assignment is recommended</span></span>
<span class='r-in'><span class='co'>#if the assumption that all cells within one clonotype</span></span>
<span class='r-in'><span class='co'>#originate from the same sample sub-group is feasible.</span></span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Alexander Yermanos.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


