<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pseudo-bulk Vignette • Platypus</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pseudo-bulk Vignette">
<meta property="og:description" content="Platypus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/quickstart.html">Get started</a>
</li>
<li>
  <a href="../vignette_index.html">Vignettes</a>
</li>
<li>
  <a href="../PlatypusDB_index.html">PlatypusDB</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://immunesim.readthedocs.io/en/latest/" class="external-link">ImmuneSIM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../cute_monotremes.html">Cute Monotremes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/alexyermanos" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/KreineVic" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/alexyermanos/Platypus" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pseudo-bulk Vignette</h1>
                        <h4 data-toc-skip class="author">Vittoria
Martinolli D’Arcy, Alexander Yermanos</h4>
            
            <h4 data-toc-skip class="date">30/05/2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alexyermanos/Platypus/tree/mastervignettes/pseudo_bulk.Rmd" class="external-link"><code>vignettes/pseudo_bulk.Rmd</code></a></small>
      <div class="hidden name"><code>pseudo_bulk.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="setup">1. Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>First of all the vgm data is obtained downloading the raw data and
converting into the VDJ_GEX_matrix structure.</p>
</div>
<div class="section level2">
<h2 id="function-usage-and-examples">2. Function usage and examples<a class="anchor" aria-label="anchor" href="#function-usage-and-examples"></a>
</h2>
<p>The pseudo_bulk_DE function takes a VGM object as an input and
performs pseudo-bulk on the data, according to criteria specified by the
User. Differential Gene Expression (DGE) analysis is then performed on
the pseudo-bulked data. The output is a data.frame or list of
data.frames containing the results of the DGE analysis for every level
of pseudo-bulking. In particular three different options are available
for pseudo-bulking the samples:</p>
<div class="section level3">
<h3 id="two-different-pseduo-bulking-categories">2.1 Two different pseduo-bulking categories<a class="anchor" aria-label="anchor" href="#two-different-pseduo-bulking-categories"></a>
</h3>
<p>Choosing one VGM column (column.group) from where to pick two samples
or two groups of samples to test for DGE. In parallel choosing another
column (column.comparison) of the VGM object from where to pick the
“levels of comparison”, meaning that the two samples picked in the first
column will be tested against each other across the categories picked in
the second column, separating the analysis and results according to each
cathegory.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_DE</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/GEX_pseudobulk.html">GEX_pseudobulk</a></span><span class="op">(</span>vgm.input<span class="op">=</span><span class="va">VGM_NP396_GP33_GP66_labelled</span>, column.group<span class="op">=</span><span class="st">"sample_id"</span>, group1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s1"</span><span class="op">)</span>, group2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"s4"</span><span class="op">)</span>, column.comparison <span class="op">=</span> <span class="st">"seurat_clusters"</span>, comparison<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">5</span>,<span class="fl">9</span><span class="op">)</span>, pool<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The output is a list of data.frames each including the results of the
DGE analysis of the respecitive comparison level.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_DE</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##              logFC   logCPM       LR        p_val        p_adj    gene</span></span>
<span><span class="co">## XIST    -10.558188 7.109666 30.28749 3.725201e-08 0.0004881503    XIST</span></span>
<span><span class="co">## TTR       6.851042 7.985671 26.41093 2.759747e-07 0.0012456126     TTR</span></span>
<span><span class="co">## UTY       9.789416 6.354439 26.34764 2.851677e-07 0.0012456126     UTY</span></span>
<span><span class="co">## CD200R1   8.945437 5.546191 21.41346 3.701633e-06 0.0097012385 CD200R1</span></span>
<span><span class="co">## DDX3Y     8.945437 5.546191 21.41346 3.701633e-06 0.0097012385   DDX3Y</span></span>
<span><span class="co">## EIF2S3Y   8.844367 5.450930 20.83781 4.998646e-06 0.0109170433 EIF2S3Y</span></span>
<span><span class="co">## COCH      8.490147 5.120463 18.85308 1.411827e-05 0.0264293922    COCH</span></span>
<span><span class="co">## DSCAM     8.421630 5.057220 18.47565 1.720893e-05 0.0281882270   DSCAM</span></span>
<span><span class="co">## GLRX5    -8.406211 5.060621 17.81171 2.438775e-05 0.0355085710   GLRX5</span></span>
<span><span class="co">## FGL2      4.815742 7.186576 16.33124 5.317987e-05 0.0640682581    FGL2</span></span>
<span><span class="co">## CDH1      8.019631 4.691409 16.30993 5.378135e-05 0.0640682581    CDH1</span></span>
<span><span class="co">## SEMA6D    5.644681 5.399101 15.44681 8.486014e-05 0.0926672692  SEMA6D</span></span>
<span><span class="co">##         comparison_id</span></span>
<span><span class="co">## XIST     comparison_2</span></span>
<span><span class="co">## TTR      comparison_2</span></span>
<span><span class="co">## UTY      comparison_2</span></span>
<span><span class="co">## CD200R1  comparison_2</span></span>
<span><span class="co">## DDX3Y    comparison_2</span></span>
<span><span class="co">## EIF2S3Y  comparison_2</span></span>
<span><span class="co">## COCH     comparison_2</span></span>
<span><span class="co">## DSCAM    comparison_2</span></span>
<span><span class="co">## GLRX5    comparison_2</span></span>
<span><span class="co">## FGL2     comparison_2</span></span>
<span><span class="co">## CDH1     comparison_2</span></span>
<span><span class="co">## SEMA6D   comparison_2</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_DE</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##               logFC   logCPM       LR        p_val        p_adj      gene</span></span>
<span><span class="co">## CCR7       8.657396 7.989358 31.12190 2.423220e-08 0.0002511667      CCR7</span></span>
<span><span class="co">## SCGB1C1    8.229789 7.573113 26.73656 2.331679e-07 0.0012083926   SCGB1C1</span></span>
<span><span class="co">## ACTA2      7.922810 7.280981 23.74721 1.098544e-06 0.0037954689     ACTA2</span></span>
<span><span class="co">## LYZ2       4.323782 8.711587 21.72485 3.146889e-06 0.0081543752      LYZ2</span></span>
<span><span class="co">## SLCO1A4    7.119907 6.558915 16.72673 4.316841e-05 0.0745734286   SLCO1A4</span></span>
<span><span class="co">## TTR        7.119907 6.558915 16.72673 4.316841e-05 0.0745734286       TTR</span></span>
<span><span class="co">## SERPINB6B -7.667660 7.352629 15.81782 6.974257e-05 0.0829426574 SERPINB6B</span></span>
<span><span class="co">## LY6C1      6.995319 6.454071 15.75703 7.201967e-05 0.0829426574     LY6C1</span></span>
<span><span class="co">## XIST      -7.710520 7.388403 16.11945 5.947026e-05 0.0829426574      XIST</span></span>
<span><span class="co">##           comparison_id</span></span>
<span><span class="co">## CCR7       comparison_9</span></span>
<span><span class="co">## SCGB1C1    comparison_9</span></span>
<span><span class="co">## ACTA2      comparison_9</span></span>
<span><span class="co">## LYZ2       comparison_9</span></span>
<span><span class="co">## SLCO1A4    comparison_9</span></span>
<span><span class="co">## TTR        comparison_9</span></span>
<span><span class="co">## SERPINB6B  comparison_9</span></span>
<span><span class="co">## LY6C1      comparison_9</span></span>
<span><span class="co">## XIST       comparison_9</span></span></code></pre>
<p>In the output object there is also a summary indicating the
correspondences between samples, groups and comparison levels.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 8 rows and 4 columns</span></span>
<span><span class="co">##                         sample_id   comparison   groups              levels</span></span>
<span><span class="co">##                       &lt;character&gt;     &lt;factor&gt; &lt;factor&gt;         &lt;character&gt;</span></span>
<span><span class="co">## s1_AAACCTGCAAGCTGGA-1          s1 comparison_3   group1 group1_comparison_3</span></span>
<span><span class="co">## s1_AAACCTGGTGCTCTTC-1          s1 comparison_2   group1 group1_comparison_2</span></span>
<span><span class="co">## s1_AAACGGGGTATTCTCT-1          s1 comparison_5   group1 group1_comparison_5</span></span>
<span><span class="co">## s1_AACTCAGCAAGCCGTC-1          s1 comparison_9   group1 group1_comparison_9</span></span>
<span><span class="co">## s4_AAACCTGGTGCTGTAT-1          s4 comparison_5   group2 group2_comparison_5</span></span>
<span><span class="co">## s4_AAACCTGGTTGACGTT-1          s4 comparison_3   group2 group2_comparison_3</span></span>
<span><span class="co">## s4_AAACCTGTCACGGTTA-1          s4 comparison_2   group2 group2_comparison_2</span></span>
<span><span class="co">## s4_AAAGATGGTGCCTGTG-1          s4 comparison_9   group2 group2_comparison_9</span></span></code></pre>
<p>pseudo_DE[[“number_de”]] reports the percentages and numbers of
defferentially expressed genes per comparison level.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"number_de"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##      number_de percentage_de</span></span>
<span><span class="co">## [1,]        86    0.27145608</span></span>
<span><span class="co">## [2,]        12    0.03787759</span></span>
<span><span class="co">## [3,]        77    0.24304788</span></span>
<span><span class="co">## [4,]         9    0.02840819</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="one-pseudo-bulking-category--pooling-samples-together-into-two-groups-">2.2 One pseudo-bulking category. Pooling samples together into two
groups.<a class="anchor" aria-label="anchor" href="#one-pseudo-bulking-category--pooling-samples-together-into-two-groups-"></a>
</h3>
<p>Choosing one VGM column only (column.group), from which to pick two
groups of samples to test for DGE and pooling together all the samples
belonging to the first group as group1 and all the samples belonging to
the second group as group2. Group1 and group2 will be tested against
each other. This option is allowed by setting pool=TRUE.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_DE_pooled</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/GEX_pseudobulk.html">GEX_pseudobulk</a></span><span class="op">(</span>vgm.input<span class="op">=</span><span class="va">VGM_NP396_GP33_GP66_labelled</span>, column.group<span class="op">=</span><span class="st">"sample_id"</span>, group1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s1"</span>, <span class="st">"s2"</span><span class="op">)</span>, group2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s4"</span>, <span class="st">"s3"</span><span class="op">)</span>, column.comparison <span class="op">=</span> <span class="cn">NULL</span>, comparison<span class="op">=</span><span class="cn">NULL</span>, pool<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pseudo_DE_pooled</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    logFC   logCPM       LR        p_val        p_adj</span></span>
<span><span class="co">## XIST          -11.038506 7.131469 324.7100 1.364543e-72 2.749965e-68</span></span>
<span><span class="co">## UTY            13.446177 5.272431 255.2942 1.820980e-57 1.834910e-53</span></span>
<span><span class="co">## EIF2S3Y         9.865577 5.641570 249.4288 3.459111e-56 2.323716e-52</span></span>
<span><span class="co">## DDX3Y           9.994586 5.171616 236.4760 2.307404e-53 1.162528e-49</span></span>
<span><span class="co">## KDM5D          11.382699 3.235047 165.9324 5.723002e-38 2.306713e-34</span></span>
<span><span class="co">## 3830403N18RIK  -8.916710 2.922212 160.7510 7.754915e-37 2.604747e-33</span></span>
<span><span class="co">##                        gene</span></span>
<span><span class="co">## XIST                   XIST</span></span>
<span><span class="co">## UTY                     UTY</span></span>
<span><span class="co">## EIF2S3Y             EIF2S3Y</span></span>
<span><span class="co">## DDX3Y                 DDX3Y</span></span>
<span><span class="co">## KDM5D                 KDM5D</span></span>
<span><span class="co">## 3830403N18RIK 3830403N18RIK</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="one-pseudo-bulking-category--separating-samples-within-the-same-group-">2.3 One pseudo-bulking category. Separating samples within the same
group.<a class="anchor" aria-label="anchor" href="#one-pseudo-bulking-category--separating-samples-within-the-same-group-"></a>
</h3>
<p>Choosing one VGM column only (column.group), from which to pick two
groups of samples to test for DGE and NOT pooling together samples
belonging to the same group. A pairwise DGE analysis between the samples
of group1 and sample of group2 is performed, therfore the order accoring
to which the samples are inserted in the input matters. This option is
allowed by setting pool=FALSE</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_DE_clusters</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/GEX_pseudobulk.html">GEX_pseudobulk</a></span><span class="op">(</span>vgm.input<span class="op">=</span><span class="va">VGM_NP396_GP33_GP66_labelled</span>, column.group<span class="op">=</span><span class="st">"seurat_clusters"</span>, group1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span><span class="op">)</span>, group2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">9</span><span class="op">)</span>, pool<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pseudo_DE_clusters</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              logFC   logCPM       LR        p_val        p_adj    gene</span></span>
<span><span class="co">## ASCL2    -8.842313 2.758881 64.00282 1.242410e-15 1.297413e-11   ASCL2</span></span>
<span><span class="co">## SOSTDC1  -6.733771 7.309354 63.91531 1.298842e-15 1.297413e-11 SOSTDC1</span></span>
<span><span class="co">## GM47957  -7.864154 2.706008 59.26261 1.379734e-14 9.188111e-11 GM47957</span></span>
<span><span class="co">## CD83     -6.224031 4.578687 54.63028 1.454767e-13 7.265833e-10    CD83</span></span>
<span><span class="co">## GM4208    7.081434 2.326647 53.25790 2.925094e-13 1.168751e-09  GM4208</span></span>
<span><span class="co">## GAL     -10.373562 1.071609 48.54325 3.230840e-12 1.075762e-08     GAL</span></span>
<span><span class="co">##         comparison_id</span></span>
<span><span class="co">## ASCL2    comparison_1</span></span>
<span><span class="co">## SOSTDC1  comparison_1</span></span>
<span><span class="co">## GM47957  comparison_1</span></span>
<span><span class="co">## CD83     comparison_1</span></span>
<span><span class="co">## GM4208   comparison_1</span></span>
<span><span class="co">## GAL      comparison_1</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_DE_clusters</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 3 columns</span></span>
<span><span class="co">##                       seurat_clusters   comparison   groups</span></span>
<span><span class="co">##                              &lt;factor&gt;     &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">## s1_AAACCTGCAAGCTGGA-1               3 comparison_2   group2</span></span>
<span><span class="co">## s1_AAACCTGCACCAGTTA-1               1 comparison_1   group2</span></span>
<span><span class="co">## s1_AAACCTGCATCGGTTA-1               6 comparison_3   group1</span></span>
<span><span class="co">## s1_AAACCTGGTGCTCTTC-1               2 comparison_1   group1</span></span>
<span><span class="co">## s1_AAACGGGAGAATGTTG-1               4 comparison_2   group1</span></span>
<span><span class="co">## s1_AACTCAGCAAGCCGTC-1               9 comparison_3   group2</span></span></code></pre>
<p>#3. Plots</p>
<p>Plotting option is not included in the function, but plots can be
generated with the following few lines of code, to be adapted according
to the data.</p>
<p>##3.1 Violin Plot</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##Violin Plot</span></span>
<span></span>
<span><span class="co"># pull top-n genes for each cluster</span></span>
<span><span class="va">top_gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">top_gs</span><span class="op">$</span><span class="va">comparison_3</span> <span class="op">&lt;-</span> <span class="va">pseudo_DE</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">9</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">top_gs</span><span class="op">$</span><span class="va">comparison_2</span> <span class="op">&lt;-</span> <span class="va">pseudo_DE</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">9</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">top_gs</span><span class="op">$</span><span class="va">comparison_5</span> <span class="op">&lt;-</span> <span class="va">pseudo_DE</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">9</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">top_gs</span><span class="op">$</span><span class="va">comparison_9</span> <span class="op">&lt;-</span> <span class="va">pseudo_DE</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">9</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># get top gene-hits common among clusters, or genes the user is interested in plotting</span></span>
<span><span class="va">top_gs</span><span class="op">$</span><span class="va">common</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"XIST"</span> ,<span class="st">"TTR"</span> ,<span class="st">"UTY"</span>, <span class="st">"DDX3Y"</span>, <span class="st">"EIF2S3Y"</span>,  <span class="st">"COCH"</span>, <span class="st">"CCR7"</span> , <span class="st">"DSCAM"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># split cells by cluster</span></span>
<span><span class="va">cs_by_k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">input</span><span class="op">)</span>,<span class="va">input</span><span class="op">@</span><span class="va">colData</span><span class="op">@</span><span class="va">listData</span><span class="op">[[</span><span class="st">"comparison"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#plot violin plot including all the comparison levels in the same plot</span></span>
<span></span>
<span><span class="va">pseudo_bulk_all_fig</span><span class="op">&lt;-</span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">input</span>, features <span class="op">=</span> <span class="va">top_gs</span><span class="op">$</span><span class="va">common</span>, </span>
<span>        x <span class="op">=</span> <span class="st">"levels"</span>, colour_by <span class="op">=</span> <span class="st">"groups"</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span>, alpha <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk_all_fig</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pseudo_bulk_files/figure-html/unnamed-chunk-9-1.png" alt="Violin Plot, including all the comparison levels in the same plot." width="960"><p class="caption">
Violin Plot, including all the comparison levels in the same plot.
</p>
</div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#otherwise plot violin plot by comparison level k</span></span>
<span></span>
<span><span class="va">gs</span> <span class="op">&lt;-</span> <span class="va">top_gs</span><span class="op">[[</span><span class="st">"comparison_3"</span><span class="op">]</span><span class="op">]</span>  <span class="co"># get top gene-hits for comparison k</span></span>
<span><span class="va">cs</span> <span class="op">&lt;-</span> <span class="va">cs_by_k</span><span class="op">[[</span><span class="st">"comparison_3"</span><span class="op">]</span><span class="op">]</span> <span class="co"># subset cells assigned to comparison k</span></span>
<span><span class="va">comparison_3.fig</span><span class="op">&lt;-</span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">input</span><span class="op">[</span>,<span class="va">cs</span><span class="op">]</span>, features <span class="op">=</span> <span class="va">gs</span>, </span>
<span>        x <span class="op">=</span> <span class="st">"levels"</span>, colour_by <span class="op">=</span> <span class="st">"groups"</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span>, alpha <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">comparison_3.fig</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pseudo_bulk_files/figure-html/unnamed-chunk-10-1.png" alt="Violin Plot, including differential gene expression of of top 9 differentially expressed genes of sample 1 (group1) against sample 4 (group2), testing cells belonging to cluster 3 only." width="700"><p class="caption">
Violin Plot, including differential gene expression of of top 9
differentially expressed genes of sample 1 (group1) against sample 4
(group2), testing cells belonging to cluster 3 only.
</p>
</div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gs</span> <span class="op">&lt;-</span> <span class="va">top_gs</span><span class="op">[[</span><span class="st">"comparison_5"</span><span class="op">]</span><span class="op">]</span>  <span class="co"># get top gene-hits for cluster k</span></span>
<span><span class="va">cs</span> <span class="op">&lt;-</span> <span class="va">cs_by_k</span><span class="op">[[</span><span class="st">"comparison_5"</span><span class="op">]</span><span class="op">]</span> <span class="co"># subset cells assigned to cluster k</span></span>
<span><span class="va">comparison_5.fig</span><span class="op">&lt;-</span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">input</span><span class="op">[</span>,<span class="va">cs</span><span class="op">]</span>, features <span class="op">=</span> <span class="va">gs</span>, </span>
<span>        x <span class="op">=</span> <span class="st">"levels"</span>, colour_by <span class="op">=</span> <span class="st">"groups"</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span>, alpha <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">comparison_5.fig</span><span class="op">)</span>   </span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pseudo_bulk_files/figure-html/unnamed-chunk-11-1.png" alt="Violin Plot, including differential gene expression of of top 9 differentially expressed genes of sample 1 (group1) against sample 4 (group2), testing cells belonging to cluster 5 only." width="700"><p class="caption">
Violin Plot, including differential gene expression of of top 9
differentially expressed genes of sample 1 (group1) against sample 4
(group2), testing cells belonging to cluster 5 only.
</p>
</div>
<p>##3.2 Heatmap</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#to match ms matrix with the summary file of pseudo_DE, a new column needs to be created   </span></span>
<span><span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">grouping</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">grouping</span><span class="op">[</span><span class="va">k</span><span class="op">]</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">sample_id</span><span class="op">[</span><span class="va">k</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">comparison</span><span class="op">[</span><span class="va">k</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span> <span class="op">}</span></span>
<span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">input</span><span class="op">@</span><span class="va">colData</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"sample_id"</span>,<span class="st">"comparison"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">ms</span> <span class="op">&lt;-</span> <span class="fu">aggregate.Matrix</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">input</span><span class="op">@</span><span class="va">assays</span><span class="op">@</span><span class="va">data</span><span class="op">@</span><span class="va">listData</span><span class="op">$</span><span class="va">logcounts</span><span class="op">)</span>, </span>
<span>    groupings <span class="op">=</span> <span class="va">g</span>, fun <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span> </span>
<span><span class="va">ms</span><span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">ms</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#To better highlight relative differences across conditions, construct the helper-function z_norm that applies a z-normalization to normalize expression values to mean 0 and standard deviation 1:</span></span>
<span></span>
<span><span class="va">z_norm</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">th</span> <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">sds</span> <span class="op">&lt;-</span> <span class="fu">rowSds</span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="va">sds</span><span class="op">[</span><span class="va">sds</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>    <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">x</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="va">sds</span><span class="op">)</span></span>
<span>    <span class="co">#x &lt;- (x - rowMeans(x, na.rm = TRUE)) / sds</span></span>
<span>    <span class="va">x</span><span class="op">[</span><span class="va">x</span> <span class="op">&gt;</span>  <span class="va">th</span><span class="op">]</span> <span class="op">&lt;-</span>  <span class="va">th</span></span>
<span>    <span class="va">x</span><span class="op">[</span><span class="va">x</span> <span class="op">&lt;</span> <span class="op">-</span><span class="va">th</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="va">th</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">#plotting function</span></span>
<span><span class="va">plot_diff_hm</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ms</span>, <span class="va">gs</span>, <span class="va">ei</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">mat</span> <span class="op">&lt;-</span> <span class="va">ms</span><span class="op">@</span><span class="va">assays</span><span class="op">@</span><span class="va">data</span><span class="op">@</span><span class="va">listData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">gs</span>, <span class="op">]</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>, <span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">grouping</span><span class="op">)</span></span>
<span>    <span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/hue_pal.html" class="external-link">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nlevels.html" class="external-link">nlevels</a></span><span class="op">(</span><span class="va">ei</span><span class="op">$</span><span class="va">groups</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">ei</span><span class="op">$</span><span class="va">groups</span><span class="op">)</span></span>
<span>    <span class="va">col_anno</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/columnAnnotation.html" class="external-link">columnAnnotation</a></span><span class="op">(</span></span>
<span>        df <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>group_id <span class="op">=</span> <span class="va">ei</span><span class="op">$</span><span class="va">groups</span><span class="op">[</span><span class="va">m</span><span class="op">]</span><span class="op">)</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>groups <span class="op">=</span> <span class="va">cols</span><span class="op">)</span>,</span>
<span>        gp <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>        show_annotation_name <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="fu">z_norm</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>, </span>
<span>        <span class="co">#column_title = k,</span></span>
<span>        name <span class="op">=</span> <span class="st">"z-normalized\nexpression"</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"royalblue"</span>, <span class="st">"cornflowerblue"</span>, <span class="st">"white"</span>, <span class="st">"gold"</span>, <span class="st">"orange"</span><span class="op">)</span>,</span>
<span>        cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        row_names_side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>        rect_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>        top_annotation <span class="op">=</span> <span class="va">col_anno</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="va">heatmap_pb</span><span class="op">&lt;-</span><span class="fu">plot_diff_hm</span><span class="op">(</span><span class="va">ms</span>, <span class="va">top_gs</span><span class="op">$</span><span class="va">common</span>, <span class="va">pseudo_DE</span><span class="op">[[</span><span class="st">"summary"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>  <span class="co"># render differential heatmap</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">heatmap_pb</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pseudo_bulk_files/figure-html/unnamed-chunk-13-1.png" alt="Heatmap, displaying normalised gene expression of of top 10 differentially expressed genes for each combination of samples (group) and clusters (comparison) analysed" width="700"><p class="caption">
Heatmap, displaying normalised gene expression of of top 10
differentially expressed genes for each combination of samples (group)
and clusters (comparison) analysed
</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alexander Yermanos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
