<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Integrating VDJ and GEX • Platypus</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Integrating VDJ and GEX">
<meta property="og:description" content="Platypus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Platypus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/quickstart.html">Get started</a>
</li>
<li>
  <a href="../vignette_index.html">Vignettes</a>
</li>
<li>
  <a href="../PlatypusDB_index.html">PlatypusDB</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://immunesim.readthedocs.io/en/latest/" class="external-link">ImmuneSIM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../cute_monotremes.html">Cute Monotremes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/alexyermanos" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/KreineVic" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/alexyermanos/Platypus" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Integrating VDJ and GEX</h1>
                        <h4 data-toc-skip class="author">Victor Kreiner,
Alexander Yermanos</h4>
            
            <h4 data-toc-skip class="date">2022-07-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alexyermanos/Platypus/tree/mastervignettes/vdjgex_integrate.Rmd" class="external-link"><code>vignettes/vdjgex_integrate.Rmd</code></a></small>
      <div class="hidden name"><code>vdjgex_integrate.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The VDJ GEX matrix function and object are central Platypus, as they
offer a format which stores VDJ and GEX in integrated formats. In most
cases, integration of VDJ and GEX is done by the VDJ GEX matrix
function. For all its parameters please refer to the VGM overview
vignette. Additional or integration post-VGM is also necessary e.g. when
new metadata or analysis results have been added to either datasets.
Further, it can be useful to itegrate expression values of selected
genes to VDJ for filtering and grouping purposes.</p>
<p>For these applications, the function VGM_integrate was designed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Platypus</span><span class="op">)</span></code></pre></div>
<p>For showcasing, we download data from T cells in the CNS of aged
mice. Analysis of this dataset is showcased in the Platypus Quickstart
vignette.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/PlatypusDB_fetch.html">PlatypusDB_fetch</a></span><span class="op">(</span>PlatypusDB.links <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"yermanos2021a//ALL"</span><span class="op">)</span>,load.to.enviroment <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## [1] "yermanos2021a__VDJGEXmatrix"</span></code></pre>
</div>
<div class="section level2">
<h2 id="usecase-1-reclonotyping-gex">2. Usecase 1: Reclonotyping GEX<a class="anchor" aria-label="anchor" href="#usecase-1-reclonotyping-gex"></a>
</h2>
<p>Several functions are available to explore GEX data by VDJ
clonotypes. To use different clonotyping strategies, we can use the
VGM_integrate function</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Reclonotype with different strategy</span>
<span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonotype.html">VDJ_clonotype</a></span><span class="op">(</span><span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
                                                     clone.strategy <span class="op">=</span> <span class="st">"cdr3.homology"</span>,
                                                     global.clonotype <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Clonotyping strategy: cdr3.homology</span></code></pre>
<pre><code><span class="co">## Processing sample 1 of 1</span></code></pre>
<pre><code><span class="co">## Backing up 10x default clonotyping in columns clonotype_id_10x and clonotype_frequency_10x before updating clonotype_id and clonotype_frequency columns</span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Reintegrate new clonotyping info to GEX</span>
<span class="va">yermanos2021a__VDJGEXmatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VGM_integrate.html">VGM_integrate</a></span><span class="op">(</span><span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Cells in GEX: 2064; cells in VDJ: 463</span>
<span class="co">##  Overlap: 455</span></code></pre>
<pre><code><span class="co">## Integrating all VDJ columns which are not in GEX to GEX</span></code></pre>
<pre><code><span class="co">## Added columns to GEX: clonotype_id; global_clonotype_frequency_cdr3.homology; global_clonotype_id_cdr3.homology; global_clonal_feature_cdr3.homology; clonotype_frequency_10x; clonotype_id; clonotype_frequency</span></code></pre>
<p>Done, now we can plot GEX clone overlay using the new clonotyping
strategy</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/VDJ_GEX_overlay_clones.html">VDJ_GEX_overlay_clones</a></span><span class="op">(</span><span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, by.sample <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Scale for 'colour' is already present. Adding another scale for 'colour',</span>
<span class="co">## which will replace the existing scale.</span></code></pre>
<p><img src="vdjgex_integrate_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="usecase-2-adding-specific-columns-from-gex-to-vdj">3. Usecase 2: Adding specific columns from GEX to VDJ<a class="anchor" aria-label="anchor" href="#usecase-2-adding-specific-columns-from-gex-to-vdj"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#add a new column to GEX by basic phenotyping</span>
<span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GEX_phenotype.html">GEX_phenotype</a></span><span class="op">(</span><span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,
                                                  default <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>

<span class="co">#transferring to VDJ</span>
<span class="va">yermanos2021a__VDJGEXmatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VGM_integrate.html">VGM_integrate</a></span><span class="op">(</span><span class="va">yermanos2021a__VDJGEXmatrix</span>, 
                                             columns.to.transfer <span class="op">=</span> <span class="st">"cell.state"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Cells in GEX: 2064; cells in VDJ: 463</span>
<span class="co">##  Overlap: 455</span></code></pre>
<pre><code><span class="co">## Added columns to VDJ: cell.state</span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#and plotting this</span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, group.by <span class="op">=</span> <span class="st">"cell.state"</span><span class="op">)</span></code></pre></div>
<p><img src="vdjgex_integrate_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="usecase-3-adding-gene-expression-values-to-vdj">4. Usecase 3: Adding gene expression values to VDJ<a class="anchor" aria-label="anchor" href="#usecase-3-adding-gene-expression-values-to-vdj"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">yermanos2021a__VDJGEXmatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VGM_integrate.html">VGM_integrate</a></span><span class="op">(</span><span class="va">yermanos2021a__VDJGEXmatrix</span>, 
                                             genes.to.VDJ <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3E"</span>,<span class="st">"CD8A"</span><span class="op">)</span>,
                                             seurat.slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Cells in GEX: 2064; cells in VDJ: 463</span>
<span class="co">##  Overlap: 455</span></code></pre>
<pre><code><span class="co">## Added columns to VDJ: CD3E; CD8A</span></code></pre>
<p>Now we can filter and show clontoype expansion based on this</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">CD8A_binary</span> <span class="op">&lt;-</span> <span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">CD8A</span> <span class="op">&gt;</span> <span class="fl">0</span>

<span class="va">expansion_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VDJ_clonal_expansion.html">VDJ_clonal_expansion</a></span><span class="op">(</span><span class="va">yermanos2021a__VDJGEXmatrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, 
                     color.by <span class="op">=</span> <span class="st">"CD8A_binary"</span>, 
                     group.by <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>

<span class="va">expansion_plot</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## [[1]]</span></code></pre>
<p><img src="vdjgex_integrate_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<p>We can see that most expanded clones in this dataset are CD8+. “none”
or blue indicates that their is no GEX information for these cells</p>
</div>
<div class="section level2">
<h2 id="session-information">5. Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.0.5 (2021-03-31)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 19044)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252   </span>
<span class="co">## [3] LC_MONETARY=German_Germany.1252 LC_NUMERIC=C                   </span>
<span class="co">## [5] LC_TIME=German_Germany.1252    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] Platypus_3.4.0     SeuratObject_4.0.4 Seurat_4.1.0       forcats_0.5.1     </span>
<span class="co">##  [5] stringr_1.4.0      purrr_0.3.4        readr_2.1.2        tidyr_1.1.3       </span>
<span class="co">##  [9] tibble_3.1.2       ggplot2_3.3.5      tidyverse_1.3.1    dplyr_1.0.7       </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] readxl_1.3.1          backports_1.4.1       systemfonts_1.0.4    </span>
<span class="co">##   [4] plyr_1.8.6            igraph_1.2.6          lazyeval_0.2.2       </span>
<span class="co">##   [7] splines_4.0.5         listenv_0.8.0         scattermore_0.7      </span>
<span class="co">##  [10] digest_0.6.27         useful_1.2.6          htmltools_0.5.2      </span>
<span class="co">##  [13] fansi_0.5.0           magrittr_2.0.1        memoise_2.0.1        </span>
<span class="co">##  [16] tensor_1.5            cluster_2.1.2         ROCR_1.0-11          </span>
<span class="co">##  [19] tzdb_0.2.0            globals_0.14.0        modelr_0.1.8         </span>
<span class="co">##  [22] matrixStats_0.59.0    pkgdown_2.0.2         spatstat.sparse_2.0-0</span>
<span class="co">##  [25] colorspace_2.0-2      rvest_1.0.2           ggrepel_0.9.1        </span>
<span class="co">##  [28] textshaping_0.3.6     haven_2.4.3           xfun_0.27            </span>
<span class="co">##  [31] crayon_1.5.0          jsonlite_1.7.2        spatstat.data_2.1-2  </span>
<span class="co">##  [34] survival_3.2-11       zoo_1.8-9             glue_1.4.2           </span>
<span class="co">##  [37] polyclip_1.10-0       gtable_0.3.0          leiden_0.3.9         </span>
<span class="co">##  [40] future.apply_1.8.1    abind_1.4-5           scales_1.1.1         </span>
<span class="co">##  [43] DBI_1.1.2             miniUI_0.1.1.1        Rcpp_1.0.7           </span>
<span class="co">##  [46] viridisLite_0.4.0     xtable_1.8-4          reticulate_1.20      </span>
<span class="co">##  [49] spatstat.core_2.2-0   htmlwidgets_1.5.4     httr_1.4.2           </span>
<span class="co">##  [52] RColorBrewer_1.1-2    ellipsis_0.3.2        ica_1.0-2            </span>
<span class="co">##  [55] farver_2.1.0          pkgconfig_2.0.3       uwot_0.1.10          </span>
<span class="co">##  [58] sass_0.4.0            dbplyr_2.1.1          deldir_0.2-10        </span>
<span class="co">##  [61] utf8_1.2.1            labeling_0.4.2        tidyselect_1.1.1     </span>
<span class="co">##  [64] rlang_0.4.10          reshape2_1.4.4        later_1.2.0          </span>
<span class="co">##  [67] munsell_0.5.0         cellranger_1.1.0      tools_4.0.5          </span>
<span class="co">##  [70] cachem_1.0.6          cli_3.1.1             generics_0.1.2       </span>
<span class="co">##  [73] broom_0.7.12          ggridges_0.5.3        evaluate_0.14        </span>
<span class="co">##  [76] fastmap_1.1.0         yaml_2.2.1            ragg_1.2.1           </span>
<span class="co">##  [79] goftest_1.2-2         knitr_1.37            fs_1.5.2             </span>
<span class="co">##  [82] fitdistrplus_1.1-6    RANN_2.6.1            pbapply_1.5-0        </span>
<span class="co">##  [85] future_1.24.0         nlme_3.1-152          mime_0.11            </span>
<span class="co">##  [88] xml2_1.3.3            compiler_4.0.5        rstudioapi_0.13      </span>
<span class="co">##  [91] plotly_4.10.0         png_0.1-7             spatstat.utils_2.2-0 </span>
<span class="co">##  [94] reprex_2.0.1          bslib_0.3.1           stringi_1.7.4        </span>
<span class="co">##  [97] highr_0.9             desc_1.4.0            lattice_0.20-44      </span>
<span class="co">## [100] Matrix_1.3-4          vctrs_0.3.8           stringdist_0.9.8     </span>
<span class="co">## [103] pillar_1.7.0          lifecycle_1.0.1       spatstat.geom_2.2-0  </span>
<span class="co">## [106] lmtest_0.9-38         jquerylib_0.1.4       RcppAnnoy_0.0.18     </span>
<span class="co">## [109] data.table_1.14.0     cowplot_1.1.1         irlba_2.3.3          </span>
<span class="co">## [112] httpuv_1.6.1          patchwork_1.1.1       R6_2.5.1             </span>
<span class="co">## [115] promises_1.2.0.1      KernSmooth_2.23-20    gridExtra_2.3        </span>
<span class="co">## [118] parallelly_1.30.0     codetools_0.2-18      MASS_7.3-54          </span>
<span class="co">## [121] assertthat_0.2.1      rprojroot_2.0.2       withr_2.4.3          </span>
<span class="co">## [124] sctransform_0.3.3     mgcv_1.8-36           parallel_4.0.5       </span>
<span class="co">## [127] hms_1.1.1             grid_4.0.5            rpart_4.1-15         </span>
<span class="co">## [130] rmarkdown_2.11        Rtsne_0.15            shiny_1.7.1          </span>
<span class="co">## [133] lubridate_1.8.0</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alexander Yermanos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
